# Practical Perl program 

## Extract sequences from genomic DNA for some specific regions

Figure 8.1 shows how to extract sequences from genomic DNA for some specific regions. For example, when we perform ChIP-seq for a transcription factor. We will get the binding sites (some regions) of transcription factor after analyzing ChIP-seq. You want to know whether there is motifs that can be recognized by the transcription factor. 

The Perl program in Figure 8.1 has two arguments (input files): genomic DNA sequences and ChIP-seq peak regions. It will print the genomic DNA sequences for each region in fasta format.  We can use `>` to generate a new file to record the result (Figure 8.1). 

The genomic DNA sequence data is saved in fasta format (See 11.1). ChIP-seq peak regions are saved in BED format (See 12.3). 

Letâ€™s go through each line in detail. 

A normal paragraph.

The above code provides an interactive HTML page (figure \@ref(fig:foo))

(ref:foo) A scatterplot of the data `cars` using **base** R graphics. 

```{r foo, fig.cap='(ref:foo)'}
plot(1)  # a scatterplot
```


## Merge overlap genomic regions

```{perl code=readLines("code_perl/genomic_coordinate_merge.pl"), eval=FALSE}
```

```{sh}
cat data/DMR_region.txt
```

```{sh}
perl code_perl/genomic_coordinate_merge.pl  data/DMR_region.txt  data/DMR_region_merged.txt 
cat data/DMR_region_merged.txt 
```
