<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>All things considered for Bioinformatics</title>
  <meta name="description" content="This is a book for people who wants to learn Bioinformatics from scratch">
  <meta name="generator" content="bookdown <!--bookdown:version--> and GitBook 2.6.7">

  <meta property="og:title" content="All things considered for Bioinformatics" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="images/cover.jpg" />
  <meta property="og:description" content="This is a book for people who wants to learn Bioinformatics from scratch" />
  <meta name="github-repo" content="xie186/Novice2Expert4Bioinformatics" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="All things considered for Bioinformatics" />
  
  <meta name="twitter:description" content="This is a book for people who wants to learn Bioinformatics from scratch" />
  <meta name="twitter:image" content="images/cover.jpg" />

<meta name="author" content="Shaojun Xie (解少俊)">


<meta name="date" content="2019-04-15">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



<!--bookdown:title:start-->
<div id="header">
<h1 class="title">All things considered for Bioinformatics</h1>
<h2 class="subtitle"><em>Essential Linux, Python, Perl, R and Statistics in life science</em></h2>
<p class="author"><em>Shaojun Xie (解少俊)</em></p>
<p class="date"><em>2019-04-15</em></p>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#preface">Preface</a></li>
<li><a href="#part-linux">(PART) Linux</a></li>
<li><a href="#why-linux"><span class="toc-section-number">1</span> Why Linux?</a><ul>
<li><a href="#what-is-linux"><span class="toc-section-number">1.1</span> What is Linux</a></li>
<li><a href="#linux-for-bioinformatics"><span class="toc-section-number">1.2</span> Linux for bioinformatics</a></li>
</ul></li>
<li><a href="#connecting-to-linux"><span class="toc-section-number">2</span> Connecting to Linux</a><ul>
<li><a href="#user-interfaces"><span class="toc-section-number">2.1</span> User interfaces</a></li>
<li><a href="#how-to-connect"><span class="toc-section-number">2.2</span> How to connect</a><ul>
<li><a href="#digitalocean-droplet"><span class="toc-section-number">2.2.1</span> DigitalOcean droplet</a></li>
<li><a href="#aws-free-tier-offering"><span class="toc-section-number">2.2.2</span> AWS Free Tier offering</a></li>
</ul></li>
</ul></li>
<li><a href="#file-system-and-commands-of-linux"><span class="toc-section-number">3</span> File system and commands of Linux</a><ul>
<li><a href="#path"><span class="toc-section-number">3.1</span> Path</a><ul>
<li><a href="#relative-and-absolute-path"><span class="toc-section-number">3.1.1</span> Relative and absolute path</a></li>
<li><a href="#comand-ls"><span class="toc-section-number">3.1.2</span> Comand <code>ls</code></a></li>
<li><a href="#command-cd"><span class="toc-section-number">3.1.3</span> Command <code>cd</code></a></li>
</ul></li>
<li><a href="#manipulations-of-files-and-directories"><span class="toc-section-number">3.2</span> Manipulations of files and directories</a><ul>
<li><a href="#command-cp"><span class="toc-section-number">3.2.1</span> Command <code>cp</code></a></li>
<li><a href="#command-mv"><span class="toc-section-number">3.2.2</span> Command <code>mv</code></a></li>
<li><a href="#command-mkdir"><span class="toc-section-number">3.2.3</span> Command <code>mkdir</code></a></li>
<li><a href="#command-rm"><span class="toc-section-number">3.2.4</span> Command ‘rm’</a></li>
</ul></li>
<li><a href="#viewing-text-files-in-linux"><span class="toc-section-number">3.3</span> Viewing text files in Linux</a><ul>
<li><a href="#command-cat"><span class="toc-section-number">3.3.1</span> Command <code>cat</code></a></li>
<li><a href="#command-more-and-less"><span class="toc-section-number">3.3.2</span> Command <code>more</code> and <code>less</code></a></li>
<li><a href="#command-head-and-tail"><span class="toc-section-number">3.3.3</span> Command <code>head</code> and <code>tail</code></a></li>
<li><a href="#auto-completion"><span class="toc-section-number">3.3.4</span> Auto-completion</a></li>
</ul></li>
<li><a href="#understand-standard-input-and-stardard-output"><span class="toc-section-number">3.4</span> Understand standard input and stardard output</a><ul>
<li><a href="#stdin"><span class="toc-section-number">3.4.1</span> STDIN</a></li>
<li><a href="#stdout"><span class="toc-section-number">3.4.2</span> STDOUT</a></li>
<li><a href="#stderr"><span class="toc-section-number">3.4.3</span> STDERR</a></li>
</ul></li>
<li><a href="#find-disk-usage-of-files-and-directories"><span class="toc-section-number">3.5</span> Find Disk Usage of Files and Directories</a></li>
<li><a href="#advanced-topic"><span class="toc-section-number">3.6</span> Advanced topic</a><ul>
<li><a href="#linux-md5sum-command"><span class="toc-section-number">3.6.1</span> Linux <code>md5sum</code> Command</a></li>
</ul></li>
<li><a href="#filtering-output"><span class="toc-section-number">3.7</span> Filtering Output</a><ul>
<li><a href="#cut"><span class="toc-section-number">3.7.1</span> <code>cut</code></a></li>
</ul></li>
<li><a href="#check-you-job-status"><span class="toc-section-number">3.8</span> Check you job status</a></li>
<li><a href="#finding-things"><span class="toc-section-number">3.9</span> Finding Things</a><ul>
<li><a href="#list-files-bigger-than-filesize-specified"><span class="toc-section-number">3.9.1</span> List files bigger than filesize specified</a></li>
</ul></li>
</ul></li>
<li><a href="#achiving-and-compressing-files"><span class="toc-section-number">4</span> Achiving and compressing files</a><ul>
<li><a href="#common-compressed-file-format"><span class="toc-section-number">4.1</span> Common compressed file format</a></li>
<li><a href="#how-to-work-with-different-format"><span class="toc-section-number">4.2</span> How to work with different format</a><ul>
<li><a href="#gz"><span class="toc-section-number">4.2.1</span> <code>*.gz</code></a></li>
<li><a href="#tar.gz"><span class="toc-section-number">4.2.2</span> <code>*.tar.gz</code></a></li>
<li><a href="#zip"><span class="toc-section-number">4.2.3</span> <code>*.zip</code></a></li>
<li><a href="#tar.bz2"><span class="toc-section-number">4.2.4</span> <code>*.tar.bz2</code></a></li>
</ul></li>
</ul></li>
<li><a href="#process-management-in-linux"><span class="toc-section-number">5</span> Process management in Linux</a><ul>
<li><a href="#top"><span class="toc-section-number">5.1</span> <code>top</code></a></li>
<li><a href="#ps"><span class="toc-section-number">5.2</span> <code>ps</code></a></li>
<li><a href="#kill"><span class="toc-section-number">5.3</span> <code>kill</code></a></li>
<li><a href="#df"><span class="toc-section-number">5.4</span> <code>df</code></a></li>
<li><a href="#advanced-topic-free"><span class="toc-section-number">5.5</span> Advanced topic <code>free</code></a></li>
</ul></li>
<li><a href="#file-transfer"><span class="toc-section-number">6</span> File transfer</a><ul>
<li><a href="#transferring-files-between-local-computer-and-linux-server"><span class="toc-section-number">6.1</span> Transferring files between local computer and Linux server</a><ul>
<li><a href="#use-command-line-tools"><span class="toc-section-number">6.1.1</span> Use command line tools</a></li>
<li><a href="#download-files"><span class="toc-section-number">6.1.2</span> Download files</a></li>
</ul></li>
</ul></li>
<li><a href="#install-bioinformatics-software-in-linux"><span class="toc-section-number">7</span> Install Bioinformatics software in Linux</a><ul>
<li><a href="#installation-from-source-code"><span class="toc-section-number">7.1</span> Installation from source code</a><ul>
<li><a href="#install-bwa"><span class="toc-section-number">7.1.1</span> Install <code>bwa</code></a></li>
<li><a href="#install-samtools"><span class="toc-section-number">7.1.2</span> Install <code>samtools</code></a></li>
<li><a href="#align-reads-to-genome-using-bwa-and-store-the-alignment-results-in-sambam-files"><span class="toc-section-number">7.1.3</span> Align reads to genome using <code>bwa</code> and store the alignment results in SAM/BAM files</a></li>
</ul></li>
<li><a href="#installing-a-precompiled-binary-executable"><span class="toc-section-number">7.2</span> Installing a precompiled binary (executable)</a><ul>
<li><a href="#install-bwa-1"><span class="toc-section-number">7.2.1</span> Install <code>bwa</code></a></li>
<li><a href="#install-with-conda-recommended-way"><span class="toc-section-number">7.2.2</span> Install with <code>conda</code> (recommended way)</a></li>
<li><a href="#install-ussing-docker-advance-topic"><span class="toc-section-number">7.2.3</span> Install ussing <code>Docker</code> (advance topic)</a></li>
</ul></li>
</ul></li>
<li><a href="#text-editor-in-linux"><span class="toc-section-number">8</span> Text editor in Linux</a><ul>
<li><a href="#basic-vi-skills"><span class="toc-section-number">8.1</span> Basic vi skills</a></li>
<li><a href="#create-new-text-file-with-vi"><span class="toc-section-number">8.2</span> Create new text file with <code>vi</code></a></li>
<li><a href="#an-example-for-using-editor-r"><span class="toc-section-number">8.3</span> An example for using editor R</a></li>
</ul></li>
<li><a href="#part-perl">(PART) Perl</a></li>
<li><a href="#first-perl-program"><span class="toc-section-number">9</span> First Perl Program</a><ul>
<li><a href="#first-program"><span class="toc-section-number">9.1</span> First Program</a></li>
</ul></li>
<li><a href="#varible-in-perl"><span class="toc-section-number">10</span> Varible in Perl</a><ul>
<li><a href="#scalar-variable"><span class="toc-section-number">10.1</span> Scalar variable</a></li>
<li><a href="#arithmetic-operations-in-perl"><span class="toc-section-number">10.2</span> Arithmetic operations in Perl</a><ul>
<li><a href="#shorthand-operations"><span class="toc-section-number">10.2.1</span> Shorthand operations</a></li>
<li><a href="#auto-increment-and-auto-decrement"><span class="toc-section-number">10.2.2</span> Auto increment and auto decrement</a></li>
</ul></li>
<li><a href="#use-strict-user-warnings-and-my"><span class="toc-section-number">10.3</span> <code>use strict</code>; <code>user warnings</code> and <code>my</code></a></li>
<li><a href="#array"><span class="toc-section-number">10.4</span> Array</a><ul>
<li><a href="#init-an-array"><span class="toc-section-number">10.4.1</span> Init an array</a></li>
<li><a href="#array-index"><span class="toc-section-number">10.4.2</span> Array index</a></li>
<li><a href="#length-of-the-array"><span class="toc-section-number">10.4.3</span> Length of the array</a></li>
<li><a href="#sort-arrays-in-perl"><span class="toc-section-number">10.4.4</span> Sort Arrays in Perl</a></li>
<li><a href="#use-push-pop-shift-and-unshift-in-perl"><span class="toc-section-number">10.4.5</span> Use <code>push</code>, <code>pop</code>, <code>shift</code> and <code>unshift</code> in Perl</a></li>
</ul></li>
<li><a href="#hash-in-perl"><span class="toc-section-number">10.5</span> Hash in Perl</a><ul>
<li><a href="#sort-keys-numerically-using-sort-ab"><span class="toc-section-number">10.5.1</span> Sort keys numerically using <code>sort {$a&lt;=&gt;$b}</code></a></li>
<li><a href="#use-exists-function-on-a-hash"><span class="toc-section-number">10.5.2</span> Use <code>exists</code> function on a hash</a></li>
</ul></li>
</ul></li>
<li><a href="#control-structure"><span class="toc-section-number">11</span> Control structure</a><ul>
<li><a href="#for-loop"><span class="toc-section-number">11.1</span> <code>for</code> loop</a></li>
<li><a href="#foreach-loop"><span class="toc-section-number">11.2</span> <code>foreach</code> loop</a></li>
<li><a href="#while-loop"><span class="toc-section-number">11.3</span> <code>while</code> loop</a></li>
<li><a href="#statement-if-else"><span class="toc-section-number">11.4</span> Statement if-else</a></li>
<li><a href="#operator-last-and-next"><span class="toc-section-number">11.5</span> Operator <code>last</code> and <code>next</code></a></li>
<li><a href="#operator-redo"><span class="toc-section-number">11.6</span> Operator <code>redo</code></a></li>
</ul></li>
<li><a href="#string-manipulation"><span class="toc-section-number">12</span> String manipulation</a><ul>
<li><a href="#string-concatenation"><span class="toc-section-number">12.1</span> String concatenation</a></li>
<li><a href="#substring-extraction"><span class="toc-section-number">12.2</span> Substring extraction</a></li>
<li><a href="#substring-search"><span class="toc-section-number">12.3</span> Substring search</a></li>
<li><a href="#split-string"><span class="toc-section-number">12.4</span> Split String</a></li>
<li><a href="#regular-expression"><span class="toc-section-number">12.5</span> Regular expression</a><ul>
<li><a href="#pattern-matching"><span class="toc-section-number">12.5.1</span> Pattern matching</a></li>
<li><a href="#pattern-substitution"><span class="toc-section-number">12.5.2</span> Pattern substitution</a></li>
<li><a href="#modifiers-to-pattern-matching-and-substitution"><span class="toc-section-number">12.5.3</span> Modifiers to pattern matching and substitution</a></li>
<li><a href="#greedy-or-non-greedy"><span class="toc-section-number">12.5.4</span> Greedy or non-greedy</a></li>
</ul></li>
</ul></li>
<li><a href="#input-and-output-in-perl"><span class="toc-section-number">13</span> Input and output in Perl</a><ul>
<li><a href="#input"><span class="toc-section-number">13.1</span> Input</a><ul>
<li><a href="#standard-input"><span class="toc-section-number">13.1.1</span> Standard input</a></li>
<li><a href="#input-from-a-file-on-the-disk"><span class="toc-section-number">13.1.2</span> Input from a file on the disk</a></li>
<li><a href="#special-variable-_"><span class="toc-section-number">13.1.3</span> Special variable <code>$_</code></a></li>
<li><a href="#the-input-record-separator"><span class="toc-section-number">13.1.4</span> The input record separator: <code>$/</code></a></li>
</ul></li>
<li><a href="#output-to-a-file-on-the-disk"><span class="toc-section-number">13.2</span> Output to a file on the disk</a></li>
<li><a href="#arguments-from-command-line"><span class="toc-section-number">13.3</span> Arguments from command line</a></li>
</ul></li>
<li><a href="#practical-perl-program"><span class="toc-section-number">14</span> Practical Perl program</a><ul>
<li><a href="#add-annotation-information-to-deseq2-results"><span class="toc-section-number">14.1</span> Add annotation information to DESeq2 results</a></li>
<li><a href="#merge-overlap-genomic-regions"><span class="toc-section-number">14.2</span> Merge overlap genomic regions</a></li>
</ul></li>
<li><a href="#perl-modules"><span class="toc-section-number">15</span> Perl modules</a><ul>
<li><a href="#what-is-a-perl-module"><span class="toc-section-number">15.1</span> What is a Perl module</a></li>
<li><a href="#how-to-install-a-perl-module"><span class="toc-section-number">15.2</span> How to install a Perl module</a></li>
<li><a href="#how-to-use-a-perl-module"><span class="toc-section-number">15.3</span> How to use a Perl module</a></li>
<li><a href="#how-to-use-bioperl-module"><span class="toc-section-number">15.4</span> How to use BioPerl module</a></li>
<li><a href="#how-to-write-a-module"><span class="toc-section-number">15.5</span> How to write a module</a></li>
</ul></li>
<li><a href="#part-statistics-and-r">(PART) Statistics and R</a></li>
<li><a href="#r-introduction"><span class="toc-section-number">16</span> R introduction</a><ul>
<li><a href="#basic-r-function"><span class="toc-section-number">16.1</span> Basic R function</a></li>
<li><a href="#producing-simple-graphs-with-r"><span class="toc-section-number">16.2</span> Producing Simple Graphs with R</a><ul>
<li><a href="#line-charts"><span class="toc-section-number">16.2.1</span> Line Charts</a></li>
</ul></li>
</ul></li>
<li><a href="#producing-simple-graphs"><span class="toc-section-number">17</span> Producing simple graphs</a><ul>
<li><a href="#producing-simple-graphs-using-r"><span class="toc-section-number">17.1</span> Producing simple graphs using R</a><ul>
<li><a href="#line-charts-1"><span class="toc-section-number">17.1.1</span> Line Charts</a></li>
<li><a href="#bar-charts"><span class="toc-section-number">17.1.2</span> Bar Charts</a></li>
<li><a href="#histograms"><span class="toc-section-number">17.1.3</span> Histograms</a></li>
<li><a href="#pie-charts"><span class="toc-section-number">17.1.4</span> Pie Charts</a></li>
<li><a href="#dotcharts"><span class="toc-section-number">17.1.5</span> Dotcharts</a></li>
<li><a href="#misc"><span class="toc-section-number">17.1.6</span> Misc</a></li>
</ul></li>
</ul></li>
<li><a href="#ggplot2"><span class="toc-section-number">18</span> <code>ggplot2</code></a><ul>
<li><a href="#ggplot2-1"><span class="toc-section-number">18.1</span> ggplot2</a></li>
<li><a href="#ggplot2-practical"><span class="toc-section-number">18.2</span> ggplot2 practical</a></li>
</ul></li>
<li><a href="#heatmap-tutorial"><span class="toc-section-number">19</span> Heatmap Tutorial</a><ul>
<li><a href="#install-pheatmap-package"><span class="toc-section-number">19.1</span> Install <code>pheatmap</code> package</a></li>
<li><a href="#draw-a-heatmap-for-gene-expression-of-rna-seq-data"><span class="toc-section-number">19.2</span> Draw a heatmap for gene expression of RNA-seq data</a></li>
<li><a href="#add-the-annotation"><span class="toc-section-number">19.3</span> Add the annotation</a></li>
<li><a href="#section"><span class="toc-section-number">19.4</span> </a></li>
<li><a href="#transfrom-the-data"><span class="toc-section-number">19.5</span> Transfrom the data</a></li>
<li><a href="#how-to-add-annotations"><span class="toc-section-number">19.6</span> How to add annotations</a></li>
<li><a href="#how-to-cut-the-trees"><span class="toc-section-number">19.7</span> How to cut the trees</a></li>
<li><a href="#how-to-get-the-cluster-information-from-the-heatmap"><span class="toc-section-number">19.8</span> How to get the cluster information from the heatmap</a></li>
<li><a href="#section-1"><span class="toc-section-number">19.9</span> </a></li>
</ul></li>
<li><a href="#preparation-of-figures-for-manuscript"><span class="toc-section-number">20</span> Preparation of figures for manuscript</a><ul>
<li><a href="#test1"><span class="toc-section-number">20.1</span> test1</a></li>
</ul></li>
<li><a href="#part-omics-data">(PART) Omics data</a></li>
<li><a href="#introduction-to-ngs"><span class="toc-section-number">21</span> Introduction to NGS</a><ul>
<li><a href="#introduction-of-biology-for-bioinformatics"><span class="toc-section-number">21.1</span> Introduction of Biology for Bioinformatics</a><ul>
<li><a href="#what-is-dna"><span class="toc-section-number">21.1.1</span> What is DNA</a></li>
<li><a href="#what-is-genome"><span class="toc-section-number">21.1.2</span> What is genome</a></li>
<li><a href="#how-to-assemble-a-genome"><span class="toc-section-number">21.1.3</span> How to assemble a genome</a></li>
<li><a href="#what-is-gene"><span class="toc-section-number">21.1.4</span> What is gene</a></li>
</ul></li>
<li><a href="#what-is-ngs"><span class="toc-section-number">21.2</span> What is NGS</a></li>
<li><a href="#application-of-ngs"><span class="toc-section-number">21.3</span> Application of NGS</a></li>
<li><a href="#section-2"><span class="toc-section-number">21.4</span> </a></li>
<li><a href="#data-file-formats-to-store-the-genomic-information"><span class="toc-section-number">21.5</span> Data file formats to store the genomic information</a><ul>
<li><a href="#fasta-file"><span class="toc-section-number">21.5.1</span> <code>fasta</code> file</a></li>
<li><a href="#gff-file"><span class="toc-section-number">21.5.2</span> GFF file</a></li>
<li><a href="#gtf-file"><span class="toc-section-number">21.5.3</span> GTF file</a></li>
<li><a href="#fastq-file"><span class="toc-section-number">21.5.4</span> FASTQ file</a></li>
<li><a href="#sambam-file"><span class="toc-section-number">21.5.5</span> SAM/BAM file</a></li>
<li><a href="#vcf-file"><span class="toc-section-number">21.5.6</span> VCF file</a></li>
<li><a href="#bed-format"><span class="toc-section-number">21.5.7</span> BED format</a></li>
<li><a href="#bedgraph-format"><span class="toc-section-number">21.5.8</span> bedGraph format</a></li>
</ul></li>
<li><a href="#usefull-links"><span class="toc-section-number">21.6</span> Usefull links:</a></li>
</ul></li>
<li><a href="#bs-seq"><span class="toc-section-number">22</span> BS-seq</a><ul>
<li><a href="#section-3"><span class="toc-section-number">22.1</span> </a></li>
<li><a href="#what-is-bs-seq"><span class="toc-section-number">22.2</span> What is BS-seq</a></li>
<li><a href="#section-4"><span class="toc-section-number">22.3</span> </a></li>
<li><a href="#how-to-analyze-bs-seq-data"><span class="toc-section-number">22.4</span> How to analyze BS-seq data</a><ul>
<li><a href="#section-5"><span class="toc-section-number">22.4.1</span> </a></li>
</ul></li>
</ul></li>
<li><a href="#start-a-project"><span class="toc-section-number">23</span> Start a project</a><ul>
<li><a href="#experimetal-design"><span class="toc-section-number">23.1</span> Experimetal design</a><ul>
<li><a href="#how-many-biological-replicates"><span class="toc-section-number">23.1.1</span> How many biological replicates</a></li>
<li><a href="#how-big"><span class="toc-section-number">23.1.2</span> How big</a></li>
<li><a href="#how-much-data-to-generate"><span class="toc-section-number">23.1.3</span> How much data to generate</a></li>
<li><a href="#section-6"><span class="toc-section-number">23.1.4</span> </a></li>
</ul></li>
</ul></li>
<li><a href="#capstone-project"><span class="toc-section-number">24</span> Capstone project:</a><ul>
<li><a href="#introduction"><span class="toc-section-number">24.1</span> Introduction</a></li>
<li><a href="#method"><span class="toc-section-number">24.2</span> Method</a></li>
<li><a href="#pipelines"><span class="toc-section-number">24.3</span> Pipelines</a></li>
<li><a href="#section-7"><span class="toc-section-number">24.4</span> </a></li>
</ul></li>
<li><a href="#practical-data-analysis-on-wgbs"><span class="toc-section-number">25</span> Practical data analysis on WGBS</a><ul>
<li><a href="#donwload-the-data"><span class="toc-section-number">25.1</span> Donwload the data</a><ul>
<li><a href="#reference"><span class="toc-section-number">25.1.1</span> Reference</a></li>
<li><a href="#now-lets-download-the-wgbs-data"><span class="toc-section-number">25.1.2</span> Now let’s download the WGBS data</a></li>
</ul></li>
<li><a href="#convert-sra-files-to-fastq-files"><span class="toc-section-number">25.2</span> Convert <code>sra</code> files to <code>fastq</code> files</a></li>
</ul></li>
<li><a href="#section-8"><span class="toc-section-number">26</span> </a><ul>
<li><a href="#including-plots"><span class="toc-section-number">26.1</span> Including Plots</a></li>
</ul></li>
<li><a href="#good-resouces-to-learn-bioinformatics"><span class="toc-section-number">27</span> Good resouces to learn Bioinformatics</a><ul>
<li><a href="#references"><span class="toc-section-number">27.1</span> References:</a></li>
</ul></li>
<li><a href="#references-1">References</a></li>
<li><a href="#crispr-cas9-screen-data-processing"><span class="toc-section-number">28</span> CRISPR-CAS9 screen data processing</a><ul>
<li><a href="#installation-of-softwares-using-conda"><span class="toc-section-number">28.1</span> Installation of softwares using <code>conda</code></a></li>
<li><a href="#an-exmaple-of-how-to-process"><span class="toc-section-number">28.2</span> An exmaple of how to process</a><ul>
<li><a href="#donwload-the-data-1"><span class="toc-section-number">28.2.1</span> Donwload the data</a></li>
<li><a href="#reference-1"><span class="toc-section-number">28.2.2</span> Reference</a></li>
</ul></li>
</ul></li>
<li><a href="#commands-for-linux-administration-advanced-topic"><span class="toc-section-number">29</span> Commands for Linux administration (Advanced topic)</a><ul>
<li><a href="#w"><span class="toc-section-number">29.1</span> <code>w</code></a></li>
<li><a href="#who"><span class="toc-section-number">29.2</span> <code>who</code></a></li>
<li><a href="#uptime"><span class="toc-section-number">29.3</span> <code>uptime</code></a></li>
<li><a href="#whoami"><span class="toc-section-number">29.4</span> <code>whoami</code></a></li>
<li><a href="#ifconfig"><span class="toc-section-number">29.5</span> <code>ifconfig</code></a></li>
<li><a href="#useradd-and-passwd"><span class="toc-section-number">29.6</span> <code>useradd</code> and <code>passwd</code></a></li>
</ul></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">All things considered for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<div id="preface" class="section level1 unnumbered">
<h1>Preface</h1>
<p><a href="" target="_blank"><img src="images/cover_fishingMan.jpg" style="display: block; margin: auto;" /></a></p>
<pre><code>## Bioconductor version 3.7 (BiocInstaller 1.30.0), ?biocLite for help</code></pre>
<pre><code>## A newer version of Bioconductor is available for this version of R,
##   ?BiocUpgrade for help</code></pre>
<pre><code>## BioC_mirror: https://bioconductor.org</code></pre>
<pre><code>## Using Bioconductor 3.7 (BiocInstaller 1.30.0), R 3.5.2 (2018-12-20).</code></pre>
<!--chapter:end:index.Rmd-->
<!--chapter:end:00_about_me_acknowledge.Rmd-->
</div>
<div id="part-linux" class="section level1 unnumbered">
<h1>(PART) Linux</h1>
</div>
<div id="why-linux" class="section level1">
<h1><span class="header-section-number">1</span> Why Linux?</h1>
<div id="what-is-linux" class="section level2">
<h2><span class="header-section-number">1.1</span> What is Linux</h2>
<p>Before the creation of Linux, Unix was developed by AT&amp;T Bell Labs in the 1960’s. It’s an operating system. Before the creation of Linux, and before the rise of Windows, the computing world was dominated by Unix (from web). After many years of evolution, Linux was created in early 1990’s.</p>
<p>In case you don’t know, Mac OS X is also a certified Unix operating system. So most of the Linux skills are applied in Mac OS X.</p>
<p>Linux is a clone of the operating system Unix, written from scratch by Linus Torvalds (Figure @ref(fig:LinuxTerminal)) with assistance from a loosely-knit team of hackers across the Net. It aims towards POSIX and Single UNIX Specification compliance (<span class="citation">Torvalds (<a href="#ref-Torvalds2015">2015</a>)</span>).</p>
<p>(ref:LinuxTerminal) An example of Linux terminal.</p>
<div class="figure" style="text-align: center">
<img src="figures/linux_terminal_example.png" alt="(ref:LinuxTerminal)" width="100%" />
<p class="caption">
(#fig:LinuxTerminal)(ref:LinuxTerminal)
</p>
</div>
<p>It has all the features you would expect in a modern fully-fledged Unix, including true multitasking, virtual memory, shared libraries, demand loading, shared copy-on-write executables, proper memory management, and multistack networking including IPv4 and IPv6. It is distributed under the GNU General Public (Torvalds, 2015).</p>
<p>Maybe it’s hard to understand what Linux or to remember the sentences mentioned above. Just know Linux is an operating system like Windows. This is enough for you to start out.</p>
<div class="figure" style="text-align: center">
<img src="figures/Linus_Torvalds_GitHub.png" alt="Linus Torvalds on GitHub" width="50%" />
<p class="caption">
(#fig:LinusTorvaldsGitHub)Linus Torvalds on GitHub
</p>
</div>
</div>
<div id="linux-for-bioinformatics" class="section level2">
<h2><span class="header-section-number">1.2</span> Linux for bioinformatics</h2>
<p>For analysis of NGS data, a large amount of software were developed for using under Linux environment. Among them, a large proportion can be only used under Linux environment.</p>
<ul>
<li>Easy to build simple pipelines
(awk, bash, piping, bash redirection, texttools)</li>
<li>Simple to install and use software development tools</li>
<li>Multiple versions of a program can be installed by the user himself and switched on/off with sourcing some scripts without being administrator</li>
<li>A lot of good scientific software is written in a non portable way for linux/unix (almost all short read aligners, samtools). This makes it necessary to use Unix for genomics.</li>
<li>Ability to perform analyses on computer clusters (important for big/long computational jobs)</li>
</ul>
<!--chapter:end:01_WhyLinux.Rmd-->
</div>
</div>
<div id="connecting-to-linux" class="section level1">
<h1><span class="header-section-number">2</span> Connecting to Linux</h1>
<div id="user-interfaces" class="section level2">
<h2><span class="header-section-number">2.1</span> User interfaces</h2>
<p>As an operating systemsm, Linux comes with two types of user interfaces: Graphical User Interface (GUI) and command line interface (shell).</p>
<p>GUI means there will be window, buttons, menus, etc. The most popular system with GUI is Windows system (Figure @ref(fig:windowsGUI)).</p>
<p>(ref:windowsGUI) Windows GUI.</p>
<div class="figure">
<img src="figures/windows_gui.png" alt="(ref:windowsGUI)"  />
<p class="caption">
(#fig:windowsGUI)(ref:windowsGUI)
</p>
</div>
<p>Command line interfaces means that you need to type the command line yourself. Usually the results will be displayed as text (Figure @ref(fig:linuxTerminalExam2)).</p>
<p>(ref:linuxTerminalExam2) An example of Linux terminal.</p>
<div class="figure">
<img src="figures/linux_terminal_exam2.png" alt="(ref:linuxTerminalExam2)"  />
<p class="caption">
(#fig:linuxTerminalExam2)(ref:linuxTerminalExam2)
</p>
</div>
<p>In Bioinformatics analysis, usually you won’t operate directly on the physical machine of the Linux server. Usually you need to connect to the Linux server via a tool, like Putty, Mobaxterm, etc.</p>
</div>
<div id="how-to-connect" class="section level2">
<h2><span class="header-section-number">2.2</span> How to connect</h2>
<p>If you want to connect to a Linux server, what you need to know first is:</p>
<ol style="list-style-type: decimal">
<li><p>IP address of your Linux server;</p></li>
<li><p>User name and password of your account;</p></li>
</ol>
<p>If you are a Mac OS X user, you can connect to a Linux server by using <code>Terminal</code>, a console program included with the operating system.</p>
<p>For Windows users, I would recommend MobaXterm for remote connection. MobatXterm is an excellent toolbox for remote connection from Windows system. It comes with an X11 server and provides many networking tools and tabbed SSH. It has all the essential UNIX commands in a single portable executable file.</p>
<p>Here I show one example of what you should do When you first open Mobaxterm. You need to follow the numbers in Figure @ref(fig:mobaxtermInit): click <code>Session</code>; then click <code>SSH</code>; type the IP or name address of the remote host, check <code>Specify username</code> if you need; click <code>OK</code>.</p>
<p>(ref:mobaxtermInit) First open of Mobaxterm.</p>
<div class="figure">
<img src="figures/mobaxterm_init.png" alt="(ref:mobaxtermInit)"  />
<p class="caption">
(#fig:mobaxtermInit)(ref:mobaxtermInit)
</p>
</div>
<p>Then you need to type the password. It’s OK that you don’t see anything when you’re typing (Figure @ref(fig:mobaxtermSavepswd)). Then click <code>Enter</code> on the keyboard.For the first time of log-in, you’ll be asked to whether to save the password or not. If you say click <code>Yes</code>, you won’t need to type the password again next time.</p>
<p>(ref:mobaxtermSavepswd) Type password and save it in Mobaxterm.</p>
<div class="figure">
<img src="figures/mobaxterm_savepassword.png" alt="(ref:mobaxtermSavepswd)"  />
<p class="caption">
(#fig:mobaxtermSavepswd)(ref:mobaxtermSavepswd)
</p>
</div>
<p>If you can find a Linux server, it’ll be very good for you to practise. If you are a student or a researcher in a university or an institute, usually you can get an account from your department.</p>
<p>If NOT, here I provided a guest account for you. Here are the user name and password:</p>
<pre><code>IP address: 198.211.107.37
User name: guest4bioinfo
Password: nobigfile</code></pre>
<p>As you can tell from the password, please do NOT upload BIG files (bigger than 2 MB).</p>
<div id="digitalocean-droplet" class="section level3">
<h3><span class="header-section-number">2.2.1</span> DigitalOcean droplet</h3>
<p>Another option is to sign up on DigitalOcean and create a droplet. DigitalOcean calls its cloud servers Droplets; each Droplet you create is a new server for your personal use. DigitalOcean has a tutorial of [<strong>How To Create Your First DigitalOcean Droplet</strong>] (<a href="https://www.digitalocean.com/community/tutorials/how-to-create-your-first-digitalocean-droplet" class="uri">https://www.digitalocean.com/community/tutorials/how-to-create-your-first-digitalocean-droplet</a>). You can get a your own Linux server with a $5 monthly payment.</p>
<p>The Linux server with IP address of <code>198.211.107.37</code> is a droplet on DigitalOcean. I pay $5 each month for this droplet.</p>
</div>
<div id="aws-free-tier-offering" class="section level3">
<h3><span class="header-section-number">2.2.2</span> AWS Free Tier offering</h3>
<p>One more way to get access to a Linux system is to take advantage of Red Hat Enterprise Linux delivered by Amazon EC2 (Elastic Compute Cloud). Red Hat and Amazon Web Services collaborate to provide official Red Hat Enterprise Linux licensed images through Amazon’s on-demand public cloud service at free or low cost.</p>
<p>The guided exercises and labs for this course were written assuming that you will set up an account with Amazon Web Services and use it to start a single, simple system running Red Hat Enterprise Linux 7. You will connect to that system securely over the internet and use it to practice commands.</p>
<p>At the time of writing, Amazon Web Services provides an AWS Free Tier offering, which gives new users free access to certain sizes of cloud instances and operating environments (including Red Hat Enterprise Linux 7) for up to 750 hours per month, for 12 months.</p>
<!--chapter:end:02_Connect2Linux.Rmd-->
</div>
</div>
</div>
<div id="file-system-and-commands-of-linux" class="section level1">
<h1><span class="header-section-number">3</span> File system and commands of Linux</h1>
<div id="path" class="section level2">
<h2><span class="header-section-number">3.1</span> Path</h2>
<p>To understand Linux file system, you can image it as a tree structure (Figure @ref(fig:linuxTreeStruc)).</p>
<p>(ref:linuxTreeStruc) Tree structure of Linux system.</p>
<div class="figure">
<img src="figures/LinuxPathTree.png" alt="(ref:linuxTreeStruc)"  />
<p class="caption">
(#fig:linuxTreeStruc)(ref:linuxTreeStruc)
</p>
</div>
<p>In Linux, a path is a unique location of a file or a directory in the file system.</p>
<p>For convenience, Linux file system is usually thought of in a tree structure. On a standard Linux system you will find the layout generally follows the scheme presented below.</p>
<p>The tree of the file system starts at the trunk or slash, indicated by a forward slash (/). This directory, containing all underlying directories and files, is also called the root directory or “the root” of the file system.</p>
<div id="relative-and-absolute-path" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Relative and absolute path</h3>
<ul>
<li><strong>Absolute path</strong></li>
</ul>
<p>An absolute path is defined as the location of a file or directory from the root directory(/). An absolute path starts from the <code>root</code> of the tree (<code>/</code>).</p>
<p>Here are some examples:</p>
<pre><code>/home/xie186
/home/xie186/perl5</code></pre>
<ul>
<li><strong>Relative path</strong></li>
</ul>
<p>Relative path is a path related to the present working directory.</p>
<pre><code>data/sample1/
../doc/</code></pre>
<p>If you want to get the <strong>absolute path</strong> based on <strong>relative path</strong>, you can use <code>readlink</code> with parameter <code>-f</code>:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="bu">pwd</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="fu">readlink</span> -f ../</a></code></pre></div>
<pre><code>## /cloud/project
## /cloud</code></pre>
<p>##Surfing in Linux file system</p>
<p>Once we enter into a Linux file system, we need to 1) know where we are; 2) how to get where we want; 3) how to know what files or directories we have in a particular path.</p>
<p>###Command <code>pwd</code></p>
<p>In order to know where we are, we need to use <code>pwd</code> command. The command <code>pwd</code> is short for “print name of current/working directory”. It will return the full path of current directory.</p>
<p>Command pwd is almost always used by itself. This means you only need to type pwd and press ENTER (Figure @ref(fig:linuxCMDpwd)).</p>
<p>(ref:linuxCMDpwd) Linux command <code>pwd</code></p>
<div class="figure" style="text-align: center">
<img src="figures/linuxCMDpwd.png" alt="ref:linuxCMDpwd" width="80%" />
<p class="caption">
(#fig:linuxCMDpwd)ref:linuxCMDpwd
</p>
</div>
</div>
<div id="comand-ls" class="section level3">
<h3><span class="header-section-number">3.1.2</span> Comand <code>ls</code></h3>
<p>After you know where you are, then you want to know what you have in that
directory, we can use command <code>ls</code> to list directory contents (Figure @ref(fig:linuxCMDls)). Its syntax is:</p>
<pre><code>ls [option]... [file]...</code></pre>
<p>(ref:linuxCMDls) Linux command <code>ls</code></p>
<div class="figure" style="text-align: center">
<img src="figures/linuxCMDls.png" alt="ref:linuxCMDls" width="80%" />
<p class="caption">
(#fig:linuxCMDls)ref:linuxCMDls
</p>
</div>
<p><code>ls</code> with no option will list files and directories in bare format. Bare format means the detailed information (type, size, modified date and time, permissions and links etc) won’t be viewed. When you use <code>ls</code> by itself (Figure @ref(fig:linuxCMDls)), it will list files and directories in the current directory.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="bu">cd</span> tables</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="fu">ls</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="bu">echo</span> <span class="st">&quot;ls -a&quot;</span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="fu">ls</span> -a </a>
<a class="sourceLine" id="cb11-6" data-line-number="6"></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="bu">echo</span> <span class="st">&quot;ls -t&quot;</span></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="fu">ls</span> -t</a></code></pre></div>
<pre><code>## 10_PerlInputOutput_bak.Rmd
## linuxPathShortcuts.tsv
## textEditorLinuxVi3modes.csv
## ls -a
## .
## ..
## 10_PerlInputOutput_bak.Rmd
## linuxPathShortcuts.tsv
## textEditorLinuxVi3modes.csv
## ls -t
## 10_PerlInputOutput_bak.Rmd
## linuxPathShortcuts.tsv
## textEditorLinuxVi3modes.csv</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="fu">ls</span> -l -a tables/</a></code></pre></div>
<pre><code>## total 24
## drwxrwxr-x  2 rstudio-user rstudio-user 4096 Apr 15 15:48 .
## drwxr-xr-x 16 rstudio-user rstudio-user 4096 Apr 15 16:33 ..
## -rw-rw-r--  1 rstudio-user rstudio-user 4139 Apr 15 15:48 10_PerlInputOutput_bak.Rmd
## -rw-rw-r--  1 rstudio-user rstudio-user  223 Apr 15 15:48 linuxPathShortcuts.tsv
## -rw-rw-r--  1 rstudio-user rstudio-user  766 Apr 15 15:48 textEditorLinuxVi3modes.csv</code></pre>
<p>Linux command options can be combined without a space between them and with a single - (dash).</p>
<p>The following command is a faster way to use the l and a options and gives the same output as the Linux command shown above.</p>
<pre><code>ls -la </code></pre>
</div>
<div id="command-cd" class="section level3">
<h3><span class="header-section-number">3.1.3</span> Command <code>cd</code></h3>
<p>Command cd is used to change the current directory. It’s syntax is:</p>
<pre><code>cd [option] [directory]</code></pre>
<p>Unlike <code>pwd</code>, when you use <code>cd</code> you usually need to provide the path (either absolute or relative path) which we want to enter.</p>
<p>If we didn’t provide any path information, we will change to home directory by default.</p>
<p>##Path shortcuts</p>
<p>In Linux, there are three commonly used path shortpaths (Table @ref(tab:linuxPathShortcuts)).</p>
<p>(ref:linuxPathShortcuts) Shortcuts of path.</p>
<table>
<caption>(#tab:linuxPathShortcuts)(ref:linuxPathShortcuts)</caption>
<thead>
<tr class="header">
<th align="center">Path</th>
<th align="center">Shortcuts</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Single dot</td>
<td align="center">.</td>
<td align="center">The current folder</td>
</tr>
<tr class="even">
<td align="center">Double dots</td>
<td align="center">..</td>
<td align="center">The folder above the current folder</td>
</tr>
<tr class="odd">
<td align="center">Tilde character</td>
<td align="center">~</td>
<td align="center">Home directory (normally the directory:/home/my_login_name)</td>
</tr>
<tr class="even">
<td align="center">Dash</td>
<td align="center">-</td>
<td align="center">Your last working directory</td>
</tr>
</tbody>
</table>
<p>Here are some examples:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="bu">pwd</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="fu">ls</span></a></code></pre></div>
<pre><code>## /home/rstudio-user
## R</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb19-1" data-line-number="1"></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="fu">ls</span> ./</a></code></pre></div>
<pre><code>## 00_about_me_acknowledge.Rmd
## 01_WhyLinux.Rmd
## 02_Connect2Linux.Rmd
## 03_FileSystemLinux.Rmd
## 04_Linux_FilteringOutputandFindingThings.Rmd
## 05_AchivingAndCompressingFiles.Rmd
## 06_procManageLinux.Rmd
## 07_fileTransfer.Rmd
## 08_InstallationOfSoftwareInLinux.Rmd
## 09_TextEditorInLinux.Rmd
## 10_FirstPerlProgram.Rmd
## 11_PerlVariableOperator.Rmd
## 13_PerlControlStructure.Rmd
## 14_StringManipulationRegExp.Rmd
## 15_PerlInputOutput.Rmd
## 16_practicalPerlProgram.Rmd
## 17_PerlModules.Rmd
## 18_R_intro.Rmd
## 19_SimpleGraphR.Rmd
## 20_GraphGgplot2.Rmd
## 21_GenerateHeatmap.Rmd
## 22_FigureManuscript.Rmd
## 23_IntroNGS.Rmd
## 24_introRNA-seq.Rmd
## 25_introChIP-seq.Rmd
## 26_introBSseq.Rmd
## 27_ExpDesign.Rmd
## 28_CapstoneProjectRNA-seq.Rmd
## 30_CapstoneProjectBS-seq.Rmd
## 30_py_why_program.Rmd
## 31_Good_resource.Rmd
## 32_Reference.Rmd
## DESCRIPTION
## LICENSE
## README.md
## Rmd_list.txt
## TODO
## _bookdown.yml
## _bookdown_files
## _build.sh
## _deploy.sh
## _output.yml
## bak
## bioinfBookXIE186.Rmd
## bioinfBookXIE186.aux
## bioinfBookXIE186.log
## bioinfBookXIE186_files
## book.bib
## bookdown-demo.Rproj
## bookdown-demo.log
## capstone_CRISPR_screen.Rmd
## cmd_linux_admin.Rmd
## code_R
## code_perl
## data
## docs
## download
## figures
## images
## index.Rmd
## index.Rmd_bak
## index.log
## lib
## order_Rmd.sh
## order_Rmd.sh.README
## packages.bib
## preamble.tex
## preamble_bak.tex
## rpres.css
## slides
## style.css
## tables
## test.pl
## test.py
## toc.css</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb21-1" data-line-number="1"></a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="co">## </span></a>
<a class="sourceLine" id="cb21-3" data-line-number="3"><span class="bu">pwd</span></a>
<a class="sourceLine" id="cb21-4" data-line-number="4"><span class="bu">cd</span> ../</a>
<a class="sourceLine" id="cb21-5" data-line-number="5"><span class="bu">pwd</span></a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="bu">cd</span> ./</a>
<a class="sourceLine" id="cb21-7" data-line-number="7"><span class="bu">pwd</span></a></code></pre></div>
<pre><code>/cloud/project
/cloud
/cloud</code></pre>
<p>Each directory has two entries in it at the start, with names . (a link to itself) and .. (a link to its parent directory). The exception, of course, is the root directory, where the .. directory also refers to the root directory.</p>
<p>Sometimes you go to a new directory and do something, then you remember that you need to go to the previous working direcotry. To get back instantly, use a dash.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="bu">echo</span> <span class="st">&quot;This is our current directory: &quot;</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="bu">pwd</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3"></a>
<a class="sourceLine" id="cb23-4" data-line-number="4"><span class="bu">echo</span> <span class="st">&quot;Let&#39;s go our home diretory: &quot;</span></a>
<a class="sourceLine" id="cb23-5" data-line-number="5"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb23-6" data-line-number="6"></a>
<a class="sourceLine" id="cb23-7" data-line-number="7"><span class="bu">echo</span> <span class="st">&quot;Check where we are: &quot;</span></a>
<a class="sourceLine" id="cb23-8" data-line-number="8"><span class="bu">pwd</span></a>
<a class="sourceLine" id="cb23-9" data-line-number="9"></a>
<a class="sourceLine" id="cb23-10" data-line-number="10"><span class="bu">echo</span> <span class="st">&quot;Let&#39;s go to your previous working direcotry: &quot;</span></a>
<a class="sourceLine" id="cb23-11" data-line-number="11"><span class="bu">cd</span> -</a>
<a class="sourceLine" id="cb23-12" data-line-number="12"><span class="bu">echo</span> <span class="st">&quot;Check where we&#39;re now: &quot;</span></a>
<a class="sourceLine" id="cb23-13" data-line-number="13"><span class="bu">pwd</span></a></code></pre></div>
<pre><code>## This is our current directory: 
## /cloud/project
## Let&#39;s go our home diretory: 
## Check where we are: 
## /home/rstudio-user
## Let&#39;s go to your previous working direcotry: 
## /cloud/project
## Check where we&#39;re now: 
## /cloud/project</code></pre>
</div>
</div>
<div id="manipulations-of-files-and-directories" class="section level2">
<h2><span class="header-section-number">3.2</span> Manipulations of files and directories</h2>
<p>In Linux, manipulations of files and directories are the most frequent work. In this section, you will learn how to copy, rename, remove, and create files and directories.</p>
<div id="command-cp" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Command <code>cp</code></h3>
<p>In Linux, command <code>cp</code> can help you copy files and directories into a target directory.</p>
</div>
<div id="command-mv" class="section level3">
<h3><span class="header-section-number">3.2.2</span> Command <code>mv</code></h3>
<p>The command <code>mv</code> is short for move (or rename) files.</p>
<div id="move-one-file" class="section level4">
<h4><span class="header-section-number">3.2.2.1</span> Move one file</h4>
<p>Here is one common example of <code>mv</code>.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb25-1" data-line-number="1"></a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="fu">mv</span> file1 directory1/</a></code></pre></div>
</div>
<div id="move-multiple-files-into-a-directory" class="section level4">
<h4><span class="header-section-number">3.2.2.2</span> Move multiple files into a directory</h4>
<div class="sourceCode" id="cb26"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="fu">mv</span> file1 file2 file3 target_direcotry/</a></code></pre></div>
</div>
<div id="move-a-directory" class="section level4">
<h4><span class="header-section-number">3.2.2.3</span> Move a directory</h4>
<div class="sourceCode" id="cb27"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="fu">mv</span> dir1</a></code></pre></div>
</div>
<div id="rename-a-file-or-a-directory" class="section level4">
<h4><span class="header-section-number">3.2.2.4</span> Rename a file or a directory</h4>
</div>
</div>
<div id="command-mkdir" class="section level3">
<h3><span class="header-section-number">3.2.3</span> Command <code>mkdir</code></h3>
<p>Command <code>mkdir</code> is short for make directory.</p>
<p>The syntax is shown as below:</p>
<pre><code>mkdir [OPTION ...] DIRECTORY ...</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="fu">mkdir</span> directory</a></code></pre></div>
<p>Multiple directories can be specified when calling <code>mkdir</code>.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="fu">mkdir</span> directory1 directory2</a></code></pre></div>
<div id="how-to-create-a-directory" class="section level4">
<h4><span class="header-section-number">3.2.3.1</span> How to create a directory</h4>
<pre><code>mkdir -p foo/bar/baz</code></pre>
<blockquote>
<p>How to defining complex directory trees with one command</p>
</blockquote>
<pre><code>mkdir -p project/{software,results,doc/{html,info,pdf},scripts}</code></pre>
<p>This will create a direcotry trees as shown below:</p>
<pre><code>$ tree project/
project/
├── doc
│   ├── html
│   ├── info
│   └── pdf
├── results
├── scripts
└── software

7 directories, 0 files</code></pre>
<p>The command line above will directories <code>foo</code>, <code>foo/bar</code>, and <code>foo/bar/baz</code> if they don’t exist.</p>
</div>
</div>
<div id="command-rm" class="section level3">
<h3><span class="header-section-number">3.2.4</span> Command ‘rm’</h3>
<p>You can use <code>rm</code> to remove both files and directories.</p>
<div id="how-to-remove-a-file-or-multiple-files" class="section level4">
<h4><span class="header-section-number">3.2.4.1</span> How to remove a file or multiple files</h4>
<pre><code>## You can remove one file. 
rm file1 
## `rm` can remove multiple files simutaneously
rm file2 file3 </code></pre>
</div>
<div id="how-to-remove-a-folder" class="section level4">
<h4><span class="header-section-number">3.2.4.2</span> How to remove a folder</h4>
<p>If a folder is empty, you can remove it using <code>rm</code> with <code>-r</code>.</p>
<pre><code>rm -r FOLDER</code></pre>
<p>If a folder is not empty, you can remove it using <code>rm</code> with <code>-r</code> and <code>-f</code>.</p>
<pre><code>mkdir test_folder
rm -r test_folder</code></pre>
</div>
</div>
</div>
<div id="viewing-text-files-in-linux" class="section level2">
<h2><span class="header-section-number">3.3</span> Viewing text files in Linux</h2>
<div id="command-cat" class="section level3">
<h3><span class="header-section-number">3.3.1</span> Command <code>cat</code></h3>
<p>The command <code>cat</code> is short for concatenate files and print on the standard output.</p>
<p>The syntax is shown as below:</p>
<pre><code>cat [OPTION]... [FILE]...</code></pre>
<p>For small text file, <code>cat</code> can be used to view the files on the standard output.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="fu">cat</span> data/testdata4linux_cmd.txt</a></code></pre></div>
<pre><code>## gene1
## gene2
## gene3
## gene4
## gene5
## gene6
## gene7
## gene8
## gene9
## gene10
## gene11
## gene12
## gene13
## gene14
## gene15
## gene16</code></pre>
<p>You can also use <code>cat</code> to merge two text files.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb40-1" data-line-number="1"><span class="fu">cat</span> file1 file2 <span class="op">&gt;</span> merged_file</a></code></pre></div>
</div>
<div id="command-more-and-less" class="section level3">
<h3><span class="header-section-number">3.3.2</span> Command <code>more</code> and <code>less</code></h3>
<p>The command <code>more</code> is old utility. When the text passed to it is too large to fit on one screen, it pages it. You can scroll down but not up.</p>
<p>The syntaxt of <code>more</code> is shown below:</p>
<pre><code>more [options] file [...]</code></pre>
<p>The command <code>less</code> was written by a man who was fed up with more’s inability to scroll backwards through a file. He turned less into an open source project and over time, various individuals added new features to it. less is massive now. That’s why some small embedded systems have more but not less. For comparison, less’s source is over 27000 lines long. more implementations are generally only a little over 2000 lines long.</p>
<p>The syntaxt of <code>less</code> is shown below:</p>
<pre><code>less [options] file [...]</code></pre>
</div>
<div id="command-head-and-tail" class="section level3">
<h3><span class="header-section-number">3.3.3</span> Command <code>head</code> and <code>tail</code></h3>
<p>The command <code>head</code> is used to output the first part of files. By default, it outputs the first 10 lines of the file.</p>
<pre><code>head [OPTION]... [FILE]...</code></pre>
<p>Here is an exmaple of printing the first 5 files of the file:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="fu">head</span> -n 5 code_perl/variable_assign.pl</a></code></pre></div>
<pre><code>## #!/usr/bin/perl
## use warnings;
## use strict;
## 
## #assign two strings to two variables</code></pre>
<p>In fact, the letter n does not even need to be used at all. Just the hyphen and the integer (with no intervening space) are sufficient to tell head how many lines to return. Thus, the following would produce the same result as the above commands:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb46-1" data-line-number="1"><span class="fu">head</span> -5 data/testdata4linux_cmd.txt</a></code></pre></div>
<pre><code>## gene1
## gene2
## gene3
## gene4
## gene5</code></pre>
<p>The command <code>tail</code> is used to output the last part of files. By default, it prints the last 10 lines of the file to standard output.</p>
<p>The syntax is shown below:</p>
<pre><code>tail [OPTION]... [FILE]...</code></pre>
<p>Here is an exmaple of printing the last 5 files of the file:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb49-1" data-line-number="1"><span class="fu">tail</span> -5 data/testdata4linux_cmd.txt</a></code></pre></div>
<pre><code>## gene12
## gene13
## gene14
## gene15
## gene16</code></pre>
<p>To view lines from a specific point in a file, you can use <code>-n +NUMBER</code> with the <code>tail</code> command. For example, here is an example of viewing the file from the 2nd line of the line.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb51-1" data-line-number="1"><span class="fu">tail</span> -n +2 data/testdata4linux_cmd.txt</a></code></pre></div>
<pre><code>## gene2
## gene3
## gene4
## gene5
## gene6
## gene7
## gene8
## gene9
## gene10
## gene11
## gene12
## gene13
## gene14
## gene15
## gene16</code></pre>
</div>
<div id="auto-completion" class="section level3">
<h3><span class="header-section-number">3.3.4</span> Auto-completion</h3>
<p>In most Shell environment, programmable completion feature will also improve your speed of typing. It permits typing a partial name of command or a partial file (or directory), then pressing <code>TAB</code> key to auto-complete the command (Figure @ref(fig:linuxAutoCompletion)). If there are more than one possible completions, then TAB will list all of them (Figure @ref(fig:linuxAutoCompletion)).</p>
<p>(ref:linuxAutoCompletion) Demonstration of programmable completion feature.</p>
<div class="figure">
<img src="figures/linuxAutoCompletion.png" alt="(ref:linuxAutoCompletion)"  />
<p class="caption">
(#fig:linuxAutoCompletion)(ref:linuxAutoCompletion)
</p>
</div>
</div>
</div>
<div id="understand-standard-input-and-stardard-output" class="section level2">
<h2><span class="header-section-number">3.4</span> Understand standard input and stardard output</h2>
<p>In the Linux environment, input and output is distributed across three streams: standard input (STDIN), standard output (STDOUT), standard error (STDERR). These three streams are also numbered: STDIN (0), STDOUT (1), STDERR (2).</p>
<div id="stdin" class="section level3">
<h3><span class="header-section-number">3.4.1</span> STDIN</h3>
<p>…
The standard input stream typically carries data from a user to a program. Programs that expect standard input usually receive input from a device, such as a keyboard. Standard input is terminated by reaching EOF (end-of-file). As described by its name, EOF indicates that there is no more data to be read.</p>
<p>To see standard input in action, run the cat program. Cat stands for concatenate, which means to link or combine something. It is commonly used to combine the contents of two files. When run on its own, cat opens a looping prompt.
…</p>
<pre><code>tail
1
2
3
`CTRL+D`
1
2
3</code></pre>
</div>
<div id="stdout" class="section level3">
<h3><span class="header-section-number">3.4.2</span> STDOUT</h3>
<p>Data that is generated by a program will be written by STDOUT. If the STDOUT is not redirected, it will output the data on to the terminal.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb54-1" data-line-number="1"><span class="va">stdout=</span><span class="st">&quot;Hello world&quot;</span></a>
<a class="sourceLine" id="cb54-2" data-line-number="2"><span class="bu">echo</span> <span class="va">$stdout</span></a></code></pre></div>
<pre><code>## Hello world</code></pre>
<p>The STDOUT can be redirected to a file. See the example below:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb56-1" data-line-number="1"><span class="va">stdout=</span><span class="st">&quot;Hello world&quot;</span></a>
<a class="sourceLine" id="cb56-2" data-line-number="2"><span class="bu">echo</span> <span class="va">$stdout</span> <span class="op">&gt;</span> data/test_output.txt</a>
<a class="sourceLine" id="cb56-3" data-line-number="3"><span class="co"># cat the data</span></a>
<a class="sourceLine" id="cb56-4" data-line-number="4"><span class="fu">cat</span> data/test_output.txt</a></code></pre></div>
<pre><code>## Hello world</code></pre>
</div>
<div id="stderr" class="section level3">
<h3><span class="header-section-number">3.4.3</span> STDERR</h3>
<p>During a program’s execution, some errors may be generated when the program fails at some parts. STDERR will help you write the errors. By default, the STDERR will be outputed onto the terminal.</p>
<p>Here is an example of STDERR</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb58-1" data-line-number="1"><span class="fu">ls</span> NOTAFILE</a></code></pre></div>
<pre><code>## ls: cannot access &#39;NOTAFILE&#39;: No such file or directory</code></pre>
</div>
</div>
<div id="find-disk-usage-of-files-and-directories" class="section level2">
<h2><span class="header-section-number">3.5</span> Find Disk Usage of Files and Directories</h2>
<p>The Linux <code>du</code> (short for Disk Usage) is a standard Unix/Linux command, used to check the information of disk usage of files and directories on a machine. The <code>du</code> command has many parameter options that can be used to get the results in many formats. The du command also displays the files and directory sizes in a recursively manner.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb60-1" data-line-number="1"><span class="fu">du</span> data/ESP6500-African_American.vcf.gz</a>
<a class="sourceLine" id="cb60-2" data-line-number="2"><span class="fu">du</span> -h data/ESP6500-African_American.vcf.gz</a></code></pre></div>
<pre><code>## 27388    data/ESP6500-African_American.vcf.gz
## 27M  data/ESP6500-African_American.vcf.gz</code></pre>
<p>To get the summary of a grand total disk usage size of an directory use the option “-s” as follows.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb62-1" data-line-number="1"><span class="fu">du</span> -sh data/</a></code></pre></div>
<pre><code>## 37M  data/</code></pre>
<p>Using “-a” flag with “du” command displays the disk usage of all the files and directories.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb64-1" data-line-number="1"><span class="fu">du</span> -ah data/</a></code></pre></div>
<pre><code>## 4.0K data/PYL10_ARATH.fasta
## 4.0K data/test_ref2.fa
## 4.0K data/test_ref.fa
## 4.0K data/test_ref_len.txt
## 4.0K data/gene_annotation.txt
## 4.0K data/WGBS_sample_information.txt
## 4.0K data/test_ref2_30.fa
## 4.0K data/DMR_region_merged.txt
## 27M  data/ESP6500-African_American.vcf.gz
## 4.0K data/DMR_region.txt
## 88K  data/maize_embryo_specific_gene_Sheet1.tsv
## 9.3M data/Arabidopsis_thaliana.TAIR10.37.gff3.gz
## 532K data/ESP6500-African_American.vcf.gz.tbi
## 4.0K data/test_output.txt
## 4.0K data/testdata4linux_cmd.txt
## 4.0K data/DEG_list.txt
## 4.0K data/README
## 37M  data/</code></pre>
</div>
<div id="advanced-topic" class="section level2">
<h2><span class="header-section-number">3.6</span> Advanced topic</h2>
<div id="linux-md5sum-command" class="section level3">
<h3><span class="header-section-number">3.6.1</span> Linux <code>md5sum</code> Command</h3>
<p><code>md5sum</code> is used to verify the integrity of files, as virtually any change to a file will cause its MD5 hash to change. Most commonly, md5sum is used to verify that a file has not changed as a result of a faulty file transfer, a disk error or non-malicious meddling. The md5sum program is included in most Unix-like operating systems.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb66-1" data-line-number="1"><span class="bu">echo</span> <span class="st">&quot;The MD5 value of index.Rmd is: &quot;</span></a>
<a class="sourceLine" id="cb66-2" data-line-number="2"><span class="ex">md5sum</span> index.Rmd</a>
<a class="sourceLine" id="cb66-3" data-line-number="3"><span class="fu">cp</span> index.Rmd index.Rmd_bak</a>
<a class="sourceLine" id="cb66-4" data-line-number="4"><span class="bu">echo</span> <span class="st">&quot;The MD5 value of index.Rmd_bak is: &quot;</span></a>
<a class="sourceLine" id="cb66-5" data-line-number="5"><span class="ex">md5sum</span> index.Rmd_bak </a>
<a class="sourceLine" id="cb66-6" data-line-number="6"><span class="bu">echo</span> <span class="st">&quot;The MD5 value of new index.Rmd_bak is: &quot;</span></a>
<a class="sourceLine" id="cb66-7" data-line-number="7"><span class="fu">head</span> index.Rmd <span class="op">&gt;</span> index.Rmd_bak</a>
<a class="sourceLine" id="cb66-8" data-line-number="8"><span class="ex">md5sum</span> index.Rmd_bak </a></code></pre></div>
<pre><code>## The MD5 value of index.Rmd is: 
## 687196e6e5bfac894b4047e4b54404e7  index.Rmd
## The MD5 value of index.Rmd_bak is: 
## 687196e6e5bfac894b4047e4b54404e7  index.Rmd_bak
## The MD5 value of new index.Rmd_bak is: 
## 6c6d75e8839891bf7ba1ab152c8f267c  index.Rmd_bak</code></pre>
<!--chapter:end:03_FileSystemLinux.Rmd-->
</div>
</div>
<div id="filtering-output" class="section level2">
<h2><span class="header-section-number">3.7</span> Filtering Output</h2>
<div id="cut" class="section level3">
<h3><span class="header-section-number">3.7.1</span> <code>cut</code></h3>
<p><code>cut</code> can be used to print selected parts of lines from each FILE to standard output.</p>
<p>cut sort uniq wc grep</p>
<p><a href="https://www.youtube.com/playlist?list=PLtK75qxsQaMLZSo7KL-PmiRarU7hrpnwK" class="uri">https://www.youtube.com/playlist?list=PLtK75qxsQaMLZSo7KL-PmiRarU7hrpnwK</a></p>
</div>
</div>
<div id="check-you-job-status" class="section level2">
<h2><span class="header-section-number">3.8</span> Check you job status</h2>
<pre><code>$ ps aux  
USER       PID  %CPU %MEM  VSZ RSS     TTY   STAT START   TIME COMMAND
timothy  29217  0.0  0.0 11916 4560 pts/21   S+   08:15   0:00 pine  
root     29505  0.0  0.0 38196 2728 ?        Ss   Mar07   0:00 sshd: can [priv]   
can      29529  0.0  0.0 38332 1904 ?        S    Mar07   0:00 sshd: can@notty  </code></pre>
<p>USER = user owning the process
PID = process ID of the process
%CPU = It is the CPU time used divided by the time the process has been running.
%MEM = ratio of the process’s resident set size to the physical memory on the machine
VSZ = virtual memory usage of entire process (in KiB)
RSS = resident set size, the non-swapped physical memory that a task has used (in KiB)
TTY = controlling tty (terminal)
STAT = multi-character process state
START = starting time or date of the process
TIME = cumulative CPU time
COMMAND = command with all its arguments</p>
</div>
<div id="finding-things" class="section level2">
<h2><span class="header-section-number">3.9</span> Finding Things</h2>
<div id="list-files-bigger-than-filesize-specified" class="section level3">
<h3><span class="header-section-number">3.9.1</span> List files bigger than filesize specified</h3>
<pre><code># To find files larger than 100MB:
find . -type f -size +100M
# If you want the current dir only:
find . -maxdepth 1 -type f -size +100M
</code></pre>
<p>References:</p>
<p><a href="https://superuser.com/questions/117913/ps-aux-output-meaning" class="uri">https://superuser.com/questions/117913/ps-aux-output-meaning</a></p>
<!--chapter:end:04_Linux_FilteringOutputandFindingThings.Rmd-->
</div>
</div>
</div>
<div id="achiving-and-compressing-files" class="section level1">
<h1><span class="header-section-number">4</span> Achiving and compressing files</h1>
<div id="common-compressed-file-format" class="section level2">
<h2><span class="header-section-number">4.1</span> Common compressed file format</h2>
<p>If you download open source software, like <code>bwa</code> and <code>ViewBS</code>, you will encouther archived files very often.</p>
<p>The common compressed file usually have the the suffix of <code>tar.gz</code> (equaivalent of <code>tgz</code>), <code>gz</code>, <code>zip</code> and <code>tar.bz2</code>.</p>
<p>For example, if we want to use <code>bowtie2</code> in Linux, we need to download the <code>bowtie2</code> software. Bowtie2 file for Linux can be downloaded form the link below:</p>
<pre><code>https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.3.3.1/bowtie2-2.3.3.1-linux-x86_64.zip</code></pre>
<p>In side the <code>zip</code> file, there are files and sub folders.</p>
<p>The files for <code>samtools v1.6</code> are archived in to a file named samtools-1.6.tar.bz2`. From the link below you can download the link:</p>
<pre><code>https://gigenet.dl.sourceforge.net/project/samtools/samtools/1.6/samtools-1.6.tar.bz2</code></pre>
</div>
<div id="how-to-work-with-different-format" class="section level2">
<h2><span class="header-section-number">4.2</span> How to work with different format</h2>
<div id="gz" class="section level3">
<h3><span class="header-section-number">4.2.1</span> <code>*.gz</code></h3>
<div id="how-to-check-the-file" class="section level4">
<h4><span class="header-section-number">4.2.1.1</span> How to check the file</h4>
<pre><code>zcat file.gz |less </code></pre>
</div>
<div id="how-to-test-if-a-gzip-file-is-valid" class="section level4">
<h4><span class="header-section-number">4.2.1.2</span> How to test if a <code>gzip</code> file is valid</h4>
<pre><code>gzip -t file.gz</code></pre>
</div>
<div id="how-decompress-a-.gz-file" class="section level4">
<h4><span class="header-section-number">4.2.1.3</span> How decompress a <code>*.gz</code> file</h4>
<ul>
<li>Decompress a <code>*.gz</code> file</li>
</ul>
<pre><code>mkdir tmp
cp tmp
cp ../data/Arabidopsis_thaliana.TAIR10.37.gff3.gz ./
gunzip Arabidopsis_thaliana.TAIR10.37.gff3.gz</code></pre>
<ul>
<li>Decompress a file and keep the original copy</li>
</ul>
<pre><code>gunzip -c file.gz &gt; file</code></pre>
</div>
</div>
<div id="tar.gz" class="section level3">
<h3><span class="header-section-number">4.2.2</span> <code>*.tar.gz</code></h3>
<div id="how-to-decompress-the-.tar.gz-file" class="section level4">
<h4><span class="header-section-number">4.2.2.1</span> How to decompress the <code>*.tar.gz</code> file</h4>
<pre><code>tar zxvf file.tar.gz</code></pre>
<ul>
<li><code>z</code> means (un)z̲ip.</li>
<li><code>x</code> means ex̲tract files from the archive.</li>
<li><code>v</code> means print the filenames v̲erbosely.</li>
<li><code>f</code> means the following argument is a f̱ilename.</li>
</ul>
</div>
<div id="how-to-view-the-content-without-extract-the-files" class="section level4">
<h4><span class="header-section-number">4.2.2.2</span> How to view the content without extract the files</h4>
<pre><code>tar -tf file.tar.gz</code></pre>
</div>
<div id="how-to-create-a-.tar.gz-file" class="section level4">
<h4><span class="header-section-number">4.2.2.3</span> How to create a <code>*.tar.gz</code> file</h4>
<pre><code>tar zcvf file_new.tar.gz file1 file2 folder1 folder2</code></pre>
</div>
<div id="how-to-create-.tar.gz-file" class="section level4">
<h4><span class="header-section-number">4.2.2.4</span> How to create <code>*.tar.gz</code> file</h4>
</div>
</div>
<div id="zip" class="section level3">
<h3><span class="header-section-number">4.2.3</span> <code>*.zip</code></h3>
<div id="how-to-unzip-a-.zip-file" class="section level4">
<h4><span class="header-section-number">4.2.3.1</span> How to unzip a <code>*.zip</code> file</h4>
<pre><code>unzip file.zip</code></pre>
</div>
<div id="how-to-create-a-.zip-file" class="section level4">
<h4><span class="header-section-number">4.2.3.2</span> How to create a <code>*.zip</code> file</h4>
<pre><code>zip file_new.zip file1 file2 folder1 folder2</code></pre>
</div>
</div>
<div id="tar.bz2" class="section level3">
<h3><span class="header-section-number">4.2.4</span> <code>*.tar.bz2</code></h3>
<div id="how-to-decompress-a-.tar.bz2-file" class="section level4">
<h4><span class="header-section-number">4.2.4.1</span> How to decompress a <code>*.tar.bz2</code> file</h4>
<div class="sourceCode" id="cb81"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb81-1" data-line-number="1"><span class="fu">mkdir</span> tmp</a>
<a class="sourceLine" id="cb81-2" data-line-number="2"><span class="bu">cd</span> tmp</a>
<a class="sourceLine" id="cb81-3" data-line-number="3"><span class="bu">pwd</span></a>
<a class="sourceLine" id="cb81-4" data-line-number="4"><span class="fu">ls</span> </a>
<a class="sourceLine" id="cb81-5" data-line-number="5"><span class="fu">wget</span> https://downloads.sourceforge.net/project/bio-bwa/bwa-0.7.17.tar.bz2</a>
<a class="sourceLine" id="cb81-6" data-line-number="6"><span class="fu">ls</span></a>
<a class="sourceLine" id="cb81-7" data-line-number="7"><span class="fu">tar</span> -vxjf bwa-0.7.17.tar.bz2</a></code></pre></div>
</div>
<div id="how-to-create-a-.tar.bz2-file" class="section level4">
<h4><span class="header-section-number">4.2.4.2</span> How to create a <code>*.tar.bz2</code> file</h4>
<pre><code>tar -cvjSf folder.tar.bz2 file1 file2 folder1 folder2</code></pre>
<!--chapter:end:05_AchivingAndCompressingFiles.Rmd-->
</div>
</div>
</div>
</div>
<div id="process-management-in-linux" class="section level1">
<h1><span class="header-section-number">5</span> Process management in Linux</h1>
<div id="top" class="section level2">
<h2><span class="header-section-number">5.1</span> <code>top</code></h2>
<p>The <code>top</code> program provides a dynamic real-time view of a running system.</p>
<p>Usually <code>top</code> is used with the option <code>-c</code>.</p>
<pre><code>top -c</code></pre>
<p>The option <code>-c</code> will let <code>top</code> to displat the full command path along with the command arguments in the <code>COMMAND</code> collumn.</p>
<p>You can also run <code>top</code> interactively. You can run <code>top</code> first and then press <code>c</code>. If you want to kill a process with <code>PID</code> of <code>186</code>, you can press <code>k</code> and then type <code>186</code> to kill the process with <code>PID</code> of 186.</p>
<p><code>man top</code> can help you get the manual of command <code>top</code>.</p>
<p>The following table explains what each column mean.</p>
<table>
<colgroup>
<col width="5%" />
<col width="94%" />
</colgroup>
<thead>
<tr class="header">
<th>Columns</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>PID</td>
<td>Process ID</td>
</tr>
<tr class="even">
<td>USER</td>
<td>Name of the effective user (owner) of the process</td>
</tr>
<tr class="odd">
<td>PR</td>
<td>Priority</td>
</tr>
<tr class="even">
<td>NI</td>
<td>Nice value</td>
</tr>
<tr class="odd">
<td>VIRT</td>
<td>Virtual memory size</td>
</tr>
<tr class="even">
<td>RES</td>
<td>Resident memory size</td>
</tr>
<tr class="odd">
<td>SHR</td>
<td>Shared memory size</td>
</tr>
<tr class="even">
<td>S</td>
<td>Process status (which could be one of the following: D (uninteruptible sleep), R (running), S (sleeping), T (traced or stopped) or Z (zombie)</td>
</tr>
<tr class="odd">
<td>%CPU</td>
<td>The share of cpu time used by the process since last update</td>
</tr>
<tr class="even">
<td>%MEM</td>
<td>Share of physical memory used</td>
</tr>
<tr class="odd">
<td>TIME+</td>
<td>Total cpu time used by the task in hundredths of a second</td>
</tr>
<tr class="even">
<td>COMMAND</td>
<td>Command name or command line (name + options)</td>
</tr>
</tbody>
</table>
</div>
<div id="ps" class="section level2">
<h2><span class="header-section-number">5.2</span> <code>ps</code></h2>
<p>The command <code>ps</code> can report a snapshot of the current processes.</p>
<p>Command <code>ps</code> is usually used with the option <code>-a</code>, <code>-u</code> and <code>-x</code>.</p>
<pre><code>ps -aux   ## can also be `ps aux`</code></pre>
<p>You can pipe the output to <code>less</code> to make it scrollable.</p>
</div>
<div id="kill" class="section level2">
<h2><span class="header-section-number">5.3</span> <code>kill</code></h2>
<p>If you want to kill a process, you can use the command <code>kill</code>.</p>
<pre><code>kill 20140418</code></pre>
</div>
<div id="df" class="section level2">
<h2><span class="header-section-number">5.4</span> <code>df</code></h2>
</div>
<div id="advanced-topic-free" class="section level2">
<h2><span class="header-section-number">5.5</span> Advanced topic <code>free</code></h2>
<p>You can use command <code>free</code> to display amount of free and used memory in the system.</p>
<pre><code>free -h</code></pre>
<p><code>-h</code> let you show all output fields automatically scaled to shortest three digit unit and display the units of print out. Following units are used.</p>
<table>
<thead>
<tr class="header">
<th>Abbreviation</th>
<th>Full Name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>B</td>
<td>Bytes</td>
</tr>
<tr class="even">
<td>K</td>
<td>Kilobytes (KB)</td>
</tr>
<tr class="odd">
<td>M</td>
<td>Megabytes(MB)</td>
</tr>
<tr class="even">
<td>G</td>
<td>Gigabytes (GB)</td>
</tr>
<tr class="odd">
<td>T</td>
<td>Terabytes (TB)</td>
</tr>
</tbody>
</table>
<!--chapter:end:06_procManageLinux.Rmd-->
</div>
</div>
<div id="file-transfer" class="section level1">
<h1><span class="header-section-number">6</span> File transfer</h1>
<div id="transferring-files-between-local-computer-and-linux-server" class="section level2">
<h2><span class="header-section-number">6.1</span> Transferring files between local computer and Linux server</h2>
<p>To transfer files between local computer and Linux sever, there are two options: 1) GUI application and 2) command line.</p>
<ul>
<li>Open FileZilla and then click <code>File</code> -&gt; <code>Site Manager</code>.</li>
</ul>
<p>GUI means there will be window, buttons, menus, etc. The most popular system with GUI is Windows system (Figure @ref(fig:filezilla_screenshot1)).</p>
<p>(ref:filezilla_screenshot1) FileZilla application.</p>
<div class="figure">
<img src="figures/filezilla_screenshot1.png" alt="(ref:filezilla_screenshot1)"  />
<p class="caption">
(#fig:filezilla_screenshot1)(ref:filezilla_screenshot1)
</p>
</div>
<div id="use-command-line-tools" class="section level3">
<h3><span class="header-section-number">6.1.1</span> Use command line tools</h3>
<p><code>rsync</code> compares the files at each end and transfers only the changed parts of changed files. When you transfer files the first timeo it behaves pretty much like scp, but for a second transfer, where most files are unchanged, it will push a lot less data than scp. It’s also a convenient way to restart failed transfers - you just reissue the same command and it will pick up where it left off the time before, whereas scp will start again from scratch.</p>
<div id="copy-files-using-rsync" class="section level4">
<h4><span class="header-section-number">6.1.1.1</span> Copy files using <code>rsync</code></h4>
<pre><code></code></pre>
</div>
<div id="copying-files-with-scp" class="section level4">
<h4><span class="header-section-number">6.1.1.2</span> Copying Files with <code>scp</code></h4>
<p>The command <code>scp</code> is short for secure copy. It can be used to copy files between hosts on a network. It uses ssh(1) for data transfer, and uses the same authentication and provides the same security as ssh(1).<code>Scp</code> will ask for passwords or passphrases if they are needed for authentication.</p>
<p>File names may contain a user and host specification to indicate that the file is to be copied to/from that host. Local file names can be made explicit using absolute or relative pathnames to avoid scp treating file names containing ‘:’ as host specifiers. Copies between two remote hosts are also permitted.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb88-1" data-line-number="1"><span class="co"># Copy the file test.pl on 198.211.107.37 to the current directory.</span></a>
<a class="sourceLine" id="cb88-2" data-line-number="2"><span class="fu">scp</span> guest4bioinfor@198.211.107.37:~/test.pl ./</a></code></pre></div>
<p>To copy files from a server to a client, you need to know where the files are located on the server. For example, to copy a single file <code>~/test.pl</code> from the server with IP address of 198.211.107.37 to the current directory.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb89-1" data-line-number="1"><span class="co"># Copy the file test.pl in the current directory to  198.211.107.37</span></a>
<a class="sourceLine" id="cb89-2" data-line-number="2"><span class="fu">scp</span> ./test.pl guest4bioinfor@198.211.107.37:~/</a></code></pre></div>
<p>To copy files from a client to a server, you need to know where the files you want to put on the server. For example, to copy a single file <code>test.pl</code> from the current folder to the HOME folder of the server with IP address of 198.211.107.37.</p>
<p>If you want to copy an entire directory recursively, you can use <code>-r</code> argument. See the example below:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb90-1" data-line-number="1"><span class="co"># Copy the file test.pl in the current directory to  198.211.107.37</span></a>
<a class="sourceLine" id="cb90-2" data-line-number="2"><span class="fu">scp</span> -r guest4bioinfor@198.211.107.37:~/bioinfo/ ./</a></code></pre></div>
</div>
</div>
<div id="download-files" class="section level3">
<h3><span class="header-section-number">6.1.2</span> Download files</h3>
<pre><code>wget &lt;url&gt;</code></pre>
<p>Resume</p>
<pre><code>wget -c &lt;url&gt;  </code></pre>
<p>Reference:</p>
<p>RH066x Fundamentals of Red Hat Enterprise Linux on edX</p>
<!--chapter:end:07_fileTransfer.Rmd-->
</div>
</div>
</div>
<div id="install-bioinformatics-software-in-linux" class="section level1">
<h1><span class="header-section-number">7</span> Install Bioinformatics software in Linux</h1>
<div id="installation-from-source-code" class="section level2">
<h2><span class="header-section-number">7.1</span> Installation from source code</h2>
<p>Nearly all of the Bioinformatics softwares will be downloaded as a compressed files. So the first thing you need to do is to uncompress the file. Then the source codes will be included in a folder. You can <code>cd</code> to the folder and <code>ls</code> the files/directories. Mostly you will find either a file named <code>README</code> or <code>INSTALL</code> or both. If you read this file to know how to install the software.</p>
<div id="install-bwa" class="section level3">
<h3><span class="header-section-number">7.1.1</span> Install <code>bwa</code></h3>
<div class="sourceCode" id="cb93"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb93-1" data-line-number="1"><span class="fu">wget</span> https://sourceforge.net/projects/bio-bwa/files/bwa-0.7.15.tar.bz2</a>
<a class="sourceLine" id="cb93-2" data-line-number="2"><span class="fu">tar</span> xjvf bwa-0.7.15.tar.bz2</a>
<a class="sourceLine" id="cb93-3" data-line-number="3"><span class="bu">cd</span> bwa-0.7.15</a>
<a class="sourceLine" id="cb93-4" data-line-number="4"><span class="fu">make</span></a></code></pre></div>
</div>
<div id="install-samtools" class="section level3">
<h3><span class="header-section-number">7.1.2</span> Install <code>samtools</code></h3>
<p>Installation of <code>Samtools</code> is one of the best representatives of how to instsall a Bioinformatics tool.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb94-1" data-line-number="1"><span class="co"># Download the source code</span></a>
<a class="sourceLine" id="cb94-2" data-line-number="2"><span class="fu">wget</span> https://iweb.dl.sourceforge.net/project/samtools/samtools/1.3.1/samtools-1.3.1.tar.bz2</a>
<a class="sourceLine" id="cb94-3" data-line-number="3"><span class="co"># Uncompress the source code</span></a>
<a class="sourceLine" id="cb94-4" data-line-number="4"><span class="fu">tar</span> xjvf samtools-1.3.1.tar.bz2</a>
<a class="sourceLine" id="cb94-5" data-line-number="5"><span class="co"># Enter the source code directory.</span></a>
<a class="sourceLine" id="cb94-6" data-line-number="6"><span class="bu">cd</span> samtools-1.3.1</a>
<a class="sourceLine" id="cb94-7" data-line-number="7"><span class="co"># Configure the build system</span></a>
<a class="sourceLine" id="cb94-8" data-line-number="8"><span class="ex">./configure</span></a>
<a class="sourceLine" id="cb94-9" data-line-number="9"><span class="co"># Build samtools</span></a>
<a class="sourceLine" id="cb94-10" data-line-number="10"><span class="fu">make</span></a>
<a class="sourceLine" id="cb94-11" data-line-number="11"><span class="co"># Become a `root` user for system-wide install:</span></a>
<a class="sourceLine" id="cb94-12" data-line-number="12"><span class="fu">su</span> root</a>
<a class="sourceLine" id="cb94-13" data-line-number="13"><span class="co"># Install `Samtools`</span></a>
<a class="sourceLine" id="cb94-14" data-line-number="14"><span class="fu">make</span> install</a></code></pre></div>

<div class="rmdtip">
<p><strong>Install <code>samtools</code> without root previledges</strong></p>
<p>By default, ‘make install’ installs samtools and the utilities under
/usr/local/bin and manual pages under /usr/local/share/man.</p>
<p>You can specify a different location to install Samtools by configuring
with –prefix=DIR or specify locations for particular parts of HTSlib by
configuring with –bindir=DIR and so on. Type ‘./configure –help’ for
the full list of such install directory options.</p>
<p>Alternatively you can specify different locations at install time by
typing ‘make prefix=DIR install’ or ‘make bindir=DIR install’ and so on.
Consult the list of prefix/exec_prefix/etc variables near the top of the
Makefile for the full list of such variables that can be overridden.</p>
<p>You can also specify a staging area by typing ‘make DESTDIR=DIR install’,
possibly in conjunction with other –prefix or prefix=DIR settings.
For example,</p>
<pre><code>make DESTDIR=/tmp/staging prefix=/opt</code></pre>
<p>would install into bin and share/man subdirectories under /tmp/staging/opt.</p>
</div>

</div>
<div id="align-reads-to-genome-using-bwa-and-store-the-alignment-results-in-sambam-files" class="section level3">
<h3><span class="header-section-number">7.1.3</span> Align reads to genome using <code>bwa</code> and store the alignment results in SAM/BAM files</h3>
<pre><code>./bwa index ref.fa
./bwa mem ref.fa read-se.fq.gz | gzip -3 &gt; aln-se.sam.gz
./bwa mem ref.fa read1.fq read2.fq | gzip -3 &gt; aln-pe.sam.gz</code></pre>
</div>
</div>
<div id="installing-a-precompiled-binary-executable" class="section level2">
<h2><span class="header-section-number">7.2</span> Installing a precompiled binary (executable)</h2>
<p>For programs that are already compiled (converted from high level source code in a language like C into machine specific code), you are often given some choices and need to determine how to download the version that has the correct CPU architecture for your machine.</p>
<div id="install-bwa-1" class="section level3">
<h3><span class="header-section-number">7.2.1</span> Install <code>bwa</code></h3>
<pre><code>wget https://downloads.sourceforge.net/project/bio-bwa/bwakit/bwakit-0.7.15_x64-linux.tar.bz2
tar xjvf bwakit-0.7.15_x64-linux.tar.bz2
cd bwa.kit/
./bwa
Program: bwa (alignment via Burrows-Wheeler transformation)
Version: 0.7.15-r1140
Contact: Heng Li &lt;lh3@sanger.ac.uk&gt;

Usage:   bwa &lt;command&gt; [options]

Command: index         index sequences in the FASTA format
         mem           BWA-MEM algorithm
         fastmap       identify super-maximal exact matches
         pemerge       merge overlapping paired ends (EXPERIMENTAL)
         aln           gapped/ungapped alignment
         samse         generate alignment (single ended)
         sampe         generate alignment (paired ended)
         bwasw         BWA-SW for long queries

         shm           manage indices in shared memory
         fa2pac        convert FASTA to PAC format
         pac2bwt       generate BWT from PAC
         pac2bwtgen    alternative algorithm for generating BWT
         bwtupdate     update .bwt to the new format
         bwt2sa        generate SA from BWT and Occ

Note: To use BWA, you need to first index the genome with `bwa index&#39;.
      There are three alignment algorithms in BWA: `mem&#39;, `bwasw&#39;, and
      `aln/samse/sampe&#39;. If you are not sure which to use, try `bwa mem&#39;
      first. Please `man ./bwa.1&#39; for the manual.
</code></pre>
</div>
<div id="install-with-conda-recommended-way" class="section level3">
<h3><span class="header-section-number">7.2.2</span> Install with <code>conda</code> (recommended way)</h3>
<div id="install-conda" class="section level4">
<h4><span class="header-section-number">7.2.2.1</span> Install <code>conda</code></h4>
<p>Go the web link here (<a href="https://conda.io/en/latest/miniconda.html" class="uri">https://conda.io/en/latest/miniconda.html</a>):</p>
<pre><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
sh Miniconda3-latest-Linux-x86_64.sh</code></pre>
</div>
<div id="install-a-software-package-in-a-envrionment" class="section level4">
<h4><span class="header-section-number">7.2.2.2</span> Install a software package in a envrionment</h4>
<p>Google-search “conda bwa”, click the first hit and copy the command lines to install <code>bwa</code>.</p>
<pre><code>conda install -c bioconda bwa 
#conda install -c bioconda/label/cf201901 bwa </code></pre>
</div>
</div>
<div id="install-ussing-docker-advance-topic" class="section level3">
<h3><span class="header-section-number">7.2.3</span> Install ussing <code>Docker</code> (advance topic)</h3>
<!--chapter:end:08_InstallationOfSoftwareInLinux.Rmd-->
</div>
</div>
</div>
<div id="text-editor-in-linux" class="section level1">
<h1><span class="header-section-number">8</span> Text editor in Linux</h1>
<p>In Linux, we sometimes need to create or edit a text file like writing a new perl script. So we need to use text editor.</p>
<p>As a newbie, someone would prefer a basic, GUI-based text editor with menus and traditional CUA key bindings. Here we recommend <a href="https://www.sublimetext.com/">Sublime</a>, <a href="https://atom.io">ATOM</a> and <a href="https://notepad-plus-plus.org/">Notepad++</a>.</p>
<p>But GUI-based text editor is not always available in Linux.</p>
<p>A powerful screen text editor <code>vi</code> (pronounced “vee-eye”) is available on nearly all Linux system. We highly recommend <code>vi</code> as a text editor, because something we’ll have to edit a text file on a system without a friendlier text editor. Once we get familiar with <code>vi</code>, we’ll find that it’s very fast and powerful.</p>
<p>But remember, it’s OK if you think this part is too difficult at the beginning. You can use either <code>Sublime</code>, <code>ATOM</code> or <code>Notepad++</code>. If you are connecting to a Linux system without <code>Sublime</code>, <code>ATOM</code> and <code>Notepad++</code>, you can write the file in a local computer and then upload the file onto Linux system.</p>
<div id="basic-vi-skills" class="section level2">
<h2><span class="header-section-number">8.1</span> Basic vi skills</h2>
<p>As <code>vi</code> uses a lot of combination of keystrokes, it may be not easy for newbies to remember all the combinations in one fell swoop. Considering this, we’ll first introduce the basic skills someone needs to know to use <code>vi</code>. We need to first understand how three modes of <code>vi</code> work and then try to remember a few basic <code>vi</code> commonds. Then we can use these skills to write Perl or R scripts in the following chaptors for Perl and R (Figure @ref(fig:workingModeVi)).</p>
<p>(ref:workingModeVi) Three modes of vi.</p>
<div class="figure">
<img src="figures/workingModeVi.png" alt="(ref:workingModeVi)"  />
<p class="caption">
(#fig:workingModeVi)(ref:workingModeVi)
</p>
</div>
</div>
<div id="create-new-text-file-with-vi" class="section level2">
<h2><span class="header-section-number">8.2</span> Create new text file with <code>vi</code></h2>
<div class="sourceCode" id="cb100"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb100-1" data-line-number="1"><span class="fu">mkdir</span> test_vi  ## generate a new folder</a>
<a class="sourceLine" id="cb100-2" data-line-number="2"><span class="bu">cd</span> test_vi     ## go into the new folder</a>
<a class="sourceLine" id="cb100-3" data-line-number="3"><span class="bu">echo</span> <span class="st">&quot;Using </span><span class="dt">\`</span><span class="st">ls</span><span class="dt">\`</span><span class="st"> we don&#39;t expect files in this folder.&quot;</span></a>
<a class="sourceLine" id="cb100-4" data-line-number="4"><span class="fu">ls</span> </a>
<a class="sourceLine" id="cb100-5" data-line-number="5"><span class="bu">echo</span> <span class="st">&quot;No file displayed!&quot;</span></a></code></pre></div>
<pre><code>## Using `ls` we don&#39;t expect files in this folder.
## No file displayed!</code></pre>
<p>Using the code above, we made a new directory named <code>test_vi</code>. We didn’t see any file.</p>
<p>If we type <code>vi test.pl</code>, an empty file and screen are created into which you may enter text because the file does not exist((Figure @ref(fig:ViNewFile))).</p>
<pre><code>vi test.pl</code></pre>
<p>(ref:ViNewFile) A screentshot of the <code>vi test.pl</code>.</p>
<div class="figure" style="text-align: center">
<img src="images/vi_new_file.png" alt="(ref:ViNewFile)" width="100%" />
<p class="caption">
(#fig:ViNewFile)(ref:ViNewFile)
</p>
</div>
<p>Now if you are in <code>vi mode</code>. To go to <code>Input mode</code>, you can type <code>i</code>, ‘a’ or ‘o’ (Figure @ref(fig:ViInpuMode)).</p>
<p>(ref:ViInpuMode) A screentshot of the <code>vi test.pl</code>.</p>
<div class="figure" style="text-align: center">
<img src="images/vi_input_mode.png" alt="(ref:ViInpuMode)" width="100%" />
<p class="caption">
(#fig:ViInpuMode)(ref:ViInpuMode)
</p>
</div>
<p>Now you can type the content (codes or other information) (@ref(fig:ViInpuType)).</p>
<p>(ref:ViInpuType) A screentshot of the <code>vi test.pl</code>.</p>
<div class="figure" style="text-align: center">
<img src="images/vi_input_type.png" alt="(ref:ViInpuType)" width="100%" />
<p class="caption">
(#fig:ViInpuType)(ref:ViInpuType)
</p>
</div>
<p>Once you are done typing. You need to go to <code>Command mode</code>(Figure @ref(fig:workingModeVi)) if you want to save and exit the file. To do this, you need to press <code>ESC</code> button on the keyboard.</p>
<p>Now we just wrote a Perl script. We can run this script.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb103-1" data-line-number="1"><span class="fu">perl</span> test.pl</a></code></pre></div>
<pre><code>## Hello Bioinformatics World!</code></pre>
</div>
<div id="an-example-for-using-editor-r" class="section level2">
<h2><span class="header-section-number">8.3</span> An example for using editor R</h2>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb105-1" data-line-number="1"><span class="kw">qnorm</span>(.<span class="dv">975</span>)</a></code></pre></div>
<pre><code>## [1] 1.959964</code></pre>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb107-1" data-line-number="1">xval&lt;-<span class="kw">seq</span>(<span class="op">-</span><span class="fl">3.2</span>,<span class="fl">3.2</span>, <span class="dt">length=</span><span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb107-2" data-line-number="2">yval&lt;-<span class="kw">dnorm</span>(xval)</a>
<a class="sourceLine" id="cb107-3" data-line-number="3"><span class="kw">plot</span>(xval, yval, <span class="dt">type=</span><span class="st">&quot;l&quot;</span>,<span class="dt">axes=</span>T,<span class="dt">lwd=</span><span class="dv">3</span>,<span class="dt">xlab=</span><span class="st">&quot;&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;&quot;</span>)</a>
<a class="sourceLine" id="cb107-4" data-line-number="4">x&lt;-<span class="kw">seq</span>(<span class="kw">qnorm</span>(.<span class="dv">975</span>), <span class="fl">3.2</span>, <span class="dt">length =</span> <span class="dv">100</span>)</a>
<a class="sourceLine" id="cb107-5" data-line-number="5"><span class="kw">polygon</span>(<span class="kw">c</span>(x,<span class="kw">rev</span>(x)), <span class="kw">c</span>(<span class="kw">dnorm</span>(x), <span class="kw">rep</span>(<span class="dv">0</span>,<span class="kw">length</span>(x))), <span class="dt">col=</span><span class="st">&quot;salmon&quot;</span>)</a>
<a class="sourceLine" id="cb107-6" data-line-number="6"><span class="kw">text</span>(<span class="kw">mean</span>(x),<span class="kw">mean</span>(<span class="kw">dnorm</span>(x))<span class="op">+</span><span class="fl">0.02</span>, <span class="st">&quot;2.5%&quot;</span>, <span class="dt">cex=</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb107-7" data-line-number="7"><span class="kw">text</span>(<span class="kw">qnorm</span>(.<span class="dv">95</span>), <span class="fl">0.01</span>, <span class="st">&quot;1.645&quot;</span>,<span class="dt">cex=</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb107-8" data-line-number="8"></a>
<a class="sourceLine" id="cb107-9" data-line-number="9">x&lt;-<span class="kw">seq</span>(<span class="op">-</span><span class="fl">3.2</span>, <span class="kw">qnorm</span>(.<span class="dv">025</span>), <span class="dt">length =</span><span class="dv">100</span>)</a>
<a class="sourceLine" id="cb107-10" data-line-number="10"><span class="kw">polygon</span>(<span class="kw">c</span>(x,<span class="kw">rev</span>(x)), <span class="kw">c</span>(<span class="kw">dnorm</span>(x), <span class="kw">rep</span>(<span class="dv">0</span>,<span class="kw">length</span>(x))), <span class="dt">col=</span><span class="st">&quot;salmon&quot;</span>)</a>
<a class="sourceLine" id="cb107-11" data-line-number="11"><span class="kw">text</span>(<span class="kw">mean</span>(x),<span class="kw">mean</span>(<span class="kw">dnorm</span>(x))<span class="op">+</span><span class="fl">0.02</span>, <span class="st">&quot;2.5%&quot;</span>, <span class="dt">cex=</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb107-12" data-line-number="12"><span class="kw">text</span>(<span class="kw">qnorm</span>(.<span class="dv">025</span>), <span class="fl">0.01</span>, <span class="st">&quot;1.645&quot;</span>,<span class="dt">cex=</span><span class="dv">2</span>)</a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-41-1.png" width="672" /></p>
<!--chapter:end:09_TextEditorInLinux.Rmd-->
</div>
</div>
<div id="part-perl" class="section level1 unnumbered">
<h1>(PART) Perl</h1>
</div>
<div id="first-perl-program" class="section level1">
<h1><span class="header-section-number">9</span> First Perl Program</h1>
<p>Scripting languages, like Perl, are very commonly used in bioinformatics. As a generous scripting language, Perl have many advantages: easy to use, free for all operating systems like Linux, designed for working with text files (tab-delimited files). It’s one of the most popular language in bioinformatics. Moreover there are many scripts and modules available. Additionally, there are a lot of resource on Internet.</p>
<div id="first-program" class="section level2">
<h2><span class="header-section-number">9.1</span> First Program</h2>
<p>As all other programming books, we begin with a “Hello world” program.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb108-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb108-2" data-line-number="2"><span class="co">#Printing a line of text “Hello, Bioinformatics”</span></a>
<a class="sourceLine" id="cb108-3" data-line-number="3"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Hello, Bioinformatics!</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<p>This program show how to display a line a text in Perl. It have several features. We go through each line in detail.</p>
<p>Line 1 is what we call shebang line. This line starts with shebang construct (<code>#!</code>). <code>/usr/bin/perl</code> indicates the path of the Perl interpreter.</p>
<p>Line 3 shows how to print a line of text in Perl. Nearly all programming language use print to display texts on the screen. Here, print is a built-in function in Perl. It print the string of characters (its arguments) between quotation marks (“” or ‘’).</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb109-1" data-line-number="1"><span class="fu">perl</span> code_perl/hello_bioinfor.pl</a></code></pre></div>
<pre><code>## Hello, Bioinformatics!</code></pre>
<p>However the characters <code>\n</code> are not displayed. Here backslash <code>\</code> is a start of an escape sequence. It changes the meaning of the character after it. The backslash <code>\</code> and <code>n</code> together (<code>\n</code>) form an escape sequence and signify a newline. Other examples are <code>\t</code> (tab) or <code>\$</code> (= print an actual dollar sign, normally a dollar sign has a special meaning). We’ll see more escape sequences in 7.1.</p>
<p>You can try to remove <code>\n</code> in the program to see what will happen. This will give you a dee per understanding of the program.</p>
<!--chapter:end:10_FirstPerlProgram.Rmd-->
</div>
</div>
<div id="varible-in-perl" class="section level1">
<h1><span class="header-section-number">10</span> Varible in Perl</h1>
<p>Perl provides three kinds of variables: <code>scalars</code>, <code>arrays</code>, and <code>hash</code>(aka <code>associative arrays</code>). The initial character of the name identifies the particular type of variable and, hence, its functionality.</p>
<table>
<thead>
<tr class="header">
<th>Type</th>
<th>Character</th>
<th>Example</th>
<th>Is a name for:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Scalar</td>
<td>$</td>
<td>$length</td>
<td>An individual value (number or string)</td>
</tr>
<tr class="even">
<td>Array</td>
<td>@</td>
<td>@gene_list</td>
<td>A list of values, keyed by number</td>
</tr>
<tr class="odd">
<td>Hash</td>
<td>%</td>
<td>%gene_annotation</td>
<td>A group of values, keyed by string</td>
</tr>
</tbody>
</table>
<div id="scalar-variable" class="section level2">
<h2><span class="header-section-number">10.1</span> Scalar variable</h2>
<p>In Perl, scalar variables can be used to store mainly two types of data: string and numbers.</p>
<p>Perl does not differentiate between a number and a string, nor does it differentiate between integers and reals.</p>
<p>In order to tell the computer what to print, we need to use variables. In Perl, the name of a <code>scalar</code> variable starts with the dollar sign <code>$</code>. You can assign either a number or a string to it.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb111-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb111-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb111-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb111-4" data-line-number="4"></a>
<a class="sourceLine" id="cb111-5" data-line-number="5"><span class="co">#assign two strings to two variables</span></a>
<a class="sourceLine" id="cb111-6" data-line-number="6"><span class="kw">my</span> <span class="dt">$dna_seq1</span> = <span class="kw">&quot;</span><span class="st">ACCTCGGTACAGTGAATGGGAAACGTAGCTGAT</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb111-7" data-line-number="7"><span class="kw">my</span> <span class="dt">$dna_seq2</span> = <span class="kw">&quot;</span><span class="st">TGCCGATCGTAATAGCTCGCTATCTAGCTCGATCGTCGTA</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb111-8" data-line-number="8"></a>
<a class="sourceLine" id="cb111-9" data-line-number="9"><span class="co">#Returns the length in characters of the value of EXPR</span></a>
<a class="sourceLine" id="cb111-10" data-line-number="10"><span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna_seq1</span>;</a>
<a class="sourceLine" id="cb111-11" data-line-number="11"><span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna_seq2</span>;</a>
<a class="sourceLine" id="cb111-12" data-line-number="12"></a>
<a class="sourceLine" id="cb111-13" data-line-number="13"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length of first DNA sequence (</span><span class="dt">$dna_seq1</span><span class="st">) is: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb111-14" data-line-number="14"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length of the Second DNA sequence (</span><span class="dt">$dna_seq2</span><span class="st">) is: </span><span class="dt">$dna_length2</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<div class="sourceCode" id="cb112"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb112-1" data-line-number="1"><span class="fu">perl</span> code_perl/variable_assign.pl</a></code></pre></div>
<pre><code>## The length of first DNA sequence (ACCTCGGTACAGTGAATGGGAAACGTAGCTGAT) is: 33
## The length of the Second DNA sequence (TGCCGATCGTAATAGCTCGCTATCTAGCTCGATCGTCGTA) is: 40</code></pre>
<p>Here are the explanations of this script.</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb114-1" data-line-number="1"><span class="co"># assign two DNA sequences to two variables</span></a>
<a class="sourceLine" id="cb114-2" data-line-number="2"><span class="kw">my</span> <span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb114-3" data-line-number="3"><span class="kw">my</span> <span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;</a></code></pre></div>
<p>The first line here is a comment. In Perl, <code>#</code> (pound sign) is the comment character. The comments can be used to document the program and improve the readability. During the execution, the comments will be ignored. The second and third lines declare two string scalar variables (<code>$dna1</code> and <code>$dna2</code>) to store two DNA sequences.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb115-1" data-line-number="1"><span class="co">#Returns the length in characters of the value of EXPR</span></a>
<a class="sourceLine" id="cb115-2" data-line-number="2"><span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna_seq1</span>; </a>
<a class="sourceLine" id="cb115-3" data-line-number="3"><span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna_seq2</span>;</a></code></pre></div>
<p>The code above has one comment line and declares two integer scalar variables to store the return values of built-on function <code>length</code>.</p>
<p>Now you can output the variables to see what are stored in them using the following code:</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb116-1" data-line-number="1"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length of first DNA sequence (</span><span class="dt">$dna_seq1</span><span class="st">) is: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb116-2" data-line-number="2"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length of the Second DNA sequence (</span><span class="dt">$dna_seq2</span><span class="st">) is: </span><span class="dt">$dna_length2</span><span class="ch">\n\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<p>This script shows you what string scalar varibles and integer scalr varibles.</p>
</div>
<div id="arithmetic-operations-in-perl" class="section level2">
<h2><span class="header-section-number">10.2</span> Arithmetic operations in Perl</h2>
<p>Most arithmetic operators are binary operators; this means they take two arguments. Unary operators only take one argument. Arithmetic operators are very simple and often transparent.</p>
<p>Here we’re mainly going to talk about basic arithmetic opertators including addition (<code>+</code>), substraction (<code>-</code>), multiplication (<code>*</code>), division (<code>/</code>) and the modulus operation (%). Modulus (<code>%</code>) returns the remainder of a division (<code>/</code>) operation.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb117-1" data-line-number="1"><span class="co">## If we know the lengths of two sequences and we want to calulate the sum of the lengths.</span></a>
<a class="sourceLine" id="cb117-2" data-line-number="2"><span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="dv">100</span>;</a>
<a class="sourceLine" id="cb117-3" data-line-number="3"><span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="dv">200</span>;</a>
<a class="sourceLine" id="cb117-4" data-line-number="4"><span class="kw">my</span> <span class="dt">$total_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_length2</span>;</a>
<a class="sourceLine" id="cb117-5" data-line-number="5"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of these two sequences is </span><span class="dt">$total_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb117-6" data-line-number="6"></a>
<a class="sourceLine" id="cb117-7" data-line-number="7"><span class="kw">my</span> <span class="dt">$length_diff</span> = <span class="dt">$dna_length1</span> - <span class="dt">$dna_length2</span>; </a>
<a class="sourceLine" id="cb117-8" data-line-number="8"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length difference is </span><span class="dt">$length_diff</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb117-9" data-line-number="9"></a>
<a class="sourceLine" id="cb117-10" data-line-number="10"><span class="kw">my</span> <span class="dt">$average_length</span> = <span class="dt">$total_length</span> / <span class="dv">2</span>;</a>
<a class="sourceLine" id="cb117-11" data-line-number="11"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The average length of two DNA sequences is </span><span class="dt">$average_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## The total length of these two sequences is 300 
## The length difference is -100
## The average length of two DNA sequences is 150</code></pre>
<p>The code above shows us how to</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb119-1" data-line-number="1"><span class="co">#Imaging the CC content of first DNA sequence is 0.5.</span></a>
<a class="sourceLine" id="cb119-2" data-line-number="2"><span class="co">#How many CC do we have in first DNA?</span></a>
<a class="sourceLine" id="cb119-3" data-line-number="3"><span class="kw">my</span> <span class="dt">$cg_content</span> = <span class="fl">0.5</span>;</a>
<a class="sourceLine" id="cb119-4" data-line-number="4"><span class="kw">my</span> <span class="dt">$cg_number</span> = <span class="dt">$dna_length1</span> <span class="kw">*</span> <span class="fl">0.5</span>;</a></code></pre></div>
<div class="sourceCode" id="cb120"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb120-1" data-line-number="1"><span class="co">#Imaging we have a 10 bps DNA sequnces,</span></a>
<a class="sourceLine" id="cb120-2" data-line-number="2"><span class="co">#how many possible DNA sequences do we have?</span></a>
<a class="sourceLine" id="cb120-3" data-line-number="3"><span class="kw">my</span> <span class="dt">$dna_nucleotide</span> = <span class="kw">&quot;</span><span class="st">ATCG</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb120-4" data-line-number="4"><span class="kw">my</span> <span class="dt">$dna_nucleo_number</span> = <span class="fu">length</span> <span class="dt">$dna_nucleotide</span>;</a>
<a class="sourceLine" id="cb120-5" data-line-number="5"><span class="kw">my</span> <span class="dt">$dna_length</span> = <span class="dv">10</span>;</a>
<a class="sourceLine" id="cb120-6" data-line-number="6"><span class="kw">my</span> <span class="dt">$possible_number</span> = <span class="dt">$dna_nucleo_number</span> *<span class="kw">*</span> <span class="dt">$dna_length</span>;</a>
<a class="sourceLine" id="cb120-7" data-line-number="7"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">We have </span><span class="dt">$possible_number</span><span class="st"> possibilities.</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## We have 1048576 possibilities.</code></pre>
<div id="shorthand-operations" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Shorthand operations</h3>
<p>The expression $x += 3; is the shorthand version of $x = $x + 3;, they have exactly the same result:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb122-1" data-line-number="1"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb122-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb122-3" data-line-number="3"></a>
<a class="sourceLine" id="cb122-4" data-line-number="4"><span class="kw">my</span> <span class="dt">$dna_length</span> = <span class="dv">10</span>;</a>
<a class="sourceLine" id="cb122-5" data-line-number="5"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb122-6" data-line-number="6"></a>
<a class="sourceLine" id="cb122-7" data-line-number="7"><span class="dt">$dna_length</span> += <span class="dv">3</span>;</a>
<a class="sourceLine" id="cb122-8" data-line-number="8"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="st"> after &#39;\$dna_length += 3&#39;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb122-9" data-line-number="9"></a>
<a class="sourceLine" id="cb122-10" data-line-number="10"><span class="dt">$dna_length</span> -= <span class="dv">3</span>;  </a>
<a class="sourceLine" id="cb122-11" data-line-number="11"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="st"> after &#39;\$dna_length -= 3&#39;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## DNA length: 10
## DNA length: 13 after &#39;$dna_length += 3&#39;
## DNA length: 10 after &#39;$dna_length -= 3&#39;</code></pre>
</div>
<div id="auto-increment-and-auto-decrement" class="section level3">
<h3><span class="header-section-number">10.2.2</span> Auto increment and auto decrement</h3>
<p><code>++</code> and <code>--</code> are provided for the auto increment and auto decrement operators. They increase and decrease respectively the value of a scalar variable by 1.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb124-1" data-line-number="1"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb124-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb124-3" data-line-number="3"></a>
<a class="sourceLine" id="cb124-4" data-line-number="4"><span class="kw">my</span> <span class="dt">$dna_length</span> = <span class="dv">10</span>;</a>
<a class="sourceLine" id="cb124-5" data-line-number="5"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb124-6" data-line-number="6"></a>
<a class="sourceLine" id="cb124-7" data-line-number="7"><span class="dt">$dna_length</span> ++;</a>
<a class="sourceLine" id="cb124-8" data-line-number="8"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="st"> after &#39;\$dna_length ++&#39;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb124-9" data-line-number="9"></a>
<a class="sourceLine" id="cb124-10" data-line-number="10"><span class="dt">$dna_length</span> --;</a>
<a class="sourceLine" id="cb124-11" data-line-number="11"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="st"> after `\$dna_length --`</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## DNA length: 10
## DNA length: 11 after &#39;$dna_length ++&#39;
## DNA length: 10 after `$dna_length --`</code></pre>
</div>
</div>
<div id="use-strict-user-warnings-and-my" class="section level2">
<h2><span class="header-section-number">10.3</span> <code>use strict</code>; <code>user warnings</code> and <code>my</code></h2>
<p>For starters, <code>use strict;</code> (and to a lesser extent, <code>use warnings;</code>) helps find typos in variable names. Even experienced programmers make such errors. A common case is forgetting to rename an instance of a variable when cleaning up or refactoring code.</p>
<p>Using <code>use strict; use warnings;</code> catches many errors sooner than they would be caught otherwise, which makes it easier to find the root causes of the errors. The root cause might be the need for an error or validation check, and that can happen regardless of programmer skill.</p>
<p>What’s good about Perl warnings is that they are rarely spurious, so there’s next to no cost to using them.</p>
<p>In the script below, <code>$dna_lenght2</code> is a typo. If you run this script, it will give you the output without any error message, although it’s not the right output.</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb126-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb126-2" data-line-number="2"></a>
<a class="sourceLine" id="cb126-3" data-line-number="3"><span class="co">#assign 2 numbers to 2 variable</span></a>
<a class="sourceLine" id="cb126-4" data-line-number="4"><span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb126-5" data-line-number="5"><span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb126-6" data-line-number="6"></a>
<a class="sourceLine" id="cb126-7" data-line-number="7"><span class="co">#Returns the length in characters of the value of EXPR</span></a>
<a class="sourceLine" id="cb126-8" data-line-number="8"><span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna1</span>;</a>
<a class="sourceLine" id="cb126-9" data-line-number="9"><span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna2</span>;</a>
<a class="sourceLine" id="cb126-10" data-line-number="10"></a>
<a class="sourceLine" id="cb126-11" data-line-number="11"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First DNA: </span><span class="dt">$dna1</span><span class="st">; Length: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb126-12" data-line-number="12"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second DNA: </span><span class="dt">$dna2</span><span class="st">; Length: </span><span class="dt">$dna_length2</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb126-13" data-line-number="13"></a>
<a class="sourceLine" id="cb126-14" data-line-number="14"><span class="co">#calculate the total length of two DNA sequences</span></a>
<a class="sourceLine" id="cb126-15" data-line-number="15"><span class="dt">$tot_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_lenght2</span>;</a>
<a class="sourceLine" id="cb126-16" data-line-number="16"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of two DNA sequences is </span><span class="dt">$tot_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<p>Let’s try to run this script:</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb127-1" data-line-number="1"><span class="fu">perl</span> code_perl/var_assign_no_strict_warnings.pl</a></code></pre></div>
<pre><code>First DNA: CTCGACCAGGACGATGAATGGGCGATGAAAATCT; Length: 34
Second DNA: CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT; Length: 48
The total length of two DNA sequences is 34 </code></pre>
<p>So the script above is supposed to output the length of two DNA sequences and the sum of the lengths.</p>
<p>In the chunk of code above, <code>$dna_lenght2</code> is an empty varible without storing any information. By defaut, Perl considers this as ZERO when doing plus operation. Although there was no error message given here, we infact have an incorrect output.</p>
<p>If we add <code>use strict</code> and <code>use warnings</code>, we need to decare each variable in the script. Let us see what will happen if we have an typo.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb129-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb129-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb129-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb129-4" data-line-number="4"></a>
<a class="sourceLine" id="cb129-5" data-line-number="5"><span class="co">#assign 2 numbers to 2 variable</span></a>
<a class="sourceLine" id="cb129-6" data-line-number="6"><span class="kw">my</span> <span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb129-7" data-line-number="7"><span class="kw">my</span> <span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb129-8" data-line-number="8"></a>
<a class="sourceLine" id="cb129-9" data-line-number="9"><span class="co">#Returns the length in characters of the value of EXPR</span></a>
<a class="sourceLine" id="cb129-10" data-line-number="10"><span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna1</span>;</a>
<a class="sourceLine" id="cb129-11" data-line-number="11"><span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna2</span>;</a>
<a class="sourceLine" id="cb129-12" data-line-number="12"></a>
<a class="sourceLine" id="cb129-13" data-line-number="13"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First DNA: </span><span class="dt">$dna1</span><span class="st">; Length: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb129-14" data-line-number="14"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second DNA: </span><span class="dt">$dna2</span><span class="st">; Length: </span><span class="dt">$dna_length2</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb129-15" data-line-number="15"></a>
<a class="sourceLine" id="cb129-16" data-line-number="16"><span class="co">#calculate the total length of two DNA sequences</span></a>
<a class="sourceLine" id="cb129-17" data-line-number="17"><span class="kw">my</span> <span class="dt">$tot_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_lenght2</span>;</a>
<a class="sourceLine" id="cb129-18" data-line-number="18"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of two DNA sequences is </span><span class="dt">$tot_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>Global symbol &quot;$dna_lenght2&quot; requires explicit package name (did you forget to declare &quot;my $dna_lenght2&quot;?) at code_perl/var_assign_strict_warnings.pl line 17.
Execution of code_perl/var_assign_strict_warnings.pl aborted due to compilation errors.</code></pre>
<p>Now if we run this script, we encounter error mesage and the script can’t be sucessfuly excuted.</p>
<pre><code>Global symbol &quot;$dna_lenght2&quot; requires explicit package name 
(did you forget to declare &quot;my $dna_lenght2&quot;?) at code_perl/
var_assign_strict_warnings.pl line 17.
Execution of code_perl/var_assign_strict_warnings.pl aborted
 due to compilation errors.</code></pre>
</div>
<div id="array" class="section level2">
<h2><span class="header-section-number">10.4</span> Array</h2>
<div id="init-an-array" class="section level3">
<h3><span class="header-section-number">10.4.1</span> Init an array</h3>
<div class="sourceCode" id="cb132"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb132-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@base_pair</span> = (<span class="kw">&#39;</span><span class="st">A</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">T</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">C</span><span class="kw">&#39;</span>, <span class="kw">&quot;</span><span class="st">C</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="st">G</span><span class="kw">&quot;</span>);</a>
<a class="sourceLine" id="cb132-2" data-line-number="2"><span class="fu">print</span> <span class="kw">&quot;&quot;</span>;</a>
<a class="sourceLine" id="cb132-3" data-line-number="3"><span class="fu">print</span> <span class="dt">@base_pair</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb132-4" data-line-number="4"><span class="fu">print</span> <span class="fu">join</span>(<span class="kw">&quot;</span><span class="ch">\t</span><span class="kw">&quot;</span>, <span class="dt">@base_pair</span>), <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## ATCCG
## A    T   C   C   G</code></pre>
</div>
<div id="array-index" class="section level3">
<h3><span class="header-section-number">10.4.2</span> Array index</h3>
<div class="sourceCode" id="cb134"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb134-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@base_pair</span> = (<span class="kw">&#39;</span><span class="st">A</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">T</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">C</span><span class="kw">&#39;</span>, <span class="kw">&quot;</span><span class="st">C</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="st">G</span><span class="kw">&quot;</span>);</a>
<a class="sourceLine" id="cb134-2" data-line-number="2"><span class="co">### Extract the first element in the array</span></a>
<a class="sourceLine" id="cb134-3" data-line-number="3"><span class="kw">my</span> <span class="dt">$first_base</span> = <span class="dt">$base_pair</span>[<span class="dv">0</span>];</a>
<a class="sourceLine" id="cb134-4" data-line-number="4"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First base is: </span><span class="dt">$first_base</span><span class="ch">\n\n</span><span class="kw">&quot;</span>; </a>
<a class="sourceLine" id="cb134-5" data-line-number="5"></a>
<a class="sourceLine" id="cb134-6" data-line-number="6"><span class="co">### Extract the last element in the array</span></a>
<a class="sourceLine" id="cb134-7" data-line-number="7"><span class="kw">my</span> <span class="dt">$last_base</span> = <span class="dt">$base_pair</span>[<span class="dv">4</span>];</a>
<a class="sourceLine" id="cb134-8" data-line-number="8"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Last base is: </span><span class="dt">$last_base</span><span class="ch">\n\n</span><span class="kw">&quot;</span>; </a>
<a class="sourceLine" id="cb134-9" data-line-number="9"></a>
<a class="sourceLine" id="cb134-10" data-line-number="10"><span class="co">### Extract the last element in the array using index `-1`</span></a>
<a class="sourceLine" id="cb134-11" data-line-number="11"><span class="kw">my</span> <span class="dt">$last_base</span> = <span class="dt">$base_pair</span>[-<span class="dv">1</span>];</a>
<a class="sourceLine" id="cb134-12" data-line-number="12"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Last base using &#39;-1&#39; is: </span><span class="dt">$last_base</span><span class="ch">\n\n</span><span class="kw">&quot;</span>; </a>
<a class="sourceLine" id="cb134-13" data-line-number="13"></a>
<a class="sourceLine" id="cb134-14" data-line-number="14"><span class="co">### Extract the last element in the array using index `$#`</span></a>
<a class="sourceLine" id="cb134-15" data-line-number="15"><span class="kw">my</span> <span class="dt">$last_base</span> = <span class="dt">$base_pair</span>[<span class="dt">$#base_pair</span>];</a>
<a class="sourceLine" id="cb134-16" data-line-number="16"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Last base using &#39;\$#&#39; is : </span><span class="dt">$last_base</span><span class="ch">\n\n</span><span class="kw">&quot;</span>; </a></code></pre></div>
<pre><code>## First base is: A
## 
## Last base is: G
## 
## Last base using &#39;-1&#39; is: G
## 
## Last base using &#39;$#&#39; is : G</code></pre>
</div>
<div id="length-of-the-array" class="section level3">
<h3><span class="header-section-number">10.4.3</span> Length of the array</h3>
<p>To get the length of the array, you can use <code>scalar @array</code> or <code>just array</code>.</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb136-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@gene_expr</span> = (<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">10</span>);</a>
<a class="sourceLine" id="cb136-2" data-line-number="2"><span class="kw">my</span> <span class="dt">$len</span> = <span class="fu">scalar</span> <span class="dt">@gene_expr</span>;  <span class="co">## or @gene_expr</span></a>
<a class="sourceLine" id="cb136-3" data-line-number="3"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array has: </span><span class="dt">@gene_expr</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb136-4" data-line-number="4"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Length of array: </span><span class="kw">&quot;</span>, <span class="dt">$len</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## Array has: 1 3 10
## Length of array: 3</code></pre>
<p>Another way is to use <code>$#array</code>. <code>$#array</code> will return the index of the last element. Since the index starts with 0, to get the length we use <code>$#array + 1</code>.</p>
<p>If the length of array is empty, <code>$#array</code> will return <code>-1</code>.</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb138-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@gene_expr</span> = ();</a>
<a class="sourceLine" id="cb138-2" data-line-number="2"><span class="kw">my</span> <span class="dt">$len</span> = <span class="dt">$#gene_expr</span> + <span class="dv">1</span>;</a>
<a class="sourceLine" id="cb138-3" data-line-number="3"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array has </span><span class="dt">$len</span><span class="st"> element</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb138-4" data-line-number="4"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Length of array: </span><span class="kw">&quot;</span>, <span class="dt">$len</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## Array has 0 element
## Length of array: 0</code></pre>
</div>
<div id="sort-arrays-in-perl" class="section level3">
<h3><span class="header-section-number">10.4.4</span> Sort Arrays in Perl</h3>
<div id="sort-alphebetically" class="section level4">
<h4><span class="header-section-number">10.4.4.1</span> Sort alphebetically</h4>
<div class="sourceCode" id="cb140"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb140-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@base_pair</span> = (<span class="kw">&#39;</span><span class="st">A</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">T</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">C</span><span class="kw">&#39;</span>, <span class="kw">&quot;</span><span class="st">C</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="st">G</span><span class="kw">&quot;</span>);</a>
<a class="sourceLine" id="cb140-2" data-line-number="2"><span class="kw">my</span> <span class="dt">@sorted_bp</span> = <span class="fu">sort</span> <span class="dt">@base_pair</span>;</a>
<a class="sourceLine" id="cb140-3" data-line-number="3"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array before sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@base_pair</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb140-4" data-line-number="4"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array after sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@sorted_bp</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## Array before sorted: A T C C G
## Array after sorted: A C C G T</code></pre>
</div>
<div id="sort-numerically" class="section level4">
<h4><span class="header-section-number">10.4.4.2</span> Sort numerically</h4>
<p>To sort an array numerically, we use <code>spaceship operator: &lt;=&gt;</code>.</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb142-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@genome_coor</span> = (<span class="dv">100</span>, <span class="dv">300</span>, <span class="dv">200</span>, <span class="dv">500</span>);</a>
<a class="sourceLine" id="cb142-2" data-line-number="2"><span class="kw">my</span> <span class="dt">@sorted_coor</span> = <span class="fu">sort</span> {<span class="dt">$a</span> &lt;=&gt; <span class="dt">$b</span>} <span class="dt">@genome_coor</span>;</a>
<a class="sourceLine" id="cb142-3" data-line-number="3"></a>
<a class="sourceLine" id="cb142-4" data-line-number="4"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array before sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@genome_coor</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb142-5" data-line-number="5"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array after sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@sorted_coor</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## Array before sorted: 100 300 200 500
## Array after sorted: 100 200 300 500</code></pre>
<p>Similarly, array can be also sorted numerically in decreasing order.</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb144-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@genome_coor</span> = (<span class="dv">100</span>, <span class="dv">300</span>, <span class="dv">200</span>, <span class="dv">500</span>);</a>
<a class="sourceLine" id="cb144-2" data-line-number="2"><span class="co">## {$a &lt;=&gt; $b} is modified as {$b &lt;=&gt; $a} </span></a>
<a class="sourceLine" id="cb144-3" data-line-number="3"><span class="kw">my</span> <span class="dt">@sorted_coor</span> = <span class="fu">sort</span> {<span class="dt">$b</span> &lt;=&gt; <span class="dt">$a</span>} <span class="dt">@genome_coor</span>;  </a>
<a class="sourceLine" id="cb144-4" data-line-number="4"></a>
<a class="sourceLine" id="cb144-5" data-line-number="5"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array before sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@genome_coor</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb144-6" data-line-number="6"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array after sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@sorted_coor</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## Array before sorted: 100 300 200 500
## Array after sorted: 500 300 200 100</code></pre>
</div>
</div>
<div id="use-push-pop-shift-and-unshift-in-perl" class="section level3">
<h3><span class="header-section-number">10.4.5</span> Use <code>push</code>, <code>pop</code>, <code>shift</code> and <code>unshift</code> in Perl</h3>
</div>
</div>
<div id="hash-in-perl" class="section level2">
<h2><span class="header-section-number">10.5</span> Hash in Perl</h2>
<p>A Perl hash is defined by key-value pairs. Perl stores elements of a hash in such an optimal way that you can look up its values based on keys very fast.</p>
<p>With the array, you use indices to access its elements. However, you must use descriptive keys to access hash’s element. A hash is sometimes referred to as an associative array.</p>
<p>Like a scalar or an array variable, a hash variable has its own prefix. A hash variable must begin with a percent sign ( %). The prefix % looks like key/value pair so remember this trick to name the hash variables.</p>
<p>The following example defines a simple hash.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb146-1" data-line-number="1"><span class="kw">my</span> <span class="dt">%gene_info</span> = (<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">ROS</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb146-2" data-line-number="2">                 <span class="kw">&quot;</span><span class="st">gene2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb146-3" data-line-number="3">                 <span class="kw">&quot;</span><span class="st">gene3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span></a>
<a class="sourceLine" id="cb146-4" data-line-number="4">                );</a>
<a class="sourceLine" id="cb146-5" data-line-number="5"></a>
<a class="sourceLine" id="cb146-6" data-line-number="6"><span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$gene_info</span><span class="st">{gene1}</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb146-7" data-line-number="7"><span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$gene_info</span><span class="st">{gene2}</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb146-8" data-line-number="8"><span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$gene_info</span><span class="st">{gene3}</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<pre><code>## ROS
## WRKY
## WRKY</code></pre>
<div class="sourceCode" id="cb148"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb148-1" data-line-number="1"><span class="kw">my</span> <span class="dt">%gene_info</span> = (<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">ROS</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb148-2" data-line-number="2">                 <span class="kw">&quot;</span><span class="st">gene2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb148-3" data-line-number="3">                 <span class="kw">&quot;</span><span class="st">gene3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span></a>
<a class="sourceLine" id="cb148-4" data-line-number="4">                );</a>
<a class="sourceLine" id="cb148-5" data-line-number="5"><span class="kw">my</span> <span class="dt">@hash_key</span>  = <span class="fu">keys</span> <span class="dt">%gene_info</span>;</a>
<a class="sourceLine" id="cb148-6" data-line-number="6"><span class="dt">@hash_key</span> = <span class="fu">sort</span> <span class="dt">@hash_key</span>;</a>
<a class="sourceLine" id="cb148-7" data-line-number="7"><span class="co">#print &quot;Keys of \%gene_info: &quot;, @hash_key, &quot;\n&quot;;</span></a>
<a class="sourceLine" id="cb148-8" data-line-number="8"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Keys of \%gene_info: </span><span class="kw">&quot;</span>, <span class="fu">join</span>(<span class="kw">&quot;</span><span class="ch">\t</span><span class="kw">&quot;</span>, <span class="fu">sort</span> <span class="fu">keys</span> <span class="dt">%gene_info</span>), <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb148-9" data-line-number="9"><span class="co">#print &quot;Keys of \%gene_info: &quot;, keys %gene_info, &quot;\n&quot;;</span></a></code></pre></div>
<pre><code>## Keys of %gene_info: gene1    gene2   gene3</code></pre>
<div class="sourceCode" id="cb150"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb150-1" data-line-number="1"><span class="kw">my</span> <span class="dt">%gene_info</span> = (<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">ROS</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb150-2" data-line-number="2">                 <span class="kw">&quot;</span><span class="st">gene2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb150-3" data-line-number="3">                 <span class="kw">&quot;</span><span class="st">gene3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span></a>
<a class="sourceLine" id="cb150-4" data-line-number="4">                );</a>
<a class="sourceLine" id="cb150-5" data-line-number="5"><span class="kw">foreach</span>(<span class="fu">sort</span> <span class="fu">keys</span> <span class="dt">%gene_info</span>){</a>
<a class="sourceLine" id="cb150-6" data-line-number="6">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The value of </span><span class="dt">$_</span><span class="st"> is: </span><span class="kw">&quot;</span>, <span class="dt">$gene_info</span>{<span class="dt">$_</span>}, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb150-7" data-line-number="7">}</a></code></pre></div>
<pre><code>## The value of gene1 is: ROS
## The value of gene2 is: WRKY
## The value of gene3 is: WRKY</code></pre>
<div id="sort-keys-numerically-using-sort-ab" class="section level3">
<h3><span class="header-section-number">10.5.1</span> Sort keys numerically using <code>sort {$a&lt;=&gt;$b}</code></h3>
<div class="sourceCode" id="cb152"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb152-1" data-line-number="1"><span class="kw">my</span> <span class="dt">%read_dep</span> = (<span class="kw">&quot;</span><span class="st">1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">100</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb152-2" data-line-number="2">                 <span class="kw">&quot;</span><span class="st">2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">200</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb152-3" data-line-number="3">                 <span class="kw">&quot;</span><span class="st">3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">50</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb152-4" data-line-number="4">                 <span class="kw">&quot;</span><span class="st">10</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">20</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb152-5" data-line-number="5">                );</a>
<a class="sourceLine" id="cb152-6" data-line-number="6"></a>
<a class="sourceLine" id="cb152-7" data-line-number="7"><span class="co">### sort in increasing order</span></a>
<a class="sourceLine" id="cb152-8" data-line-number="8"><span class="kw">foreach</span>(<span class="fu">sort</span> {<span class="dt">$a</span>&lt;=&gt;<span class="dt">$b</span>} <span class="fu">keys</span> <span class="dt">%read_dep</span>){</a>
<a class="sourceLine" id="cb152-9" data-line-number="9">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The value of </span><span class="dt">$_</span><span class="st"> is: </span><span class="kw">&quot;</span>, <span class="dt">$read_dep</span>{<span class="dt">$_</span>}, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb152-10" data-line-number="10">}</a>
<a class="sourceLine" id="cb152-11" data-line-number="11"></a>
<a class="sourceLine" id="cb152-12" data-line-number="12"><span class="fu">print</span> <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">Sort in descending order: </span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb152-13" data-line-number="13"><span class="co">### sort {$b&lt;=&gt;$a} if you want to sort in descending order python</span></a>
<a class="sourceLine" id="cb152-14" data-line-number="14"><span class="kw">foreach</span>(<span class="fu">sort</span> {<span class="dt">$b</span>&lt;=&gt;<span class="dt">$a</span>} <span class="fu">keys</span> <span class="dt">%read_dep</span>){</a>
<a class="sourceLine" id="cb152-15" data-line-number="15">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The value of </span><span class="dt">$_</span><span class="st"> is: </span><span class="kw">&quot;</span>, <span class="dt">$read_dep</span>{<span class="dt">$_</span>}, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb152-16" data-line-number="16">}</a></code></pre></div>
<pre><code>## The value of 1 is: 100
## The value of 2 is: 200
## The value of 3 is: 50
## The value of 10 is: 20
## 
## Sort in descending order: 
## The value of 10 is: 20
## The value of 3 is: 50
## The value of 2 is: 200
## The value of 1 is: 100</code></pre>
</div>
<div id="use-exists-function-on-a-hash" class="section level3">
<h3><span class="header-section-number">10.5.2</span> Use <code>exists</code> function on a hash</h3>
<p>Given an expression that specifies an element of a hash, <code>exists</code> returns true if the specified element in the hash has ever been initialized, even if the corresponding value is undefined.</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb154-1" data-line-number="1"><span class="kw">my</span> <span class="dt">%gene_info</span> = (<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">ROS</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb154-2" data-line-number="2">                 <span class="kw">&quot;</span><span class="st">gene2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>,</a>
<a class="sourceLine" id="cb154-3" data-line-number="3">                 <span class="kw">&quot;</span><span class="st">gene3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span></a>
<a class="sourceLine" id="cb154-4" data-line-number="4">                );</a>
<a class="sourceLine" id="cb154-5" data-line-number="5"><span class="kw">if</span>(<span class="fu">exists</span> <span class="dt">$gene_info</span>{<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>}){</a>
<a class="sourceLine" id="cb154-6" data-line-number="6">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The key &#39;gene1&#39; exists in \%gene_info</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb154-7" data-line-number="7">}</a></code></pre></div>
<pre><code>## The key &#39;gene1&#39; exists in %gene_info</code></pre>
<!--chapter:end:11_PerlVariableOperator.Rmd-->
</div>
</div>
</div>
<div id="control-structure" class="section level1">
<h1><span class="header-section-number">11</span> Control structure</h1>
<p>….
Perl is an iterative language in which control flows from the first statement in the program to the last statement unless something interrupts. Some of the things that can interrupt this linear flow are conditional branches and loop structures. Perl offers approximately a dozen such constructs, which are described below. The basic form will be shown for each followed by a partial example.
….</p>
<div id="for-loop" class="section level2">
<h2><span class="header-section-number">11.1</span> <code>for</code> loop</h2>
<div class="sourceCode" id="cb156"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb156-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@gene_expr</span> = (<span class="dv">2</span>,<span class="dv">6</span>,<span class="dv">8</span>, <span class="dv">9</span>);</a>
<a class="sourceLine" id="cb156-2" data-line-number="2"></a>
<a class="sourceLine" id="cb156-3" data-line-number="3"><span class="co"># This is an example of  for loop </span></a>
<a class="sourceLine" id="cb156-4" data-line-number="4"><span class="kw">for</span>(<span class="kw">my</span> <span class="dt">$i</span>=<span class="dv">0</span>; <span class="dt">$i</span>&lt;<span class="dt">@gene_expr</span>; ++<span class="dt">$i</span>){</a>
<a class="sourceLine" id="cb156-5" data-line-number="5">    <span class="kw">my</span> <span class="dt">$tem_var</span>= <span class="dt">$gene_expr</span>[<span class="dt">$i</span>]/<span class="dv">2</span>;</a>
<a class="sourceLine" id="cb156-6" data-line-number="6"></a>
<a class="sourceLine" id="cb156-7" data-line-number="7">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">At </span><span class="dt">$i</span><span class="st"> place, the number devided by 2 equals: </span><span class="dt">$tem_var</span><span class="ch">\n</span><span class="kw">&quot;</span>;    </a>
<a class="sourceLine" id="cb156-8" data-line-number="8">}</a></code></pre></div>
<pre><code>## At 0 place, the number devided by 2 equals: 1
## At 1 place, the number devided by 2 equals: 3
## At 2 place, the number devided by 2 equals: 4
## At 3 place, the number devided by 2 equals: 4.5</code></pre>
</div>
<div id="foreach-loop" class="section level2">
<h2><span class="header-section-number">11.2</span> <code>foreach</code> loop</h2>
<div class="sourceCode" id="cb158"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb158-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@gene_expr</span> = (<span class="dv">2</span>,<span class="dv">6</span>,<span class="dv">8</span>, <span class="dv">9</span>);</a>
<a class="sourceLine" id="cb158-2" data-line-number="2"></a>
<a class="sourceLine" id="cb158-3" data-line-number="3"><span class="kw">my</span> <span class="dt">$j</span> = <span class="dv">0</span>;</a>
<a class="sourceLine" id="cb158-4" data-line-number="4"><span class="kw">foreach</span>(<span class="dt">@gene_expr</span>){</a>
<a class="sourceLine" id="cb158-5" data-line-number="5">    <span class="kw">my</span> <span class="dt">$tem_var</span>= <span class="dt">$gene_expr</span>[<span class="dt">$j</span>]/<span class="dv">2</span>;</a>
<a class="sourceLine" id="cb158-6" data-line-number="6">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">At </span><span class="dt">$j</span><span class="st"> place, the number devided by 2 equals: </span><span class="dt">$tem_var</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb158-7" data-line-number="7">    <span class="co">#$j = $j +1</span></a>
<a class="sourceLine" id="cb158-8" data-line-number="8">    ++<span class="dt">$j</span>;</a>
<a class="sourceLine" id="cb158-9" data-line-number="9">}</a></code></pre></div>
<pre><code>## At 0 place, the number devided by 2 equals: 1
## At 1 place, the number devided by 2 equals: 3
## At 2 place, the number devided by 2 equals: 4
## At 3 place, the number devided by 2 equals: 4.5</code></pre>
</div>
<div id="while-loop" class="section level2">
<h2><span class="header-section-number">11.3</span> <code>while</code> loop</h2>
<div class="sourceCode" id="cb160"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb160-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@gene_expr</span> = (<span class="dv">2</span>,<span class="dv">6</span>,<span class="dv">8</span>, <span class="dv">9</span>);</a>
<a class="sourceLine" id="cb160-2" data-line-number="2"></a>
<a class="sourceLine" id="cb160-3" data-line-number="3"><span class="kw">my</span> <span class="dt">$k</span> = <span class="dv">0</span>;</a>
<a class="sourceLine" id="cb160-4" data-line-number="4"><span class="kw">while</span>(<span class="dt">$k</span>&lt;<span class="dt">@gene_expr</span>){</a>
<a class="sourceLine" id="cb160-5" data-line-number="5">    <span class="kw">my</span> <span class="dt">$tem_var</span>= <span class="dt">$gene_expr</span>[<span class="dt">$k</span>]/<span class="dv">2</span>;</a>
<a class="sourceLine" id="cb160-6" data-line-number="6">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">At </span><span class="dt">$k</span><span class="st"> place, the number devided by 2 equals: </span><span class="dt">$tem_var</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb160-7" data-line-number="7">    ++<span class="dt">$k</span>;</a>
<a class="sourceLine" id="cb160-8" data-line-number="8">}</a></code></pre></div>
<pre><code>## At 0 place, the number devided by 2 equals: 1
## At 1 place, the number devided by 2 equals: 3
## At 2 place, the number devided by 2 equals: 4
## At 3 place, the number devided by 2 equals: 4.5</code></pre>
</div>
<div id="statement-if-else" class="section level2">
<h2><span class="header-section-number">11.4</span> Statement if-else</h2>
<div class="sourceCode" id="cb162"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb162-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb162-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb162-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb162-4" data-line-number="4"></a>
<a class="sourceLine" id="cb162-5" data-line-number="5"><span class="kw">my</span> <span class="dt">@gene_exp_lev</span> = (<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">9</span>, <span class="dv">10</span>);</a>
<a class="sourceLine" id="cb162-6" data-line-number="6"></a>
<a class="sourceLine" id="cb162-7" data-line-number="7"><span class="co"># for (my $i = 0; $i &lt; @gene_exp_lev; $i++) {</span></a>
<a class="sourceLine" id="cb162-8" data-line-number="8"><span class="kw">for</span> (<span class="kw">my</span> <span class="dt">$i</span> = <span class="dv">0</span>; <span class="dt">$i</span> &lt; <span class="fu">scalar</span> <span class="dt">@gene_exp_lev</span>; <span class="dt">$i</span>++) {</a>
<a class="sourceLine" id="cb162-9" data-line-number="9">    <span class="kw">if</span> (<span class="dt">$gene_exp_lev</span>[<span class="dt">$i</span>] &gt; <span class="dv">4</span>) {</a>
<a class="sourceLine" id="cb162-10" data-line-number="10">        <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Index </span><span class="dt">$i</span><span class="st">: </span><span class="dt">$gene_exp_lev</span><span class="st">[</span><span class="dt">$i</span><span class="st">]</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb162-11" data-line-number="11">    }</a>
<a class="sourceLine" id="cb162-12" data-line-number="12">}</a></code></pre></div>
<pre><code>Index 1: 5
Index 4: 9
Index 5: 10</code></pre>
</div>
<div id="operator-last-and-next" class="section level2">
<h2><span class="header-section-number">11.5</span> Operator <code>last</code> and <code>next</code></h2>
</div>
<div id="operator-redo" class="section level2">
<h2><span class="header-section-number">11.6</span> Operator <code>redo</code></h2>
<p>Before we use <code>redo</code>, first let’s see what is <code>BLOCK</code> in Perl. In Perl, a <code>BLOCK</code> by itself (labeled or not) is semantically equivalent to a loop that executes once. Thus you can use any of the loop control statements in it to leave or restart the block.</p>
<p>The redo command restarts the loop block without evaluating the conditional again.</p>
<p>Let’s say we have some postitions on the genome where a few</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb164-1" data-line-number="1"><span class="kw">my</span> <span class="dt">@read_depth</span> = (<span class="dv">8</span>, <span class="dv">9</span>, <span class="dv">10</span>, <span class="dv">7</span>, <span class="dv">10</span>, <span class="dv">7</span>, <span class="dv">7</span>);</a></code></pre></div>
<!--chapter:end:13_PerlControlStructure.Rmd-->
</div>
</div>
<div id="string-manipulation" class="section level1">
<h1><span class="header-section-number">12</span> String manipulation</h1>
<div id="string-concatenation" class="section level2">
<h2><span class="header-section-number">12.1</span> String concatenation</h2>
<p>Dot (<code>.</code>) can be used to concatenate two strings together.</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb165-1" data-line-number="1"><span class="co"># concatenate two strings together and assing to $z</span></a>
<a class="sourceLine" id="cb165-2" data-line-number="2"><span class="dt">$z</span> = <span class="dt">$x</span> . <span class="dt">$y</span>;</a>
<a class="sourceLine" id="cb165-3" data-line-number="3"><span class="co"># Append $y to $x</span></a>
<a class="sourceLine" id="cb165-4" data-line-number="4"><span class="dt">$x</span> = <span class="dt">$x</span> . <span class="dt">$y</span>;</a>
<a class="sourceLine" id="cb165-5" data-line-number="5"><span class="co"># Append $y to $x</span></a>
<a class="sourceLine" id="cb165-6" data-line-number="6"><span class="dt">$x</span> .= <span class="dt">$y</span>;</a></code></pre></div>
<p>A more convenient way is to use operator <code>.=</code> to append one variable to another.</p>
<p>As is any other assignments in Perl, if you see an assignment written this way <code>$x = $x</code> op expr, where op stands for any operator and expr stands for the rest of the statement, you can make a shorter version by moving the op to the front of the assignment, e.g., <code>$x op= expr</code>. The string concatenation operator <code>.</code> is just one possible op among many others such as <code>+</code>, <code>-</code>, <code>*</code> and <code>/</code>.</p>
<div class="sourceCode" id="cb166"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb166-1" data-line-number="1"><span class="kw">my</span> <span class="dt">$x</span> = <span class="dv">5</span>;</a>
<a class="sourceLine" id="cb166-2" data-line-number="2"><span class="kw">my</span> <span class="dt">$y</span> = <span class="dv">6</span>;</a>
<a class="sourceLine" id="cb166-3" data-line-number="3"><span class="co"># Add $y to $x</span></a>
<a class="sourceLine" id="cb166-4" data-line-number="4"><span class="dt">$x</span> = <span class="dt">$x</span> + <span class="dt">$y</span>;</a>
<a class="sourceLine" id="cb166-5" data-line-number="5"><span class="co"># Add $y to $x</span></a>
<a class="sourceLine" id="cb166-6" data-line-number="6"><span class="dt">$x</span> += <span class="dt">$y</span>;</a></code></pre></div>
</div>
<div id="substring-extraction" class="section level2">
<h2><span class="header-section-number">12.2</span> Substring extraction</h2>
</div>
<div id="substring-search" class="section level2">
<h2><span class="header-section-number">12.3</span> Substring search</h2>
<p>Index</p>
</div>
<div id="split-string" class="section level2">
<h2><span class="header-section-number">12.4</span> Split String</h2>
<p>split</p>
<p>join</p>
</div>
<div id="regular-expression" class="section level2">
<h2><span class="header-section-number">12.5</span> Regular expression</h2>
<div id="pattern-matching" class="section level3">
<h3><span class="header-section-number">12.5.1</span> Pattern matching</h3>
</div>
<div id="pattern-substitution" class="section level3">
<h3><span class="header-section-number">12.5.2</span> Pattern substitution</h3>
</div>
<div id="modifiers-to-pattern-matching-and-substitution" class="section level3">
<h3><span class="header-section-number">12.5.3</span> Modifiers to pattern matching and substitution</h3>
</div>
<div id="greedy-or-non-greedy" class="section level3">
<h3><span class="header-section-number">12.5.4</span> Greedy or non-greedy</h3>
<!--chapter:end:14_StringManipulationRegExp.Rmd-->
</div>
</div>
</div>
<div id="input-and-output-in-perl" class="section level1">
<h1><span class="header-section-number">13</span> Input and output in Perl</h1>
<div id="input" class="section level2">
<h2><span class="header-section-number">13.1</span> Input</h2>
<div id="standard-input" class="section level3">
<h3><span class="header-section-number">13.1.1</span> Standard input</h3>
</div>
<div id="input-from-a-file-on-the-disk" class="section level3">
<h3><span class="header-section-number">13.1.2</span> Input from a file on the disk</h3>
<div class="sourceCode" id="cb167"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb167-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb167-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb167-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb167-4" data-line-number="4"></a>
<a class="sourceLine" id="cb167-5" data-line-number="5"><span class="kw">my</span> <span class="dt">$fasta</span> = <span class="kw">&quot;</span><span class="st">./data/test_ref.fa</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb167-6" data-line-number="6"></a>
<a class="sourceLine" id="cb167-7" data-line-number="7"><span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="dt">$!</span><span class="st">: </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb167-8" data-line-number="8"><span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$line</span> = <span class="kw">&lt;FASTA&gt;</span>){</a>
<a class="sourceLine" id="cb167-9" data-line-number="9">    <span class="fu">chomp</span> <span class="dt">$line</span>;</a>
<a class="sourceLine" id="cb167-10" data-line-number="10">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$line</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb167-11" data-line-number="11">}</a>
<a class="sourceLine" id="cb167-12" data-line-number="12"><span class="fu">close</span> FASTA;</a></code></pre></div>
<div class="sourceCode" id="cb168"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb168-1" data-line-number="1"><span class="fu">perl</span> code_perl/open_file.pl</a></code></pre></div>
<pre><code>## &gt;chr1
## ACGCTAGCTAGTCAGTCGATCGT
## CGTAGCTAGCTAG
## &gt;chr2
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTA
## &gt;chr3
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTA</code></pre>
</div>
<div id="special-variable-_" class="section level3">
<h3><span class="header-section-number">13.1.3</span> Special variable <code>$_</code></h3>
<div class="sourceCode" id="cb170"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb170-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb170-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb170-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb170-4" data-line-number="4"></a>
<a class="sourceLine" id="cb170-5" data-line-number="5"><span class="kw">my</span> <span class="dt">$fasta</span> = <span class="kw">&quot;</span><span class="st">./data/test_ref.fa</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb170-6" data-line-number="6"></a>
<a class="sourceLine" id="cb170-7" data-line-number="7"><span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="dt">$!</span><span class="st">: </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb170-8" data-line-number="8"><span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$line</span> = <span class="kw">&lt;FASTA&gt;</span>){</a>
<a class="sourceLine" id="cb170-9" data-line-number="9">    <span class="fu">chomp</span> <span class="dt">$line</span>;</a>
<a class="sourceLine" id="cb170-10" data-line-number="10">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$line</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb170-11" data-line-number="11">}</a>
<a class="sourceLine" id="cb170-12" data-line-number="12"><span class="fu">close</span> FASTA;</a></code></pre></div>
<p>When you are using <code>while</code> and <code>foreach</code>, the content will be assigned to <code>$_</code> if you don’t assign it explicitly to any variable.</p>
<p>For some build-in functions (<code>chomp</code>, <code>length</code>, <code>split</code> and <code>print</code>),</p>
<p>I do NOT suggest you to use <code>$_</code> because <code>$_</code> will reduce the readerbility of the code.</p>
<p>In Perl, $_ is a powerful
In Perl, several functions and operators use this variable as a default, in case no parameter is explicitly used. In general, I’d say you should NOT see $_ in real code. I think the whole point of $_ is that you don’t have to write it explicitly.</p>
</div>
<div id="the-input-record-separator" class="section level3">
<h3><span class="header-section-number">13.1.4</span> The input record separator: <code>$/</code></h3>
<p><code>$/</code> is the input record separator. By default, <code>$/</code> equals newline (<code>\n</code>).</p>
<p>You could actually change the value of <code>$/</code>. For example, by assigning the greate-than ‘&gt;’ like this: <code>$/ = &quot;&gt;&quot;</code>;. Then every call to the read-line operator <code>$chunk = &lt;FILEHANDLE&gt;</code> will read in all the characters up-to and including the first ‘&gt;’.</p>
<p>We could also assign longer strings to <code>$/</code> and then that would be the input record separator.</p>
<div id="how-to-process-fasta-file-by-changing" class="section level4">
<h4><span class="header-section-number">13.1.4.1</span> How to process <code>fasta</code> file by changing <code>$/</code></h4>
<p>FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes (A, T, G, C, etc.).</p>
<p>In <code>fasta</code> format, each sequence begins with a single-line description, followed by lines of sequence data. The description line is distinguished from the sequence data by a greater-than (“&gt;”) symbol in the first column.</p>
<p>Here is an example of fasta file (Figure @ref(fig:fastaPYL)).</p>
<p>(ref:fastaPYL) Protein sequence of PYL10 in Arabidopsis.</p>
<div class="figure">
<img src="figures/fasta_pyl10_ara.png" alt="(ref:fastaPYL)"  />
<p class="caption">
(#fig:fastaPYL)(ref:fastaPYL)
</p>
</div>
<div class="sourceCode" id="cb171"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb171-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb171-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb171-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb171-4" data-line-number="4"></a>
<a class="sourceLine" id="cb171-5" data-line-number="5"><span class="kw">my</span> <span class="dt">$fasta</span> = <span class="kw">&quot;</span><span class="st">./data/test_ref.fa</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb171-6" data-line-number="6"></a>
<a class="sourceLine" id="cb171-7" data-line-number="7"><span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="dt">$!</span><span class="st">: </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb171-8" data-line-number="8"><span class="co"># Set the input record separator</span></a>
<a class="sourceLine" id="cb171-9" data-line-number="9"><span class="dt">$/</span> = <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">&gt;</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb171-10" data-line-number="10"><span class="co"># Print header</span></a>
<a class="sourceLine" id="cb171-11" data-line-number="11"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Chrom</span><span class="ch">\t</span><span class="st">Length</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb171-12" data-line-number="12"><span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$chunk</span> = <span class="kw">&lt;FASTA&gt;</span>){</a>
<a class="sourceLine" id="cb171-13" data-line-number="13">    <span class="co"># Remove &gt; in the first chunk</span></a>
<a class="sourceLine" id="cb171-14" data-line-number="14">    <span class="dt">$chunk</span> =~ <span class="kw">s/</span><span class="ch">^</span><span class="ot">&gt;</span><span class="kw">//</span>;</a>
<a class="sourceLine" id="cb171-15" data-line-number="15">    <span class="co"># Remove \n&gt; from the end of the chunk</span></a>
<a class="sourceLine" id="cb171-16" data-line-number="16">    <span class="fu">chomp</span> <span class="dt">$chunk</span>;</a>
<a class="sourceLine" id="cb171-17" data-line-number="17">    <span class="co"># Assign ID and seq to two variables</span></a>
<a class="sourceLine" id="cb171-18" data-line-number="18">    <span class="kw">my</span> (<span class="dt">$chrom</span>, <span class="dt">$seq</span>) = <span class="fu">split</span>(<span class="kw">/</span><span class="bn">\n</span><span class="kw">/</span>, <span class="dt">$chunk</span>, <span class="dv">2</span>);</a>
<a class="sourceLine" id="cb171-19" data-line-number="19">    <span class="co"># Remove \n in the string</span></a>
<a class="sourceLine" id="cb171-20" data-line-number="20">    <span class="dt">$seq</span> =~ <span class="kw">s/</span><span class="bn">\n</span><span class="kw">//g</span>;</a>
<a class="sourceLine" id="cb171-21" data-line-number="21">    <span class="co"># Calculate the length of sequence.</span></a>
<a class="sourceLine" id="cb171-22" data-line-number="22">    <span class="kw">my</span> <span class="dt">$seq_length</span> = <span class="fu">length</span> <span class="dt">$seq</span>;</a>
<a class="sourceLine" id="cb171-23" data-line-number="23">    <span class="co"># Print ID and length of the sequence.</span></a>
<a class="sourceLine" id="cb171-24" data-line-number="24">    <span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$chrom</span><span class="ch">\t</span><span class="dt">$seq_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb171-25" data-line-number="25">}</a>
<a class="sourceLine" id="cb171-26" data-line-number="26"><span class="fu">close</span> FASTA;</a></code></pre></div>
<p>You can consider the content in the fasta file as a string (Figure @ref(fig:fastaString)) sperarated by <code>\n&gt;</code>.</p>
<p>We set the input record separator <code>$/</code> as <code>\n&gt;</code>.</p>
<p>In the first cycle of the <code>while</code> loop, the first chunk is <code>&gt;chr1\nACGCTAGCTAGTCAGTCGATCGT\nCGTAGCTAGCTAG\n&gt;</code>.</p>
<p>You should notice that there is a leading <code>&gt;</code> in the first chunk. So you need to use <code>s/^&gt;//</code> to remove the leading <code>&gt;</code>.</p>
<p>To remove the trailing <code>\n&gt;</code>, you can use <code>chomp $chunk;</code>. Then the string is:</p>
<pre><code>chr1\nACGCTAGCTAGTCAGTCGATCGT\nCGTAGCTAGCTAG</code></pre>
<p>The string before first <code>\n</code> is the sequence ID. The string after the first <code>\n</code> is the sequence. To get the sequnce ID and sequnce and assign to two variables, you can use <code>split(/\n/, $chunk, 2)</code>. Here LIMIT is set to 2, it represents the maximum number (here 2) of fields into which the EXPR may be split.</p>
<p>The goal of this code is to output the sequence IDs and their lengths. Currently the string of the variable (<code>$seq</code>) is <code>ACGCTAGCTAGTCAGTCGATCGT\nCGTAGCTAGCTAG</code>. To do this, you first need to remove <code>\n</code> in the sequence by using <code>$seq =~ s/\n//g;</code>. Now the string of the variable (<code>$seq</code>) is <code>ACGCTAGCTAGTCAGTCGATCGTCGTAGCTAGCTAG</code>. Then built-in function <code>length</code> can be used to calculate the length of the sequence. The function <code>print</code> is used to output the sequnce ID and the length of the sequence.</p>
<p>In the second cycles of the <code>while</code> loop, the chunk is
<code>chr2\nCTGCGGGCTAAATCGATCGATCG\nGTACGTACGAGCTAGCTA\n&gt;</code>.</p>
<p>Unlike the first chunk, there is no leading <code>&gt;</code> in the second chunk. So the code <code>s/^&gt;//</code> will do nothing for the second chunk.</p>
<p>The following steps is the same as you can do for the first cycle.</p>
<p>In the third cycle of the <code>while</code> loop, you can do the same thing on the third sequence.</p>
<p>(ref:fastaString) Fasta string.</p>
<div class="figure">
<img src="figures/fasta_string.png" alt="(ref:fastaString)"  />
<p class="caption">
(#fig:fastaString)(ref:fastaString)
</p>
</div>
<div class="sourceCode" id="cb173"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb173-1" data-line-number="1"><span class="fu">perl</span> code_perl/fa_seq_len.pl</a></code></pre></div>
<pre><code>## Chrom    Length
## chr1 36
## chr2 41
## chr3 41</code></pre>
</div>
</div>
</div>
<div id="output-to-a-file-on-the-disk" class="section level2">
<h2><span class="header-section-number">13.2</span> Output to a file on the disk</h2>
<div class="sourceCode" id="cb175"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb175-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb175-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb175-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb175-4" data-line-number="4"></a>
<a class="sourceLine" id="cb175-5" data-line-number="5"><span class="kw">my</span> <span class="dt">$fasta</span>   = <span class="kw">&quot;</span><span class="st">./data/test_ref.fa</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb175-6" data-line-number="6"><span class="kw">my</span> <span class="dt">$out_len</span> = <span class="kw">&quot;</span><span class="st">./data/test_ref_len.txt</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb175-7" data-line-number="7"></a>
<a class="sourceLine" id="cb175-8" data-line-number="8"><span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">Can&#39;t open file for reading: </span><span class="dt">$!</span><span class="st"> </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb175-9" data-line-number="9"><span class="fu">open</span> OUT, <span class="kw">&quot;</span><span class="st">+&gt;</span><span class="dt">$out_len</span><span class="kw">&quot;</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">Can&#39;t open file for writing: </span><span class="dt">$!</span><span class="st"> </span><span class="dt">$out_len</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb175-10" data-line-number="10"><span class="co"># Set the input record separator</span></a>
<a class="sourceLine" id="cb175-11" data-line-number="11"><span class="dt">$/</span> = <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">&gt;</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb175-12" data-line-number="12"><span class="co"># Print header</span></a>
<a class="sourceLine" id="cb175-13" data-line-number="13"><span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="st">Chrom</span><span class="ch">\t</span><span class="st">Length</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb175-14" data-line-number="14"><span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$chunk</span> = <span class="kw">&lt;FASTA&gt;</span>){</a>
<a class="sourceLine" id="cb175-15" data-line-number="15">    <span class="co"># Remove &gt; in the first chunk</span></a>
<a class="sourceLine" id="cb175-16" data-line-number="16">    <span class="dt">$chunk</span> =~ <span class="kw">s/</span><span class="ch">^</span><span class="ot">&gt;</span><span class="kw">//</span>;</a>
<a class="sourceLine" id="cb175-17" data-line-number="17">    <span class="co"># Remove \n&gt; from the end of the chunk</span></a>
<a class="sourceLine" id="cb175-18" data-line-number="18">    <span class="fu">chomp</span> <span class="dt">$chunk</span>;</a>
<a class="sourceLine" id="cb175-19" data-line-number="19">    <span class="co"># Assign ID and seq to two variables</span></a>
<a class="sourceLine" id="cb175-20" data-line-number="20">    <span class="kw">my</span> (<span class="dt">$chrom</span>, <span class="dt">$seq</span>) = <span class="fu">split</span>(<span class="kw">/</span><span class="bn">\n</span><span class="kw">/</span>, <span class="dt">$chunk</span>, <span class="dv">2</span>);</a>
<a class="sourceLine" id="cb175-21" data-line-number="21">    <span class="co"># Remove \n in the string</span></a>
<a class="sourceLine" id="cb175-22" data-line-number="22">    <span class="dt">$seq</span> =~ <span class="kw">s/</span><span class="bn">\n</span><span class="kw">//g</span>;</a>
<a class="sourceLine" id="cb175-23" data-line-number="23">    <span class="co"># Calculate the length of sequence.</span></a>
<a class="sourceLine" id="cb175-24" data-line-number="24">    <span class="kw">my</span> <span class="dt">$seq_length</span> = <span class="fu">length</span> <span class="dt">$seq</span>;</a>
<a class="sourceLine" id="cb175-25" data-line-number="25">    <span class="co"># Print ID and length of the sequence.</span></a>
<a class="sourceLine" id="cb175-26" data-line-number="26">    <span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="dt">$chrom</span><span class="ch">\t</span><span class="dt">$seq_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb175-27" data-line-number="27">}</a>
<a class="sourceLine" id="cb175-28" data-line-number="28"><span class="fu">close</span> FASTA;</a>
<a class="sourceLine" id="cb175-29" data-line-number="29"><span class="fu">close</span> OUT;</a></code></pre></div>
<div class="sourceCode" id="cb176"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb176-1" data-line-number="1"><span class="fu">perl</span> code_perl/fa_seq_len_out.pl</a>
<a class="sourceLine" id="cb176-2" data-line-number="2"><span class="fu">cat</span> ./data/test_ref_len.txt</a></code></pre></div>
<pre><code>## Chrom    Length
## chr1 36
## chr2 41
## chr3 41</code></pre>
</div>
<div id="arguments-from-command-line" class="section level2">
<h2><span class="header-section-number">13.3</span> Arguments from command line</h2>
<p>Imagine you have 5 fasta files and you want to calculate sequence lengths for all the 5 files, if you modify the file each time you run the code, it will be very tedious.</p>
<p>Perls provides an array called <code>@ARGV</code>. <code>@ARGV</code> holds all the arguments from the command line. The first one will be <code>$ARGV[0]</code>.</p>
<p><code>@ARGV</code> will automatically hold all the arguments. If no arguments are provided, <code>@ARGV</code> will be empty.</p>
<p>The following example shows you how it looks like in real code.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb178-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb178-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb178-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb178-4" data-line-number="4"></a>
<a class="sourceLine" id="cb178-5" data-line-number="5"><span class="kw">my</span> (<span class="dt">$fasta</span>, <span class="dt">$out_len</span>) = <span class="dt">@ARGV</span>; </a>
<a class="sourceLine" id="cb178-6" data-line-number="6"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First arguments \$ARGV[0] : </span><span class="dt">$ARGV</span><span class="st">[0]</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb178-7" data-line-number="7"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second arguments \$ARGV[0]: </span><span class="dt">$ARGV</span><span class="st">[1]</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb178-8" data-line-number="8"></a>
<a class="sourceLine" id="cb178-9" data-line-number="9"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The variable \$fasta:   </span><span class="dt">$fasta</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb178-10" data-line-number="10"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The variable \$out_len: </span><span class="dt">$out_len</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a></code></pre></div>
<div class="sourceCode" id="cb179"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb179-1" data-line-number="1"><span class="fu">perl</span> code_perl/test_argv.pl test_ref.fa test_ref_len.txt</a></code></pre></div>
<pre><code>## First arguments $ARGV[0] : test_ref.fa
## Second arguments $ARGV[0]: test_ref_len.txt
## The variable $fasta:   test_ref.fa
## The variable $out_len: test_ref_len.txt</code></pre>
<p>In the above example, the first argument is <code>$ARGV[0]</code> which stores the file name: test_ref.fa. The second one stores the file name: test_ref_len.txt. It’s highly recommended to assign the values in <code>@ARGV</code> to some variables that the readers (including yourself) can understand the variables based on the names.</p>
<p>Here I’ll show you how to use ARGV using an example. The code below will show you what to do if you want to filter a fasta file based on the sequence length.</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb181-1" data-line-number="1"><span class="kw">#!/usr/bin/perl</span></a>
<a class="sourceLine" id="cb181-2" data-line-number="2"><span class="fu">use</span> <span class="kw">warnings</span>;</a>
<a class="sourceLine" id="cb181-3" data-line-number="3"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb181-4" data-line-number="4"></a>
<a class="sourceLine" id="cb181-5" data-line-number="5"><span class="kw">my</span> (<span class="dt">$fasta</span>, <span class="dt">$out_len</span>) = <span class="dt">@ARGV</span>; </a>
<a class="sourceLine" id="cb181-6" data-line-number="6"></a>
<a class="sourceLine" id="cb181-7" data-line-number="7"><span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">Can&#39;t open file for reading: </span><span class="dt">$!</span><span class="st"> </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb181-8" data-line-number="8"><span class="fu">open</span> OUT, <span class="kw">&quot;</span><span class="st">+&gt;</span><span class="dt">$out_len</span><span class="kw">&quot;</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">Can&#39;t open file for writing: </span><span class="dt">$!</span><span class="st"> </span><span class="dt">$out_len</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb181-9" data-line-number="9"><span class="co"># Set the input record separator</span></a>
<a class="sourceLine" id="cb181-10" data-line-number="10"><span class="dt">$/</span> = <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">&gt;</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb181-11" data-line-number="11"><span class="co"># Print header</span></a>
<a class="sourceLine" id="cb181-12" data-line-number="12"><span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="st">The input file name is: </span><span class="dt">$fasta</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb181-13" data-line-number="13"><span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="st">Chrom</span><span class="ch">\t</span><span class="st">Length</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb181-14" data-line-number="14"><span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$chunk</span> = <span class="kw">&lt;FASTA&gt;</span>){</a>
<a class="sourceLine" id="cb181-15" data-line-number="15">    <span class="co"># Remove &gt; in the first chunk</span></a>
<a class="sourceLine" id="cb181-16" data-line-number="16">    <span class="dt">$chunk</span> =~ <span class="kw">s/</span><span class="ch">^</span><span class="ot">&gt;</span><span class="kw">//</span>;</a>
<a class="sourceLine" id="cb181-17" data-line-number="17">    <span class="co"># Remove \n&gt; from the end of the chunk</span></a>
<a class="sourceLine" id="cb181-18" data-line-number="18">    <span class="fu">chomp</span> <span class="dt">$chunk</span>;</a>
<a class="sourceLine" id="cb181-19" data-line-number="19">    <span class="co"># Assign ID and seq to two variables</span></a>
<a class="sourceLine" id="cb181-20" data-line-number="20">    <span class="kw">my</span> (<span class="dt">$chrom</span>, <span class="dt">$seq</span>) = <span class="fu">split</span>(<span class="kw">/</span><span class="bn">\n</span><span class="kw">/</span>, <span class="dt">$chunk</span>, <span class="dv">2</span>);</a>
<a class="sourceLine" id="cb181-21" data-line-number="21">    <span class="co"># Remove \n in the string</span></a>
<a class="sourceLine" id="cb181-22" data-line-number="22">    <span class="dt">$seq</span> =~ <span class="kw">s/</span><span class="bn">\n</span><span class="kw">//g</span>;</a>
<a class="sourceLine" id="cb181-23" data-line-number="23">    <span class="co"># Calculate the length of sequence.</span></a>
<a class="sourceLine" id="cb181-24" data-line-number="24">    <span class="kw">my</span> <span class="dt">$seq_length</span> = <span class="fu">length</span> <span class="dt">$seq</span>;</a>
<a class="sourceLine" id="cb181-25" data-line-number="25">    <span class="co"># Print ID and length of the sequence.</span></a>
<a class="sourceLine" id="cb181-26" data-line-number="26">    <span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="dt">$chrom</span><span class="ch">\t</span><span class="dt">$seq_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb181-27" data-line-number="27">}</a>
<a class="sourceLine" id="cb181-28" data-line-number="28"><span class="fu">close</span> FASTA;</a>
<a class="sourceLine" id="cb181-29" data-line-number="29"><span class="fu">close</span> OUT;</a></code></pre></div>
<p>For example if you want to remove the sequences that is shorter than 30 bps, you can use the following example.</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb182-1" data-line-number="1"><span class="bu">printf</span> <span class="st">&quot;Before filtering:\n&quot;</span></a>
<a class="sourceLine" id="cb182-2" data-line-number="2"><span class="fu">cat</span> ./data/test_ref2.fa</a>
<a class="sourceLine" id="cb182-3" data-line-number="3"><span class="fu">perl</span> code_perl/fa_seq_len_out_argv_fil.pl ./data/test_ref2.fa 30 ./data/test_ref2_30.fa</a>
<a class="sourceLine" id="cb182-4" data-line-number="4"></a>
<a class="sourceLine" id="cb182-5" data-line-number="5"><span class="bu">printf</span> <span class="st">&quot;\nAfter filtering: \n&quot;</span></a>
<a class="sourceLine" id="cb182-6" data-line-number="6"><span class="fu">cat</span> ./data/test_ref2_30.fa </a></code></pre></div>
<pre><code>## Before filtering:
## &gt;chr1
## ACGCTAGCTAGTCAGTCGATCGT
## CGTAGCTAGCTAG
## &gt;chr2
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTAA
## &gt;chr3
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAG
## &gt;chr4
## CTGCGGGCTAAATCAGCTAA
## &gt;chr5
## CTGCTCGATCGATCGACGAGCTA
## GCTA
## After filtering: 
## The input file name is: ./data/test_ref2.fa
## Chrom    Length
## &gt;chr1 36
## ACGCTAGCTAGTCAGTCGATCGTCGTAGCTAGCTAG
## &gt;chr2 42
## CTGCGGGCTAAATCGATCGATCGGTACGTACGAGCTAGCTAA
## &gt;chr3 34
## CTGCGGGCTAAATCGATCGATCGGTACGTACGAG</code></pre>
<p>If you have multiple files, you can just change the file names in the command line. It’s very handy. You can also change the length cutoff in the command line.</p>
<!--chapter:end:15_PerlInputOutput.Rmd-->
</div>
</div>
<div id="practical-perl-program" class="section level1">
<h1><span class="header-section-number">14</span> Practical Perl program</h1>
<div id="add-annotation-information-to-deseq2-results" class="section level2">
<h2><span class="header-section-number">14.1</span> Add annotation information to DESeq2 results</h2>
<p>Imaging we have a table that stores that differentially expressed genes information.</p>
<p>It includes three columns (tab-delimited):</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb184-1" data-line-number="1"><span class="fu">cat</span> data/DEG_list.txt</a></code></pre></div>
<pre><code>## #gene_id log2fc  p-val
## gene1    2   0.01
## gene2    3   0.04
## gene3    -2  0.06
## gene4    -8  0.001</code></pre>
<p>We have another table which has the annotation information of each gene:</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb186-1" data-line-number="1"><span class="fu">cat</span> data/gene_annotation.txt</a></code></pre></div>
<pre><code>## gene1    ROS
## gene2    WRKY
## gene3    ZmCCT</code></pre>
</div>
<div id="merge-overlap-genomic-regions" class="section level2">
<h2><span class="header-section-number">14.2</span> Merge overlap genomic regions</h2>
<div class="sourceCode" id="cb188"><pre class="sourceCode perl"><code class="sourceCode perl"><a class="sourceLine" id="cb188-1" data-line-number="1"><span class="kw">#!/usr/bin/perl -w</span></a>
<a class="sourceLine" id="cb188-2" data-line-number="2"><span class="fu">use</span> <span class="kw">strict</span>;</a>
<a class="sourceLine" id="cb188-3" data-line-number="3"></a>
<a class="sourceLine" id="cb188-4" data-line-number="4"><span class="kw">my</span> (<span class="dt">$DMR</span>, <span class="dt">$out</span>) = <span class="dt">@ARGV</span>;</a>
<a class="sourceLine" id="cb188-5" data-line-number="5"></a>
<a class="sourceLine" id="cb188-6" data-line-number="6"><span class="fu">open</span> DMR, <span class="dt">$DMR</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">File not found: </span><span class="dt">$!</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb188-7" data-line-number="7"><span class="fu">open</span> OUT, <span class="kw">&quot;</span><span class="st">+&gt;</span><span class="dt">$out</span><span class="kw">&quot;</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="dt">$!</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb188-8" data-line-number="8"></a>
<a class="sourceLine" id="cb188-9" data-line-number="9"><span class="co">#read  a  line,</span></a>
<a class="sourceLine" id="cb188-10" data-line-number="10"><span class="co"># go to next line</span></a>
<a class="sourceLine" id="cb188-11" data-line-number="11"><span class="co">#     check overlap</span></a>
<a class="sourceLine" id="cb188-12" data-line-number="12"><span class="co">#      if overlap: merge two regions. Then go to the next line.</span></a>
<a class="sourceLine" id="cb188-13" data-line-number="13"><span class="co">#      if not overlap: print whatever we have now; then start from a line</span></a>
<a class="sourceLine" id="cb188-14" data-line-number="14"><span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$region</span> = <span class="kw">&lt;DMR&gt;</span>){</a>
<a class="sourceLine" id="cb188-15" data-line-number="15">    <span class="fu">chomp</span> <span class="dt">$region</span>;</a>
<a class="sourceLine" id="cb188-16" data-line-number="16">    <span class="kw">my</span> (<span class="dt">$chrom1</span>, <span class="dt">$start1</span>, <span class="dt">$end1</span>) = <span class="fu">split</span>(<span class="kw">/</span><span class="ot">\t</span><span class="kw">/</span>, <span class="dt">$region</span>);</a>
<a class="sourceLine" id="cb188-17" data-line-number="17">    BLOCK:{</a>
<a class="sourceLine" id="cb188-18" data-line-number="18">        <span class="kw">my</span> <span class="dt">$new_region</span> = <span class="kw">&lt;DMR&gt;</span>;</a>
<a class="sourceLine" id="cb188-19" data-line-number="19"></a>
<a class="sourceLine" id="cb188-20" data-line-number="20">        <span class="kw">if</span>(!<span class="dt">$new_region</span>){</a>
<a class="sourceLine" id="cb188-21" data-line-number="21">            <span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="dt">$chrom1</span><span class="ch">\t</span><span class="dt">$start1</span><span class="ch">\t</span><span class="dt">$end1</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb188-22" data-line-number="22">        }<span class="kw">else</span>{</a>
<a class="sourceLine" id="cb188-23" data-line-number="23">            <span class="fu">chomp</span> <span class="dt">$new_region</span>;</a>
<a class="sourceLine" id="cb188-24" data-line-number="24">            <span class="kw">my</span> (<span class="dt">$chrom2</span>, <span class="dt">$start2</span>, <span class="dt">$end2</span>) = <span class="fu">split</span>(<span class="kw">/</span><span class="ot">\t</span><span class="kw">/</span>, <span class="dt">$new_region</span>);</a>
<a class="sourceLine" id="cb188-25" data-line-number="25">             <span class="co">## start1===============end1</span></a>
<a class="sourceLine" id="cb188-26" data-line-number="26">             <span class="co">##           start2=====================end2</span></a>
<a class="sourceLine" id="cb188-27" data-line-number="27">             <span class="co">## start1===============================new_end1</span></a>
<a class="sourceLine" id="cb188-28" data-line-number="28">            <span class="kw">if</span>(<span class="dt">$chrom2</span> <span class="kw">eq</span> <span class="dt">$chrom1</span> &amp;&amp; <span class="dt">$start2</span> &gt;= <span class="dt">$start1</span> &amp;&amp; <span class="dt">$start2</span> &lt;=<span class="dt">$end1</span>){</a>
<a class="sourceLine" id="cb188-29" data-line-number="29">                <span class="dt">$end1</span> = <span class="dt">$end2</span>;</a>
<a class="sourceLine" id="cb188-30" data-line-number="30">                <span class="fu">redo</span> BLOCK;</a>
<a class="sourceLine" id="cb188-31" data-line-number="31">            }<span class="kw">else</span>{</a>
<a class="sourceLine" id="cb188-32" data-line-number="32">                <span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="dt">$chrom1</span><span class="ch">\t</span><span class="dt">$start1</span><span class="ch">\t</span><span class="dt">$end1</span><span class="ch">\n</span><span class="kw">&quot;</span>;</a>
<a class="sourceLine" id="cb188-33" data-line-number="33">                (<span class="dt">$chrom1</span>, <span class="dt">$start1</span>, <span class="dt">$end1</span>) = (<span class="dt">$chrom2</span>, <span class="dt">$start2</span>, <span class="dt">$end2</span>);</a>
<a class="sourceLine" id="cb188-34" data-line-number="34">                <span class="fu">redo</span> BLOCK;</a>
<a class="sourceLine" id="cb188-35" data-line-number="35">            }</a>
<a class="sourceLine" id="cb188-36" data-line-number="36">        }</a>
<a class="sourceLine" id="cb188-37" data-line-number="37">    }</a>
<a class="sourceLine" id="cb188-38" data-line-number="38">}</a></code></pre></div>
<div class="sourceCode" id="cb189"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb189-1" data-line-number="1"><span class="fu">cat</span> data/DMR_region.txt</a></code></pre></div>
<pre><code>## chr1 100 200
## chr1 150 250
## chr1 200 300
## chr1 500 600
## chr1 550 650
## chr2 300 800
## chr2 400 1000
## chr3 500 1000</code></pre>
<div class="sourceCode" id="cb191"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb191-1" data-line-number="1"><span class="fu">perl</span> code_perl/genomic_coordinate_merge.pl  data/DMR_region.txt  data/DMR_region_merged.txt </a>
<a class="sourceLine" id="cb191-2" data-line-number="2"><span class="fu">cat</span> data/DMR_region_merged.txt </a></code></pre></div>
<pre><code>## chr1 100 300
## chr1 500 650
## chr2 300 1000
## chr3 500 1000</code></pre>
<!--chapter:end:16_practicalPerlProgram.Rmd-->
</div>
</div>
<div id="perl-modules" class="section level1">
<h1><span class="header-section-number">15</span> Perl modules</h1>
<div id="what-is-a-perl-module" class="section level2">
<h2><span class="header-section-number">15.1</span> What is a Perl module</h2>
<p>Perl modules are a set of related functions in a library file. They are specifically designed to be reusable by other modules or programs. There are more than 100,000 modules ready for you to use on the Comprehensive Perl Archive Network.</p>
<p>Most Perl modules are written in Perl, some use XS (they are written in C) so require a C compiler. Modules may have dependencies on other modules (almost always on CPAN) and cannot be installed without them (or without a specific version of them). Many modules on CPAN now require a recent version of Perl (version 5.8 or above).</p>
<p>The reason that we need to use Perl module is Perl module can largely reduce our coding work.</p>
</div>
<div id="how-to-install-a-perl-module" class="section level2">
<h2><span class="header-section-number">15.2</span> How to install a Perl module</h2>
<p>Here we’ll only discuss how to install Perl module in Linux system. The easies way I think is to use <code>cpanm</code>.</p>
<p>Here is how you can do this. First go to webpage and download the <code>cpanm</code> source code.</p>
<pre><code>## Web link:https://raw.githubusercontent.com/miyagawa/cpanminus/master/cpanm

chmod 755 cpanm

./cpanm Bio::Seq
</code></pre>
<p>NOTE</p>

<div class="rmdtip">
<p><strong>Possible problem when using <code>cpanm</code></strong></p>
<p>You may encouter the following error message. This is a bug from Perl. You cna run <code>yum install perl-CPAN</code> resolved the issue.</p>
<p>Here is the ERROR MESSAGE:</p>
!
! Can’t write to /usr/local/share/perl5 and /usr/local/bin: Installing modules to /home/xie186/perl5
! To turn off this warning, you have to do one of the following:
! - run me as a root or with –sudo option (to install to /usr/local/share/perl5 and /usr/local/bin)
! - Configure local::lib in your existing shell to set PERL_MM_OPT etc.
! - Install local::lib by running the following commands
!
! cpanm –local-lib=~/perl5 local::lib &amp;&amp; eval $(perl -I ~/perl5/lib/perl5/ -Mlocal::lib)
!
–&gt; Working on Bio::Seq
Fetching <a href="http://www.cpan.org/authors/id/C/CJ/CJFIELDS/BioPerl-1.007001.tar.gz" class="uri">http://www.cpan.org/authors/id/C/CJ/CJFIELDS/BioPerl-1.007001.tar.gz</a> … OK
==&gt; Found dependencies: Module::Build, ExtUtils::Install
–&gt; Working on Module::Build
Fetching <a href="http://www.cpan.org/authors/id/L/LE/LEONT/Module-Build-0.4222.tar.gz" class="uri">http://www.cpan.org/authors/id/L/LE/LEONT/Module-Build-0.4222.tar.gz</a> … OK
==&gt; Found dependencies: Module::Metadata, version, CPAN::Meta, Perl::OSType
–&gt; Working on Module::Metadata
Fetching <a href="http://www.cpan.org/authors/id/E/ET/ETHER/Module-Metadata-1.000033.tar.gz" class="uri">http://www.cpan.org/authors/id/E/ET/ETHER/Module-Metadata-1.000033.tar.gz</a> … OK
==&gt; Found dependencies: ExtUtils::MakeMaker
–&gt; Working on ExtUtils::MakeMaker
Fetching <a href="http://www.cpan.org/authors/id/B/BI/BINGOS/ExtUtils-MakeMaker-7.26.tar.gz" class="uri">http://www.cpan.org/authors/id/B/BI/BINGOS/ExtUtils-MakeMaker-7.26.tar.gz</a> … OK
Configuring ExtUtils-MakeMaker-7.26 … OK
Can’t locate ExtUtils/Manifest.pm in <span class="citation">(<span class="citeproc-not-found" data-reference-id="INC"><strong>???</strong></span>)</span> (<span class="citation">(<span class="citeproc-not-found" data-reference-id="INC"><strong>???</strong></span>)</span> contains: FatPacked::26160008=HASH(0x18f2b88) /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5 .) at ./cpanm line 132.
</div>

</div>
<div id="how-to-use-a-perl-module" class="section level2">
<h2><span class="header-section-number">15.3</span> How to use a Perl module</h2>
</div>
<div id="how-to-use-bioperl-module" class="section level2">
<h2><span class="header-section-number">15.4</span> How to use BioPerl module</h2>
</div>
<div id="how-to-write-a-module" class="section level2">
<h2><span class="header-section-number">15.5</span> How to write a module</h2>
<!--chapter:end:17_PerlModules.Rmd-->
</div>
</div>
<div id="part-statistics-and-r" class="section level1 unnumbered">
<h1>(PART) Statistics and R</h1>
</div>
<div id="r-introduction" class="section level1">
<h1><span class="header-section-number">16</span> R introduction</h1>
<div id="basic-r-function" class="section level2">
<h2><span class="header-section-number">16.1</span> Basic R function</h2>
</div>
<div id="producing-simple-graphs-with-r" class="section level2">
<h2><span class="header-section-number">16.2</span> Producing Simple Graphs with R</h2>
<p>The credit of this section goes to Dr. Frank McCown (<span class="citation">Frank McCown (<a href="#ref-simpleGraphR">2006</a>)</span>).</p>
<div id="line-charts" class="section level3">
<h3><span class="header-section-number">16.2.1</span> Line Charts</h3>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb194-1" data-line-number="1"><span class="co"># Define the cars vector with 5 values</span></a>
<a class="sourceLine" id="cb194-2" data-line-number="2">cars &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">4</span>, <span class="dv">9</span>)</a>
<a class="sourceLine" id="cb194-3" data-line-number="3"></a>
<a class="sourceLine" id="cb194-4" data-line-number="4"><span class="co"># Graph the cars vector with all defaults</span></a>
<a class="sourceLine" id="cb194-5" data-line-number="5"><span class="kw">plot</span>(cars)</a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-92-1.png" width="672" /></p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb195-1" data-line-number="1">?plot</a></code></pre></div>
<p>Let’s add a title, a line to connect the points, and some color:</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb196-1" data-line-number="1"><span class="co"># Define the cars vector with 5 values</span></a>
<a class="sourceLine" id="cb196-2" data-line-number="2">cars &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">4</span>, <span class="dv">9</span>)</a>
<a class="sourceLine" id="cb196-3" data-line-number="3"></a>
<a class="sourceLine" id="cb196-4" data-line-number="4"><span class="co"># Graph cars using blue points overlayed by a line </span></a>
<a class="sourceLine" id="cb196-5" data-line-number="5"><span class="kw">plot</span>(cars, <span class="dt">type=</span><span class="st">&quot;o&quot;</span>, <span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</a>
<a class="sourceLine" id="cb196-6" data-line-number="6"></a>
<a class="sourceLine" id="cb196-7" data-line-number="7"><span class="co"># Create a title with a red, bold/italic font</span></a>
<a class="sourceLine" id="cb196-8" data-line-number="8"><span class="kw">title</span>(<span class="dt">main=</span><span class="st">&quot;Autos&quot;</span>, <span class="dt">col.main=</span><span class="st">&quot;red&quot;</span>, <span class="dt">font.main=</span><span class="dv">4</span>)</a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-93-1.png" width="672" /></p>
<p>Now let’s add a red line for trucks and specify the y-axis range directly so it will be large enough to fit the truck data:</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb197-1" data-line-number="1"><span class="co"># Define 2 vectors</span></a>
<a class="sourceLine" id="cb197-2" data-line-number="2">cars &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">4</span>, <span class="dv">9</span>)</a>
<a class="sourceLine" id="cb197-3" data-line-number="3">trucks &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">12</span>)</a>
<a class="sourceLine" id="cb197-4" data-line-number="4"></a>
<a class="sourceLine" id="cb197-5" data-line-number="5"><span class="co"># Graph cars using a y axis that ranges from 0 to 12</span></a>
<a class="sourceLine" id="cb197-6" data-line-number="6"><span class="kw">plot</span>(cars, <span class="dt">type=</span><span class="st">&quot;o&quot;</span>, <span class="dt">col=</span><span class="st">&quot;blue&quot;</span>, <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">12</span>))</a>
<a class="sourceLine" id="cb197-7" data-line-number="7"></a>
<a class="sourceLine" id="cb197-8" data-line-number="8"><span class="co"># Graph trucks with red dashed line and square points</span></a>
<a class="sourceLine" id="cb197-9" data-line-number="9"><span class="kw">lines</span>(trucks, <span class="dt">type=</span><span class="st">&quot;o&quot;</span>, <span class="dt">pch=</span><span class="dv">22</span>, <span class="dt">lty=</span><span class="dv">2</span>, <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a>
<a class="sourceLine" id="cb197-10" data-line-number="10"></a>
<a class="sourceLine" id="cb197-11" data-line-number="11"><span class="co"># Create a title with a red, bold/italic font</span></a>
<a class="sourceLine" id="cb197-12" data-line-number="12"><span class="kw">title</span>(<span class="dt">main=</span><span class="st">&quot;Autos&quot;</span>, <span class="dt">col.main=</span><span class="st">&quot;red&quot;</span>, <span class="dt">font.main=</span><span class="dv">4</span>)</a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-94-1.png" width="672" /></p>
<!--chapter:end:18_R_intro.Rmd-->
</div>
</div>
</div>
<div id="producing-simple-graphs" class="section level1">
<h1><span class="header-section-number">17</span> Producing simple graphs</h1>
<div id="producing-simple-graphs-using-r" class="section level2">
<h2><span class="header-section-number">17.1</span> Producing simple graphs using R</h2>
<div id="line-charts-1" class="section level3">
<h3><span class="header-section-number">17.1.1</span> Line Charts</h3>
</div>
<div id="bar-charts" class="section level3">
<h3><span class="header-section-number">17.1.2</span> Bar Charts</h3>
</div>
<div id="histograms" class="section level3">
<h3><span class="header-section-number">17.1.3</span> Histograms</h3>
</div>
<div id="pie-charts" class="section level3">
<h3><span class="header-section-number">17.1.4</span> Pie Charts</h3>
</div>
<div id="dotcharts" class="section level3">
<h3><span class="header-section-number">17.1.5</span> Dotcharts</h3>
</div>
<div id="misc" class="section level3">
<h3><span class="header-section-number">17.1.6</span> Misc</h3>
<!--chapter:end:19_SimpleGraphR.Rmd-->
</div>
</div>
</div>
<div id="ggplot2" class="section level1">
<h1><span class="header-section-number">18</span> <code>ggplot2</code></h1>
<div id="ggplot2-1" class="section level2">
<h2><span class="header-section-number">18.1</span> ggplot2</h2>
</div>
<div id="ggplot2-practical" class="section level2">
<h2><span class="header-section-number">18.2</span> ggplot2 practical</h2>
<!--chapter:end:20_GraphGgplot2.Rmd-->
</div>
</div>
<div id="heatmap-tutorial" class="section level1">
<h1><span class="header-section-number">19</span> Heatmap Tutorial</h1>
<p><a href="https://goo.gl/KLZ7N0">Data link</a></p>
<div id="install-pheatmap-package" class="section level2">
<h2><span class="header-section-number">19.1</span> Install <code>pheatmap</code> package</h2>
<pre><code>install.packages(&quot;pheatmap&quot;)</code></pre>
</div>
<div id="draw-a-heatmap-for-gene-expression-of-rna-seq-data" class="section level2">
<h2><span class="header-section-number">19.2</span> Draw a heatmap for gene expression of RNA-seq data</h2>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb199-1" data-line-number="1"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb199-2" data-line-number="2">gene_exp &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/maize_embryo_specific_gene_Sheet1.tsv&quot;</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb199-3" data-line-number="3"><span class="kw">pheatmap</span>(gene_exp)</a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-95-1.png" width="672" /></p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb200-1" data-line-number="1"><span class="kw">pheatmap</span>(<span class="kw">log2</span>(gene_exp <span class="op">+</span><span class="st"> </span><span class="fl">0.000001</span>), <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>)</a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-95-2.png" width="672" /></p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb201-1" data-line-number="1"><span class="kw">pheatmap</span>(<span class="kw">log2</span>(gene_exp <span class="op">+</span><span class="st"> </span><span class="fl">0.01</span>), <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>, <span class="dt">show_rownames =</span> T, <span class="dt">show_colnames =</span> F)</a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-95-3.png" width="672" /></p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb202-1" data-line-number="1"><span class="kw">pheatmap</span>(<span class="kw">log2</span>(gene_exp <span class="op">+</span><span class="st"> </span><span class="fl">0.01</span>), <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>, <span class="dt">show_rownames =</span> F, <span class="dt">show_colnames =</span> T)</a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-95-4.png" width="672" /></p>
</div>
<div id="add-the-annotation" class="section level2">
<h2><span class="header-section-number">19.3</span> Add the annotation</h2>
</div>
<div id="section" class="section level2">
<h2><span class="header-section-number">19.4</span> </h2>
</div>
<div id="transfrom-the-data" class="section level2">
<h2><span class="header-section-number">19.5</span> Transfrom the data</h2>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb203-1" data-line-number="1"><span class="kw">pheatmap</span>(<span class="kw">log2</span>(gene_exp <span class="op">+</span><span class="st"> </span><span class="fl">0.01</span>)) </a></code></pre></div>
<p><img src="bioinfBookXIE186_files/figure-html/unnamed-chunk-96-1.png" width="672" /></p>
</div>
<div id="how-to-add-annotations" class="section level2">
<h2><span class="header-section-number">19.6</span> How to add annotations</h2>
</div>
<div id="how-to-cut-the-trees" class="section level2">
<h2><span class="header-section-number">19.7</span> How to cut the trees</h2>
</div>
<div id="how-to-get-the-cluster-information-from-the-heatmap" class="section level2">
<h2><span class="header-section-number">19.8</span> How to get the cluster information from the heatmap</h2>
</div>
<div id="section-1" class="section level2">
<h2><span class="header-section-number">19.9</span> </h2>
<!--chapter:end:21_GenerateHeatmap.Rmd-->
</div>
</div>
<div id="preparation-of-figures-for-manuscript" class="section level1">
<h1><span class="header-section-number">20</span> Preparation of figures for manuscript</h1>
<p>Inkscape is an open source software which can be used to arrange, crop, and annotate your images; bring in graphs and charts; draw diagrams; and export the final figure in whatever format the journal wants.</p>
<div id="test1" class="section level2">
<h2><span class="header-section-number">20.1</span> test1</h2>
<!--chapter:end:22_FigureManuscript.Rmd-->
</div>
</div>
<div id="part-omics-data" class="section level1 unnumbered">
<h1>(PART) Omics data</h1>
</div>
<div id="introduction-to-ngs" class="section level1">
<h1><span class="header-section-number">21</span> Introduction to NGS</h1>
<div id="introduction-of-biology-for-bioinformatics" class="section level2">
<h2><span class="header-section-number">21.1</span> Introduction of Biology for Bioinformatics</h2>
<div id="what-is-dna" class="section level3">
<h3><span class="header-section-number">21.1.1</span> What is DNA</h3>
</div>
<div id="what-is-genome" class="section level3">
<h3><span class="header-section-number">21.1.2</span> What is genome</h3>
</div>
<div id="how-to-assemble-a-genome" class="section level3">
<h3><span class="header-section-number">21.1.3</span> How to assemble a genome</h3>
</div>
<div id="what-is-gene" class="section level3">
<h3><span class="header-section-number">21.1.4</span> What is gene</h3>
</div>
</div>
<div id="what-is-ngs" class="section level2">
<h2><span class="header-section-number">21.2</span> What is NGS</h2>
<p>Genetic information is stored within the chemical structure of the molecule DNA. DNA can be transcribed into RNA. Then RNA can be translated into protein.</p>
<p>NGS (Next geneneration sequencing) is not an accurate term. An more accurate term maybe high throughput sequencing. NGS is specifically used to refer to sequencing technologies after Sanger sequencing (1st generation) and single molecular sequencing (3rd sequencing, e.g. PacBio) and nanopore based sequencing (4th generation, e.g. Oxford Nanopore).</p>
</div>
<div id="application-of-ngs" class="section level2">
<h2><span class="header-section-number">21.3</span> Application of NGS</h2>
</div>
<div id="section-2" class="section level2">
<h2><span class="header-section-number">21.4</span> </h2>
</div>
<div id="data-file-formats-to-store-the-genomic-information" class="section level2">
<h2><span class="header-section-number">21.5</span> Data file formats to store the genomic information</h2>
<div id="fasta-file" class="section level3">
<h3><span class="header-section-number">21.5.1</span> <code>fasta</code> file</h3>
</div>
<div id="gff-file" class="section level3">
<h3><span class="header-section-number">21.5.2</span> GFF file</h3>
</div>
<div id="gtf-file" class="section level3">
<h3><span class="header-section-number">21.5.3</span> GTF file</h3>
</div>
<div id="fastq-file" class="section level3">
<h3><span class="header-section-number">21.5.4</span> FASTQ file</h3>
</div>
<div id="sambam-file" class="section level3">
<h3><span class="header-section-number">21.5.5</span> SAM/BAM file</h3>
</div>
<div id="vcf-file" class="section level3">
<h3><span class="header-section-number">21.5.6</span> VCF file</h3>
</div>
<div id="bed-format" class="section level3">
<h3><span class="header-section-number">21.5.7</span> BED format</h3>
</div>
<div id="bedgraph-format" class="section level3">
<h3><span class="header-section-number">21.5.8</span> bedGraph format</h3>
</div>
</div>
<div id="usefull-links" class="section level2">
<h2><span class="header-section-number">21.6</span> Usefull links:</h2>
<p><a href="http://www.ecseq.com/support/ngs/trimming-adapter-sequences-is-it-necessary" class="uri">http://www.ecseq.com/support/ngs/trimming-adapter-sequences-is-it-necessary</a></p>
<p><a href="http://www.gendx.com/illumina-adapter-ligation-librx" class="uri">http://www.gendx.com/illumina-adapter-ligation-librx</a></p>
<p><a href="http://veleta.rosety.com/plasmid.html" class="uri">http://veleta.rosety.com/plasmid.html</a></p>
<!--chapter:end:23_IntroNGS.Rmd-->
<!--chapter:end:24_introRNA-seq.Rmd-->
<!--chapter:end:25_introChIP-seq.Rmd-->
</div>
</div>
<div id="bs-seq" class="section level1">
<h1><span class="header-section-number">22</span> BS-seq</h1>
<div id="section-3" class="section level2">
<h2><span class="header-section-number">22.1</span> </h2>
<p><a href="https://www.slideshare.net/secret/Da9IOe8wLsaF8V" class="uri">https://www.slideshare.net/secret/Da9IOe8wLsaF8V</a></p>
</div>
<div id="what-is-bs-seq" class="section level2">
<h2><span class="header-section-number">22.2</span> What is BS-seq</h2>
</div>
<div id="section-4" class="section level2">
<h2><span class="header-section-number">22.3</span> </h2>
</div>
<div id="how-to-analyze-bs-seq-data" class="section level2">
<h2><span class="header-section-number">22.4</span> How to analyze BS-seq data</h2>
<div id="section-5" class="section level3">
<h3><span class="header-section-number">22.4.1</span> </h3>
<p>MINI REVIEW: Statistical methods for detecting differentially methylated loci and regions</p>
<p>Strategies for analyzing bisulfite sequencing data</p>
<!--chapter:end:26_introBSseq.Rmd-->
</div>
</div>
</div>
<div id="start-a-project" class="section level1">
<h1><span class="header-section-number">23</span> Start a project</h1>
<div id="experimetal-design" class="section level2">
<h2><span class="header-section-number">23.1</span> Experimetal design</h2>
<p>Before</p>
<div id="how-many-biological-replicates" class="section level3">
<h3><span class="header-section-number">23.1.1</span> How many biological replicates</h3>
</div>
<div id="how-big" class="section level3">
<h3><span class="header-section-number">23.1.2</span> How big</h3>
</div>
<div id="how-much-data-to-generate" class="section level3">
<h3><span class="header-section-number">23.1.3</span> How much data to generate</h3>
</div>
<div id="section-6" class="section level3">
<h3><span class="header-section-number">23.1.4</span> </h3>
<p>Where to start?</p>
<p>There are many ways to start to talk about Bioinformatics analysis. Here we’ll go from whether there is a reference genome or not. Let’s take RNA-seq as an example.</p>
<!--chapter:end:27_ExpDesign.Rmd-->
</div>
</div>
</div>
<div id="capstone-project" class="section level1">
<h1><span class="header-section-number">24</span> Capstone project:</h1>
<div id="introduction" class="section level2">
<h2><span class="header-section-number">24.1</span> Introduction</h2>
</div>
<div id="method" class="section level2">
<h2><span class="header-section-number">24.2</span> Method</h2>
</div>
<div id="pipelines" class="section level2">
<h2><span class="header-section-number">24.3</span> Pipelines</h2>
</div>
<div id="section-7" class="section level2">
<h2><span class="header-section-number">24.4</span> </h2>
<!--chapter:end:28_CapstoneProjectRNA-seq.Rmd-->
</div>
</div>
<div id="practical-data-analysis-on-wgbs" class="section level1">
<h1><span class="header-section-number">25</span> Practical data analysis on WGBS</h1>
<p>We are going to use the data in this <a href="https://www.ncbi.nlm.nih.gov//geo/query/acc.cgi?acc=GSE66905">link</a>.</p>
<p>In this link, it has multile WGBS from different genotypes of yeast. For simplicity, we are going to only pick 3 samples: <code>EV strain 1</code>,</p>
<p>We are going to use the data from these three samples to illustrate how to analyze WGBS data (Figure @ref(fig:wgbs_bg)).</p>
<div class="figure">
<img src="figures/elife_yeast_paper.jpg" alt="Study background of WGBS data"  />
<p class="caption">
(#fig:wgbs_bg)Study background of WGBS data
</p>
</div>
<p>Let’s walk through these workflow step-by-step.</p>
<div id="donwload-the-data" class="section level2">
<h2><span class="header-section-number">25.1</span> Donwload the data</h2>
<p>First we need to download the reference genome (in fasta file) and the genome annotation file (in gff/gtf file).</p>
<div id="reference" class="section level3">
<h3><span class="header-section-number">25.1.1</span> Reference</h3>
<p>Reference link: <a href="http://useast.ensembl.org/Saccharomyces_cerevisiae/Info/Index" class="uri">http://useast.ensembl.org/Saccharomyces_cerevisiae/Info/Index</a></p>
<pre><code>wget ftp://ftp.ensembl.org/pub/release-95/fasta/saccharomyces_cerevisiae/dna/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa.gz
wget ftp://ftp.ensembl.org/pub/release-95/gff3/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.95.gff3.gz
</code></pre>
</div>
<div id="now-lets-download-the-wgbs-data" class="section level3">
<h3><span class="header-section-number">25.1.2</span> Now let’s download the WGBS data</h3>
<p>Based on the GEO <a href="https://www.ncbi.nlm.nih.gov//geo/query/acc.cgi?acc=GSE66905">link</a> here, we can use the command lines below to download the <code>sra</code> files.</p>
<p>(ref:wgbs_sample) Sample list for WGBS data</p>
<table>
<caption>(#tab:wgbs_sample)(ref:wgbs_sample)</caption>
<thead>
<tr class="header">
<th align="center">SRR_number</th>
<th align="center">Sample_name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">SRR1916129</td>
<td align="center">EV1</td>
</tr>
<tr class="even">
<td align="center">SRR1916134</td>
<td align="center">3bstrain1</td>
</tr>
<tr class="odd">
<td align="center">SRR1916142</td>
<td align="center">set1rep1</td>
</tr>
</tbody>
</table>
<pre><code>wget -O EV1.sra ftp://ftp.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR191/SRR1916129/SRR1916129.sra
wget -O 3bstrain1.sra  ftp://ftp.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR191/SRR1916134/SRR1916134.sra
wget -O set1_rep1.sra ftp://ftp.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR191/SRR1916142/SRR1916142.sra
</code></pre>
</div>
</div>
<div id="convert-sra-files-to-fastq-files" class="section level2">
<h2><span class="header-section-number">25.2</span> Convert <code>sra</code> files to <code>fastq</code> files</h2>
<pre><code></code></pre>
<!--chapter:end:30_CapstoneProjectBS-seq.Rmd-->
<ol style="list-style-type: decimal">
<li><p>why program</p></li>
<li><p>how to program</p></li>
<li><p>programming language</p></li>
</ol>
</div>
</div>
<div id="section-8" class="section level1">
<h1><span class="header-section-number">26</span> </h1>
<p>This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <a href="http://rmarkdown.rstudio.com" class="uri">http://rmarkdown.rstudio.com</a>.</p>
<p>When you click the <strong>Knit</strong> button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb207-1" data-line-number="1"><span class="kw">summary</span>(cars)</a></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##     1.0     3.0     4.0     4.6     6.0     9.0</code></pre>
<div id="including-plots" class="section level2">
<h2><span class="header-section-number">26.1</span> Including Plots</h2>
<p>You can also embed plots, for example:</p>
<p><img src="bioinfBookXIE186_files/figure-html/pressure-1.png" width="672" /></p>
<p>Note that the <code>echo = FALSE</code> parameter was added to the code chunk to prevent printing of the R code that generated the plot.</p>
<!--chapter:end:30_py_why_program.Rmd-->
</div>
</div>
<div id="good-resouces-to-learn-bioinformatics" class="section level1">
<h1><span class="header-section-number">27</span> Good resouces to learn Bioinformatics</h1>
<div id="references" class="section level2">
<h2><span class="header-section-number">27.1</span> References:</h2>
<p><a href="https://www.bioinformatics.babraham.ac.uk/training.html" class="uri">https://www.bioinformatics.babraham.ac.uk/training.html</a></p>
<!--chapter:end:31_Good_resource.Rmd-->
</div>
</div>
<div id="references-1" class="section level1 unnumbered">
<h1>References</h1>
<!--chapter:end:32_Reference.Rmd-->
</div>
<div id="crispr-cas9-screen-data-processing" class="section level1">
<h1><span class="header-section-number">28</span> CRISPR-CAS9 screen data processing</h1>
<div id="installation-of-softwares-using-conda" class="section level2">
<h2><span class="header-section-number">28.1</span> Installation of softwares using <code>conda</code></h2>
<pre><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
conda config --add channels conda-forge
conda config --add channels bioconda
conda create -n mageck-vispr mageck mageck-vispr python=3</code></pre>
</div>
<div id="an-exmaple-of-how-to-process" class="section level2">
<h2><span class="header-section-number">28.2</span> An exmaple of how to process</h2>
<p>CRISPR Screen.
For the pooled genome-wide CRISPR screen, 1 × 108 LNCaP cells were infected with the pooled lentiviral GeCKO v2 library at a multiplicity of infection of 0.5. After 3 d of puromycin selection, half of the surviving cells were stored as 0-d control samples, and the rest of cells were cultured for an additional 2 wk. PCR was performed on genomic DNA to construct the sequencing library. Each library was sequenced at 30∼40 million reads to achieve ∼300× average coverage over the CRISPR library and data were analyzed by MAGeCK and MAGeCK-VISPR.</p>
<div id="donwload-the-data-1" class="section level3">
<h3><span class="header-section-number">28.2.1</span> Donwload the data</h3>
<pre><code>#!/usr/bin/env bash
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR231/007/SRR2318477/SRR2318477.fastq.gz -o CRISPR-0d.fastq.gz &amp;
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR231/006/SRR2318476/SRR2318476.fastq.gz -o CRISPR-plasmid.fastq.gz &amp;
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR231/008/SRR2318478/SRR2318478.fastq.gz -o CRISPR-DHT.fastq.gz &amp;</code></pre>
</div>
<div id="reference-1" class="section level3">
<h3><span class="header-section-number">28.2.2</span> Reference</h3>
<p>Fei T, Chen Y, Xiao T, Li W et al. Genome-wide CRISPR screen identifies HNRNPL as a prostate cancer dependency regulating RNA splicing. Proc Natl Acad Sci U S A 2017 Jun 27;114(26):E5207-E5215. PMID: 28611215</p>
<!--chapter:end:capstone_CRISPR_screen.Rmd-->
</div>
</div>
</div>
<div id="commands-for-linux-administration-advanced-topic" class="section level1">
<h1><span class="header-section-number">29</span> Commands for Linux administration (Advanced topic)</h1>
<div id="w" class="section level2">
<h2><span class="header-section-number">29.1</span> <code>w</code></h2>
</div>
<div id="who" class="section level2">
<h2><span class="header-section-number">29.2</span> <code>who</code></h2>
</div>
<div id="uptime" class="section level2">
<h2><span class="header-section-number">29.3</span> <code>uptime</code></h2>
<p>In Linux uptime command shows since how long your system is running and the number of users are currently logged in and also displays load average for 1,5 and 15 minutes intervals.</p>
</div>
<div id="whoami" class="section level2">
<h2><span class="header-section-number">29.4</span> <code>whoami</code></h2>
<div class="sourceCode" id="cb211"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb211-1" data-line-number="1"><span class="fu">whoami</span></a></code></pre></div>
<pre><code>## rstudio-user</code></pre>
</div>
<div id="ifconfig" class="section level2">
<h2><span class="header-section-number">29.5</span> <code>ifconfig</code></h2>
<pre><code>ifconfig</code></pre>
</div>
<div id="useradd-and-passwd" class="section level2">
<h2><span class="header-section-number">29.6</span> <code>useradd</code> and <code>passwd</code></h2>
<pre><code>## Need to have root access
adduser tabomics 
</code></pre>
<pre><code>## add a user to a specified group
adduser tabomics -g bioinf</code></pre>
<!--chapter:end:cmd_linux_admin.Rmd-->
<div id="refs" class="references">
<div id="ref-simpleGraphR">
<p>Frank McCown. 2006. <em>Producing Simple Graphs with R</em>. Searcy, AR: Harding University. <a href="https://www.harding.edu/fmccown/r/">https://www.harding.edu/fmccown/r/</a>.</p>
</div>
<div id="ref-Torvalds2015">
<p>Torvalds, Linus. 2015. “Linux kernel release 4.x.” <a href="https://github.com/torvalds/linux">https://github.com/torvalds/linux</a>.</p>
</div>
</div>
</div>
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
    </div>
  </div>
<!--bookdown:config-->

</body>

</html>
