\documentclass[]{book}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={All things considered for Bioinformatics},
            pdfauthor={Shaojun Xie (解少俊)},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{natbib}
\bibliographystyle{apalike}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\providecommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{All things considered for Bioinformatics}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
  \subtitle{Essential Linux, Python, Perl, R and Statistics in life science}
  \author{Shaojun Xie (解少俊)}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
      \predate{\centering\large\emph}
  \postdate{\par}
    \date{2020-03-20}

\usepackage{booktabs}

\makeatletter
\newenvironment{kframe}{%
\medskip{}
\setlength{\fboxsep}{.8em}
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\renewenvironment{Shaded}{\begin{kframe}}{\end{kframe}}

\newenvironment{rmdblock}[1]
  {
  \begin{itemize}
  \renewcommand{\labelitemi}{
    \raisebox{-.7\height}[0pt][0pt]{
      {\setkeys{Gin}{width=3em,keepaspectratio}\includegraphics{images/#1}}
    }
  }
  \setlength{\fboxsep}{1em}
  \begin{kframe}
  \item
  }
  {
  \end{kframe}
  \end{itemize}
  }
\newenvironment{rmdnote}
  {\begin{rmdblock}{note}}
  {\end{rmdblock}}
\newenvironment{rmdcaution}
  {\begin{rmdblock}{caution}}
  {\end{rmdblock}}
\newenvironment{rmdimportant}
  {\begin{rmdblock}{important}}
  {\end{rmdblock}}
\newenvironment{rmdtip}
  {\begin{rmdblock}{tip}}
  {\end{rmdblock}}
\newenvironment{rmdwarning}
  {\begin{rmdblock}{warning}}
  {\end{rmdblock}}

\let\BeginKnitrBlock\begin \let\EndKnitrBlock\end
\begin{document}
\maketitle

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\hypertarget{preface}{%
\chapter*{Preface}\label{preface}}
\addcontentsline{toc}{chapter}{Preface}

\begin{verbatim}
## Bioconductor version 3.7 (BiocInstaller 1.30.0), ?biocLite for help
\end{verbatim}

\begin{verbatim}
## A newer version of Bioconductor is available for this version of R,
##   ?BiocUpgrade for help
\end{verbatim}

\begin{verbatim}
## BioC_mirror: https://bioconductor.org
\end{verbatim}

\begin{verbatim}
## Using Bioconductor 3.7 (BiocInstaller 1.30.0), R 3.5.3 (2019-03-11).
\end{verbatim}

\hypertarget{part-linux}{%
\part{Linux}\label{part-linux}}

\hypertarget{why-linux}{%
\chapter{Why Linux?}\label{why-linux}}

\hypertarget{what-is-linux}{%
\section{What is Linux}\label{what-is-linux}}

Before the creation of Linux, Unix was developed by AT\&T Bell Labs in the 1960's. It's an operating system. Before the creation of Linux, and before the rise of Windows, the computing world was dominated by Unix (from web). After many years of evolution, Linux was created in early 1990's.

In case you don't know, Mac OS X is also a certified Unix operating system. So most of the Linux skills are applied in Mac OS X.

Linux is a clone of the operating system Unix, written from scratch by Linus Torvalds (Figure \ref{fig:LinuxTerminal}) with assistance from a loosely-knit team of hackers across the Net. It aims towards POSIX and Single UNIX Specification compliance (\citet{Torvalds2015}).



\begin{figure}

{\centering \includegraphics[width=1\linewidth]{figures/linux_terminal_example} 

}

\caption{An example of Linux terminal.}\label{fig:LinuxTerminal}
\end{figure}

It has all the features you would expect in a modern fully-fledged Unix, including true multitasking, virtual memory, shared libraries, demand loading, shared copy-on-write executables, proper memory management, and multistack networking including IPv4 and IPv6. It is distributed under the GNU General Public (Torvalds, 2015).

Maybe it's hard to understand what Linux or to remember the sentences mentioned above. Just know Linux is an operating system like Windows. This is enough for you to start out.

\begin{figure}

{\centering \includegraphics[width=0.5\linewidth]{figures/Linus_Torvalds_GitHub} 

}

\caption{Linus Torvalds on GitHub}\label{fig:LinusTorvaldsGitHub}
\end{figure}

\hypertarget{linux-for-bioinformatics}{%
\section{Linux for bioinformatics}\label{linux-for-bioinformatics}}

For analysis of NGS data, a large amount of software were developed for using under Linux environment. Among them, a large proportion can be only used under Linux environment.

\begin{itemize}
\tightlist
\item
  Easy to build simple pipelines
  (awk, bash, piping, bash redirection, texttools)
\item
  Simple to install and use software development tools
\item
  Multiple versions of a program can be installed by the user himself and switched on/off with sourcing some scripts without being administrator
\item
  A lot of good scientific software is written in a non portable way for linux/unix (almost all short read aligners, samtools). This makes it necessary to use Unix for genomics.
\item
  Ability to perform analyses on computer clusters (important for big/long computational jobs)
\end{itemize}

\hypertarget{connecting-to-linux}{%
\chapter{Connecting to Linux}\label{connecting-to-linux}}

\hypertarget{user-interfaces}{%
\section{User interfaces}\label{user-interfaces}}

As an operating systemsm, Linux comes with two types of user interfaces: Graphical User Interface (GUI) and command line interface (shell).

GUI means there will be window, buttons, menus, etc. The most popular system with GUI is Windows system (Figure \ref{fig:windowsGUI}).



\begin{figure}
\centering
\includegraphics{figures/windows_gui.png}
\caption{\label{fig:windowsGUI}Windows GUI.}
\end{figure}

Command line interfaces means that you need to type the command line yourself. Usually the results will be displayed as text (Figure \ref{fig:linuxTerminalExam2}).



\begin{figure}
\centering
\includegraphics{figures/linux_terminal_exam2.png}
\caption{\label{fig:linuxTerminalExam2}An example of Linux terminal.}
\end{figure}

In Bioinformatics analysis, usually you won't operate directly on the physical machine of the Linux server. Usually you need to connect to the Linux server via a tool, like Putty, Mobaxterm, etc.

\hypertarget{how-to-connect}{%
\section{How to connect}\label{how-to-connect}}

If you want to connect to a Linux server, what you need to know first is:

\begin{enumerate}
\def\labelenumi{\arabic{enumi})}
\item
  IP address of your Linux server;
\item
  User name and password of your account;
\end{enumerate}

If you are a Mac OS X user, you can connect to a Linux server by using \texttt{Terminal}, a console program included with the operating system.

\begin{quote}
Mac OS system itself is also a UNIX system. Majority of the command lines in Linux will also work in Mac OS.
\end{quote}

For Windows users, I would recommend MobaXterm for remote connection. MobatXterm is an excellent toolbox for remote connection from Windows system. It comes with an X11 server and provides many networking tools and tabbed SSH. It has all the essential UNIX commands in a single portable executable file.

Here I show one example of what you should do When you first open Mobaxterm. You need to follow the numbers in Figure \ref{fig:mobaxtermInit}: click \texttt{Session}; then click \texttt{SSH}; type the IP or name address of the remote host, check \texttt{Specify\ username} if you need; click \texttt{OK}.



\begin{figure}
\centering
\includegraphics{figures/mobaxterm_init.png}
\caption{\label{fig:mobaxtermInit}First open of Mobaxterm.}
\end{figure}

Then you need to type the password. It's OK that you don't see anything when you're typing (Figure \ref{fig:mobaxtermSavepswd}). Then click \texttt{Enter} on the keyboard.For the first time of log-in, you'll be asked to whether to save the password or not. If you say click \texttt{Yes}, you won't need to type the password again next time.



\begin{figure}
\centering
\includegraphics{figures/mobaxterm_savepassword.png}
\caption{\label{fig:mobaxtermSavepswd}Type password and save it in Mobaxterm.}
\end{figure}

If you can find a Linux server, it'll be very good for you to practise. If you are a student or a researcher in a university or an institute, usually you can get an account from your department.

If NOT, here I provided a guest account for you. Here are the user name and password:

\begin{verbatim}
IP address: 198.211.107.37
User name: guest4bioinfo
Password: nobigfile
\end{verbatim}

As you can tell from the password, please do NOT upload BIG files (bigger than 2 MB).

\hypertarget{alternative-ways-to-gain-access-to-a-linux-server}{%
\subsection{Alternative ways to gain access to a Linux Server}\label{alternative-ways-to-gain-access-to-a-linux-server}}

\hypertarget{mac-os}{%
\subsubsection{Mac OS}\label{mac-os}}

\hypertarget{dual-boot-ubuntu-and-windows}{%
\subsubsection{Dual boot Ubuntu and Windows}\label{dual-boot-ubuntu-and-windows}}

If you are a windows user, you can set up \texttt{{[}WindowsDualBoot{]}}(\url{https://rstudio.cloud/spaces/15199/project/311887}).

\hypertarget{digitalocean-droplet}{%
\subsubsection{DigitalOcean droplet}\label{digitalocean-droplet}}

Another option is to sign up on DigitalOcean and create a droplet. DigitalOcean calls its cloud servers Droplets; each Droplet you create is a new server for your personal use. DigitalOcean has a tutorial of {[}\textbf{How To Create Your First DigitalOcean Droplet}{]} (\url{https://www.digitalocean.com/community/tutorials/how-to-create-your-first-digitalocean-droplet}). You can get a your own Linux server with a \$5 monthly payment.

The Linux server with IP address of \texttt{198.211.107.37} is a droplet on DigitalOcean. I pay \$5 each month for this droplet.

\hypertarget{aws-free-tier-offering}{%
\subsubsection{AWS Free Tier offering}\label{aws-free-tier-offering}}

One more way to get access to a Linux system is to take advantage of Red Hat Enterprise Linux delivered by Amazon EC2 (Elastic Compute Cloud). Red Hat and Amazon Web Services collaborate to provide official Red Hat Enterprise Linux licensed images through Amazon's on-demand public cloud service at free or low cost.

The guided exercises and labs for this course were written assuming that you will set up an account with Amazon Web Services and use it to start a single, simple system running Red Hat Enterprise Linux 7. You will connect to that system securely over the internet and use it to practice commands.

At the time of writing, Amazon Web Services provides an AWS Free Tier offering, which gives new users free access to certain sizes of cloud instances and operating environments (including Red Hat Enterprise Linux 7) for up to 750 hours per month, for 12 months.

\hypertarget{navigating-in-linux-file-system}{%
\chapter{Navigating in Linux file system}\label{navigating-in-linux-file-system}}

You are in your home directory after you log into the system and are directed to the shell command prompt. This section will show you hot to explore Linux file system using shell commands.

To start, you need to take a tour of what the Linux filesystem looks like so you know where you are going.

\hypertarget{path}{%
\section{Path}\label{path}}

To understand Linux file system, you can image it as a tree structure (Figure \ref{fig:linuxTreeStruc}).



\begin{figure}
\centering
\includegraphics{figures/LinuxPathTree.png}
\caption{\label{fig:linuxTreeStruc}Tree structure of Linux system.}
\end{figure}

In Linux, a path is a unique location of a file or a directory in the file system.

For convenience, Linux file system is usually thought of in a tree structure. On a standard Linux system you will find the layout generally follows the scheme presented below.

The tree of the file system starts at the trunk or slash, indicated by a forward slash (/). This directory, containing all underlying directories and files, is also called the root directory or ``the root'' of the file system.

\hypertarget{relative-and-absolute-path}{%
\subsection{Relative and absolute path}\label{relative-and-absolute-path}}

\begin{itemize}
\tightlist
\item
  \textbf{Absolute path}
\end{itemize}

An absolute path is defined as the location of a file or directory from the root directory(/). An absolute path starts from the \texttt{root} of the tree (\texttt{/}).

Here are some examples:

\begin{verbatim}
/home/xie186
/home/xie186/perl5
\end{verbatim}

\begin{itemize}
\tightlist
\item
  \textbf{Relative path}
\end{itemize}

Relative path is a path related to the present working directory.

\begin{verbatim}
data/sample1/
../doc/
\end{verbatim}

If you want to get the \textbf{absolute path} based on \textbf{relative path}, you can use \texttt{readlink} with parameter \texttt{-f}:

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{pwd}
\FunctionTok{readlink}\NormalTok{ -f ../}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## /cloud/project
## /cloud
\end{verbatim}

\hypertarget{surfing-in-linux-file-system}{%
\section{Surfing in Linux file system}\label{surfing-in-linux-file-system}}

Once we enter into a Linux file system, we need to 1) know where we are; 2) how to get where we want; 3) how to know what files or directories we have in a particular path.

\hypertarget{check-where-you-are-using-command-pwd}{%
\subsection{\texorpdfstring{Check where you are using command \texttt{pwd}}{Check where you are using command pwd}}\label{check-where-you-are-using-command-pwd}}

In order to know where we are, we need to use \texttt{pwd} command. The command \texttt{pwd} is short for ``print name of current/working directory''. It will return the full path of current directory.

Command pwd is almost always used by itself. This means you only need to type pwd and press ENTER (Figure \ref{fig:linuxCMDpwd}).



\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{figures/linuxCMDpwd} 

}

\caption{ref:linuxCMDpwd}\label{fig:linuxCMDpwd}
\end{figure}

\hypertarget{listing-the-contents-using-command-ls}{%
\subsection{\texorpdfstring{Listing the contents using command \texttt{ls}}{Listing the contents using command ls}}\label{listing-the-contents-using-command-ls}}

After you know where you are, then you want to know what you have in that
directory, we can use command \texttt{ls} to list directory contents (Figure \ref{fig:linuxCMDls}). Its syntax is:

\begin{verbatim}
ls [option]... [file]...
\end{verbatim}



\begin{figure}

{\centering \includegraphics[width=0.8\linewidth]{figures/linuxCMDls} 

}

\caption{ref:linuxCMDls}\label{fig:linuxCMDls}
\end{figure}

\texttt{ls} with no option will list files and directories in bare format. Bare format means the detailed information (type, size, modified date and time, permissions and links etc) won't be viewed. When you use \texttt{ls} by itself (Figure \ref{fig:linuxCMDls}), it will list files and directories in the current directory.

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{cd}\NormalTok{ tables}
\FunctionTok{ls}

\BuiltInTok{echo} \StringTok{"ls -a"}
\FunctionTok{ls}\NormalTok{ -a }

\BuiltInTok{echo} \StringTok{"ls -t"}
\FunctionTok{ls}\NormalTok{ -t}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 10_PerlInputOutput_bak.Rmd
## linuxPathShortcuts.tsv
## regexp_perl.tsv
## textEditorLinuxVi3modes.csv
## ls -a
## .
## ..
## 10_PerlInputOutput_bak.Rmd
## linuxPathShortcuts.tsv
## regexp_perl.tsv
## textEditorLinuxVi3modes.csv
## ls -t
## regexp_perl.tsv
## 10_PerlInputOutput_bak.Rmd
## linuxPathShortcuts.tsv
## textEditorLinuxVi3modes.csv
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ls}\NormalTok{ -l -a tables/}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## total 28
## drwxrwxr-x  2 rstudio-user rstudio-user 4096 May 18  2019 .
## drwxr-xr-x 17 rstudio-user rstudio-user 4096 Mar 20 17:58 ..
## -rw-rw-r--  1 rstudio-user rstudio-user 4139 Apr 15  2019 10_PerlInputOutput_bak.Rmd
## -rw-rw-r--  1 rstudio-user rstudio-user  223 Apr 15  2019 linuxPathShortcuts.tsv
## -rw-rw-r--  1 rstudio-user rstudio-user  259 May 18  2019 regexp_perl.tsv
## -rw-rw-r--  1 rstudio-user rstudio-user  766 Apr 15  2019 textEditorLinuxVi3modes.csv
\end{verbatim}

Linux command options can be combined without a space between them and with a single - (dash).

The following command is a faster way to use the l and a options and gives the same output as the Linux command shown above.

\begin{verbatim}
ls -la 
\end{verbatim}

\hypertarget{change-directory-using-command-cd}{%
\subsection{\texorpdfstring{Change directory using command \texttt{cd}}{Change directory using command cd}}\label{change-directory-using-command-cd}}

Command cd is used to change the current directory. It's syntax is:

\begin{verbatim}
cd [option] [directory]
\end{verbatim}

Unlike \texttt{pwd}, when you use \texttt{cd} you usually need to provide the path (either absolute or relative path) which we want to enter.

If we didn't provide any path information, we will change to home directory by default.

\hypertarget{path-shortcuts}{%
\section{Path shortcuts}\label{path-shortcuts}}

In Linux, there are three commonly used path shortpaths (Table \ref{tab:linuxPathShortcuts}).



\begin{table}[t]

\caption{\label{tab:linuxPathShortcuts}Shortcuts of path.}
\centering
\begin{tabular}{c|c|c}
\hline
Path & Shortcuts & Description\\
\hline
Single dot & . & The current folder\\
\hline
Double dots & .. & The folder above the current folder\\
\hline
Tilde character & \textasciitilde{} & Home directory (normally the directory:/home/my\_login\_name)\\
\hline
Dash & - & Your last working directory\\
\hline
\end{tabular}
\end{table}

Here are some examples:

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{cd}\NormalTok{ ~}
\BuiltInTok{pwd}
\FunctionTok{ls}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## /home/rstudio-user
## R
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]

\FunctionTok{ls}\NormalTok{ ./}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 00_about_me_acknowledge.Rmd
## 01_WhyLinux.Rmd
## 02_Connect2Linux.Rmd
## 03_FileSystemLinux.Rmd
## 04_Linux_FilteringOutputandFindingThings.Rmd
## 05_AchivingAndCompressingFiles.Rmd
## 06_procManageLinux.Rmd
## 07_fileTransfer.Rmd
## 08_InstallationOfSoftwareInLinux.Rmd
## 09_TextEditorInLinux.Rmd
## 100_Perl_oneliner.Rmd
## 10_FirstPerlProgram.Rmd
## 11_PerlVariableOperator.Rmd
## 13_PerlControlStructure.Rmd
## 14_StringManipulationRegExp.Rmd
## 15_PerlInputOutput.Rmd
## 16_practicalPerlProgram.Rmd
## 17_PerlModules.Rmd
## 18_R_intro.Rmd
## 20_GraphGgplot2.Rmd
## 21_GenerateHeatmap.Rmd
## 22_FigureManuscript.Rmd
## 23_IntroNGS.Rmd
## 24_introRNA-seq.Rmd
## 25_introChIP-seq.Rmd
## 26_introBSseq.Rmd
## 27_ExpDesign.Rmd
## 28_CapstoneProjectRNA-seq.Rmd
## 29_R_data.table.Rmd
## 30_CapstoneProjectBS-seq.Rmd
## 30_py_why_program.Rmd
## 31_Good_resource.Rmd
## 32_Reference.Rmd
## 41_basic_statistics_R.Rmd
## DESCRIPTION
## LICENSE
## README.md
## Rmd_list.txt
## TODO
## _bookdown.yml
## _bookdown_files
## _build.sh
## _deploy.sh
## _output.yml
## bak
## bioinfBookXIE186.Rmd
## bioinfBookXIE186_files
## book.bib
## bookdown-demo.Rproj
## bookdown-demo.log
## code_R
## code_perl
## code_python
## data
## docs
## download
## figures
## images
## index.Rmd
## index.Rmd_bak
## index.log
## lib
## order_Rmd.sh
## order_Rmd.sh.README
## packages.bib
## preamble.tex
## preamble_bak.tex
## python_first_python.Rmd
## rpres.css
## slides
## style.css
## tables
## test.pl
## toc.css
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]

\CommentTok{## }
\BuiltInTok{pwd}
\BuiltInTok{cd}\NormalTok{ ../}
\BuiltInTok{pwd}
\BuiltInTok{cd}\NormalTok{ ./}
\BuiltInTok{pwd}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
/cloud/project
/cloud
/cloud
\end{verbatim}

Each directory has two entries in it at the start, with names . (a link to itself) and .. (a link to its parent directory). The exception, of course, is the root directory, where the .. directory also refers to the root directory.

Sometimes you go to a new directory and do something, then you remember that you need to go to the previous working direcotry. To get back instantly, use a dash.

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{echo} \StringTok{"This is our current directory: "}
\BuiltInTok{pwd}

\BuiltInTok{echo} \StringTok{"Let's go our home diretory: "}
\BuiltInTok{cd}\NormalTok{ ~}

\BuiltInTok{echo} \StringTok{"Check where we are: "}
\BuiltInTok{pwd}

\BuiltInTok{echo} \StringTok{"Let's go to your previous working direcotry: "}
\BuiltInTok{cd}\NormalTok{ -}
\BuiltInTok{echo} \StringTok{"Check where we're now: "}
\BuiltInTok{pwd}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## This is our current directory: 
## /cloud/project
## Let's go our home diretory: 
## Check where we are: 
## /home/rstudio-user
## Let's go to your previous working direcotry: 
## /cloud/project
## Check where we're now: 
## /cloud/project
\end{verbatim}

\hypertarget{manipulations-of-files-and-directories}{%
\section{Manipulations of files and directories}\label{manipulations-of-files-and-directories}}

In Linux, manipulations of files and directories are the most frequent work. In this section, you will learn how to copy, rename, remove, and create files and directories.

\hypertarget{command-cp}{%
\subsection{\texorpdfstring{Command \texttt{cp}}{Command cp}}\label{command-cp}}

In Linux, command \texttt{cp} can help you copy files and directories into a target directory.

\hypertarget{command-mv}{%
\subsection{\texorpdfstring{Command \texttt{mv}}{Command mv}}\label{command-mv}}

The command \texttt{mv} is short for move (or rename) files.

\hypertarget{move-one-file}{%
\subsubsection{Move one file}\label{move-one-file}}

Here is one common example of \texttt{mv}.

\begin{Shaded}
\begin{Highlighting}[]

\FunctionTok{mv}\NormalTok{ file1 directory1/}
\end{Highlighting}
\end{Shaded}

\hypertarget{move-multiple-files-into-a-directory}{%
\subsubsection{Move multiple files into a directory}\label{move-multiple-files-into-a-directory}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{mv}\NormalTok{ file1 file2 file3 target_direcotry/}
\end{Highlighting}
\end{Shaded}

\hypertarget{move-a-directory}{%
\subsubsection{Move a directory}\label{move-a-directory}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{mv}\NormalTok{ dir1}
\end{Highlighting}
\end{Shaded}

\hypertarget{rename-a-file-or-a-directory}{%
\subsubsection{Rename a file or a directory}\label{rename-a-file-or-a-directory}}

\hypertarget{command-mkdir}{%
\subsection{\texorpdfstring{Command \texttt{mkdir}}{Command mkdir}}\label{command-mkdir}}

Command \texttt{mkdir} is short for make directory.

The syntax is shown as below:

\begin{verbatim}
mkdir [OPTION ...] DIRECTORY ...
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{mkdir}\NormalTok{ directory}
\end{Highlighting}
\end{Shaded}

Multiple directories can be specified when calling \texttt{mkdir}.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{mkdir}\NormalTok{ directory1 directory2}
\end{Highlighting}
\end{Shaded}

\hypertarget{how-to-create-a-directory}{%
\subsubsection{How to create a directory}\label{how-to-create-a-directory}}

\begin{verbatim}
mkdir -p foo/bar/baz
\end{verbatim}

\begin{quote}
How to defining complex directory trees with one command
\end{quote}

\begin{verbatim}
mkdir -p project/{software,results,doc/{html,info,pdf},scripts}
\end{verbatim}

This will create a direcotry trees as shown below:

\begin{verbatim}
$ tree project/
project/
├── doc
│   ├── html
│   ├── info
│   └── pdf
├── results
├── scripts
└── software

7 directories, 0 files
\end{verbatim}

The command line above will directories \texttt{foo}, \texttt{foo/bar}, and \texttt{foo/bar/baz} if they don't exist.

\hypertarget{command-rm}{%
\subsection{Command `rm'}\label{command-rm}}

You can use \texttt{rm} to remove both files and directories.

\hypertarget{how-to-remove-a-file-or-multiple-files}{%
\subsubsection{How to remove a file or multiple files}\label{how-to-remove-a-file-or-multiple-files}}

\begin{verbatim}
## You can remove one file. 
rm file1 
## `rm` can remove multiple files simutaneously
rm file2 file3 
\end{verbatim}

\hypertarget{how-to-remove-a-folder}{%
\subsubsection{How to remove a folder}\label{how-to-remove-a-folder}}

If a folder is empty, you can remove it using \texttt{rm} with \texttt{-r}.

\begin{verbatim}
rm -r FOLDER
\end{verbatim}

If a folder is not empty, you can remove it using \texttt{rm} with \texttt{-r} and \texttt{-f}.

\begin{verbatim}
mkdir test_folder
rm -r test_folder
\end{verbatim}

\hypertarget{viewing-text-files-in-linux}{%
\section{Viewing text files in Linux}\label{viewing-text-files-in-linux}}

\hypertarget{command-cat}{%
\subsection{\texorpdfstring{Command \texttt{cat}}{Command cat}}\label{command-cat}}

The command \texttt{cat} is short for concatenate files and print on the standard output.

The syntax is shown as below:

\begin{verbatim}
cat [OPTION]... [FILE]...
\end{verbatim}

For small text file, \texttt{cat} can be used to view the files on the standard output.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{cat}\NormalTok{ data/testdata4linux_cmd.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene1
## gene2
## gene3
## gene4
## gene5
## gene6
## gene7
## gene8
## gene9
## gene10
## gene11
## gene12
## gene13
## gene14
## gene15
## gene16
\end{verbatim}

You can also use \texttt{cat} to merge two text files.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{cat}\NormalTok{ file1 file2 }\OperatorTok{>}\NormalTok{ merged_file}
\end{Highlighting}
\end{Shaded}

\hypertarget{command-more-and-less}{%
\subsection{\texorpdfstring{Command \texttt{more} and \texttt{less}}{Command more and less}}\label{command-more-and-less}}

The command \texttt{more} is old utility. When the text passed to it is too large to fit on one screen, it pages it. You can scroll down but not up.

The syntaxt of \texttt{more} is shown below:

\begin{verbatim}
more [options] file [...]
\end{verbatim}

The command \texttt{less} was written by a man who was fed up with more's inability to scroll backwards through a file. He turned less into an open source project and over time, various individuals added new features to it. less is massive now. That's why some small embedded systems have more but not less. For comparison, less's source is over 27000 lines long. more implementations are generally only a little over 2000 lines long.

The syntaxt of \texttt{less} is shown below:

\begin{verbatim}
less [options] file [...]
\end{verbatim}

\hypertarget{command-head-and-tail}{%
\subsection{\texorpdfstring{Command \texttt{head} and \texttt{tail}}{Command head and tail}}\label{command-head-and-tail}}

The command \texttt{head} is used to output the first part of files. By default, it outputs the first 10 lines of the file.

\begin{verbatim}
head [OPTION]... [FILE]...
\end{verbatim}

Here is an exmaple of printing the first 5 files of the file:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{ -n 5 code_perl/variable_assign.pl}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## #!/usr/bin/perl
## use warnings;
## use strict;
## 
## #assign two strings to two variables
\end{verbatim}

In fact, the letter n does not even need to be used at all. Just the hyphen and the integer (with no intervening space) are sufficient to tell head how many lines to return. Thus, the following would produce the same result as the above commands:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{head}\NormalTok{ -5 data/testdata4linux_cmd.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene1
## gene2
## gene3
## gene4
## gene5
\end{verbatim}

The command \texttt{tail} is used to output the last part of files. By default, it prints the last 10 lines of the file to standard output.

The syntax is shown below:

\begin{verbatim}
tail [OPTION]... [FILE]...
\end{verbatim}

Here is an exmaple of printing the last 5 files of the file:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tail}\NormalTok{ -5 data/testdata4linux_cmd.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene12
## gene13
## gene14
## gene15
## gene16
\end{verbatim}

To view lines from a specific point in a file, you can use \texttt{-n\ +NUMBER} with the \texttt{tail} command. For example, here is an example of viewing the file from the 2nd line of the line.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tail}\NormalTok{ -n +2 data/testdata4linux_cmd.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene2
## gene3
## gene4
## gene5
## gene6
## gene7
## gene8
## gene9
## gene10
## gene11
## gene12
## gene13
## gene14
## gene15
## gene16
\end{verbatim}

\hypertarget{auto-completion}{%
\subsection{Auto-completion}\label{auto-completion}}

In most Shell environment, programmable completion feature will also improve your speed of typing. It permits typing a partial name of command or a partial file (or directory), then pressing \texttt{TAB} key to auto-complete the command (Figure \ref{fig:linuxAutoCompletion}). If there are more than one possible completions, then TAB will list all of them (Figure \ref{fig:linuxAutoCompletion}).



\begin{figure}
\centering
\includegraphics{figures/linuxAutoCompletion.png}
\caption{\label{fig:linuxAutoCompletion}Demonstration of programmable completion feature.}
\end{figure}

\hypertarget{understand-standard-input-and-stardard-output}{%
\section{Understand standard input and stardard output}\label{understand-standard-input-and-stardard-output}}

In the Linux environment, input and output is distributed across three streams: standard input (STDIN), standard output (STDOUT), standard error (STDERR). These three streams are also numbered: STDIN (0), STDOUT (1), STDERR (2).

\hypertarget{stdin}{%
\subsection{STDIN}\label{stdin}}

\ldots{}
The standard input stream typically carries data from a user to a program. Programs that expect standard input usually receive input from a device, such as a keyboard. Standard input is terminated by reaching EOF (end-of-file). As described by its name, EOF indicates that there is no more data to be read.

To see standard input in action, run the cat program. Cat stands for concatenate, which means to link or combine something. It is commonly used to combine the contents of two files. When run on its own, cat opens a looping prompt.
\ldots{}

\begin{verbatim}
tail
1
2
3
`CTRL+D`
1
2
3
\end{verbatim}

\hypertarget{stdout}{%
\subsection{STDOUT}\label{stdout}}

Data that is generated by a program will be written by STDOUT. If the STDOUT is not redirected, it will output the data on to the terminal.

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{stdout=}\StringTok{"Hello world"}
\BuiltInTok{echo} \VariableTok{$stdout}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Hello world
\end{verbatim}

The STDOUT can be redirected to a file. See the example below:

\begin{Shaded}
\begin{Highlighting}[]
\VariableTok{stdout=}\StringTok{"Hello world"}
\BuiltInTok{echo} \VariableTok{$stdout} \OperatorTok{>}\NormalTok{ data/test_output.txt}
\CommentTok{# cat the data}
\FunctionTok{cat}\NormalTok{ data/test_output.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Hello world
\end{verbatim}

\hypertarget{stderr}{%
\subsection{STDERR}\label{stderr}}

During a program's execution, some errors may be generated when the program fails at some parts. STDERR will help you write the errors. By default, the STDERR will be outputed onto the terminal.

Here is an example of STDERR

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ls}\NormalTok{ NOTAFILE}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ls: cannot access 'NOTAFILE': No such file or directory
\end{verbatim}

\hypertarget{find-disk-usage-of-files-and-directories}{%
\section{Find Disk Usage of Files and Directories}\label{find-disk-usage-of-files-and-directories}}

The Linux \texttt{du} (short for Disk Usage) is a standard Unix/Linux command, used to check the information of disk usage of files and directories on a machine. The \texttt{du} command has many parameter options that can be used to get the results in many formats. The du command also displays the files and directory sizes in a recursively manner.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{du}\NormalTok{ data/ESP6500-African_American.vcf.gz}
\FunctionTok{du}\NormalTok{ -h data/ESP6500-African_American.vcf.gz}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 27388    data/ESP6500-African_American.vcf.gz
## 27M  data/ESP6500-African_American.vcf.gz
\end{verbatim}

To get the summary of a grand total disk usage size of an directory use the option ``-s'' as follows.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{du}\NormalTok{ -sh data/}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 37M  data/
\end{verbatim}

Using ``-a'' flag with ``du'' command displays the disk usage of all the files and directories.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{du}\NormalTok{ -ah data/}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 4.0K data/PYL10_ARATH.fasta
## 4.0K data/test_ref2.fa
## 4.0K data/test_ref.fa
## 4.0K data/test_ref_len.txt
## 4.0K data/gene_annotation.txt
## 8.0K data/Pneumonia_china_2020.RDS
## 4.0K data/WGBS_sample_information.txt
## 0    data/regexp_perl.txt
## 4.0K data/test_ref2_30.fa
## 4.0K data/DMR_region_merged.txt
## 27M  data/ESP6500-African_American.vcf.gz
## 8.0K data/WGBS_example_data/EV1.fastq
## 12K  data/WGBS_example_data
## 4.0K data/DMR_region.txt
## 88K  data/maize_embryo_specific_gene_Sheet1.tsv
## 9.3M data/Arabidopsis_thaliana.TAIR10.37.gff3.gz
## 532K data/ESP6500-African_American.vcf.gz.tbi
## 4.0K data/test_output.txt
## 4.0K data/testdata4linux_cmd.txt
## 4.0K data/DEG_list.txt
## 4.0K data/README
## 37M  data/
\end{verbatim}

\hypertarget{advanced-topic}{%
\section{Advanced topic}\label{advanced-topic}}

\hypertarget{linux-md5sum-command}{%
\subsection{\texorpdfstring{Linux \texttt{md5sum} Command}{Linux md5sum Command}}\label{linux-md5sum-command}}

\texttt{md5sum} is used to verify the integrity of files, as virtually any change to a file will cause its MD5 hash to change. Most commonly, md5sum is used to verify that a file has not changed as a result of a faulty file transfer, a disk error or non-malicious meddling. The md5sum program is included in most Unix-like operating systems.

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{echo} \StringTok{"The MD5 value of index.Rmd is: "}
\ExtensionTok{md5sum}\NormalTok{ index.Rmd}
\FunctionTok{cp}\NormalTok{ index.Rmd index.Rmd_bak}
\BuiltInTok{echo} \StringTok{"The MD5 value of index.Rmd_bak is: "}
\ExtensionTok{md5sum}\NormalTok{ index.Rmd_bak }
\BuiltInTok{echo} \StringTok{"The MD5 value of new index.Rmd_bak is: "}
\FunctionTok{head}\NormalTok{ index.Rmd }\OperatorTok{>}\NormalTok{ index.Rmd_bak}
\ExtensionTok{md5sum}\NormalTok{ index.Rmd_bak }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## The MD5 value of index.Rmd is: 
## c33df2aa9b5eb7f181fb7b35f2df8fbc  index.Rmd
## The MD5 value of index.Rmd_bak is: 
## c33df2aa9b5eb7f181fb7b35f2df8fbc  index.Rmd_bak
## The MD5 value of new index.Rmd_bak is: 
## 6c6d75e8839891bf7ba1ab152c8f267c  index.Rmd_bak
\end{verbatim}

\hypertarget{file-content-filtering}{%
\chapter{File content filtering}\label{file-content-filtering}}

\hypertarget{file-filtering}{%
\section{File Filtering}\label{file-filtering}}

\hypertarget{column-filtering}{%
\subsection{Column filtering}\label{column-filtering}}

\hypertarget{cut}{%
\subsection{\texorpdfstring{\texttt{cut}}{cut}}\label{cut}}

\texttt{cut} can be used to print selected parts of lines from each FILE to standard output.

cut sort uniq wc grep

\url{https://www.youtube.com/playlist?list=PLtK75qxsQaMLZSo7KL-PmiRarU7hrpnwK}

\hypertarget{row-filtering}{%
\subsection{Row filtering}\label{row-filtering}}

\hypertarget{grep}{%
\subsubsection{\texorpdfstring{\texttt{grep}}{grep}}\label{grep}}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

The \texttt{grep} command which stands for ``global regular expression print,'' processes text line by line and prints any lines which match a specified pattern. The grep command is used to search text or searches the given file for lines containing a match to the given strings or words. By default, grep displays the matching lines.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{grep} \StringTok{'WRKY'}\NormalTok{ data/gene_annotation.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene2 WRKY
## gene4 WRKY1
## gene5 WRKY2
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{grep} \StringTok{'WRKY'}\NormalTok{ data/gene_annotation.txt }\KeywordTok{|}\FunctionTok{wc}\NormalTok{ -l }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{grep}\NormalTok{ -i }\StringTok{'WRKY'}\NormalTok{ data/gene_annotation.txt }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene2 WRKY
## gene4 WRKY1
## gene5 WRKY2
## gene6 wrky
\end{verbatim}

If you want to search for a word, and avoid matching substrings use `-w `option.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{grep} \StringTok{'gene1'}\NormalTok{ data/gene_annotation.txt }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene1    ROS
## gene10  MCU
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{grep}\NormalTok{ -w }\StringTok{'gene1'}\NormalTok{ data/gene_annotation.txt }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene1    ROS
\end{verbatim}

\hypertarget{awk}{%
\subsubsection{\texorpdfstring{\texttt{awk}}{awk}}\label{awk}}

\hypertarget{finding-things}{%
\section{Finding Things}\label{finding-things}}

\hypertarget{find-files-with-pattern-matching}{%
\subsection{Find files with pattern matching}\label{find-files-with-pattern-matching}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## Find any files with "Linux" and ".Rmd" in the file names}
\FunctionTok{find}\NormalTok{ . -type f -name }\StringTok{"*Linux*.Rmd"}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ./slides/slides01_Linux.Rmd
## ./04_Linux_FilteringOutputandFindingThings.Rmd
## ./01_WhyLinux.Rmd
## ./02_Connect2Linux.Rmd
## ./09_TextEditorInLinux.Rmd
## ./08_InstallationOfSoftwareInLinux.Rmd
## ./06_procManageLinux.Rmd
## ./03_FileSystemLinux.Rmd
\end{verbatim}

\hypertarget{count-file-numbers-in-a-folder-and-its-subdirectories}{%
\subsection{Count file numbers in a folder and its subdirectories}\label{count-file-numbers-in-a-folder-and-its-subdirectories}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{find}\NormalTok{ . -type f }\KeywordTok{|} \FunctionTok{wc}\NormalTok{ -l}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 1608
\end{verbatim}

\hypertarget{list-files-bigger-than-filesize-specified}{%
\subsection{List files bigger than filesize specified}\label{list-files-bigger-than-filesize-specified}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#To find files larger than 10MB:}
\FunctionTok{find}\NormalTok{ . -type f -size +100M}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ./.git/objects/pack/pack-b6b6179d7bb7190bc99389e158b2a8e4cbd98d2f.pack
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# If you want the current dir only:}
\FunctionTok{find}\NormalTok{ . -maxdepth 1 -type f -size +1M}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ./download
\end{verbatim}

\hypertarget{find-files-and-do-someting}{%
\subsection{Find files and do someting}\label{find-files-and-do-someting}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{find}\NormalTok{ . -type f -name }\StringTok{"*fa*.pl"}\NormalTok{ -exec ls -l }\DataTypeTok{\{\}}\NormalTok{ +}\KeywordTok{;} 
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## -rw-rw-r-- 1 rstudio-user rstudio-user 642 Apr 15  2019 ./code_perl/fa_seq_len.pl
## -rw-rw-r-- 1 rstudio-user rstudio-user 809 Apr 15  2019 ./code_perl/fa_seq_len_out.pl
## -rw-rw-r-- 1 rstudio-user rstudio-user 810 Apr 15  2019 ./code_perl/fa_seq_len_out_argv.pl
## -rw-rw-r-- 1 rstudio-user rstudio-user 879 Apr 15  2019 ./code_perl/fa_seq_len_out_argv_fil.pl
\end{verbatim}

\hypertarget{check-you-job-status}{%
\section{Check you job status}\label{check-you-job-status}}

\begin{verbatim}
$ ps aux  
USER       PID  %CPU %MEM  VSZ RSS     TTY   STAT START   TIME COMMAND
timothy  29217  0.0  0.0 11916 4560 pts/21   S+   08:15   0:00 pine  
root     29505  0.0  0.0 38196 2728 ?        Ss   Mar07   0:00 sshd: can [priv]   
can      29529  0.0  0.0 38332 1904 ?        S    Mar07   0:00 sshd: can@notty  
\end{verbatim}

USER = user owning the process
PID = process ID of the process
\%CPU = It is the CPU time used divided by the time the process has been running.
\%MEM = ratio of the process's resident set size to the physical memory on the machine
VSZ = virtual memory usage of entire process (in KiB)
RSS = resident set size, the non-swapped physical memory that a task has used (in KiB)
TTY = controlling tty (terminal)
STAT = multi-character process state
START = starting time or date of the process
TIME = cumulative CPU time
COMMAND = command with all its arguments

References:

\url{https://superuser.com/questions/117913/ps-aux-output-meaning}

\hypertarget{achiving-and-compressing-files}{%
\chapter{Achiving and compressing files}\label{achiving-and-compressing-files}}

\hypertarget{common-compressed-file-format}{%
\section{Common compressed file format}\label{common-compressed-file-format}}

If you download open source software, like \texttt{bwa} and \texttt{ViewBS}, you will encouther archived files very often.

The common compressed file usually have the the suffix of \texttt{tar.gz} (equaivalent of \texttt{tgz}), \texttt{gz}, \texttt{zip} and \texttt{tar.bz2}.

For example, if we want to use \texttt{bowtie2} in Linux, we need to download the \texttt{bowtie2} software. Bowtie2 file for Linux can be downloaded form the link below:

\begin{verbatim}
https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.3.3.1/bowtie2-2.3.3.1-linux-x86_64.zip
\end{verbatim}

In side the \texttt{zip} file, there are files and sub folders.

The files for \texttt{samtools\ v1.6} are archived in to a file named samtools-1.6.tar.bz2`. From the link below you can download the link:

\begin{verbatim}
https://gigenet.dl.sourceforge.net/project/samtools/samtools/1.6/samtools-1.6.tar.bz2
\end{verbatim}

\hypertarget{how-to-work-with-different-format}{%
\section{How to work with different format}\label{how-to-work-with-different-format}}

\hypertarget{gz}{%
\subsection{\texorpdfstring{\texttt{*.gz}}{*.gz}}\label{gz}}

\hypertarget{how-to-check-the-file}{%
\subsubsection{How to check the file}\label{how-to-check-the-file}}

\begin{verbatim}
zcat file.gz |less 
\end{verbatim}

\hypertarget{how-to-test-if-a-gzip-file-is-valid}{%
\subsubsection{\texorpdfstring{How to test if a \texttt{gzip} file is valid}{How to test if a gzip file is valid}}\label{how-to-test-if-a-gzip-file-is-valid}}

\begin{verbatim}
gzip -t file.gz
\end{verbatim}

\hypertarget{how-decompress-a-.gz-file}{%
\subsubsection{\texorpdfstring{How decompress a \texttt{*.gz} file}{How decompress a *.gz file}}\label{how-decompress-a-.gz-file}}

\begin{itemize}
\tightlist
\item
  Decompress a \texttt{*.gz} file
\end{itemize}

\begin{verbatim}
mkdir tmp
cp tmp
cp ../data/Arabidopsis_thaliana.TAIR10.37.gff3.gz ./
gunzip Arabidopsis_thaliana.TAIR10.37.gff3.gz
\end{verbatim}

\begin{itemize}
\tightlist
\item
  Decompress a file and keep the original copy
\end{itemize}

\begin{verbatim}
gunzip -c file.gz > file
\end{verbatim}

\hypertarget{tar.gz}{%
\subsection{\texorpdfstring{\texttt{*.tar.gz}}{*.tar.gz}}\label{tar.gz}}

\hypertarget{how-to-decompress-the-.tar.gz-file}{%
\subsubsection{\texorpdfstring{How to decompress the \texttt{*.tar.gz} file}{How to decompress the *.tar.gz file}}\label{how-to-decompress-the-.tar.gz-file}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tar}\NormalTok{ zxvf file.tar.gz}
\end{Highlighting}
\end{Shaded}

\begin{itemize}
\tightlist
\item
  \texttt{z} means (un)z̲ip.
\item
  \texttt{x} means ex̲tract files from the archive.
\item
  \texttt{v} means print the filenames v̲erbosely.
\item
  \texttt{f} means the following argument is a f̱ilename.
\end{itemize}

\hypertarget{how-to-view-the-content-without-extract-the-files}{%
\subsubsection{How to view the content without extract the files}\label{how-to-view-the-content-without-extract-the-files}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tar}\NormalTok{ -tf file.tar.gz}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tar}\NormalTok{ -tvf file.tar.gz}
\end{Highlighting}
\end{Shaded}

\hypertarget{how-to-create-a-.tar.gz-file}{%
\subsubsection{\texorpdfstring{How to create a \texttt{*.tar.gz} file}{How to create a *.tar.gz file}}\label{how-to-create-a-.tar.gz-file}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{tar}\NormalTok{ zcvf file_new.tar.gz file1 file2 folder1 folder2}
\end{Highlighting}
\end{Shaded}

\hypertarget{how-to-extract-a-specific-file-from-a-.tar.gz-file}{%
\subsubsection{\texorpdfstring{How to extract a specific file from a \texttt{*.tar.gz} file}{How to extract a specific file from a *.tar.gz file}}\label{how-to-extract-a-specific-file-from-a-.tar.gz-file}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#tar -xvf \{tarball.tar\} \{path/to/file\}}
\FunctionTok{tar}\NormalTok{ -zxvf config.tar.gz etc/default/sysstat}
\end{Highlighting}
\end{Shaded}

\hypertarget{zip}{%
\subsection{\texorpdfstring{\texttt{*.zip}}{*.zip}}\label{zip}}

\hypertarget{how-to-unzip-a-.zip-file}{%
\subsubsection{\texorpdfstring{How to unzip a \texttt{*.zip} file}{How to unzip a *.zip file}}\label{how-to-unzip-a-.zip-file}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{unzip}\NormalTok{ file.zip}
\end{Highlighting}
\end{Shaded}

\hypertarget{how-to-create-a-.zip-file}{%
\subsubsection{\texorpdfstring{How to create a \texttt{*.zip} file}{How to create a *.zip file}}\label{how-to-create-a-.zip-file}}

\begin{verbatim}
zip file_new.zip file1 file2 folder1 folder2
\end{verbatim}

\hypertarget{tar.bz2}{%
\subsection{\texorpdfstring{\texttt{*.tar.bz2}}{*.tar.bz2}}\label{tar.bz2}}

\hypertarget{how-to-decompress-a-.tar.bz2-file}{%
\subsubsection{\texorpdfstring{How to decompress a \texttt{*.tar.bz2} file}{How to decompress a *.tar.bz2 file}}\label{how-to-decompress-a-.tar.bz2-file}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{mkdir}\NormalTok{ tmp}
\BuiltInTok{cd}\NormalTok{ tmp}
\BuiltInTok{pwd}
\FunctionTok{ls} 
\FunctionTok{wget}\NormalTok{ https://downloads.sourceforge.net/project/bio-bwa/bwa-0.7.17.tar.bz2}
\FunctionTok{ls}
\FunctionTok{tar}\NormalTok{ -jxvf bwa-0.7.17.tar.bz2}
\end{Highlighting}
\end{Shaded}

\hypertarget{how-to-create-a-.tar.bz2-file}{%
\subsubsection{\texorpdfstring{How to create a \texttt{*.tar.bz2} file}{How to create a *.tar.bz2 file}}\label{how-to-create-a-.tar.bz2-file}}

\begin{verbatim}
tar -cvjSf folder.tar.bz2 file1 file2 folder1 folder2
\end{verbatim}

\hypertarget{process-management-in-linux}{%
\chapter{Process management in Linux}\label{process-management-in-linux}}

\hypertarget{top}{%
\section{\texorpdfstring{\texttt{top}}{top}}\label{top}}

The \texttt{top} program provides a dynamic real-time view of a running system.

Usually \texttt{top} is used with the option \texttt{-c}.

\begin{verbatim}
top -c
\end{verbatim}

The option \texttt{-c} will let \texttt{top} to displat the full command path along with the command arguments in the \texttt{COMMAND} collumn.

You can also run \texttt{top} interactively. You can run \texttt{top} first and then press \texttt{c}. If you want to kill a process with \texttt{PID} of \texttt{186}, you can press \texttt{k} and then type \texttt{186} to kill the process with \texttt{PID} of 186.

\texttt{man\ top} can help you get the manual of command \texttt{top}.

The following table explains what each column mean.

\begin{longtable}[]{@{}ll@{}}
\toprule
\begin{minipage}[b]{0.05\columnwidth}\raggedright
Columns\strut
\end{minipage} & \begin{minipage}[b]{0.89\columnwidth}\raggedright
Description\strut
\end{minipage}\tabularnewline
\midrule
\endhead
\begin{minipage}[t]{0.05\columnwidth}\raggedright
PID\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Process ID\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
USER\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Name of the effective user (owner) of the process\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
PR\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Priority\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
NI\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Nice value\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
VIRT\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Virtual memory size\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
RES\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Resident memory size\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
SHR\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Shared memory size\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
S\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Process status (which could be one of the following: D (uninteruptible sleep), R (running), S (sleeping), T (traced or stopped) or Z (zombie)\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
\%CPU\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
The share of cpu time used by the process since last update\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
\%MEM\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Share of physical memory used\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
TIME+\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Total cpu time used by the task in hundredths of a second\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.05\columnwidth}\raggedright
COMMAND\strut
\end{minipage} & \begin{minipage}[t]{0.89\columnwidth}\raggedright
Command name or command line (name + options)\strut
\end{minipage}\tabularnewline
\bottomrule
\end{longtable}

\hypertarget{ps}{%
\section{\texorpdfstring{\texttt{ps}}{ps}}\label{ps}}

The command \texttt{ps} can report a snapshot of the current processes.

Command \texttt{ps} is usually used with the option \texttt{-a}, \texttt{-u} and \texttt{-x}.

\begin{verbatim}
ps -aux   ## can also be `ps aux`
\end{verbatim}

You can pipe the output to \texttt{less} to make it scrollable.

\hypertarget{kill}{%
\section{\texorpdfstring{\texttt{kill}}{kill}}\label{kill}}

If you want to kill a process, you can use the command \texttt{kill}.

\begin{verbatim}
kill 20140418
\end{verbatim}

\hypertarget{df}{%
\section{\texorpdfstring{\texttt{df}}{df}}\label{df}}

\hypertarget{advanced-topic-free}{%
\section{\texorpdfstring{Advanced topic \texttt{free}}{Advanced topic free}}\label{advanced-topic-free}}

You can use command \texttt{free} to display amount of free and used memory in the system.

\begin{verbatim}
free -h
\end{verbatim}

\texttt{-h} let you show all output fields automatically scaled to shortest three digit unit and display the units of print out. Following units are used.

\begin{longtable}[]{@{}ll@{}}
\toprule
Abbreviation & Full Name\tabularnewline
\midrule
\endhead
B & Bytes\tabularnewline
K & Kilobytes (KB)\tabularnewline
M & Megabytes(MB)\tabularnewline
G & Gigabytes (GB)\tabularnewline
T & Terabytes (TB)\tabularnewline
\bottomrule
\end{longtable}

\hypertarget{commands-for-linux-administration-advanced-topic}{%
\section{Commands for Linux administration (Advanced topic)}\label{commands-for-linux-administration-advanced-topic}}

\hypertarget{w}{%
\section{\texorpdfstring{\texttt{w}}{w}}\label{w}}

\hypertarget{who}{%
\section{\texorpdfstring{\texttt{who}}{who}}\label{who}}

\hypertarget{uptime}{%
\section{\texorpdfstring{\texttt{uptime}}{uptime}}\label{uptime}}

In Linux uptime command shows since how long your system is running and the number of users are currently logged in and also displays load average for 1,5 and 15 minutes intervals.

\hypertarget{whoami}{%
\section{\texorpdfstring{\texttt{whoami}}{whoami}}\label{whoami}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{whoami}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## rstudio-user
\end{verbatim}

\hypertarget{ifconfig}{%
\section{\texorpdfstring{\texttt{ifconfig}}{ifconfig}}\label{ifconfig}}

\begin{verbatim}
ifconfig
\end{verbatim}

\hypertarget{useradd-and-passwd}{%
\section{\texorpdfstring{\texttt{useradd} and \texttt{passwd}}{useradd and passwd}}\label{useradd-and-passwd}}

\begin{verbatim}
## Need to have root access
adduser superomics 
\end{verbatim}

\begin{verbatim}
## add a user to a specified group
adduser superomics -g bioinf
\end{verbatim}

\hypertarget{file-transfer}{%
\chapter{File transfer}\label{file-transfer}}

\hypertarget{transferring-files-between-local-computer-and-linux-server}{%
\section{Transferring files between local computer and Linux server}\label{transferring-files-between-local-computer-and-linux-server}}

To transfer files between local computer and Linux sever, there are two options: 1) GUI application and 2) command line.

\begin{itemize}
\tightlist
\item
  Open FileZilla and then click \texttt{File} -\textgreater{} \texttt{Site\ Manager}.
\end{itemize}

GUI means there will be window, buttons, menus, etc. The most popular system with GUI is Windows system (Figure @ref(fig:filezilla\_screenshot1)).

(ref:filezilla\_screenshot1) FileZilla application.

\begin{figure}
\centering
\includegraphics{figures/filezilla_screenshot1.png}
\caption{(\#fig:filezilla\_screenshot1)(ref:filezilla\_screenshot1)}
\end{figure}

\hypertarget{use-command-line-tools}{%
\subsection{Use command line tools}\label{use-command-line-tools}}

\texttt{rsync} compares the files at each end and transfers only the changed parts of changed files. When you transfer files the first timeo it behaves pretty much like scp, but for a second transfer, where most files are unchanged, it will push a lot less data than scp. It's also a convenient way to restart failed transfers - you just reissue the same command and it will pick up where it left off the time before, whereas scp will start again from scratch.

\hypertarget{copy-files-using-rsync}{%
\subsubsection{\texorpdfstring{Copy files using \texttt{rsync}}{Copy files using rsync}}\label{copy-files-using-rsync}}

\begin{verbatim}
\end{verbatim}

\hypertarget{copying-files-with-scp}{%
\subsubsection{\texorpdfstring{Copying Files with \texttt{scp}}{Copying Files with scp}}\label{copying-files-with-scp}}

The command \texttt{scp} is short for secure copy. It can be used to copy files between hosts on a network. It uses ssh(1) for data transfer, and uses the same authentication and provides the same security as ssh(1).\texttt{Scp} will ask for passwords or passphrases if they are needed for authentication.

File names may contain a user and host specification to indicate that the file is to be copied to/from that host. Local file names can be made explicit using absolute or relative pathnames to avoid scp treating file names containing `:' as host specifiers. Copies between two remote hosts are also permitted.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Copy the file test.pl on 198.211.107.37 to the current directory.}
\FunctionTok{scp}\NormalTok{ guest4bioinfor@198.211.107.37:~/test.pl ./}
\end{Highlighting}
\end{Shaded}

To copy files from a server to a client, you need to know where the files are located on the server. For example, to copy a single file \texttt{\textasciitilde{}/test.pl} from the server with IP address of 198.211.107.37 to the current directory.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Copy the file test.pl in the current directory to  198.211.107.37}
\FunctionTok{scp}\NormalTok{ ./test.pl guest4bioinfor@198.211.107.37:~/}
\end{Highlighting}
\end{Shaded}

To copy files from a client to a server, you need to know where the files you want to put on the server. For example, to copy a single file \texttt{test.pl} from the current folder to the HOME folder of the server with IP address of 198.211.107.37.

If you want to copy an entire directory recursively, you can use \texttt{-r} argument. See the example below:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Copy the file test.pl in the current directory to  198.211.107.37}
\FunctionTok{scp}\NormalTok{ -r guest4bioinfor@198.211.107.37:~/bioinfo/ ./}
\end{Highlighting}
\end{Shaded}

\hypertarget{download-files}{%
\subsection{Download files}\label{download-files}}

\begin{verbatim}
wget <url>
\end{verbatim}

Resume

\begin{verbatim}
wget -c <url>  
\end{verbatim}

Reference:

RH066x Fundamentals of Red Hat Enterprise Linux on edX

\hypertarget{install-bioinformatics-software-in-linux}{%
\chapter{Install Bioinformatics software in Linux}\label{install-bioinformatics-software-in-linux}}

\hypertarget{installation-from-source-code}{%
\section{Installation from source code}\label{installation-from-source-code}}

Nearly all of the Bioinformatics softwares will be downloaded as a compressed files. So the first thing you need to do is to uncompress the file. Then the source codes will be included in a folder. You can \texttt{cd} to the folder and \texttt{ls} the files/directories. Mostly you will find either a file named \texttt{README} or \texttt{INSTALL} or both. If you read this file to know how to install the software.

\hypertarget{install-bwa}{%
\subsection{\texorpdfstring{Install \texttt{bwa}}{Install bwa}}\label{install-bwa}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{wget}\NormalTok{ https://sourceforge.net/projects/bio-bwa/files/bwa-0.7.15.tar.bz2}
\FunctionTok{tar}\NormalTok{ xjvf bwa-0.7.15.tar.bz2}
\BuiltInTok{cd}\NormalTok{ bwa-0.7.15}
\FunctionTok{make}
\end{Highlighting}
\end{Shaded}

\hypertarget{install-samtools}{%
\subsection{\texorpdfstring{Install \texttt{samtools}}{Install samtools}}\label{install-samtools}}

Installation of \texttt{Samtools} is one of the best representatives of how to instsall a Bioinformatics tool.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Download the source code}
\FunctionTok{wget}\NormalTok{ https://iweb.dl.sourceforge.net/project/samtools/samtools/1.3.1/samtools-1.3.1.tar.bz2}
\CommentTok{# Uncompress the source code}
\FunctionTok{tar}\NormalTok{ xjvf samtools-1.3.1.tar.bz2}
\CommentTok{# Enter the source code directory.}
\BuiltInTok{cd}\NormalTok{ samtools-1.3.1}
\CommentTok{# Configure the build system}
\ExtensionTok{./configure}
\CommentTok{# Build samtools}
\FunctionTok{make}
\CommentTok{# Become a `root` user for system-wide install:}
\FunctionTok{su}\NormalTok{ root}
\CommentTok{# Install `Samtools`}
\FunctionTok{make}\NormalTok{ install}
\end{Highlighting}
\end{Shaded}

\BeginKnitrBlock{rmdtip}
\textbf{Install \texttt{samtools} without root previledges}

By default, `make install' installs samtools and the utilities under
/usr/local/bin and manual pages under /usr/local/share/man.

You can specify a different location to install Samtools by configuring
with --prefix=DIR or specify locations for particular parts of HTSlib by
configuring with --bindir=DIR and so on. Type `./configure --help' for
the full list of such install directory options.

Alternatively you can specify different locations at install time by
typing `make prefix=DIR install' or `make bindir=DIR install' and so on.
Consult the list of prefix/exec\_prefix/etc variables near the top of the
Makefile for the full list of such variables that can be overridden.

You can also specify a staging area by typing `make DESTDIR=DIR install',
possibly in conjunction with other --prefix or prefix=DIR settings.
For example,

\begin{verbatim}
make DESTDIR=/tmp/staging prefix=/opt
\end{verbatim}

would install into bin and share/man subdirectories under /tmp/staging/opt.
\EndKnitrBlock{rmdtip}

\hypertarget{align-reads-to-genome-using-bwa-and-store-the-alignment-results-in-sambam-files}{%
\subsection{\texorpdfstring{Align reads to genome using \texttt{bwa} and store the alignment results in SAM/BAM files}{Align reads to genome using bwa and store the alignment results in SAM/BAM files}}\label{align-reads-to-genome-using-bwa-and-store-the-alignment-results-in-sambam-files}}

\begin{verbatim}
./bwa index ref.fa
./bwa mem ref.fa read-se.fq.gz | gzip -3 > aln-se.sam.gz
./bwa mem ref.fa read1.fq read2.fq | gzip -3 > aln-pe.sam.gz
\end{verbatim}

\hypertarget{installing-a-precompiled-binary-executable}{%
\section{Installing a precompiled binary (executable)}\label{installing-a-precompiled-binary-executable}}

For programs that are already compiled (converted from high level source code in a language like C into machine specific code), you are often given some choices and need to determine how to download the version that has the correct CPU architecture for your machine.

\hypertarget{install-bwa-1}{%
\subsection{\texorpdfstring{Install \texttt{bwa}}{Install bwa}}\label{install-bwa-1}}

\begin{verbatim}
wget https://downloads.sourceforge.net/project/bio-bwa/bwakit/bwakit-0.7.15_x64-linux.tar.bz2
tar xjvf bwakit-0.7.15_x64-linux.tar.bz2
cd bwa.kit/
./bwa
Program: bwa (alignment via Burrows-Wheeler transformation)
Version: 0.7.15-r1140
Contact: Heng Li <lh3@sanger.ac.uk>

Usage:   bwa <command> [options]

Command: index         index sequences in the FASTA format
         mem           BWA-MEM algorithm
         fastmap       identify super-maximal exact matches
         pemerge       merge overlapping paired ends (EXPERIMENTAL)
         aln           gapped/ungapped alignment
         samse         generate alignment (single ended)
         sampe         generate alignment (paired ended)
         bwasw         BWA-SW for long queries

         shm           manage indices in shared memory
         fa2pac        convert FASTA to PAC format
         pac2bwt       generate BWT from PAC
         pac2bwtgen    alternative algorithm for generating BWT
         bwtupdate     update .bwt to the new format
         bwt2sa        generate SA from BWT and Occ

Note: To use BWA, you need to first index the genome with `bwa index'.
      There are three alignment algorithms in BWA: `mem', `bwasw', and
      `aln/samse/sampe'. If you are not sure which to use, try `bwa mem'
      first. Please `man ./bwa.1' for the manual.
\end{verbatim}

\hypertarget{install-with-conda-recommended-way}{%
\subsection{\texorpdfstring{Install with \texttt{conda} (recommended way)}{Install with conda (recommended way)}}\label{install-with-conda-recommended-way}}

\hypertarget{install-conda}{%
\subsubsection{\texorpdfstring{Install \texttt{conda}}{Install conda}}\label{install-conda}}

Go the web link here (\url{https://conda.io/en/latest/miniconda.html}):

\begin{verbatim}
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
sh Miniconda3-latest-Linux-x86_64.sh
\end{verbatim}

\hypertarget{install-a-software-package-in-a-envrionment}{%
\subsubsection{Install a software package in a envrionment}\label{install-a-software-package-in-a-envrionment}}

Google-search ``conda bwa'', click the first hit and copy the command lines to install \texttt{bwa}.

\begin{verbatim}
conda install -c bioconda bwa 
#conda install -c bioconda/label/cf201901 bwa 
\end{verbatim}

\hypertarget{install-ussing-docker-advanced-topic}{%
\subsection{\texorpdfstring{Install ussing \texttt{Docker} (Advanced topic)}{Install ussing Docker (Advanced topic)}}\label{install-ussing-docker-advanced-topic}}

\hypertarget{text-editor-in-linux}{%
\chapter{Text editor in Linux}\label{text-editor-in-linux}}

In Linux, we sometimes need to create or edit a text file like writing a new perl script. So we need to use text editor.

As a newbie, someone would prefer a basic, GUI-based text editor with menus and traditional CUA key bindings. Here we recommend \href{https://www.sublimetext.com/}{Sublime}, \href{https://atom.io}{ATOM} and \href{https://notepad-plus-plus.org/}{Notepad++}.

But GUI-based text editor is not always available in Linux.

A powerful screen text editor \texttt{vi} (pronounced ``vee-eye'') is available on nearly all Linux system. We highly recommend \texttt{vi} as a text editor, because something we'll have to edit a text file on a system without a friendlier text editor. Once we get familiar with \texttt{vi}, we'll find that it's very fast and powerful.

But remember, it's OK if you think this part is too difficult at the beginning. You can use either \texttt{Sublime}, \texttt{ATOM} or \texttt{Notepad++}. If you are connecting to a Linux system without \texttt{Sublime}, \texttt{ATOM} and \texttt{Notepad++}, you can write the file in a local computer and then upload the file onto Linux system.

\hypertarget{basic-vi-skills}{%
\section{Basic vi skills}\label{basic-vi-skills}}

As \texttt{vi} uses a lot of combination of keystrokes, it may be not easy for newbies to remember all the combinations in one fell swoop. Considering this, we'll first introduce the basic skills someone needs to know to use \texttt{vi}. We need to first understand how three modes of \texttt{vi} work and then try to remember a few basic \texttt{vi} commonds. Then we can use these skills to write Perl or R scripts in the following chaptors for Perl and R (Figure \ref{fig:workingModeVi}).



\begin{figure}
\centering
\includegraphics{figures/workingModeVi.png}
\caption{\label{fig:workingModeVi}Three modes of vi.}
\end{figure}

\hypertarget{create-new-text-file-with-vi}{%
\section{\texorpdfstring{Create new text file with \texttt{vi}}{Create new text file with vi}}\label{create-new-text-file-with-vi}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{mkdir}\NormalTok{ test_vi  ## generate a new folder}
\BuiltInTok{cd}\NormalTok{ test_vi     ## go into the new folder}
\BuiltInTok{echo} \StringTok{"Using }\DataTypeTok{\textbackslash{}`}\StringTok{ls}\DataTypeTok{\textbackslash{}`}\StringTok{ we don't expect files in this folder."}
\FunctionTok{ls} 
\BuiltInTok{echo} \StringTok{"No file displayed!"}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Using `ls` we don't expect files in this folder.
## No file displayed!
\end{verbatim}

Using the code above, we made a new directory named \texttt{test\_vi}. We didn't see any file.

If we type \texttt{vi\ test.pl}, an empty file and screen are created into which you may enter text because the file does not exist((Figure \ref{fig:ViNewFile})).

\begin{verbatim}
vi test.pl
\end{verbatim}



\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/vi_new_file} 

}

\caption{A screentshot of the \texttt{vi\ test.pl}.}\label{fig:ViNewFile}
\end{figure}

Now if you are in \texttt{vi\ mode}. To go to \texttt{Input\ mode}, you can type \texttt{i}, `a' or `o' (Figure \ref{fig:ViInpuMode}).



\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/vi_input_mode} 

}

\caption{A screentshot of the \texttt{vi\ test.pl}.}\label{fig:ViInpuMode}
\end{figure}

Now you can type the content (codes or other information) (\ref{fig:ViInpuType}).



\begin{figure}

{\centering \includegraphics[width=1\linewidth]{images/vi_input_type} 

}

\caption{A screentshot of the \texttt{vi\ test.pl}.}\label{fig:ViInpuType}
\end{figure}

Once you are done typing. You need to go to \texttt{Command\ mode}(Figure \ref{fig:workingModeVi}) if you want to save and exit the file. To do this, you need to press \texttt{ESC} button on the keyboard.

Now we just wrote a Perl script. We can run this script.

\begin{Shaded}
\begin{Highlighting}[]

\FunctionTok{perl}\NormalTok{ test.pl}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Hello Bioinformatics World!
\end{verbatim}

\hypertarget{an-example-for-using-editor-r}{%
\section{An example for using editor R}\label{an-example-for-using-editor-r}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{qnorm}\NormalTok{(.}\DecValTok{975}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1.959964
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{xval<-}\KeywordTok{seq}\NormalTok{(}\OperatorTok{-}\FloatTok{3.2}\NormalTok{,}\FloatTok{3.2}\NormalTok{, }\DataTypeTok{length=}\DecValTok{1000}\NormalTok{)}
\NormalTok{yval<-}\KeywordTok{dnorm}\NormalTok{(xval)}
\KeywordTok{plot}\NormalTok{(xval, yval, }\DataTypeTok{type=}\StringTok{"l"}\NormalTok{,}\DataTypeTok{axes=}\NormalTok{T,}\DataTypeTok{lwd=}\DecValTok{3}\NormalTok{,}\DataTypeTok{xlab=}\StringTok{""}\NormalTok{,}\DataTypeTok{ylab=}\StringTok{""}\NormalTok{)}
\NormalTok{x<-}\KeywordTok{seq}\NormalTok{(}\KeywordTok{qnorm}\NormalTok{(.}\DecValTok{975}\NormalTok{), }\FloatTok{3.2}\NormalTok{, }\DataTypeTok{length =} \DecValTok{100}\NormalTok{)}
\KeywordTok{polygon}\NormalTok{(}\KeywordTok{c}\NormalTok{(x,}\KeywordTok{rev}\NormalTok{(x)), }\KeywordTok{c}\NormalTok{(}\KeywordTok{dnorm}\NormalTok{(x), }\KeywordTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{,}\KeywordTok{length}\NormalTok{(x))), }\DataTypeTok{col=}\StringTok{"salmon"}\NormalTok{)}
\KeywordTok{text}\NormalTok{(}\KeywordTok{mean}\NormalTok{(x),}\KeywordTok{mean}\NormalTok{(}\KeywordTok{dnorm}\NormalTok{(x))}\OperatorTok{+}\FloatTok{0.02}\NormalTok{, }\StringTok{"2.5%"}\NormalTok{, }\DataTypeTok{cex=}\DecValTok{2}\NormalTok{)}
\KeywordTok{text}\NormalTok{(}\KeywordTok{qnorm}\NormalTok{(.}\DecValTok{95}\NormalTok{), }\FloatTok{0.01}\NormalTok{, }\StringTok{"1.645"}\NormalTok{,}\DataTypeTok{cex=}\DecValTok{2}\NormalTok{)}

\NormalTok{x<-}\KeywordTok{seq}\NormalTok{(}\OperatorTok{-}\FloatTok{3.2}\NormalTok{, }\KeywordTok{qnorm}\NormalTok{(.}\DecValTok{025}\NormalTok{), }\DataTypeTok{length =}\DecValTok{100}\NormalTok{)}
\KeywordTok{polygon}\NormalTok{(}\KeywordTok{c}\NormalTok{(x,}\KeywordTok{rev}\NormalTok{(x)), }\KeywordTok{c}\NormalTok{(}\KeywordTok{dnorm}\NormalTok{(x), }\KeywordTok{rep}\NormalTok{(}\DecValTok{0}\NormalTok{,}\KeywordTok{length}\NormalTok{(x))), }\DataTypeTok{col=}\StringTok{"salmon"}\NormalTok{)}
\KeywordTok{text}\NormalTok{(}\KeywordTok{mean}\NormalTok{(x),}\KeywordTok{mean}\NormalTok{(}\KeywordTok{dnorm}\NormalTok{(x))}\OperatorTok{+}\FloatTok{0.02}\NormalTok{, }\StringTok{"2.5%"}\NormalTok{, }\DataTypeTok{cex=}\DecValTok{2}\NormalTok{)}
\KeywordTok{text}\NormalTok{(}\KeywordTok{qnorm}\NormalTok{(.}\DecValTok{025}\NormalTok{), }\FloatTok{0.01}\NormalTok{, }\StringTok{"1.645"}\NormalTok{,}\DataTypeTok{cex=}\DecValTok{2}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-58-1.pdf}

\hypertarget{part-perl}{%
\part{Perl}\label{part-perl}}

\hypertarget{first-perl-program}{%
\chapter{First Perl Program}\label{first-perl-program}}

Scripting languages, like Perl, are very commonly used in bioinformatics. As a generous scripting language, Perl have many advantages: easy to use, free for all operating systems like Linux, designed for working with text files (tab-delimited files). It's one of the most popular language in bioinformatics. Moreover there are many scripts and modules available. Additionally, there are a lot of resource on Internet.

\hypertarget{first-program}{%
\section{First Program}\label{first-program}}

As all other programming books, we begin with a ``Hello world'' program.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\CommentTok{#Printing a line of text “Hello, Bioinformatics”}
\FunctionTok{print} \KeywordTok{"}\StringTok{Hello, Bioinformatics!}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

This program show how to display a line a text in Perl. It have several features. We go through each line in detail.

Line 1 is what we call shebang line. This line starts with shebang construct (\texttt{\#!}). \texttt{/usr/bin/perl} indicates the path of the Perl interpreter.

Line 3 shows how to print a line of text in Perl. Nearly all programming language use print to display texts on the screen. Here, print is a built-in function in Perl. It print the string of characters (its arguments) between quotation marks (``'' or `').

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{perl}\NormalTok{ code_perl/hello_bioinfor.pl}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Hello, Bioinformatics!
\end{verbatim}

However the characters \texttt{\textbackslash{}n} are not displayed. Here backslash \texttt{\textbackslash{}} is a start of an escape sequence. It changes the meaning of the character after it. The backslash \texttt{\textbackslash{}} and \texttt{n} together (\texttt{\textbackslash{}n}) form an escape sequence and signify a newline. Other examples are \texttt{\textbackslash{}t} (tab) or \texttt{\textbackslash{}\$} (= print an actual dollar sign, normally a dollar sign has a special meaning). We'll see more escape sequences in 7.1.

You can try to remove \texttt{\textbackslash{}n} in the program to see what will happen. This will give you a dee per understanding of the program.

\hypertarget{perl-one-liner}{%
\chapter{Perl one-liner}\label{perl-one-liner}}

\hypertarget{escape-single-quote}{%
\section{Escape single quote}\label{escape-single-quote}}

You can't use single quotes alone. You need to escape them correctly using `'' This works:

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{echo} \StringTok{"a,b"} \KeywordTok{|} \FunctionTok{perl}\NormalTok{ -F}\StringTok{','}\NormalTok{ -lane }\StringTok{'print "'}\DataTypeTok{\textbackslash{}'}\StringTok{'$F[0]'}\DataTypeTok{\textbackslash{}'}\StringTok{'";'}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 'a'
\end{verbatim}

The \texttt{\textbackslash{}047} is an octal escape. it's actually just a single quote. when working with embedded quotes, it's sometimes easier to just write \textbackslash{}047 rather than something like \texttt{\textquotesingle{}\textbackslash{}\textquotesingle{}\textquotesingle{}}.

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{echo} \StringTok{"a,b"} \KeywordTok{|} \FunctionTok{perl}\NormalTok{ -F}\StringTok{','}\NormalTok{ -lane }\StringTok{'print "\textbackslash{}047$F[0]\textbackslash{}047";'}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 'a'
\end{verbatim}

\hypertarget{varible-in-perl}{%
\chapter{Varible in Perl}\label{varible-in-perl}}

Perl provides three kinds of variables: \texttt{scalars}, \texttt{arrays}, and \texttt{hash}(aka \texttt{associative\ arrays}). The initial character of the name identifies the particular type of variable and, hence, its functionality.

\begin{longtable}[]{@{}llll@{}}
\toprule
Type & Character & Example & Is a name for:\tabularnewline
\midrule
\endhead
Scalar & \$ & \$length & An individual value (number or string)\tabularnewline
Array & @ & @gene\_list & A list of values, keyed by number\tabularnewline
Hash & \% & \%gene\_annotation & A group of values, keyed by string\tabularnewline
\bottomrule
\end{longtable}

\hypertarget{scalar-variable}{%
\section{Scalar variable}\label{scalar-variable}}

In Perl, scalar variables can be used to store mainly two types of data: string and numbers.

Perl does not differentiate between a number and a string, nor does it differentiate between integers and reals.

In order to tell the computer what to print, we need to use variables. In Perl, the name of a \texttt{scalar} variable starts with the dollar sign \texttt{\$}. You can assign either a number or a string to it.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\CommentTok{#assign two strings to two variables}
\KeywordTok{my} \DataTypeTok{$dna_seq1}\NormalTok{ = }\KeywordTok{"}\StringTok{ACCTCGGTACAGTGAATGGGAAACGTAGCTGAT}\KeywordTok{"}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$dna_seq2}\NormalTok{ = }\KeywordTok{"}\StringTok{TGCCGATCGTAATAGCTCGCTATCTAGCTCGATCGTCGTA}\KeywordTok{"}\NormalTok{;}

\CommentTok{#Returns the length in characters of the value of EXPR}
\KeywordTok{my} \DataTypeTok{$dna_length1}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna_seq1}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$dna_length2}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna_seq2}\NormalTok{;}

\FunctionTok{print} \KeywordTok{"}\StringTok{The length of first DNA sequence (}\DataTypeTok{$dna_seq1}\StringTok{) is: }\DataTypeTok{$dna_length1}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{The length of the Second DNA sequence (}\DataTypeTok{$dna_seq2}\StringTok{) is: }\DataTypeTok{$dna_length2}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{perl}\NormalTok{ code_perl/variable_assign.pl}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## The length of first DNA sequence (ACCTCGGTACAGTGAATGGGAAACGTAGCTGAT) is: 33
## The length of the Second DNA sequence (TGCCGATCGTAATAGCTCGCTATCTAGCTCGATCGTCGTA) is: 40
\end{verbatim}

Here are the explanations of this script.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# assign two DNA sequences to two variables}
\KeywordTok{my} \DataTypeTok{$dna1}\NormalTok{ = }\KeywordTok{"}\StringTok{CTCGACCAGGACGATGAATGGGCGATGAAAATCT}\KeywordTok{"}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$dna2}\NormalTok{ = }\KeywordTok{"}\StringTok{CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

The first line here is a comment. In Perl, \texttt{\#} (pound sign) is the comment character. The comments can be used to document the program and improve the readability. During the execution, the comments will be ignored. The second and third lines declare two string scalar variables (\texttt{\$dna1} and \texttt{\$dna2}) to store two DNA sequences.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Returns the length in characters of the value of EXPR}
\KeywordTok{my} \DataTypeTok{$dna_length1}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna_seq1}\NormalTok{; }
\KeywordTok{my} \DataTypeTok{$dna_length2}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna_seq2}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

The code above has one comment line and declares two integer scalar variables to store the return values of built-on function \texttt{length}.

Now you can output the variables to see what are stored in them using the following code:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print} \KeywordTok{"}\StringTok{The length of first DNA sequence (}\DataTypeTok{$dna_seq1}\StringTok{) is: }\DataTypeTok{$dna_length1}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{The length of the Second DNA sequence (}\DataTypeTok{$dna_seq2}\StringTok{) is: }\DataTypeTok{$dna_length2}\CharTok{\textbackslash{}n\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

This script shows you what string scalar varibles and integer scalr varibles.

\hypertarget{arithmetic-operations-in-perl}{%
\section{Arithmetic operations in Perl}\label{arithmetic-operations-in-perl}}

Most arithmetic operators are binary operators; this means they take two arguments. Unary operators only take one argument. Arithmetic operators are very simple and often transparent.

Here we're mainly going to talk about basic arithmetic opertators including addition (\texttt{+}), substraction (\texttt{-}), multiplication (\texttt{*}), division (\texttt{/}) and the modulus operation (\%). Modulus (\texttt{\%}) returns the remainder of a division (\texttt{/}) operation.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## If we know the lengths of two sequences and we want to calulate the sum of the lengths.}
\KeywordTok{my} \DataTypeTok{$dna_length1}\NormalTok{ = }\DecValTok{100}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$dna_length2}\NormalTok{ = }\DecValTok{200}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$total_length}\NormalTok{ = }\DataTypeTok{$dna_length1}\NormalTok{ + }\DataTypeTok{$dna_length2}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{The total length of these two sequences is }\DataTypeTok{$total_length}\StringTok{ }\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$length_diff}\NormalTok{ = }\DataTypeTok{$dna_length1}\NormalTok{ - }\DataTypeTok{$dna_length2}\NormalTok{; }
\FunctionTok{print} \KeywordTok{"}\StringTok{The length difference is }\DataTypeTok{$length_diff}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$average_length}\NormalTok{ = }\DataTypeTok{$total_length}\NormalTok{ / }\DecValTok{2}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{The average length of two DNA sequences is }\DataTypeTok{$average_length}\StringTok{ }\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## The total length of these two sequences is 300 
## The length difference is -100
## The average length of two DNA sequences is 150
\end{verbatim}

The code above shows us how to

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Imaging the CC content of first DNA sequence is 0.5.}
\CommentTok{#How many CC do we have in first DNA?}
\KeywordTok{my} \DataTypeTok{$cg_content}\NormalTok{ = }\FloatTok{0.5}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$cg_number}\NormalTok{ = }\DataTypeTok{$dna_length1} \KeywordTok{*} \FloatTok{0.5}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Imaging we have a 10 bps DNA sequnces,}
\CommentTok{#how many possible DNA sequences do we have?}
\KeywordTok{my} \DataTypeTok{$dna_nucleotide}\NormalTok{ = }\KeywordTok{"}\StringTok{ATCG}\KeywordTok{"}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$dna_nucleo_number}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna_nucleotide}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$dna_length}\NormalTok{ = }\DecValTok{10}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$possible_number}\NormalTok{ = }\DataTypeTok{$dna_nucleo_number}\NormalTok{ *}\KeywordTok{*} \DataTypeTok{$dna_length}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{We have }\DataTypeTok{$possible_number}\StringTok{ possibilities.}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## We have 1048576 possibilities.
\end{verbatim}

\hypertarget{shorthand-operations}{%
\subsection{Shorthand operations}\label{shorthand-operations}}

The expression \$x += 3; is the shorthand version of \$x = \$x + 3;, they have exactly the same result:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$dna_length}\NormalTok{ = }\DecValTok{10}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{DNA length: }\DataTypeTok{$dna_length}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\DataTypeTok{$dna_length}\NormalTok{ += }\DecValTok{3}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{DNA length: }\DataTypeTok{$dna_length}\StringTok{ after '\textbackslash{}$dna_length += 3'}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\DataTypeTok{$dna_length}\NormalTok{ -= }\DecValTok{3}\NormalTok{;  }
\FunctionTok{print} \KeywordTok{"}\StringTok{DNA length: }\DataTypeTok{$dna_length}\StringTok{ after '\textbackslash{}$dna_length -= 3'}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DNA length: 10
## DNA length: 13 after '$dna_length += 3'
## DNA length: 10 after '$dna_length -= 3'
\end{verbatim}

\hypertarget{auto-increment-and-auto-decrement}{%
\subsection{Auto increment and auto decrement}\label{auto-increment-and-auto-decrement}}

\texttt{++} and \texttt{-\/-} are provided for the auto increment and auto decrement operators. They increase and decrease respectively the value of a scalar variable by 1.

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$dna_length}\NormalTok{ = }\DecValTok{10}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{DNA length: }\DataTypeTok{$dna_length}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\DataTypeTok{$dna_length}\NormalTok{ ++;}
\FunctionTok{print} \KeywordTok{"}\StringTok{DNA length: }\DataTypeTok{$dna_length}\StringTok{ after '\textbackslash{}$dna_length ++'}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\DataTypeTok{$dna_length}\NormalTok{ --;}
\FunctionTok{print} \KeywordTok{"}\StringTok{DNA length: }\DataTypeTok{$dna_length}\StringTok{ after `\textbackslash{}$dna_length --`}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## DNA length: 10
## DNA length: 11 after '$dna_length ++'
## DNA length: 10 after `$dna_length --`
\end{verbatim}

\hypertarget{use-strict-user-warnings-and-my}{%
\section{\texorpdfstring{\texttt{use\ strict}; \texttt{user\ warnings} and \texttt{my}}{use strict; user warnings and my}}\label{use-strict-user-warnings-and-my}}

For starters, \texttt{use\ strict;} (and to a lesser extent, \texttt{use\ warnings;}) helps find typos in variable names. Even experienced programmers make such errors. A common case is forgetting to rename an instance of a variable when cleaning up or refactoring code.

Using \texttt{use\ strict;\ use\ warnings;} catches many errors sooner than they would be caught otherwise, which makes it easier to find the root causes of the errors. The root cause might be the need for an error or validation check, and that can happen regardless of programmer skill.

What's good about Perl warnings is that they are rarely spurious, so there's next to no cost to using them.

In the script below, \texttt{\$dna\_lenght2} is a typo. If you run this script, it will give you the output without any error message, although it's not the right output.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}

\CommentTok{#assign 2 numbers to 2 variable}
\DataTypeTok{$dna1}\NormalTok{ = }\KeywordTok{"}\StringTok{CTCGACCAGGACGATGAATGGGCGATGAAAATCT}\KeywordTok{"}\NormalTok{;}
\DataTypeTok{$dna2}\NormalTok{ = }\KeywordTok{"}\StringTok{CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT}\KeywordTok{"}\NormalTok{;}

\CommentTok{#Returns the length in characters of the value of EXPR}
\DataTypeTok{$dna_length1}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna1}\NormalTok{;}
\DataTypeTok{$dna_length2}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna2}\NormalTok{;}

\FunctionTok{print} \KeywordTok{"}\StringTok{First DNA: }\DataTypeTok{$dna1}\StringTok{; Length: }\DataTypeTok{$dna_length1}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Second DNA: }\DataTypeTok{$dna2}\StringTok{; Length: }\DataTypeTok{$dna_length2}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\CommentTok{#calculate the total length of two DNA sequences}
\DataTypeTok{$tot_length}\NormalTok{ = }\DataTypeTok{$dna_length1}\NormalTok{ + }\DataTypeTok{$dna_lenght2}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{The total length of two DNA sequences is }\DataTypeTok{$tot_length}\StringTok{ }\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

Let's try to run this script:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{perl}\NormalTok{ code_perl/var_assign_no_strict_warnings.pl}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
First DNA: CTCGACCAGGACGATGAATGGGCGATGAAAATCT; Length: 34
Second DNA: CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT; Length: 48
The total length of two DNA sequences is 34 
\end{verbatim}

So the script above is supposed to output the length of two DNA sequences and the sum of the lengths.

In the chunk of code above, \texttt{\$dna\_lenght2} is an empty varible without storing any information. By defaut, Perl considers this as ZERO when doing plus operation. Although there was no error message given here, we infact have an incorrect output.

If we add \texttt{use\ strict} and \texttt{use\ warnings}, we need to decare each variable in the script. Let us see what will happen if we have an typo.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\CommentTok{#assign 2 numbers to 2 variable}
\KeywordTok{my} \DataTypeTok{$dna1}\NormalTok{ = }\KeywordTok{"}\StringTok{CTCGACCAGGACGATGAATGGGCGATGAAAATCT}\KeywordTok{"}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$dna2}\NormalTok{ = }\KeywordTok{"}\StringTok{CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT}\KeywordTok{"}\NormalTok{;}

\CommentTok{#Returns the length in characters of the value of EXPR}
\KeywordTok{my} \DataTypeTok{$dna_length1}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna1}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$dna_length2}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$dna2}\NormalTok{;}

\FunctionTok{print} \KeywordTok{"}\StringTok{First DNA: }\DataTypeTok{$dna1}\StringTok{; Length: }\DataTypeTok{$dna_length1}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Second DNA: }\DataTypeTok{$dna2}\StringTok{; Length: }\DataTypeTok{$dna_length2}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\CommentTok{#calculate the total length of two DNA sequences}
\KeywordTok{my} \DataTypeTok{$tot_length}\NormalTok{ = }\DataTypeTok{$dna_length1}\NormalTok{ + }\DataTypeTok{$dna_lenght2}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{The total length of two DNA sequences is }\DataTypeTok{$tot_length}\StringTok{ }\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Global symbol "$dna_lenght2" requires explicit package name (did you forget to declare "my $dna_lenght2"?) at code_perl/var_assign_strict_warnings.pl line 17.
Execution of code_perl/var_assign_strict_warnings.pl aborted due to compilation errors.
\end{verbatim}

Now if we run this script, we encounter error mesage and the script can't be sucessfuly excuted.

\begin{verbatim}
Global symbol "$dna_lenght2" requires explicit package name 
(did you forget to declare "my $dna_lenght2"?) at code_perl/
var_assign_strict_warnings.pl line 17.
Execution of code_perl/var_assign_strict_warnings.pl aborted
 due to compilation errors.
\end{verbatim}

\hypertarget{array}{%
\section{Array}\label{array}}

\hypertarget{init-an-array}{%
\subsection{Init an array}\label{init-an-array}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@base_pair}\NormalTok{ = (}\KeywordTok{'}\StringTok{A}\KeywordTok{'}\NormalTok{, }\KeywordTok{'}\StringTok{T}\KeywordTok{'}\NormalTok{, }\KeywordTok{'}\StringTok{C}\KeywordTok{'}\NormalTok{, }\KeywordTok{"}\StringTok{C}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\StringTok{G}\KeywordTok{"}\NormalTok{);}
\FunctionTok{print} \KeywordTok{""}\NormalTok{;}
\FunctionTok{print} \DataTypeTok{@base_pair}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \FunctionTok{join}\NormalTok{(}\KeywordTok{"}\CharTok{\textbackslash{}t}\KeywordTok{"}\NormalTok{, }\DataTypeTok{@base_pair}\NormalTok{), }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ATCCG
## A    T   C   C   G
\end{verbatim}

\hypertarget{array-index}{%
\subsection{Array index}\label{array-index}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@base_pair}\NormalTok{ = (}\KeywordTok{'}\StringTok{A}\KeywordTok{'}\NormalTok{, }\KeywordTok{'}\StringTok{T}\KeywordTok{'}\NormalTok{, }\KeywordTok{'}\StringTok{C}\KeywordTok{'}\NormalTok{, }\KeywordTok{"}\StringTok{C}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\StringTok{G}\KeywordTok{"}\NormalTok{);}
\CommentTok{### Extract the first element in the array}
\KeywordTok{my} \DataTypeTok{$first_base}\NormalTok{ = }\DataTypeTok{$base_pair}\NormalTok{[}\DecValTok{0}\NormalTok{];}
\FunctionTok{print} \KeywordTok{"}\StringTok{First base is: }\DataTypeTok{$first_base}\CharTok{\textbackslash{}n\textbackslash{}n}\KeywordTok{"}\NormalTok{; }

\CommentTok{### Extract the last element in the array}
\KeywordTok{my} \DataTypeTok{$last_base}\NormalTok{ = }\DataTypeTok{$base_pair}\NormalTok{[}\DecValTok{4}\NormalTok{];}
\FunctionTok{print} \KeywordTok{"}\StringTok{Last base is: }\DataTypeTok{$last_base}\CharTok{\textbackslash{}n\textbackslash{}n}\KeywordTok{"}\NormalTok{; }

\CommentTok{### Extract the last element in the array using index `-1`}
\KeywordTok{my} \DataTypeTok{$last_base}\NormalTok{ = }\DataTypeTok{$base_pair}\NormalTok{[-}\DecValTok{1}\NormalTok{];}
\FunctionTok{print} \KeywordTok{"}\StringTok{Last base using '-1' is: }\DataTypeTok{$last_base}\CharTok{\textbackslash{}n\textbackslash{}n}\KeywordTok{"}\NormalTok{; }

\CommentTok{### Extract the last element in the array using index `$#`}
\KeywordTok{my} \DataTypeTok{$last_base}\NormalTok{ = }\DataTypeTok{$base_pair}\NormalTok{[}\DataTypeTok{$#base_pair}\NormalTok{];}
\FunctionTok{print} \KeywordTok{"}\StringTok{Last base using '\textbackslash{}$#' is : }\DataTypeTok{$last_base}\CharTok{\textbackslash{}n\textbackslash{}n}\KeywordTok{"}\NormalTok{; }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## First base is: A
## 
## Last base is: G
## 
## Last base using '-1' is: G
## 
## Last base using '$#' is : G
\end{verbatim}

\hypertarget{length-of-the-array}{%
\subsection{Length of the array}\label{length-of-the-array}}

To get the length of the array, you can use \texttt{scalar\ @array} or \texttt{just\ array}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@gene_expr}\NormalTok{ = (}\DecValTok{1}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{10}\NormalTok{);}
\KeywordTok{my} \DataTypeTok{$len}\NormalTok{ = }\FunctionTok{scalar} \DataTypeTok{@gene_expr}\NormalTok{;  }\CommentTok{## or @gene_expr}
\FunctionTok{print} \KeywordTok{"}\StringTok{Array has: }\DataTypeTok{@gene_expr}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Length of array: }\KeywordTok{"}\NormalTok{, }\DataTypeTok{$len}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Array has: 1 3 10
## Length of array: 3
\end{verbatim}

Another way is to use \texttt{\$\#array}. \texttt{\$\#array} will return the index of the last element. Since the index starts with 0, to get the length we use \texttt{\$\#array\ +\ 1}.

If the length of array is empty, \texttt{\$\#array} will return \texttt{-1}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@gene_expr}\NormalTok{ = ();}
\KeywordTok{my} \DataTypeTok{$len}\NormalTok{ = }\DataTypeTok{$#gene_expr}\NormalTok{ + }\DecValTok{1}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Array has }\DataTypeTok{$len}\StringTok{ element}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Length of array: }\KeywordTok{"}\NormalTok{, }\DataTypeTok{$len}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Array has 0 element
## Length of array: 0
\end{verbatim}

\hypertarget{sort-arrays-in-perl}{%
\subsection{Sort Arrays in Perl}\label{sort-arrays-in-perl}}

\hypertarget{sort-alphebetically}{%
\subsubsection{Sort alphebetically}\label{sort-alphebetically}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@base_pair}\NormalTok{ = (}\KeywordTok{'}\StringTok{A}\KeywordTok{'}\NormalTok{, }\KeywordTok{'}\StringTok{T}\KeywordTok{'}\NormalTok{, }\KeywordTok{'}\StringTok{C}\KeywordTok{'}\NormalTok{, }\KeywordTok{"}\StringTok{C}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\StringTok{G}\KeywordTok{"}\NormalTok{);}
\KeywordTok{my} \DataTypeTok{@sorted_bp}\NormalTok{ = }\FunctionTok{sort} \DataTypeTok{@base_pair}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Array before sorted: }\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\DataTypeTok{@base_pair}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Array after sorted: }\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\DataTypeTok{@sorted_bp}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Array before sorted: A T C C G
## Array after sorted: A C C G T
\end{verbatim}

\hypertarget{sort-numerically}{%
\subsubsection{Sort numerically}\label{sort-numerically}}

To sort an array numerically, we use \texttt{spaceship\ operator:\ \textless{}=\textgreater{}}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@genome_coor}\NormalTok{ = (}\DecValTok{100}\NormalTok{, }\DecValTok{300}\NormalTok{, }\DecValTok{200}\NormalTok{, }\DecValTok{500}\NormalTok{);}
\KeywordTok{my} \DataTypeTok{@sorted_coor}\NormalTok{ = }\FunctionTok{sort}\NormalTok{ \{}\DataTypeTok{$a}\NormalTok{ <=> }\DataTypeTok{$b}\NormalTok{\} }\DataTypeTok{@genome_coor}\NormalTok{;}

\FunctionTok{print} \KeywordTok{"}\StringTok{Array before sorted: }\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\DataTypeTok{@genome_coor}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Array after sorted: }\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\DataTypeTok{@sorted_coor}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Array before sorted: 100 300 200 500
## Array after sorted: 100 200 300 500
\end{verbatim}

Similarly, array can be also sorted numerically in decreasing order.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@genome_coor}\NormalTok{ = (}\DecValTok{100}\NormalTok{, }\DecValTok{300}\NormalTok{, }\DecValTok{200}\NormalTok{, }\DecValTok{500}\NormalTok{);}
\CommentTok{## \{$a <=> $b\} is modified as \{$b <=> $a\} }
\KeywordTok{my} \DataTypeTok{@sorted_coor}\NormalTok{ = }\FunctionTok{sort}\NormalTok{ \{}\DataTypeTok{$b}\NormalTok{ <=> }\DataTypeTok{$a}\NormalTok{\} }\DataTypeTok{@genome_coor}\NormalTok{;  }

\FunctionTok{print} \KeywordTok{"}\StringTok{Array before sorted: }\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\DataTypeTok{@genome_coor}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Array after sorted: }\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\DataTypeTok{@sorted_coor}\KeywordTok{"}\NormalTok{, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Array before sorted: 100 300 200 500
## Array after sorted: 500 300 200 100
\end{verbatim}

\hypertarget{use-push-pop-shift-and-unshift-in-perl}{%
\subsection{\texorpdfstring{Use \texttt{push}, \texttt{pop}, \texttt{shift} and \texttt{unshift} in Perl}{Use push, pop, shift and unshift in Perl}}\label{use-push-pop-shift-and-unshift-in-perl}}

\hypertarget{hash-in-perl}{%
\section{Hash in Perl}\label{hash-in-perl}}

A Perl hash is defined by key-value pairs. Perl stores elements of a hash in such an optimal way that you can look up its values based on keys very fast.

With the array, you use indices to access its elements. However, you must use descriptive keys to access hash's element. A hash is sometimes referred to as an associative array.

Like a scalar or an array variable, a hash variable has its own prefix. A hash variable must begin with a percent sign ( \%). The prefix \% looks like key/value pair so remember this trick to name the hash variables.

The following example defines a simple hash.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{%gene_info}\NormalTok{ = (}\KeywordTok{"}\StringTok{gene1}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{ROS}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{gene2}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{WRKY}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{gene3}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{WRKY}\KeywordTok{"}
\NormalTok{                );}

\FunctionTok{print} \KeywordTok{"}\DataTypeTok{$gene_info}\StringTok{\{gene1\}}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\DataTypeTok{$gene_info}\StringTok{\{gene2\}}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\DataTypeTok{$gene_info}\StringTok{\{gene3\}}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ROS
## WRKY
## WRKY
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{%gene_info}\NormalTok{ = (}\KeywordTok{"}\StringTok{gene1}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{ROS}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{gene2}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{WRKY}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{gene3}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{WRKY}\KeywordTok{"}
\NormalTok{                );}
\KeywordTok{my} \DataTypeTok{@hash_key}\NormalTok{  = }\FunctionTok{keys} \DataTypeTok{%gene_info}\NormalTok{;}
\DataTypeTok{@hash_key}\NormalTok{ = }\FunctionTok{sort} \DataTypeTok{@hash_key}\NormalTok{;}
\CommentTok{#print "Keys of \textbackslash{}%gene_info: ", @hash_key, "\textbackslash{}n";}
\FunctionTok{print} \KeywordTok{"}\StringTok{Keys of \textbackslash{}%gene_info: }\KeywordTok{"}\NormalTok{, }\FunctionTok{join}\NormalTok{(}\KeywordTok{"}\CharTok{\textbackslash{}t}\KeywordTok{"}\NormalTok{, }\FunctionTok{sort} \FunctionTok{keys} \DataTypeTok{%gene_info}\NormalTok{), }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\CommentTok{#print "Keys of \textbackslash{}%gene_info: ", keys %gene_info, "\textbackslash{}n";}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Keys of %gene_info: gene1    gene2   gene3
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{%gene_info}\NormalTok{ = (}\KeywordTok{"}\StringTok{gene1}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{ROS}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{gene2}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{WRKY}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{gene3}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{WRKY}\KeywordTok{"}
\NormalTok{                );}
\KeywordTok{foreach}\NormalTok{(}\FunctionTok{sort} \FunctionTok{keys} \DataTypeTok{%gene_info}\NormalTok{)\{}
    \FunctionTok{print} \KeywordTok{"}\StringTok{The value of }\DataTypeTok{$_}\StringTok{ is: }\KeywordTok{"}\NormalTok{, }\DataTypeTok{$gene_info}\NormalTok{\{}\DataTypeTok{$_}\NormalTok{\}, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## The value of gene1 is: ROS
## The value of gene2 is: WRKY
## The value of gene3 is: WRKY
\end{verbatim}

\hypertarget{sort-keys-numerically-using-sort-ab}{%
\subsection{\texorpdfstring{Sort keys numerically using \texttt{sort\ \{\$a\textless{}=\textgreater{}\$b\}}}{Sort keys numerically using sort \{\$a\textless{}=\textgreater{}\$b\}}}\label{sort-keys-numerically-using-sort-ab}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{%read_dep}\NormalTok{ = (}\KeywordTok{"}\StringTok{1}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{100}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{2}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{200}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{3}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{50}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{10}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{20}\KeywordTok{"}\NormalTok{,}
\NormalTok{                );}

\CommentTok{### sort in increasing order}
\KeywordTok{foreach}\NormalTok{(}\FunctionTok{sort}\NormalTok{ \{}\DataTypeTok{$a}\NormalTok{<=>}\DataTypeTok{$b}\NormalTok{\} }\FunctionTok{keys} \DataTypeTok{%read_dep}\NormalTok{)\{}
    \FunctionTok{print} \KeywordTok{"}\StringTok{The value of }\DataTypeTok{$_}\StringTok{ is: }\KeywordTok{"}\NormalTok{, }\DataTypeTok{$read_dep}\NormalTok{\{}\DataTypeTok{$_}\NormalTok{\}, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}

\FunctionTok{print} \KeywordTok{"}\CharTok{\textbackslash{}n}\StringTok{Sort in descending order: }\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\CommentTok{### sort \{$b<=>$a\} if you want to sort in descending order python}
\KeywordTok{foreach}\NormalTok{(}\FunctionTok{sort}\NormalTok{ \{}\DataTypeTok{$b}\NormalTok{<=>}\DataTypeTok{$a}\NormalTok{\} }\FunctionTok{keys} \DataTypeTok{%read_dep}\NormalTok{)\{}
    \FunctionTok{print} \KeywordTok{"}\StringTok{The value of }\DataTypeTok{$_}\StringTok{ is: }\KeywordTok{"}\NormalTok{, }\DataTypeTok{$read_dep}\NormalTok{\{}\DataTypeTok{$_}\NormalTok{\}, }\KeywordTok{"}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## The value of 1 is: 100
## The value of 2 is: 200
## The value of 3 is: 50
## The value of 10 is: 20
## 
## Sort in descending order: 
## The value of 10 is: 20
## The value of 3 is: 50
## The value of 2 is: 200
## The value of 1 is: 100
\end{verbatim}

\hypertarget{use-exists-function-on-a-hash}{%
\subsection{\texorpdfstring{Use \texttt{exists} function on a hash}{Use exists function on a hash}}\label{use-exists-function-on-a-hash}}

Given an expression that specifies an element of a hash, \texttt{exists} returns true if the specified element in the hash has ever been initialized, even if the corresponding value is undefined.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{%gene_info}\NormalTok{ = (}\KeywordTok{"}\StringTok{gene1}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{ROS}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{gene2}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{WRKY}\KeywordTok{"}\NormalTok{,}
                 \KeywordTok{"}\StringTok{gene3}\KeywordTok{"}\NormalTok{=>}\KeywordTok{"}\StringTok{WRKY}\KeywordTok{"}
\NormalTok{                );}
\KeywordTok{if}\NormalTok{(}\FunctionTok{exists} \DataTypeTok{$gene_info}\NormalTok{\{}\KeywordTok{"}\StringTok{gene1}\KeywordTok{"}\NormalTok{\})\{}
    \FunctionTok{print} \KeywordTok{"}\StringTok{The key 'gene1' exists in \textbackslash{}%gene_info}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## The key 'gene1' exists in %gene_info
\end{verbatim}

\hypertarget{control-structure}{%
\chapter{Control structure}\label{control-structure}}

\ldots{}.
Perl is an iterative language in which control flows from the first statement in the program to the last statement unless something interrupts. Some of the things that can interrupt this linear flow are conditional branches and loop structures. Perl offers approximately a dozen such constructs, which are described below. The basic form will be shown for each followed by a partial example.
\ldots{}.

\hypertarget{for-loop}{%
\section{\texorpdfstring{\texttt{for} loop}{for loop}}\label{for-loop}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@gene_expr}\NormalTok{ = (}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{,}\DecValTok{8}\NormalTok{, }\DecValTok{9}\NormalTok{);}

\CommentTok{# This is an example of  for loop }
\KeywordTok{for}\NormalTok{(}\KeywordTok{my} \DataTypeTok{$i}\NormalTok{=}\DecValTok{0}\NormalTok{; }\DataTypeTok{$i}\NormalTok{<}\DataTypeTok{@gene_expr}\NormalTok{; ++}\DataTypeTok{$i}\NormalTok{)\{}
    \KeywordTok{my} \DataTypeTok{$tem_var}\NormalTok{= }\DataTypeTok{$gene_expr}\NormalTok{[}\DataTypeTok{$i}\NormalTok{]/}\DecValTok{2}\NormalTok{;}

    \FunctionTok{print} \KeywordTok{"}\StringTok{At }\DataTypeTok{$i}\StringTok{ place, the number devided by 2 equals: }\DataTypeTok{$tem_var}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;    }
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## At 0 place, the number devided by 2 equals: 1
## At 1 place, the number devided by 2 equals: 3
## At 2 place, the number devided by 2 equals: 4
## At 3 place, the number devided by 2 equals: 4.5
\end{verbatim}

\hypertarget{foreach-loop}{%
\section{\texorpdfstring{\texttt{foreach} loop}{foreach loop}}\label{foreach-loop}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@gene_expr}\NormalTok{ = (}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{,}\DecValTok{8}\NormalTok{, }\DecValTok{9}\NormalTok{);}

\KeywordTok{my} \DataTypeTok{$j}\NormalTok{ = }\DecValTok{0}\NormalTok{;}
\KeywordTok{foreach}\NormalTok{(}\DataTypeTok{@gene_expr}\NormalTok{)\{}
    \KeywordTok{my} \DataTypeTok{$tem_var}\NormalTok{= }\DataTypeTok{$gene_expr}\NormalTok{[}\DataTypeTok{$j}\NormalTok{]/}\DecValTok{2}\NormalTok{;}
    \FunctionTok{print} \KeywordTok{"}\StringTok{At }\DataTypeTok{$j}\StringTok{ place, the number devided by 2 equals: }\DataTypeTok{$tem_var}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
    \CommentTok{#$j = $j +1}
\NormalTok{    ++}\DataTypeTok{$j}\NormalTok{;}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## At 0 place, the number devided by 2 equals: 1
## At 1 place, the number devided by 2 equals: 3
## At 2 place, the number devided by 2 equals: 4
## At 3 place, the number devided by 2 equals: 4.5
\end{verbatim}

\hypertarget{while-loop}{%
\section{\texorpdfstring{\texttt{while} loop}{while loop}}\label{while-loop}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@gene_expr}\NormalTok{ = (}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{,}\DecValTok{8}\NormalTok{, }\DecValTok{9}\NormalTok{);}

\KeywordTok{my} \DataTypeTok{$k}\NormalTok{ = }\DecValTok{0}\NormalTok{;}
\KeywordTok{while}\NormalTok{(}\DataTypeTok{$k}\NormalTok{<}\DataTypeTok{@gene_expr}\NormalTok{)\{}
    \KeywordTok{my} \DataTypeTok{$tem_var}\NormalTok{= }\DataTypeTok{$gene_expr}\NormalTok{[}\DataTypeTok{$k}\NormalTok{]/}\DecValTok{2}\NormalTok{;}
    \FunctionTok{print} \KeywordTok{"}\StringTok{At }\DataTypeTok{$k}\StringTok{ place, the number devided by 2 equals: }\DataTypeTok{$tem_var}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{    ++}\DataTypeTok{$k}\NormalTok{;}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## At 0 place, the number devided by 2 equals: 1
## At 1 place, the number devided by 2 equals: 3
## At 2 place, the number devided by 2 equals: 4
## At 3 place, the number devided by 2 equals: 4.5
\end{verbatim}

\hypertarget{statement-if-else}{%
\section{Statement if-else}\label{statement-if-else}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{@gene_exp_lev}\NormalTok{ = (}\DecValTok{1}\NormalTok{, }\DecValTok{5}\NormalTok{, }\DecValTok{3}\NormalTok{, }\DecValTok{4}\NormalTok{, }\DecValTok{9}\NormalTok{, }\DecValTok{10}\NormalTok{);}

\CommentTok{# for (my $i = 0; $i < @gene_exp_lev; $i++) \{}
\KeywordTok{for}\NormalTok{ (}\KeywordTok{my} \DataTypeTok{$i}\NormalTok{ = }\DecValTok{0}\NormalTok{; }\DataTypeTok{$i}\NormalTok{ < }\FunctionTok{scalar} \DataTypeTok{@gene_exp_lev}\NormalTok{; }\DataTypeTok{$i}\NormalTok{++) \{}
    \KeywordTok{if}\NormalTok{ (}\DataTypeTok{$gene_exp_lev}\NormalTok{[}\DataTypeTok{$i}\NormalTok{] > }\DecValTok{4}\NormalTok{) \{}
        \FunctionTok{print} \KeywordTok{"}\StringTok{Index }\DataTypeTok{$i}\StringTok{: }\DataTypeTok{$gene_exp_lev}\StringTok{[}\DataTypeTok{$i}\StringTok{]}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{    \}}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
Index 1: 5
Index 4: 9
Index 5: 10
\end{verbatim}

\hypertarget{operator-last-and-next}{%
\section{\texorpdfstring{Operator \texttt{last} and \texttt{next}}{Operator last and next}}\label{operator-last-and-next}}

\hypertarget{operator-redo}{%
\section{\texorpdfstring{Operator \texttt{redo}}{Operator redo}}\label{operator-redo}}

Before we use \texttt{redo}, first let's see what is \texttt{BLOCK} in Perl. In Perl, a \texttt{BLOCK} by itself (labeled or not) is semantically equivalent to a loop that executes once. Thus you can use any of the loop control statements in it to leave or restart the block.

The redo command restarts the loop block without evaluating the conditional again.

Let's say we have some postitions on the genome where a few

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{@read_depth}\NormalTok{ = (}\DecValTok{8}\NormalTok{, }\DecValTok{9}\NormalTok{, }\DecValTok{10}\NormalTok{, }\DecValTok{7}\NormalTok{, }\DecValTok{10}\NormalTok{, }\DecValTok{7}\NormalTok{, }\DecValTok{7}\NormalTok{);}
\end{Highlighting}
\end{Shaded}

\hypertarget{string-manipulation}{%
\chapter{String manipulation}\label{string-manipulation}}

\hypertarget{string-concatenation}{%
\section{String concatenation}\label{string-concatenation}}

Dot (\texttt{.}) can be used to concatenate two strings together.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# concatenate two strings together and assing to $z}
\DataTypeTok{$z}\NormalTok{ = }\DataTypeTok{$x}\NormalTok{ . }\DataTypeTok{$y}\NormalTok{;}
\CommentTok{# Append $y to $x}
\DataTypeTok{$x}\NormalTok{ = }\DataTypeTok{$x}\NormalTok{ . }\DataTypeTok{$y}\NormalTok{;}
\CommentTok{# Append $y to $x}
\DataTypeTok{$x}\NormalTok{ .= }\DataTypeTok{$y}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

A more convenient way is to use operator \texttt{.=} to append one variable to another.

As is any other assignments in Perl, if you see an assignment written this way \texttt{\$x\ =\ \$x} op expr, where op stands for any operator and expr stands for the rest of the statement, you can make a shorter version by moving the op to the front of the assignment, e.g., \texttt{\$x\ op=\ expr}. The string concatenation operator \texttt{.} is just one possible op among many others such as \texttt{+}, \texttt{-}, \texttt{*} and \texttt{/}.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{my} \DataTypeTok{$x}\NormalTok{ = }\DecValTok{5}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$y}\NormalTok{ = }\DecValTok{6}\NormalTok{;}
\CommentTok{# Add $y to $x}
\DataTypeTok{$x}\NormalTok{ = }\DataTypeTok{$x}\NormalTok{ + }\DataTypeTok{$y}\NormalTok{;}
\CommentTok{# Add $y to $x}
\DataTypeTok{$x}\NormalTok{ += }\DataTypeTok{$y}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\hypertarget{substring-extraction}{%
\section{Substring extraction}\label{substring-extraction}}

\hypertarget{substring-search}{%
\section{Substring search}\label{substring-search}}

Index

\hypertarget{split-string}{%
\section{Split String}\label{split-string}}

split

join

\hypertarget{regular-expression}{%
\section{Regular expression}\label{regular-expression}}

A regular expresion is a string of characters that defines the pattern or patterns you are searching. Usually this pattern is used by string searching algorithms for `find' or `find and replace' operations on strings or for input validation.

You can use pattern binding operators \texttt{=\textasciitilde{}} and \texttt{!\textasciitilde{}}. The first operator is a test and assignment operator.

In Perl, there are three three scenarios you may use regular expression:

\begin{itemize}
\item
  Pattern matching: \texttt{m//}
\item
  Pattern substitution: \texttt{s///}
\item
  Modifiers to pattern matching and substitution: \texttt{tr///}
\end{itemize}

In each case above, the forward slashes is used as delimiters for regular expression specified by you.

\hypertarget{pattern-matching}{%
\subsection{Pattern matching}\label{pattern-matching}}

In Perl, \texttt{m//} is used to match a string (could be sequence in Bioinformatics) to a regular expression. For example, to match a mRNA sequence \texttt{\$mRNA} against the mRNA sequence `\$mRNA'

The match operator, m//, is used to match a string or statement to a regular expression. For example, to match the character sequence ``foo'' against the scalar \$bar, you might use a statement like this −

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}

\DataTypeTok{$mRNA}\NormalTok{ = }\KeywordTok{"}\StringTok{ATG}\KeywordTok{"}\NormalTok{;}
\KeywordTok{if}\NormalTok{ (}\DataTypeTok{$bar}\NormalTok{ =~ }\KeywordTok{/}\OtherTok{foo}\KeywordTok{/}\NormalTok{) \{}
   \FunctionTok{print} \KeywordTok{"}\StringTok{First time is matching}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\} }\KeywordTok{else}\NormalTok{ \{}
   \FunctionTok{print} \KeywordTok{"}\StringTok{First time is not matching}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## First time is not matching
\end{verbatim}

Several special variables also refer back to portions of the previous match.

(ref:regexp\_perl) Several special variables

\textbackslash{}begin\{table\}{[}t{]}

\textbackslash{}caption\{(\#tab:regexp\_perl)(ref:regexp\_perl)\}
\centering

\begin{tabular}{c|c}
\hline
Special.variables & Description\\
\hline
\$+ & Whatever the last bracket match matched\\
\hline
\$\& & The entire matched string\\
\hline
\$` & Everything before the matched string\\
\hline
\$' & Everything after the marched string\\
\hline
\$\textasciicircum{}N & Whatever was matched by the most-recently closed group (submatch)\\
\hline
\end{tabular}

\textbackslash{}end\{table\}

\hypertarget{pattern-substitution}{%
\subsection{Pattern substitution}\label{pattern-substitution}}

\hypertarget{modifiers-to-pattern-matching-and-substitution}{%
\subsection{Modifiers to pattern matching and substitution}\label{modifiers-to-pattern-matching-and-substitution}}

\hypertarget{greedy-or-non-greedy}{%
\subsection{Greedy or non-greedy}\label{greedy-or-non-greedy}}

\hypertarget{practical-perl-for-regular-expresssion-advanced}{%
\subsection{Practical Perl for regular expresssion (Advanced)}\label{practical-perl-for-regular-expresssion-advanced}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl -l}

\CommentTok{# http://perlmonks.org/?node_id=1146191}

\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$sequence}\NormalTok{ = }\KeywordTok{'}\StringTok{AATGGTTTCTCCCATCTCTCCATCGGCATAAAAATACAGAATGATCTAACGAA}\KeywordTok{'}\NormalTok{;}

\KeywordTok{while}\NormalTok{( }\DataTypeTok{$sequence}\NormalTok{ =~ }\KeywordTok{/}\OtherTok{ATG}\KeywordTok{/g}\NormalTok{ )\{}
    \CommentTok{## Post match: $':}
    \CommentTok{## The string following whatever was matched by the last successful pattern match }
    \KeywordTok{my} \DataTypeTok{$rest}\NormalTok{ = }\DataTypeTok{$'}\NormalTok{;}
    \KeywordTok{while}\NormalTok{(}\DataTypeTok{$rest}\NormalTok{ =~ }\KeywordTok{/}\CharTok{(}\OtherTok{TAG}\CharTok{|}\OtherTok{TAA}\CharTok{|}\OtherTok{TGA}\CharTok{)}\KeywordTok{/g}\NormalTok{)\{}
        \KeywordTok{my} \DataTypeTok{$output}\NormalTok{ = }\KeywordTok{'}\StringTok{ATG}\KeywordTok{'}\NormalTok{ . }\DataTypeTok{$`}\NormalTok{ . }\DataTypeTok{$1}\NormalTok{;}
        \FunctionTok{print} \DataTypeTok{$output}\NormalTok{;}
\NormalTok{    \}}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ATGGTTTCTCCCATCTCTCCATCGGCATAA
## ATGGTTTCTCCCATCTCTCCATCGGCATAAAAATACAGAATGA
## ATGGTTTCTCCCATCTCTCCATCGGCATAAAAATACAGAATGATCTAA
## ATGATCTAA
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}

\CommentTok{# http://perlmonks.org/?node_id=1146191}

\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$sequence}\NormalTok{ = }\KeywordTok{'}\StringTok{AATGGTTTCTCCCATCTCTCCATCGGCATAAAAATACAGAATGATCTAACGAA}\KeywordTok{'}\NormalTok{;}

\DataTypeTok{$sequence}\NormalTok{ =~ }\KeywordTok{/}\CharTok{(}\OtherTok{ATG.}\CharTok{*?(?:}\OtherTok{TAG}\CharTok{|}\OtherTok{TAA}\CharTok{|}\OtherTok{TGA}\CharTok{))(??\{print "$1\textbackslash{}n" if (length $1)}\OtherTok{%3 == 0\}}\CharTok{)}\KeywordTok{/}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ATGGTTTCTCCCATCTCTCCATCGGCATAA
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}

\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$sequence}\NormalTok{ = }\KeywordTok{'}\StringTok{AATGGTTTCTCCCATCTCTCCATCGGCATAAAAATACAGAATGATCTAACGAA}\KeywordTok{'}\NormalTok{;}

\KeywordTok{while}\NormalTok{(}\DataTypeTok{$sequence}\NormalTok{ =~ }\KeywordTok{/}\OtherTok{ATG}\CharTok{(([}\BaseNTok{ATGC}\CharTok{]\{3\})+?)(}\OtherTok{TAG}\CharTok{|}\OtherTok{TAA}\CharTok{|}\OtherTok{TGA}\CharTok{)}\KeywordTok{/g}\NormalTok{)\{}
    \FunctionTok{print} \KeywordTok{"}\StringTok{ATG}\DataTypeTok{$1}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
    \FunctionTok{print} \KeywordTok{"}\DataTypeTok{$2}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ATGGTTTCTCCCATCTCTCCATCGGCA
## GCAATGATC
## ATC
\end{verbatim}

We start with matching the start codon ATG. Then the thing we want to match
is inside a parenthesis and it is at least one group of three arbitrary nucleotides
({[}ATGC{]}\{3\}). Finally we want to end with either of the three stop codons. Now
if we look by eye, we can see that there are two such matches in the string:
GGAATACGCGGA and CACCACGACGCCATAATT, but if we run the script it will only find
one sequence starting after the first start codon and ending with the last stop codon:
GGAATACGCGGATAAGGCGAAATGCACCACGACGCCATAATT. This is because the + character
matches as many as possible. It is greedy. To stop it from being greedy we need
to add a ? sign after it:

\hypertarget{input-and-output-in-perl}{%
\chapter{Input and output in Perl}\label{input-and-output-in-perl}}

\hypertarget{input}{%
\section{Input}\label{input}}

\hypertarget{standard-input}{%
\subsection{Standard input}\label{standard-input}}

\hypertarget{input-from-a-file-on-the-disk}{%
\subsection{Input from a file on the disk}\label{input-from-a-file-on-the-disk}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$fasta}\NormalTok{ = }\KeywordTok{"}\StringTok{./data/test_ref.fa}\KeywordTok{"}\NormalTok{;}

\FunctionTok{open}\NormalTok{ FASTA, }\DataTypeTok{$fasta} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\DataTypeTok{$!}\StringTok{: }\DataTypeTok{$fasta}\KeywordTok{"}\NormalTok{;}
\KeywordTok{while}\NormalTok{(}\KeywordTok{my} \DataTypeTok{$line}\NormalTok{ = }\KeywordTok{<FASTA>}\NormalTok{)\{}
    \FunctionTok{chomp} \DataTypeTok{$line}\NormalTok{;}
    \FunctionTok{print} \KeywordTok{"}\DataTypeTok{$line}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\FunctionTok{close}\NormalTok{ FASTA;}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{perl}\NormalTok{ code_perl/open_file.pl}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## >chr1
## ACGCTAGCTAGTCAGTCGATCGT
## CGTAGCTAGCTAG
## >chr2
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTA
## >chr3
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTA
\end{verbatim}

\hypertarget{special-variable-_}{%
\subsection{\texorpdfstring{Special variable \texttt{\$\_}}{Special variable \$\_}}\label{special-variable-_}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$fasta}\NormalTok{ = }\KeywordTok{"}\StringTok{./data/test_ref.fa}\KeywordTok{"}\NormalTok{;}

\FunctionTok{open}\NormalTok{ FASTA, }\DataTypeTok{$fasta} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\DataTypeTok{$!}\StringTok{: }\DataTypeTok{$fasta}\KeywordTok{"}\NormalTok{;}
\KeywordTok{while}\NormalTok{(}\KeywordTok{my} \DataTypeTok{$line}\NormalTok{ = }\KeywordTok{<FASTA>}\NormalTok{)\{}
    \FunctionTok{chomp} \DataTypeTok{$line}\NormalTok{;}
    \FunctionTok{print} \KeywordTok{"}\DataTypeTok{$line}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\FunctionTok{close}\NormalTok{ FASTA;}
\end{Highlighting}
\end{Shaded}

When you are using \texttt{while} and \texttt{foreach}, the content will be assigned to \texttt{\$\_} if you don't assign it explicitly to any variable.

For some build-in functions (\texttt{chomp}, \texttt{length}, \texttt{split} and \texttt{print}),

I do NOT suggest you to use \texttt{\$\_} because \texttt{\$\_} will reduce the readerbility of the code.

In Perl, \$\_ is a powerful
In Perl, several functions and operators use this variable as a default, in case no parameter is explicitly used. In general, I'd say you should NOT see \$\_ in real code. I think the whole point of \$\_ is that you don't have to write it explicitly.

\hypertarget{the-input-record-separator}{%
\subsection{\texorpdfstring{The input record separator: \texttt{\$/}}{The input record separator: \$/}}\label{the-input-record-separator}}

\texttt{\$/} is the input record separator. By default, \texttt{\$/} equals newline (\texttt{\textbackslash{}n}).

You could actually change the value of \texttt{\$/}. For example, by assigning the greate-than `\textgreater{}' like this: \texttt{\$/\ =\ "\textgreater{}"};. Then every call to the read-line operator \texttt{\$chunk\ =\ \textless{}FILEHANDLE\textgreater{}} will read in all the characters up-to and including the first `\textgreater{}'.

We could also assign longer strings to \texttt{\$/} and then that would be the input record separator.

\hypertarget{how-to-process-fasta-file-by-changing}{%
\subsubsection{\texorpdfstring{How to process \texttt{fasta} file by changing \texttt{\$/}}{How to process fasta file by changing \$/}}\label{how-to-process-fasta-file-by-changing}}

FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes (A, T, G, C, etc.).

In \texttt{fasta} format, each sequence begins with a single-line description, followed by lines of sequence data. The description line is distinguished from the sequence data by a greater-than (``\textgreater{}'') symbol in the first column.

Here is an example of fasta file (Figure \ref{fig:fastaPYL}).



\begin{figure}
\centering
\includegraphics{figures/fasta_pyl10_ara.png}
\caption{\label{fig:fastaPYL}Protein sequence of PYL10 in Arabidopsis.}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$fasta}\NormalTok{ = }\KeywordTok{"}\StringTok{./data/test_ref.fa}\KeywordTok{"}\NormalTok{;}

\FunctionTok{open}\NormalTok{ FASTA, }\DataTypeTok{$fasta} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\DataTypeTok{$!}\StringTok{: }\DataTypeTok{$fasta}\KeywordTok{"}\NormalTok{;}
\CommentTok{# Set the input record separator}
\DataTypeTok{$/}\NormalTok{ = }\KeywordTok{"}\CharTok{\textbackslash{}n}\StringTok{>}\KeywordTok{"}\NormalTok{;}
\CommentTok{# Print header}
\FunctionTok{print} \KeywordTok{"}\StringTok{Chrom}\CharTok{\textbackslash{}t}\StringTok{Length}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\KeywordTok{while}\NormalTok{(}\KeywordTok{my} \DataTypeTok{$chunk}\NormalTok{ = }\KeywordTok{<FASTA>}\NormalTok{)\{}
    \CommentTok{# Remove > in the first chunk}
    \DataTypeTok{$chunk}\NormalTok{ =~ }\KeywordTok{s/}\CharTok{^}\OtherTok{>}\KeywordTok{//}\NormalTok{;}
    \CommentTok{# Remove \textbackslash{}n> from the end of the chunk}
    \FunctionTok{chomp} \DataTypeTok{$chunk}\NormalTok{;}
    \CommentTok{# Assign ID and seq to two variables}
    \KeywordTok{my}\NormalTok{ (}\DataTypeTok{$chrom}\NormalTok{, }\DataTypeTok{$seq}\NormalTok{) = }\FunctionTok{split}\NormalTok{(}\KeywordTok{/}\BaseNTok{\textbackslash{}n}\KeywordTok{/}\NormalTok{, }\DataTypeTok{$chunk}\NormalTok{, }\DecValTok{2}\NormalTok{);}
    \CommentTok{# Remove \textbackslash{}n in the string}
    \DataTypeTok{$seq}\NormalTok{ =~ }\KeywordTok{s/}\BaseNTok{\textbackslash{}n}\KeywordTok{//g}\NormalTok{;}
    \CommentTok{# Calculate the length of sequence.}
    \KeywordTok{my} \DataTypeTok{$seq_length}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$seq}\NormalTok{;}
    \CommentTok{# Print ID and length of the sequence.}
    \FunctionTok{print} \KeywordTok{"}\DataTypeTok{$chrom}\CharTok{\textbackslash{}t}\DataTypeTok{$seq_length}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\FunctionTok{close}\NormalTok{ FASTA;}
\end{Highlighting}
\end{Shaded}

You can consider the content in the fasta file as a string (Figure \ref{fig:fastaString}) sperarated by \texttt{\textbackslash{}n\textgreater{}}.

We set the input record separator \texttt{\$/} as \texttt{\textbackslash{}n\textgreater{}}.

In the first cycle of the \texttt{while} loop, the first chunk is \texttt{\textgreater{}chr1\textbackslash{}nACGCTAGCTAGTCAGTCGATCGT\textbackslash{}nCGTAGCTAGCTAG\textbackslash{}n\textgreater{}}.

You should notice that there is a leading \texttt{\textgreater{}} in the first chunk. So you need to use \texttt{s/\^{}\textgreater{}//} to remove the leading \texttt{\textgreater{}}.

To remove the trailing \texttt{\textbackslash{}n\textgreater{}}, you can use \texttt{chomp\ \$chunk;}. Then the string is:

\begin{verbatim}
chr1\nACGCTAGCTAGTCAGTCGATCGT\nCGTAGCTAGCTAG
\end{verbatim}

The string before first \texttt{\textbackslash{}n} is the sequence ID. The string after the first \texttt{\textbackslash{}n} is the sequence. To get the sequnce ID and sequnce and assign to two variables, you can use \texttt{split(/\textbackslash{}n/,\ \$chunk,\ 2)}. Here LIMIT is set to 2, it represents the maximum number (here 2) of fields into which the EXPR may be split.

The goal of this code is to output the sequence IDs and their lengths. Currently the string of the variable (\texttt{\$seq}) is \texttt{ACGCTAGCTAGTCAGTCGATCGT\textbackslash{}nCGTAGCTAGCTAG}. To do this, you first need to remove \texttt{\textbackslash{}n} in the sequence by using \texttt{\$seq\ =\textasciitilde{}\ s/\textbackslash{}n//g;}. Now the string of the variable (\texttt{\$seq}) is \texttt{ACGCTAGCTAGTCAGTCGATCGTCGTAGCTAGCTAG}. Then built-in function \texttt{length} can be used to calculate the length of the sequence. The function \texttt{print} is used to output the sequnce ID and the length of the sequence.

In the second cycles of the \texttt{while} loop, the chunk is
\texttt{chr2\textbackslash{}nCTGCGGGCTAAATCGATCGATCG\textbackslash{}nGTACGTACGAGCTAGCTA\textbackslash{}n\textgreater{}}.

Unlike the first chunk, there is no leading \texttt{\textgreater{}} in the second chunk. So the code \texttt{s/\^{}\textgreater{}//} will do nothing for the second chunk.

The following steps is the same as you can do for the first cycle.

In the third cycle of the \texttt{while} loop, you can do the same thing on the third sequence.



\begin{figure}
\centering
\includegraphics{figures/fasta_string.png}
\caption{\label{fig:fastaString}Fasta string.}
\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{perl}\NormalTok{ code_perl/fa_seq_len.pl}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Chrom    Length
## chr1 36
## chr2 41
## chr3 41
\end{verbatim}

\hypertarget{output-to-a-file-on-the-disk}{%
\section{Output to a file on the disk}\label{output-to-a-file-on-the-disk}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\KeywordTok{my} \DataTypeTok{$fasta}\NormalTok{   = }\KeywordTok{"}\StringTok{./data/test_ref.fa}\KeywordTok{"}\NormalTok{;}
\KeywordTok{my} \DataTypeTok{$out_len}\NormalTok{ = }\KeywordTok{"}\StringTok{./data/test_ref_len.txt}\KeywordTok{"}\NormalTok{;}

\FunctionTok{open}\NormalTok{ FASTA, }\DataTypeTok{$fasta} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\StringTok{Can't open file for reading: }\DataTypeTok{$!}\StringTok{ }\DataTypeTok{$fasta}\KeywordTok{"}\NormalTok{;}
\FunctionTok{open}\NormalTok{ OUT, }\KeywordTok{"}\StringTok{+>}\DataTypeTok{$out_len}\KeywordTok{"} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\StringTok{Can't open file for writing: }\DataTypeTok{$!}\StringTok{ }\DataTypeTok{$out_len}\KeywordTok{"}\NormalTok{;}
\CommentTok{# Set the input record separator}
\DataTypeTok{$/}\NormalTok{ = }\KeywordTok{"}\CharTok{\textbackslash{}n}\StringTok{>}\KeywordTok{"}\NormalTok{;}
\CommentTok{# Print header}
\FunctionTok{print}\NormalTok{ OUT }\KeywordTok{"}\StringTok{Chrom}\CharTok{\textbackslash{}t}\StringTok{Length}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\KeywordTok{while}\NormalTok{(}\KeywordTok{my} \DataTypeTok{$chunk}\NormalTok{ = }\KeywordTok{<FASTA>}\NormalTok{)\{}
    \CommentTok{# Remove > in the first chunk}
    \DataTypeTok{$chunk}\NormalTok{ =~ }\KeywordTok{s/}\CharTok{^}\OtherTok{>}\KeywordTok{//}\NormalTok{;}
    \CommentTok{# Remove \textbackslash{}n> from the end of the chunk}
    \FunctionTok{chomp} \DataTypeTok{$chunk}\NormalTok{;}
    \CommentTok{# Assign ID and seq to two variables}
    \KeywordTok{my}\NormalTok{ (}\DataTypeTok{$chrom}\NormalTok{, }\DataTypeTok{$seq}\NormalTok{) = }\FunctionTok{split}\NormalTok{(}\KeywordTok{/}\BaseNTok{\textbackslash{}n}\KeywordTok{/}\NormalTok{, }\DataTypeTok{$chunk}\NormalTok{, }\DecValTok{2}\NormalTok{);}
    \CommentTok{# Remove \textbackslash{}n in the string}
    \DataTypeTok{$seq}\NormalTok{ =~ }\KeywordTok{s/}\BaseNTok{\textbackslash{}n}\KeywordTok{//g}\NormalTok{;}
    \CommentTok{# Calculate the length of sequence.}
    \KeywordTok{my} \DataTypeTok{$seq_length}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$seq}\NormalTok{;}
    \CommentTok{# Print ID and length of the sequence.}
    \FunctionTok{print}\NormalTok{ OUT }\KeywordTok{"}\DataTypeTok{$chrom}\CharTok{\textbackslash{}t}\DataTypeTok{$seq_length}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\FunctionTok{close}\NormalTok{ FASTA;}
\FunctionTok{close}\NormalTok{ OUT;}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{perl}\NormalTok{ code_perl/fa_seq_len_out.pl}
\FunctionTok{cat}\NormalTok{ ./data/test_ref_len.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Chrom    Length
## chr1 36
## chr2 41
## chr3 41
\end{verbatim}

\hypertarget{arguments-from-command-line}{%
\section{Arguments from command line}\label{arguments-from-command-line}}

Imagine you have 5 fasta files and you want to calculate sequence lengths for all the 5 files, if you modify the file each time you run the code, it will be very tedious.

Perls provides an array called \texttt{@ARGV}. \texttt{@ARGV} holds all the arguments from the command line. The first one will be \texttt{\$ARGV{[}0{]}}.

\texttt{@ARGV} will automatically hold all the arguments. If no arguments are provided, \texttt{@ARGV} will be empty.

The following example shows you how it looks like in real code.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\KeywordTok{my}\NormalTok{ (}\DataTypeTok{$fasta}\NormalTok{, }\DataTypeTok{$out_len}\NormalTok{) = }\DataTypeTok{@ARGV}\NormalTok{; }
\FunctionTok{print} \KeywordTok{"}\StringTok{First arguments \textbackslash{}$ARGV[0] : }\DataTypeTok{$ARGV}\StringTok{[0]}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{Second arguments \textbackslash{}$ARGV[0]: }\DataTypeTok{$ARGV}\StringTok{[1]}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}

\FunctionTok{print} \KeywordTok{"}\StringTok{The variable \textbackslash{}$fasta:   }\DataTypeTok{$fasta}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print} \KeywordTok{"}\StringTok{The variable \textbackslash{}$out_len: }\DataTypeTok{$out_len}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{perl}\NormalTok{ code_perl/test_argv.pl test_ref.fa test_ref_len.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## First arguments $ARGV[0] : test_ref.fa
## Second arguments $ARGV[0]: test_ref_len.txt
## The variable $fasta:   test_ref.fa
## The variable $out_len: test_ref_len.txt
\end{verbatim}

In the above example, the first argument is \texttt{\$ARGV{[}0{]}} which stores the file name: test\_ref.fa. The second one stores the file name: test\_ref\_len.txt. It's highly recommended to assign the values in \texttt{@ARGV} to some variables that the readers (including yourself) can understand the variables based on the names.

Here I'll show you how to use ARGV using an example. The code below will show you what to do if you want to filter a fasta file based on the sequence length.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl}
\FunctionTok{use} \KeywordTok{warnings}\NormalTok{;}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\KeywordTok{my}\NormalTok{ (}\DataTypeTok{$fasta}\NormalTok{, }\DataTypeTok{$out_len}\NormalTok{) = }\DataTypeTok{@ARGV}\NormalTok{; }

\FunctionTok{open}\NormalTok{ FASTA, }\DataTypeTok{$fasta} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\StringTok{Can't open file for reading: }\DataTypeTok{$!}\StringTok{ }\DataTypeTok{$fasta}\KeywordTok{"}\NormalTok{;}
\FunctionTok{open}\NormalTok{ OUT, }\KeywordTok{"}\StringTok{+>}\DataTypeTok{$out_len}\KeywordTok{"} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\StringTok{Can't open file for writing: }\DataTypeTok{$!}\StringTok{ }\DataTypeTok{$out_len}\KeywordTok{"}\NormalTok{;}
\CommentTok{# Set the input record separator}
\DataTypeTok{$/}\NormalTok{ = }\KeywordTok{"}\CharTok{\textbackslash{}n}\StringTok{>}\KeywordTok{"}\NormalTok{;}
\CommentTok{# Print header}
\FunctionTok{print}\NormalTok{ OUT }\KeywordTok{"}\StringTok{The input file name is: }\DataTypeTok{$fasta}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\FunctionTok{print}\NormalTok{ OUT }\KeywordTok{"}\StringTok{Chrom}\CharTok{\textbackslash{}t}\StringTok{Length}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\KeywordTok{while}\NormalTok{(}\KeywordTok{my} \DataTypeTok{$chunk}\NormalTok{ = }\KeywordTok{<FASTA>}\NormalTok{)\{}
    \CommentTok{# Remove > in the first chunk}
    \DataTypeTok{$chunk}\NormalTok{ =~ }\KeywordTok{s/}\CharTok{^}\OtherTok{>}\KeywordTok{//}\NormalTok{;}
    \CommentTok{# Remove \textbackslash{}n> from the end of the chunk}
    \FunctionTok{chomp} \DataTypeTok{$chunk}\NormalTok{;}
    \CommentTok{# Assign ID and seq to two variables}
    \KeywordTok{my}\NormalTok{ (}\DataTypeTok{$chrom}\NormalTok{, }\DataTypeTok{$seq}\NormalTok{) = }\FunctionTok{split}\NormalTok{(}\KeywordTok{/}\BaseNTok{\textbackslash{}n}\KeywordTok{/}\NormalTok{, }\DataTypeTok{$chunk}\NormalTok{, }\DecValTok{2}\NormalTok{);}
    \CommentTok{# Remove \textbackslash{}n in the string}
    \DataTypeTok{$seq}\NormalTok{ =~ }\KeywordTok{s/}\BaseNTok{\textbackslash{}n}\KeywordTok{//g}\NormalTok{;}
    \CommentTok{# Calculate the length of sequence.}
    \KeywordTok{my} \DataTypeTok{$seq_length}\NormalTok{ = }\FunctionTok{length} \DataTypeTok{$seq}\NormalTok{;}
    \CommentTok{# Print ID and length of the sequence.}
    \FunctionTok{print}\NormalTok{ OUT }\KeywordTok{"}\DataTypeTok{$chrom}\CharTok{\textbackslash{}t}\DataTypeTok{$seq_length}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{\}}
\FunctionTok{close}\NormalTok{ FASTA;}
\FunctionTok{close}\NormalTok{ OUT;}
\end{Highlighting}
\end{Shaded}

For example if you want to remove the sequences that is shorter than 30 bps, you can use the following example.

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{printf} \StringTok{"Before filtering:\textbackslash{}n"}
\FunctionTok{cat}\NormalTok{ ./data/test_ref2.fa}
\FunctionTok{perl}\NormalTok{ code_perl/fa_seq_len_out_argv_fil.pl ./data/test_ref2.fa 30 ./data/test_ref2_30.fa}

\BuiltInTok{printf} \StringTok{"\textbackslash{}nAfter filtering: \textbackslash{}n"}
\FunctionTok{cat}\NormalTok{ ./data/test_ref2_30.fa }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Before filtering:
## >chr1
## ACGCTAGCTAGTCAGTCGATCGT
## CGTAGCTAGCTAG
## >chr2
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTAA
## >chr3
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAG
## >chr4
## CTGCGGGCTAAATCAGCTAA
## >chr5
## CTGCTCGATCGATCGACGAGCTA
## GCTA
## After filtering: 
## The input file name is: ./data/test_ref2.fa
## Chrom    Length
## >chr1 36
## ACGCTAGCTAGTCAGTCGATCGTCGTAGCTAGCTAG
## >chr2 42
## CTGCGGGCTAAATCGATCGATCGGTACGTACGAGCTAGCTAA
## >chr3 34
## CTGCGGGCTAAATCGATCGATCGGTACGTACGAG
\end{verbatim}

If you have multiple files, you can just change the file names in the command line. It's very handy. You can also change the length cutoff in the command line.

\hypertarget{practical-perl-program}{%
\chapter{Practical Perl program}\label{practical-perl-program}}

\hypertarget{add-annotation-information-to-deseq2-results}{%
\section{Add annotation information to DESeq2 results}\label{add-annotation-information-to-deseq2-results}}

Imaging we have a table that stores that differentially expressed genes information.

It includes three columns (tab-delimited):

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{cat}\NormalTok{ data/DEG_list.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## #gene_id log2fc  p-val
## gene1    2   0.01
## gene2    3   0.04
## gene3    -2  0.06
## gene4    -8  0.001
\end{verbatim}

We have another table which has the annotation information of each gene:

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{cat}\NormalTok{ data/gene_annotation.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## gene_id gene_shortname
## gene1    ROS
## gene2 WRKY
## gene3 ZmCCT
## gene4 WRKY1
## gene5 WRKY2
## gene6 wrky
## gene10  MCU
\end{verbatim}

\hypertarget{merge-overlap-genomic-regions}{%
\section{Merge overlap genomic regions}\label{merge-overlap-genomic-regions}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/perl -w}
\FunctionTok{use} \KeywordTok{strict}\NormalTok{;}

\KeywordTok{my}\NormalTok{ (}\DataTypeTok{$DMR}\NormalTok{, }\DataTypeTok{$out}\NormalTok{) = }\DataTypeTok{@ARGV}\NormalTok{;}

\FunctionTok{open}\NormalTok{ DMR, }\DataTypeTok{$DMR} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\StringTok{File not found: }\DataTypeTok{$!}\KeywordTok{"}\NormalTok{;}
\FunctionTok{open}\NormalTok{ OUT, }\KeywordTok{"}\StringTok{+>}\DataTypeTok{$out}\KeywordTok{"} \KeywordTok{or} \FunctionTok{die} \KeywordTok{"}\DataTypeTok{$!}\KeywordTok{"}\NormalTok{;}

\CommentTok{#read  a  line,}
\CommentTok{# go to next line}
\CommentTok{#     check overlap}
\CommentTok{#      if overlap: merge two regions. Then go to the next line.}
\CommentTok{#      if not overlap: print whatever we have now; then start from a line}
\KeywordTok{while}\NormalTok{(}\KeywordTok{my} \DataTypeTok{$region}\NormalTok{ = }\KeywordTok{<DMR>}\NormalTok{)\{}
    \FunctionTok{chomp} \DataTypeTok{$region}\NormalTok{;}
    \KeywordTok{my}\NormalTok{ (}\DataTypeTok{$chrom1}\NormalTok{, }\DataTypeTok{$start1}\NormalTok{, }\DataTypeTok{$end1}\NormalTok{) = }\FunctionTok{split}\NormalTok{(}\KeywordTok{/}\OtherTok{\textbackslash{}t}\KeywordTok{/}\NormalTok{, }\DataTypeTok{$region}\NormalTok{);}
\NormalTok{    BLOCK:\{}
        \KeywordTok{my} \DataTypeTok{$new_region}\NormalTok{ = }\KeywordTok{<DMR>}\NormalTok{;}

        \KeywordTok{if}\NormalTok{(!}\DataTypeTok{$new_region}\NormalTok{)\{}
            \FunctionTok{print}\NormalTok{ OUT }\KeywordTok{"}\DataTypeTok{$chrom1}\CharTok{\textbackslash{}t}\DataTypeTok{$start1}\CharTok{\textbackslash{}t}\DataTypeTok{$end1}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{        \}}\KeywordTok{else}\NormalTok{\{}
            \FunctionTok{chomp} \DataTypeTok{$new_region}\NormalTok{;}
            \KeywordTok{my}\NormalTok{ (}\DataTypeTok{$chrom2}\NormalTok{, }\DataTypeTok{$start2}\NormalTok{, }\DataTypeTok{$end2}\NormalTok{) = }\FunctionTok{split}\NormalTok{(}\KeywordTok{/}\OtherTok{\textbackslash{}t}\KeywordTok{/}\NormalTok{, }\DataTypeTok{$new_region}\NormalTok{);}
             \CommentTok{## start1===============end1}
             \CommentTok{##           start2=====================end2}
             \CommentTok{## start1===============================new_end1}
            \KeywordTok{if}\NormalTok{(}\DataTypeTok{$chrom2} \KeywordTok{eq} \DataTypeTok{$chrom1}\NormalTok{ && }\DataTypeTok{$start2}\NormalTok{ >= }\DataTypeTok{$start1}\NormalTok{ && }\DataTypeTok{$start2}\NormalTok{ <=}\DataTypeTok{$end1}\NormalTok{)\{}
                \DataTypeTok{$end1}\NormalTok{ = }\DataTypeTok{$end2}\NormalTok{;}
                \FunctionTok{redo}\NormalTok{ BLOCK;}
\NormalTok{            \}}\KeywordTok{else}\NormalTok{\{}
                \FunctionTok{print}\NormalTok{ OUT }\KeywordTok{"}\DataTypeTok{$chrom1}\CharTok{\textbackslash{}t}\DataTypeTok{$start1}\CharTok{\textbackslash{}t}\DataTypeTok{$end1}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{;}
\NormalTok{                (}\DataTypeTok{$chrom1}\NormalTok{, }\DataTypeTok{$start1}\NormalTok{, }\DataTypeTok{$end1}\NormalTok{) = (}\DataTypeTok{$chrom2}\NormalTok{, }\DataTypeTok{$start2}\NormalTok{, }\DataTypeTok{$end2}\NormalTok{);}
                \FunctionTok{redo}\NormalTok{ BLOCK;}
\NormalTok{            \}}
\NormalTok{        \}}
\NormalTok{    \}}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{cat}\NormalTok{ data/DMR_region.txt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## chr1 100 200
## chr1 150 250
## chr1 200 300
## chr1 500 600
## chr1 550 650
## chr2 300 800
## chr2 400 1000
## chr3 500 1000
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{perl}\NormalTok{ code_perl/genomic_coordinate_merge.pl  data/DMR_region.txt  data/DMR_region_merged.txt }
\FunctionTok{cat}\NormalTok{ data/DMR_region_merged.txt }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## chr1 100 300
## chr1 500 650
## chr2 300 1000
## chr3 500 1000
\end{verbatim}

\hypertarget{perl-modules}{%
\chapter{Perl modules}\label{perl-modules}}

\hypertarget{what-is-a-perl-module}{%
\section{What is a Perl module}\label{what-is-a-perl-module}}

Perl modules are a set of related functions in a library file. They are specifically designed to be reusable by other modules or programs. There are more than 100,000 modules ready for you to use on the Comprehensive Perl Archive Network.

Most Perl modules are written in Perl, some use XS (they are written in C) so require a C compiler. Modules may have dependencies on other modules (almost always on CPAN) and cannot be installed without them (or without a specific version of them). Many modules on CPAN now require a recent version of Perl (version 5.8 or above).

The reason that we need to use Perl module is Perl module can largely reduce our coding work.

\hypertarget{how-to-install-a-perl-module}{%
\section{How to install a Perl module}\label{how-to-install-a-perl-module}}

Here we'll only discuss how to install Perl module in Linux system. The easies way I think is to use \texttt{cpanm}.

Here is how you can do this. First go to webpage and download the \texttt{cpanm} source code.

\begin{verbatim}
## Web link:https://raw.githubusercontent.com/miyagawa/cpanminus/master/cpanm

chmod 755 cpanm

./cpanm Bio::Seq
\end{verbatim}

NOTE

\BeginKnitrBlock{rmdtip}
\textbf{Possible problem when using \texttt{cpanm}}

You may encouter the following error message. This is a bug from Perl. You cna run \texttt{yum\ install\ perl-CPAN} resolved the issue.

Here is the ERROR MESSAGE:

!
! Can't write to /usr/local/share/perl5 and /usr/local/bin: Installing modules to /home/xie186/perl5
! To turn off this warning, you have to do one of the following:
! - run me as a root or with --sudo option (to install to /usr/local/share/perl5 and /usr/local/bin)
! - Configure local::lib in your existing shell to set PERL\_MM\_OPT etc.
! - Install local::lib by running the following commands
!
! cpanm --local-lib=\textasciitilde{}/perl5 local::lib \&\& eval \$(perl -I \textasciitilde{}/perl5/lib/perl5/ -Mlocal::lib)
!
--\textgreater{} Working on Bio::Seq
Fetching \url{http://www.cpan.org/authors/id/C/CJ/CJFIELDS/BioPerl-1.007001.tar.gz} \ldots{} OK
==\textgreater{} Found dependencies: Module::Build, ExtUtils::Install
--\textgreater{} Working on Module::Build
Fetching \url{http://www.cpan.org/authors/id/L/LE/LEONT/Module-Build-0.4222.tar.gz} \ldots{} OK
==\textgreater{} Found dependencies: Module::Metadata, version, CPAN::Meta, Perl::OSType
--\textgreater{} Working on Module::Metadata
Fetching \url{http://www.cpan.org/authors/id/E/ET/ETHER/Module-Metadata-1.000033.tar.gz} \ldots{} OK
==\textgreater{} Found dependencies: ExtUtils::MakeMaker
--\textgreater{} Working on ExtUtils::MakeMaker
Fetching \url{http://www.cpan.org/authors/id/B/BI/BINGOS/ExtUtils-MakeMaker-7.26.tar.gz} \ldots{} OK
Configuring ExtUtils-MakeMaker-7.26 \ldots{} OK
Can't locate ExtUtils/Manifest.pm in \citet{INC} (\citet{INC} contains: FatPacked::26160008=HASH(0x18f2b88) /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor\_perl /usr/share/perl5/vendor\_perl /usr/lib64/perl5 /usr/share/perl5 .) at ./cpanm line 132.
\EndKnitrBlock{rmdtip}

\hypertarget{how-to-use-a-perl-module}{%
\section{How to use a Perl module}\label{how-to-use-a-perl-module}}

\hypertarget{how-to-use-bioperl-module}{%
\section{How to use BioPerl module}\label{how-to-use-bioperl-module}}

\hypertarget{how-to-write-a-module}{%
\section{How to write a module}\label{how-to-write-a-module}}

\hypertarget{part-statistics-and-r}{%
\part{Statistics and R}\label{part-statistics-and-r}}

\hypertarget{r-introduction}{%
\chapter{R introduction}\label{r-introduction}}

\hypertarget{basic-r-function}{%
\section{Basic R function}\label{basic-r-function}}

\hypertarget{producing-simple-graphs-with-r}{%
\section{Producing Simple Graphs with R}\label{producing-simple-graphs-with-r}}

The credit of this section goes to Dr.~Frank McCown (\citet{simpleGraphR}).

\hypertarget{line-charts}{%
\subsection{Line Charts}\label{line-charts}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Define the gene_expr_level vector with 5 values}
\NormalTok{gene_expr_level <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{8}\NormalTok{, }\DecValTok{20}\NormalTok{, }\DecValTok{20}\NormalTok{, }\DecValTok{100}\NormalTok{, }\DecValTok{120}\NormalTok{)}

\CommentTok{# Graph the gene_expr_fpkm vector with all defaults}
\KeywordTok{plot}\NormalTok{(gene_expr_level)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-115-1.pdf}

Let's add a title, a line to connect the points, and some color:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Define the gene_expr_level vector with 5 values}
\NormalTok{geneX_expr <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{8}\NormalTok{, }\DecValTok{20}\NormalTok{, }\DecValTok{20}\NormalTok{, }\DecValTok{100}\NormalTok{, }\DecValTok{120}\NormalTok{)}

\CommentTok{# Graph cars using blue points overlayed by a line }
\KeywordTok{plot}\NormalTok{(geneX_expr, }\DataTypeTok{type=}\StringTok{"o"}\NormalTok{, }\DataTypeTok{col=}\StringTok{"blue"}\NormalTok{)}

\CommentTok{# Create a title with a red, bold/italic font}
\KeywordTok{title}\NormalTok{(}\DataTypeTok{main=}\StringTok{"GeneX"}\NormalTok{, }\DataTypeTok{col.main=}\StringTok{"red"}\NormalTok{, }\DataTypeTok{font.main=}\DecValTok{4}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-116-1.pdf}

Now let's add a red line for trucks and specify the y-axis range directly so it will be large enough to fit the truck data:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Define the gene_expr_level vector with 5 values}
\NormalTok{geneX_expr <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{8}\NormalTok{, }\DecValTok{20}\NormalTok{, }\DecValTok{20}\NormalTok{, }\DecValTok{100}\NormalTok{, }\DecValTok{120}\NormalTok{)}
\NormalTok{geneY_expr <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\DecValTok{300}\NormalTok{, }\DecValTok{280}\NormalTok{, }\DecValTok{20}\NormalTok{, }\DecValTok{10}\NormalTok{, }\DecValTok{12}\NormalTok{)}

\CommentTok{# Graph cars using blue points overlayed by a line }
\KeywordTok{plot}\NormalTok{(geneX_expr, }\DataTypeTok{type=}\StringTok{"o"}\NormalTok{, }\DataTypeTok{col=}\StringTok{"blue"}\NormalTok{, }\DataTypeTok{ylim=}\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{300}\NormalTok{))}
\CommentTok{# Graph trucks with red dashed line and square points}
\KeywordTok{lines}\NormalTok{(geneY_expr, }\DataTypeTok{type=}\StringTok{"o"}\NormalTok{, }\DataTypeTok{pch=}\DecValTok{22}\NormalTok{, }\DataTypeTok{lty=}\DecValTok{2}\NormalTok{, }\DataTypeTok{col=}\StringTok{"red"}\NormalTok{)}
\CommentTok{# Create a title with a red, bold/italic font}
\KeywordTok{title}\NormalTok{(}\DataTypeTok{main=}\StringTok{"Gene expresion level"}\NormalTok{, }\DataTypeTok{col.main=}\StringTok{"red"}\NormalTok{, }\DataTypeTok{font.main=}\DecValTok{4}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-117-1.pdf}

\hypertarget{xxx}{%
\section{XXX}\label{xxx}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fruit =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"apple"}\NormalTok{, }\StringTok{"apple"}\NormalTok{, }\StringTok{"pear"}\NormalTok{, }\StringTok{"orange"}\NormalTok{)}
\NormalTok{fruit }\OperatorTok{==}\StringTok{ "apple"} 
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1]  TRUE  TRUE FALSE FALSE
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fruit =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"apple"}\NormalTok{, }\StringTok{"apple"}\NormalTok{, }\StringTok{"pear"}\NormalTok{, }\StringTok{"orange"}\NormalTok{)}
\KeywordTok{which}\NormalTok{(fruit }\OperatorTok{==}\StringTok{ "apple"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1 2
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fruit =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"apple"}\NormalTok{, }\StringTok{"apple"}\NormalTok{, }\StringTok{"pear"}\NormalTok{, }\StringTok{"orange"}\NormalTok{)}
\KeywordTok{which}\NormalTok{(fruit }\OperatorTok{==}\StringTok{ "apple"} \OperatorTok{|}\StringTok{ }\NormalTok{fruit }\OperatorTok{==}\StringTok{ "pear"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1 2 3
\end{verbatim}

\hypertarget{logic-and}{%
\section{\texorpdfstring{Logic \texttt{\&\&} and \texttt{\textbar{}}}{Logic \&\& and \textbar{}}}\label{logic-and}}

The short answer is that \&\& and \textbar{}\textbar{} only ever return a single (scalar, length-1 vector) TRUE or FALSE value, whereas \textbar{} and \& return a vector after doing element-by-element comparisons.

The only place in R you routinely use a scalar TRUE/FALSE value is in the conditional of an if statement, so you'll often see \&\& or \textbar{}\textbar{} used in idioms like:

if (length(x) \textgreater{} 0 \&\& any(is.na(x))) \{ do.something() \}

In most other instances you'll be working with vectors and use \& and \textbar{} instead.

\hypertarget{list-as-dictionary}{%
\section{List as dictionary}\label{list-as-dictionary}}

the list type is a good approximation. You can use names() on your list to set and retrieve the `keys':

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{foo <-}\StringTok{ }\KeywordTok{vector}\NormalTok{(}\DataTypeTok{mode=}\StringTok{"list"}\NormalTok{, }\DataTypeTok{length=}\DecValTok{3}\NormalTok{)}
\KeywordTok{names}\NormalTok{(foo) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"tic"}\NormalTok{, }\StringTok{"tac"}\NormalTok{, }\StringTok{"toe"}\NormalTok{)}
\NormalTok{foo[[}\DecValTok{1}\NormalTok{]] <-}\StringTok{ }\DecValTok{12}\NormalTok{; foo[[}\DecValTok{2}\NormalTok{]] <-}\StringTok{ }\DecValTok{22}\NormalTok{; foo[[}\DecValTok{3}\NormalTok{]] <-}\StringTok{ }\DecValTok{33}
\NormalTok{foo}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $tic
## [1] 12
## 
## $tac
## [1] 22
## 
## $toe
## [1] 33
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{names}\NormalTok{(foo)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "tic" "tac" "toe"
\end{verbatim}

\hypertarget{parsing-arguments-as-string}{%
\section{Parsing arguments as string}\label{parsing-arguments-as-string}}

\hypertarget{string-as-xlim}{%
\subsection{String as xlim}\label{string-as-xlim}}

\hypertarget{how-to-access-data-frame-column-using-variable}{%
\subsection{How to access data frame column using variable}\label{how-to-access-data-frame-column-using-variable}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{a =}\StringTok{ "col1"}
\NormalTok{b =}\StringTok{ "col2"}
\NormalTok{d =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{a=}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{),}\DataTypeTok{b=}\KeywordTok{c}\NormalTok{(}\DecValTok{4}\NormalTok{,}\DecValTok{5}\NormalTok{,}\DecValTok{6}\NormalTok{))}
\KeywordTok{colnames}\NormalTok{(d) <-}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"col1"}\NormalTok{, }\StringTok{"col2"}\NormalTok{)}
\NormalTok{d[[a]]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 1 2 3
\end{verbatim}

This is useful when you parse a variable from the command line

\hypertarget{how-to-create-a-formula-from-a}{%
\subsection{How to create a formula from a}\label{how-to-create-a-formula-from-a}}

It can be useful to create a formula from a string. This often occurs in functions where the formula arguments are passed in as strings.

It can be useful to create a formula from a string. This often occurs in functions where the formula arguments are passed in as strings.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{design1 =}\StringTok{ "diet"}
\NormalTok{design2 =}\StringTok{ "age"}
\CommentTok{## `~ diet + age`}
\KeywordTok{as.formula}\NormalTok{(}\KeywordTok{paste0}\NormalTok{(}\StringTok{"~ "}\NormalTok{ , design1, }\StringTok{" + "}\NormalTok{, design2))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## ~diet + age
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{cat}\NormalTok{(}\KeywordTok{readLines}\NormalTok{(}\StringTok{'code_R/parse_aug_as.formula.R'}\NormalTok{), }\DataTypeTok{sep =} \StringTok{'}\CharTok{\textbackslash{}n}\StringTok{'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
argv <- commandArgs(trailingOnly = T)

level1 <- argv[1]
level2 <- argv[2]

#First, build a simple data frame with time as a factor and Time as a continuous,
#numeric variable. The two variables look alike when you print the data frame.
#But, if you summarize the data, you see that they are different.
d <- data.frame(level1 = factor(1:4), level2 = 1:4)
colnames(d)<-c(level1, level2)
summary(d)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{/opt/R/3.5.3/lib/R/bin/Rscript}\NormalTok{ code_R/parse_aug_as.formula.R time Time}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  time       Time     
##  1:1   Min.   :1.00  
##  2:1   1st Qu.:1.75  
##  3:1   Median :2.50  
##  4:1   Mean   :2.50  
##        3rd Qu.:3.25  
##        Max.   :4.00
\end{verbatim}

\hypertarget{ggplot2}{%
\chapter{\texorpdfstring{\texttt{ggplot2}}{ggplot2}}\label{ggplot2}}

\hypertarget{ggplot2-1}{%
\section{ggplot2}\label{ggplot2-1}}

\hypertarget{ggplot2-practical}{%
\section{ggplot2 practical}\label{ggplot2-practical}}

\hypertarget{heatmap-tutorial}{%
\chapter{Heatmap Tutorial}\label{heatmap-tutorial}}

\href{https://goo.gl/KLZ7N0}{Data link}

\hypertarget{install-pheatmap-package}{%
\section{\texorpdfstring{Install \texttt{pheatmap} package}{Install pheatmap package}}\label{install-pheatmap-package}}

\begin{verbatim}
install.packages("pheatmap")
\end{verbatim}

\hypertarget{draw-a-heatmap-for-gene-expression-of-rna-seq-data}{%
\section{Draw a heatmap for gene expression of RNA-seq data}\label{draw-a-heatmap-for-gene-expression-of-rna-seq-data}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(pheatmap)}
\NormalTok{gene_exp <-}\StringTok{ }\KeywordTok{read.table}\NormalTok{(}\StringTok{"data/maize_embryo_specific_gene_Sheet1.tsv"}\NormalTok{, }\DataTypeTok{header=}\NormalTok{T, }\DataTypeTok{row.names=}\DecValTok{1}\NormalTok{)}
\KeywordTok{pheatmap}\NormalTok{(gene_exp)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-127-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(ggplot2)}
\CommentTok{# Set the theme for all the following plots.}
\KeywordTok{theme_set}\NormalTok{(}\KeywordTok{theme_bw}\NormalTok{(}\DataTypeTok{base_size =} \DecValTok{16}\NormalTok{))}

\NormalTok{dat <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{values =} \KeywordTok{as.numeric}\NormalTok{(}\KeywordTok{unlist}\NormalTok{(gene_exp)))}
\KeywordTok{summary}\NormalTok{(dat)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      values        
##  Min.   :    0.00  
##  1st Qu.:    0.00  
##  Median :    0.40  
##  Mean   :   38.91  
##  3rd Qu.:    6.22  
##  Max.   :25534.25
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{ggplot}\NormalTok{(dat, }\KeywordTok{aes}\NormalTok{(values)) }\OperatorTok{+}\StringTok{ }\KeywordTok{geom_density}\NormalTok{(}\DataTypeTok{bw =} \StringTok{"SJ"}\NormalTok{) }\OperatorTok{+}\StringTok{ }\KeywordTok{xlim}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{100}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning: Removed 958 rows containing non-finite values (stat_density).
\end{verbatim}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-128-1.pdf}

The data is skewed.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{log2}\NormalTok{(gene_exp }\OperatorTok{+}\StringTok{ }\FloatTok{0.000001}\NormalTok{), }\DataTypeTok{scale=}\StringTok{"row"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-129-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{log2}\NormalTok{(gene_exp }\OperatorTok{+}\StringTok{ }\FloatTok{0.01}\NormalTok{), }\DataTypeTok{scale=}\StringTok{"row"}\NormalTok{, }\DataTypeTok{show_rownames =}\NormalTok{ T, }\DataTypeTok{show_colnames =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-129-2.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{log2}\NormalTok{(gene_exp }\OperatorTok{+}\StringTok{ }\FloatTok{0.01}\NormalTok{), }\DataTypeTok{scale=}\StringTok{"row"}\NormalTok{, }\DataTypeTok{show_rownames =}\NormalTok{ F, }\DataTypeTok{show_colnames =}\NormalTok{ T)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-129-3.pdf}

\hypertarget{add-the-annotation}{%
\section{Add the annotation}\label{add-the-annotation}}

\hypertarget{section}{%
\section{}\label{section}}

\hypertarget{transfrom-the-data}{%
\section{Transfrom the data}\label{transfrom-the-data}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{log2}\NormalTok{(gene_exp }\OperatorTok{+}\StringTok{ }\FloatTok{0.01}\NormalTok{), }\DataTypeTok{show_rownames =}\NormalTok{ F) }
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-130-1.pdf}

\hypertarget{how-to-add-annotations}{%
\section{How to add annotations}\label{how-to-add-annotations}}

\hypertarget{how-to-cut-the-trees}{%
\section{How to cut the trees}\label{how-to-cut-the-trees}}

\hypertarget{how-to-get-the-cluster-information-from-the-heatmap}{%
\section{How to get the cluster information from the heatmap}\label{how-to-get-the-cluster-information-from-the-heatmap}}

\hypertarget{change-color}{%
\section{Change color}\label{change-color}}

\hypertarget{use-brewer}{%
\subsection{Use Brewer}\label{use-brewer}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{col.pal <-}\StringTok{ }\NormalTok{RColorBrewer}\OperatorTok{::}\KeywordTok{brewer.pal}\NormalTok{(}\DecValTok{9}\NormalTok{, }\StringTok{"Reds"}\NormalTok{)}
\NormalTok{col.pal }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "#FFF5F0" "#FEE0D2" "#FCBBA1" "#FC9272" "#FB6A4A" "#EF3B2C" "#CB181D"
## [8] "#A50F15" "#67000D"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{log2}\NormalTok{(gene_exp }\OperatorTok{+}\StringTok{ }\FloatTok{0.01}\NormalTok{), }\DataTypeTok{scale=}\StringTok{"row"}\NormalTok{,}
         \DataTypeTok{show_rownames =}\NormalTok{ F, }\DataTypeTok{show_colnames =}\NormalTok{ T)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-131-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{pheatmap}\NormalTok{(}\KeywordTok{log2}\NormalTok{(gene_exp }\OperatorTok{+}\StringTok{ }\FloatTok{0.01}\NormalTok{), }\DataTypeTok{scale=}\StringTok{"row"}\NormalTok{, }\DataTypeTok{color =}\NormalTok{ col.pal,}
         \DataTypeTok{show_rownames =}\NormalTok{ F, }\DataTypeTok{show_colnames =}\NormalTok{ T)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-131-2.pdf}

\hypertarget{preparation-of-figures-for-manuscript}{%
\chapter{Preparation of figures for manuscript}\label{preparation-of-figures-for-manuscript}}

Inkscape is an open source software which can be used to arrange, crop, and annotate your images; bring in graphs and charts; draw diagrams; and export the final figure in whatever format the journal wants.

\hypertarget{test1}{%
\section{test1}\label{test1}}

\hypertarget{part-omics-data}{%
\part{Omics data}\label{part-omics-data}}

\hypertarget{introduction-to-ngs}{%
\chapter{Introduction to NGS}\label{introduction-to-ngs}}

Next-generation sequencing (NGS) is one of the fundamental technological developments of the decade in life sciences. Whole genome sequencing (WGS), RAD-Seq, RNA-Seq, Chip-Seq, and several other technologies are routinely used to investigate important biological problems. These are also called high-throughput sequencing technologies, and with good reason: they generate vast amounts of data that needs to be processed. NGS is the main reason that computational biology has become a big-data discipline. More than anything else, this is a field that requires strong bioinformatics techniques.

As this is not an introductory book, you are expected to know at least what FASTA, FASTQ, Binary Alignment Map (BAM), and Variant Call Format (VCF) files are. I will also make use of the basic genomic terminology without introducing it (such as exomes, nonsynonymous mutations, and so on). You are required to be familiar with basic Python. We will leverage this knowledge to introduce the fundamental libraries in Python to perform the NGS analysis. Here, we will follow the flow of a standard bioinformatics pipeline.

However, before we delve into real data from a real project, let's get comfortable with accessing existing genomic databases and basic sequence processing---a simple start before the storm.

\hypertarget{introduction-of-biology-for-bioinformatics}{%
\section{Introduction of Biology for Bioinformatics}\label{introduction-of-biology-for-bioinformatics}}

\hypertarget{what-is-dna}{%
\subsection{What is DNA}\label{what-is-dna}}

\hypertarget{what-is-genome}{%
\subsection{What is genome}\label{what-is-genome}}

\hypertarget{how-to-assemble-a-genome}{%
\subsection{How to assemble a genome}\label{how-to-assemble-a-genome}}

\hypertarget{what-is-gene}{%
\subsection{What is gene}\label{what-is-gene}}

\hypertarget{what-is-ngs}{%
\section{What is NGS}\label{what-is-ngs}}

Genetic information is stored within the chemical structure of the molecule DNA. DNA can be transcribed into RNA. Then RNA can be translated into protein.

NGS (Next geneneration sequencing) is not an accurate term. An more accurate term maybe high throughput sequencing. NGS is specifically used to refer to sequencing technologies after Sanger sequencing (1st generation) and single molecular sequencing (3rd sequencing, e.g.~PacBio) and nanopore based sequencing (4th generation, e.g.~Oxford Nanopore).

\hypertarget{application-of-ngs}{%
\section{Application of NGS}\label{application-of-ngs}}

\hypertarget{section-1}{%
\section{}\label{section-1}}

\hypertarget{data-file-formats-to-store-the-genomic-information}{%
\section{Data file formats to store the genomic information}\label{data-file-formats-to-store-the-genomic-information}}

\hypertarget{fasta-file}{%
\subsection{\texorpdfstring{\texttt{fasta} file}{fasta file}}\label{fasta-file}}

FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes. A sequence in FASTA format begins with a single-line description, followed by lines of sequence data. The description line is distinguished from the sequence data by a greater-than (``\textgreater{}'') symbol in the first column. It is recommended that all lines of text be shorter than 80 characters in length.

A sequence in FASTA format begins with a single-line description, followed by lines of sequence data. The description line (defline) is distinguished from the sequence data by a greater-than (``\textgreater{}'') symbol at the beginning. It is recommended that all lines of text be shorter than 80 characters in length. An example sequence in FASTA format is:

\begin{verbatim}
>gi|186681228|ref|YP_001864424.1| phycoerythrobilin:ferredoxin oxidoreductase
MNSERSDVTLYQPFLDYAIAYMRSRLDLEPYPIPTGFESNSAVVGKGKNQEEVVTTSYAFQTAKLRQIRA
AHVQGGNSLQVLNFVIFPHLNYDLPFFGADLVTLPGGHLIALDMQPLFRDDSAYQAKYTEPILPIFHAHQ
QHLSWGGDFPEEAQPFFSPAFLWTRPQETAVVETQVFAAFKDYLKAYLDFVEQAEAVTDSQNLVAIKQAQ
LRYLRYRAEKDPARGMFKRFYGAEWTEEYIHGFLFDLERKLTVVK
\end{verbatim}

\hypertarget{gff-file}{%
\subsection{GFF file}\label{gff-file}}

The GFF (gene-finding format, generic feature format or general feature format) is a file format used for describing genes and other features of DNA, RNA and protein sequences. The filename extension associated with such files is .GFF and the content type associated with them is text/x-gff3. There are two versions of the GFF file format in general use. We here focus on GFF3.

are tabular files with 9 fields per line, separated by tabs. They all share the same structure for the first 7 fields, while differing in the content and format of the ninth field. The general structure is as follows:

\begin{longtable}[]{@{}lll@{}}
\toprule
\begin{minipage}[b]{0.03\columnwidth}\raggedright
Position index\strut
\end{minipage} & \begin{minipage}[b]{0.03\columnwidth}\raggedright
Position name\strut
\end{minipage} & \begin{minipage}[b]{0.85\columnwidth}\raggedright
Description\strut
\end{minipage}\tabularnewline
\midrule
\endhead
\begin{minipage}[t]{0.03\columnwidth}\raggedright
1\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
sequence\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
The name of the sequence where the feature is located.\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.03\columnwidth}\raggedright
2\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
source\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
Keyword identifying the source of the feature, like a program (e.g.~Augustus or RepeatMasker) or an organization (like TAIR).\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.03\columnwidth}\raggedright
3\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
feature\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
The feature type name, like ``gene'' or ``exon''. In a well structured GFF file, all the children features always follow their parents in a single block (so all exons of a transcript are put after their parent ``transcript'' feature line and before any other parent transcript line). In GFF3, all features and their relationships should be compatible with the standards released by the Sequence Ontology Project.\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.03\columnwidth}\raggedright
4\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
start\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
Genomic start of the feature, with a 1-base offset. This is in contrast with other 0-offset half-open sequence formats, like BED files.\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.03\columnwidth}\raggedright
5\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
end\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
Genomic end of the feature, with a 1-base offset. This is the same end coordinate as it is in 0-offset half-open sequence formats, like BED files.{[}citation needed{]}\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.03\columnwidth}\raggedright
6\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
score\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
Numeric value that generally indicates the confidence of the source on the annotated feature. A value of ``.'' (a dot) is used to define a null value.\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.03\columnwidth}\raggedright
7\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
strand\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
Single character that indicates the Sense (molecular biology) strand of the feature; it can assume the values of ``+'' (positive, or 5'-\textgreater{}3'), ``-'', (negative, or 3'-\textgreater{}5'), ``.'' (undetermined).\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.03\columnwidth}\raggedright
8\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
phase\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
phase of CDS features; it can be either one of 0, 1, 2 (for CDS features) or ``.'' (for everything else). See the section below for a detailed explanation.\strut
\end{minipage}\tabularnewline
\begin{minipage}[t]{0.03\columnwidth}\raggedright
9\strut
\end{minipage} & \begin{minipage}[t]{0.03\columnwidth}\raggedright
Attributes.\strut
\end{minipage} & \begin{minipage}[t]{0.85\columnwidth}\raggedright
All the other information pertaining to this feature. The format, structure and content of this field is the one which varies the most between the three competing file formats.\strut
\end{minipage}\tabularnewline
\bottomrule
\end{longtable}

\hypertarget{reference}{%
\subsubsection{Reference}\label{reference}}

Generic Feature Format Version 3 (GFF3): \url{https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md}

General feature format wikipage: \url{https://en.wikipedia.org/wiki/General_feature_format}

\hypertarget{gtf-file}{%
\subsection{GTF file}\label{gtf-file}}

\hypertarget{fastq-file}{%
\subsection{FASTQ file}\label{fastq-file}}

\hypertarget{sambam-file}{%
\subsection{SAM/BAM file}\label{sambam-file}}

\hypertarget{vcf-file}{%
\subsection{VCF file}\label{vcf-file}}

\hypertarget{bed-format}{%
\subsection{BED format}\label{bed-format}}

\hypertarget{bedgraph-format}{%
\subsection{bedGraph format}\label{bedgraph-format}}

\hypertarget{usefull-links}{%
\section{Usefull links:}\label{usefull-links}}

\url{http://www.ecseq.com/support/ngs/trimming-adapter-sequences-is-it-necessary}

\url{http://www.gendx.com/illumina-adapter-ligation-librx}

\url{http://veleta.rosety.com/plasmid.html}

\hypertarget{bs-seq}{%
\chapter{BS-seq}\label{bs-seq}}

\hypertarget{section-2}{%
\section{}\label{section-2}}

\url{https://www.slideshare.net/secret/Da9IOe8wLsaF8V}

\hypertarget{what-is-bs-seq}{%
\section{What is BS-seq}\label{what-is-bs-seq}}

\hypertarget{section-3}{%
\section{}\label{section-3}}

\hypertarget{how-to-analyze-bs-seq-data}{%
\section{How to analyze BS-seq data}\label{how-to-analyze-bs-seq-data}}

\hypertarget{section-4}{%
\subsection{}\label{section-4}}

MINI REVIEW: Statistical methods for detecting differentially methylated loci and regions

Strategies for analyzing bisulfite sequencing data

\hypertarget{start-a-project}{%
\chapter{Start a project}\label{start-a-project}}

\hypertarget{experimetal-design}{%
\section{Experimetal design}\label{experimetal-design}}

Before

\hypertarget{how-many-biological-replicates}{%
\subsection{How many biological replicates}\label{how-many-biological-replicates}}

\hypertarget{how-big}{%
\subsection{How big}\label{how-big}}

\hypertarget{how-much-data-to-generate}{%
\subsection{How much data to generate}\label{how-much-data-to-generate}}

\hypertarget{section-5}{%
\subsection{}\label{section-5}}

Where to start?

There are many ways to start to talk about Bioinformatics analysis. Here we'll go from whether there is a reference genome or not. Let's take RNA-seq as an example.

\hypertarget{capstone-project}{%
\chapter{Capstone project:}\label{capstone-project}}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

\hypertarget{method}{%
\section{Method}\label{method}}

\hypertarget{pipelines}{%
\section{Pipelines}\label{pipelines}}

\hypertarget{section-6}{%
\section{}\label{section-6}}

\hypertarget{data.table}{%
\chapter{data.table}\label{data.table}}

\hypertarget{split-data.table-into-chunks-in-a-list}{%
\section{Split data.table into chunks in a list}\label{split-data.table-into-chunks-in-a-list}}

Split method for data.table. Faster and more flexible. Be aware that processing list of data.tables will be generally much slower than manipulation in single data.table by group using by argument, read more on data.table.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(data.table)}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{123}\NormalTok{)}
\NormalTok{dt =}\StringTok{ }\KeywordTok{data.table}\NormalTok{(}\DataTypeTok{x1 =} \KeywordTok{rep}\NormalTok{(letters[}\DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{], }\DecValTok{6}\NormalTok{), }
                \DataTypeTok{x2 =} \KeywordTok{rep}\NormalTok{(letters[}\DecValTok{3}\OperatorTok{:}\DecValTok{5}\NormalTok{], }\DecValTok{4}\NormalTok{), }
                \DataTypeTok{x3 =} \KeywordTok{rep}\NormalTok{(letters[}\DecValTok{5}\OperatorTok{:}\DecValTok{8}\NormalTok{], }\DecValTok{3}\NormalTok{), }
                \DataTypeTok{y =} \KeywordTok{rnorm}\NormalTok{(}\DecValTok{12}\NormalTok{))}
\NormalTok{dt =}\StringTok{ }\NormalTok{dt[}\KeywordTok{sample}\NormalTok{(.N)]}
\NormalTok{df =}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(dt)}
\NormalTok{df}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    x1 x2 x3           y
## 1   b  d  h -1.26506123
## 2   b  e  h  0.35981383
## 3   b  e  f  1.71506499
## 4   b  c  f -0.44566197
## 5   a  e  g  1.55870831
## 6   b  d  f -0.23017749
## 7   a  e  e -0.68685285
## 8   a  d  e  0.12928774
## 9   a  d  g  1.22408180
## 10  b  c  h  0.07050839
## 11  a  c  e -0.56047565
## 12  a  c  g  0.46091621
\end{verbatim}

\hypertarget{nested-list-using-flatten-arguments}{%
\subsection{\texorpdfstring{nested list using \texttt{flatten} arguments}{nested list using flatten arguments}}\label{nested-list-using-flatten-arguments}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{new_list <-}\StringTok{ }\KeywordTok{split}\NormalTok{(dt, }\DataTypeTok{by=}\KeywordTok{c}\NormalTok{(}\StringTok{"x1"}\NormalTok{, }\StringTok{"x2"}\NormalTok{))}
\NormalTok{new_list}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $b.d
##    x1 x2 x3          y
## 1:  b  d  h -1.2650612
## 2:  b  d  f -0.2301775
## 
## $b.e
##    x1 x2 x3         y
## 1:  b  e  h 0.3598138
## 2:  b  e  f 1.7150650
## 
## $b.c
##    x1 x2 x3           y
## 1:  b  c  f -0.44566197
## 2:  b  c  h  0.07050839
## 
## $a.e
##    x1 x2 x3          y
## 1:  a  e  g  1.5587083
## 2:  a  e  e -0.6868529
## 
## $a.d
##    x1 x2 x3         y
## 1:  a  d  e 0.1292877
## 2:  a  d  g 1.2240818
## 
## $a.c
##    x1 x2 x3          y
## 1:  a  c  e -0.5604756
## 2:  a  c  g  0.4609162
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{new_list <-}\StringTok{ }\KeywordTok{split}\NormalTok{(dt, }\DataTypeTok{by=}\KeywordTok{c}\NormalTok{(}\StringTok{"x1"}\NormalTok{, }\StringTok{"x2"}\NormalTok{), }\DataTypeTok{flatten=}\OtherTok{FALSE}\NormalTok{)}
\NormalTok{new_list}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## $b
## $b$d
##    x1 x2 x3          y
## 1:  b  d  h -1.2650612
## 2:  b  d  f -0.2301775
## 
## $b$e
##    x1 x2 x3         y
## 1:  b  e  h 0.3598138
## 2:  b  e  f 1.7150650
## 
## $b$c
##    x1 x2 x3           y
## 1:  b  c  f -0.44566197
## 2:  b  c  h  0.07050839
## 
## 
## $a
## $a$e
##    x1 x2 x3          y
## 1:  a  e  g  1.5587083
## 2:  a  e  e -0.6868529
## 
## $a$d
##    x1 x2 x3         y
## 1:  a  d  e 0.1292877
## 2:  a  d  g 1.2240818
## 
## $a$c
##    x1 x2 x3          y
## 1:  a  c  e -0.5604756
## 2:  a  c  g  0.4609162
\end{verbatim}

\hypertarget{example}{%
\subsection{Example}\label{example}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{dt_example =}\StringTok{ }\KeywordTok{data.table}\NormalTok{(}\DataTypeTok{group =} \KeywordTok{rep}\NormalTok{(}\KeywordTok{c}\NormalTok{(}\StringTok{"group1"}\NormalTok{, }\StringTok{"group2"}\NormalTok{), }\DecValTok{4}\NormalTok{), }
                \DataTypeTok{gene =} \KeywordTok{c}\NormalTok{(letters[}\DecValTok{1}\OperatorTok{:}\DecValTok{4}\NormalTok{], letters[}\DecValTok{3}\OperatorTok{:}\DecValTok{6}\NormalTok{]))}
\NormalTok{dt}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##     x1 x2 x3           y
##  1:  b  d  h -1.26506123
##  2:  b  e  h  0.35981383
##  3:  b  e  f  1.71506499
##  4:  b  c  f -0.44566197
##  5:  a  e  g  1.55870831
##  6:  b  d  f -0.23017749
##  7:  a  e  e -0.68685285
##  8:  a  d  e  0.12928774
##  9:  a  d  g  1.22408180
## 10:  b  c  h  0.07050839
## 11:  a  c  e -0.56047565
## 12:  a  c  g  0.46091621
\end{verbatim}

\hypertarget{crate-a-matrix-from-data.table}{%
\subsubsection{Crate a matrix from data.table}\label{crate-a-matrix-from-data.table}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(UpSetR)}
\NormalTok{list_group =}\StringTok{ }\KeywordTok{split}\NormalTok{(dt_example[}\OperatorTok{-}\KeywordTok{which}\NormalTok{(}\KeywordTok{names}\NormalTok{(df)}\OperatorTok{==}\StringTok{"z"}\NormalTok{)], }\DataTypeTok{by=}\StringTok{"group"}\NormalTok{,  }\DataTypeTok{drop=}\OtherTok{TRUE}\NormalTok{)}
\NormalTok{list_group}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## named list()
\end{verbatim}

\hypertarget{practical-data-analysis-on-wgbs}{%
\chapter{Practical data analysis on WGBS}\label{practical-data-analysis-on-wgbs}}

We are going to use the data in this \href{https://www.ncbi.nlm.nih.gov//geo/query/acc.cgi?acc=GSE66905}{link}.

In this link, it has multile WGBS from different genotypes of yeast. For simplicity, we are going to only pick 3 samples: \texttt{EV\ strain\ 1\ (EV1)}, \texttt{dnmt3b\ strain\ 1\ (3bstrain1)} and \texttt{set1\ replicate1\ (set1rep1)}

We are going to use the data from these three samples to illustrate how to analyze WGBS data (Figure @ref(fig:wgbs\_bg)).

\begin{figure}
\centering
\includegraphics{figures/elife_yeast_paper.jpg}
\caption{(\#fig:wgbs\_bg)Study background of WGBS data}
\end{figure}

Let's walk through these workflow step-by-step.

\hypertarget{download-the-wgbs-data}{%
\section{Download the WGBS data}\label{download-the-wgbs-data}}

Based on the GEO \href{https://www.ncbi.nlm.nih.gov//geo/query/acc.cgi?acc=GSE66905}{link} here, we can use the command lines below to download the \texttt{sra} files.

(ref:wgbs\_sample) Sample list for WGBS data

\textbackslash{}begin\{table\}{[}t{]}

\textbackslash{}caption\{(\#tab:wgbs\_sample)(ref:wgbs\_sample)\}
\centering

\begin{tabular}{c|c}
\hline
SRR\_number & Sample\_name\\
\hline
SRR1916129 & EV1\\
\hline
SRR1916134 & 3bstrain1\\
\hline
SRR1916142 & set1rep1\\
\hline
\end{tabular}

\textbackslash{}end\{table\}

\begin{verbatim}
wget -O EV1.sra ftp://ftp.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR191/SRR1916129/SRR1916129.sra
wget -O 3bstrain1.sra  ftp://ftp.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR191/SRR1916134/SRR1916134.sra
wget -O set1_rep1.sra ftp://ftp.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/SRR191/SRR1916142/SRR1916142.sra
\end{verbatim}

\hypertarget{convert-sra-files-to-fastq-files}{%
\section{\texorpdfstring{Convert \texttt{sra} files to \texttt{fastq} files}{Convert sra files to fastq files}}\label{convert-sra-files-to-fastq-files}}

\begin{verbatim}
fastq-dump 3bstrain1.sra 
fastq-dump EV1.sra 
fastq-dump set1rep1.sra 
\end{verbatim}

This step will generate three files suffixed with \texttt{.fastq}:

\begin{itemize}
\tightlist
\item
  3bstrain1.fastq
\item
  EV1.fastq
\item
  set1rep1.fastq
\end{itemize}

\hypertarget{quality-control-of-the-reads}{%
\section{Quality control of the reads}\label{quality-control-of-the-reads}}

\hypertarget{quality-checking}{%
\subsection{Quality checking}\label{quality-checking}}

\begin{verbatim}
fastqc 3bstrain1.fastq
fastqc EV1.fastq 
fastqc set1rep1.fastq
\end{verbatim}

\hypertarget{quality-control-using-trim_galore}{%
\subsection{\texorpdfstring{Quality control using \texttt{trim\_galore}}{Quality control using trim\_galore}}\label{quality-control-using-trim_galore}}

The default value for \texttt{-q} (or \texttt{-\/-quality}) is 20 meaning \texttt{Trim\ low-quality\ ends\ from\ reads}. After we remove the low quality bases, a read could be very short. So \texttt{-\/-length\ 75} can be used to discard any reads that are shorter than 75 bps after quality control.

\begin{verbatim}
trim_galore --length 75 EV1.fastq 
## Equivalent of:
trim_galore -q 20 --length 75 EV1.fastq 

trim_galore --length 75 3bstrain1.fastq
trim_galore --length 75 set1rep1.fastq 
\end{verbatim}

Taken \texttt{EV1.fastq} as an exmaple, two files will be generated: \texttt{EV1\_trimmed.fa} and \texttt{EV1.fastq\_trimming\_report.txt}.

\hypertarget{donwload-reference-genome}{%
\section{Donwload reference genome}\label{donwload-reference-genome}}

You need to download the reference genome (in fasta file) and the genome annotation file (in gff/gtf file).

Here is the link for reference genome: \url{http://useast.ensembl.org/Saccharomyces_cerevisiae/Info/Index}

\begin{verbatim}
wget ftp://ftp.ensembl.org/pub/release-95/fasta/saccharomyces_cerevisiae/dna/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa.gz
wget ftp://ftp.ensembl.org/pub/release-95/gff3/saccharomyces_cerevisiae/Saccharomyces_cerevisiae.R64-1-1.95.gff3.gz
\end{verbatim}

\hypertarget{build-bismark-index-for-bismark-alignment}{%
\section{\texorpdfstring{Build bismark index for \texttt{Bismark} alignment}{Build bismark index for Bismark alignment}}\label{build-bismark-index-for-bismark-alignment}}

\begin{verbatim}
mkdir bismark_index/

zcat Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa.gz > bismark_index/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa
\end{verbatim}

\begin{verbatim}
bismark_genome_preparation bismark_index/
\end{verbatim}

This script \texttt{bismark\_genome\_preparation} needs to be run only once to prepare the genome of interest for bisulfite alignments. You need to specify a directory containing the genome you want to align your reads against (please be aware that the bismark\_genome\_preparation script expects FastA files in this folder (with either .fa or .fasta extension, single or multiple sequence entries per file). Bismark will create two individual folders within this directory, one for a C-\textgreater{}T converted genome and the other one for the G-\textgreater{}A converted genome. After creating C-\textgreater{}T and G-\textgreater{}A versions of the genome they will be indexed in parallel using the indexer bowtie2-build (or hisat2-build). Once both C-\textgreater{}T and G-\textgreater{}A genome indices have been created you do not need to use the genome preparation script again (unless you want to align against a different genome\ldots{}).

\begin{quote}
Please note that Bowtie 2 and HISAT2 indexes are not compatible! To create a genome index for use with HISAT2 the option --hisat2 needs to be included as well.
\end{quote}

\hypertarget{bismark-alignment}{%
\section{Bismark alignment}\label{bismark-alignment}}

\begin{verbatim}
## make sure bowtie2 and samtools have been installed.
bismark ../ref/bismark_index/ 3bstrain1.fastq 
bismark ../ref/bismark_index/ EV1.fastq 
bismark ../ref/bismark_index/ set1rep1.fastq
\end{verbatim}

\hypertarget{python}{%
\chapter{Python}\label{python}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  why program
\item
  how to program
\item
  programming language
\end{enumerate}

\hypertarget{section-7}{%
\chapter{}\label{section-7}}

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see \url{http://rmarkdown.rstudio.com}.

When you click the \textbf{Knit} button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{summary}\NormalTok{(cars)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      speed           dist       
##  Min.   : 4.0   Min.   :  2.00  
##  1st Qu.:12.0   1st Qu.: 26.00  
##  Median :15.0   Median : 36.00  
##  Mean   :15.4   Mean   : 42.98  
##  3rd Qu.:19.0   3rd Qu.: 56.00  
##  Max.   :25.0   Max.   :120.00
\end{verbatim}

\hypertarget{including-plots}{%
\section{Including Plots}\label{including-plots}}

You can also embed plots, for example:

\includegraphics{bioinfBookXIE186_files/figure-latex/pressure-1.pdf}

Note that the \texttt{echo\ =\ FALSE} parameter was added to the code chunk to prevent printing of the R code that generated the plot.

\hypertarget{good-resouces-to-learn-bioinformatics}{%
\chapter{Good resouces to learn Bioinformatics}\label{good-resouces-to-learn-bioinformatics}}

\hypertarget{references}{%
\section{References:}\label{references}}

\url{https://www.bioinformatics.babraham.ac.uk/training.html}

\hypertarget{basic-statistics}{%
\chapter{Basic statistics}\label{basic-statistics}}

\hypertarget{probability-distribution}{%
\section{Probability distribution}\label{probability-distribution}}

\hypertarget{geometirc-distribution}{%
\subsection{Geometirc distribution}\label{geometirc-distribution}}

The geometric distribution is the distribution of the number of trials needed to get get the first success in repeated independent Bernolli trails.

\hypertarget{bionomial-distribution}{%
\subsection{Bionomial distribution}\label{bionomial-distribution}}

The binomial distribution is the distribution of the number of successes (X) in a fixed number (\emph{n}) if independent Bernolli trails.

\hypertarget{negative-binomial-distribution}{%
\subsection{Negative binomial distribution}\label{negative-binomial-distribution}}

The negative bionomial distribution distribution is the distribution of the number of trials needed (X) to get the \_r\_th success (r)(\citet{jbstatistics}).

\[
\left(\begin{array}{l}
{x-1} \\
{r-1}
\end{array}\right) p^{r-1}(1-p)^{(x-1)-(r-1)}
\]

The probabilty of the rth success that occures on the xth trial is:

\[
\begin{aligned}
P(X=x) &=p \times\left(\begin{array}{c}
{x-1} \\
{r-1}
\end{array}\right) p^{r-1}(1-p)^{(x-1)-(r-1)} \\
&=\left(\begin{array}{c}
{x-1} \\
{r-1}
\end{array}\right) p^{r}(1-p)^{x-r}
\end{aligned}
\]

We say Random varaible X is distributed \[X \sim N B(r, p)\]. The mean is: \[\mu=\frac{r}{p}\]. The variance is \[\sigma^{2}=\frac{r(1-p)}{p^{2}}\].

\emph{X} is the count of independent Bernoulli trials required to achieve the rth successful trial when the probability of success is constant \emph{p}.

R function \texttt{dnbinom} gives the density. \texttt{dnbbinom(x,\ size,\ prob)} calculate the probability that a number of failures \emph{x} occurs before \emph{r}-th success, in a sequence of Bernoulli trials, for which the probability of individual success is p.

Example 1 (\citet{NBblogMichael}):

An oil company has a \texttt{p\ =\ 0.20} chance of striking oil when drilling a well. What is the probability the company drills x = 7 wells to strike oil r = 3 times?

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{r =}\StringTok{ }\DecValTok{3}
\NormalTok{p =}\StringTok{ }\FloatTok{0.20}
\NormalTok{n =}\StringTok{ }\DecValTok{7} \OperatorTok{-}\StringTok{ }\NormalTok{r}
\CommentTok{# exact}
\KeywordTok{dnbinom}\NormalTok{(}\DataTypeTok{x =}\NormalTok{ n, }\DataTypeTok{size =}\NormalTok{ r, }\DataTypeTok{prob =}\NormalTok{ p)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.049152
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# simulated}
\KeywordTok{mean}\NormalTok{(}\KeywordTok{rnbinom}\NormalTok{(}\DataTypeTok{n =} \DecValTok{10000}\NormalTok{, }\DataTypeTok{size =}\NormalTok{ r, }\DataTypeTok{prob =}\NormalTok{ p) }\OperatorTok{==}\StringTok{ }\NormalTok{n)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 0.05
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{barplot}\NormalTok{(}\KeywordTok{dnbinom}\NormalTok{(}\DecValTok{1}\OperatorTok{:}\DecValTok{25}\NormalTok{,}\DecValTok{2}\NormalTok{,}\FloatTok{0.5}\NormalTok{), }\DataTypeTok{col=}\StringTok{"grey"}\NormalTok{, }\DataTypeTok{names.arg=}\DecValTok{1}\OperatorTok{:}\DecValTok{25}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-139-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressPackageStartupMessages}\NormalTok{(}\KeywordTok{library}\NormalTok{(dplyr))}
\KeywordTok{suppressPackageStartupMessages}\NormalTok{(}\KeywordTok{library}\NormalTok{(ggplot2))}

\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{x =} \DecValTok{0}\OperatorTok{:}\DecValTok{10}\NormalTok{, }\DataTypeTok{prob =} \KeywordTok{dnbinom}\NormalTok{(}\DataTypeTok{x =} \DecValTok{0}\OperatorTok{:}\DecValTok{10}\NormalTok{, }\DataTypeTok{size =}\NormalTok{ r, }\DataTypeTok{prob =}\NormalTok{ p)) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{mutate}\NormalTok{(}\DataTypeTok{Failures =} \KeywordTok{ifelse}\NormalTok{(x }\OperatorTok{==}\StringTok{ }\NormalTok{n, n, }\StringTok{"other"}\NormalTok{)) }\OperatorTok{%>%}
\KeywordTok{ggplot}\NormalTok{(}\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x =} \KeywordTok{factor}\NormalTok{(x), }\DataTypeTok{y =}\NormalTok{ prob, }\DataTypeTok{fill =}\NormalTok{ Failures)) }\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_col}\NormalTok{() }\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_text}\NormalTok{(}
    \KeywordTok{aes}\NormalTok{(}\DataTypeTok{label =} \KeywordTok{round}\NormalTok{(prob,}\DecValTok{2}\NormalTok{), }\DataTypeTok{y =}\NormalTok{ prob }\OperatorTok{+}\StringTok{ }\FloatTok{0.01}\NormalTok{),}
    \DataTypeTok{position =} \KeywordTok{position_dodge}\NormalTok{(}\FloatTok{0.9}\NormalTok{),}
    \DataTypeTok{size =} \DecValTok{3}\NormalTok{,}
    \DataTypeTok{vjust =} \DecValTok{0}
\NormalTok{  ) }\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \StringTok{"Probability of r = 3 Successes in X = 7 Trials"}\NormalTok{,}
       \DataTypeTok{subtitle =} \StringTok{"NB(3,.2)"}\NormalTok{,}
       \DataTypeTok{x =} \StringTok{"Failed Trials (X - r)"}\NormalTok{,}
       \DataTypeTok{y =} \StringTok{"Probability"}\NormalTok{) }
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-140-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#What is the expected number of trials to achieve r = 3 successes when the probability of success is p = 0.2?}
\NormalTok{r =}\StringTok{ }\DecValTok{3}
\NormalTok{p =}\StringTok{ }\FloatTok{0.20}
\CommentTok{# mean}
\CommentTok{# exact}
\NormalTok{r }\OperatorTok{/}\StringTok{ }\NormalTok{p}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 15
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# simulated}
\KeywordTok{mean}\NormalTok{(}\KeywordTok{rnbinom}\NormalTok{(}\DataTypeTok{n =} \DecValTok{10000}\NormalTok{, }\DataTypeTok{size =} \DecValTok{3}\NormalTok{, }\DataTypeTok{prob =}\NormalTok{ p)) }\OperatorTok{+}\StringTok{ }\NormalTok{r}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 14.9629
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Variance}
\CommentTok{# exact}
\NormalTok{r }\OperatorTok{*}\StringTok{ }\NormalTok{(}\DecValTok{1} \OperatorTok{-}\StringTok{ }\NormalTok{p) }\OperatorTok{/}\StringTok{ }\NormalTok{p}\OperatorTok{^}\DecValTok{2}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 60
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# simulated}
\KeywordTok{var}\NormalTok{(}\KeywordTok{rnbinom}\NormalTok{(}\DataTypeTok{n =} \DecValTok{100000}\NormalTok{, }\DataTypeTok{size =}\NormalTok{ r, }\DataTypeTok{prob =}\NormalTok{ p))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 60.21722
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressPackageStartupMessages}\NormalTok{(}\KeywordTok{library}\NormalTok{(dplyr))}
\KeywordTok{suppressPackageStartupMessages}\NormalTok{(}\KeywordTok{library}\NormalTok{(ggplot2))}

\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{x =} \DecValTok{1}\OperatorTok{:}\DecValTok{20}\NormalTok{, }
           \DataTypeTok{pmf =} \KeywordTok{dnbinom}\NormalTok{(}\DataTypeTok{x =} \DecValTok{1}\OperatorTok{:}\DecValTok{20}\NormalTok{, }\DataTypeTok{size =}\NormalTok{ r, }\DataTypeTok{prob =}\NormalTok{ p),}
           \DataTypeTok{cdf =} \KeywordTok{pnbinom}\NormalTok{(}\DataTypeTok{q =} \DecValTok{1}\OperatorTok{:}\DecValTok{20}\NormalTok{, }\DataTypeTok{size =}\NormalTok{ r, }\DataTypeTok{prob =}\NormalTok{ p, }\DataTypeTok{lower.tail =} \OtherTok{TRUE}\NormalTok{)) }\OperatorTok{%>%}
\KeywordTok{ggplot}\NormalTok{(}\KeywordTok{aes}\NormalTok{(}\DataTypeTok{x =} \KeywordTok{factor}\NormalTok{(x), }\DataTypeTok{y =}\NormalTok{ cdf)) }\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_col}\NormalTok{() }\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_text}\NormalTok{(}
    \KeywordTok{aes}\NormalTok{(}\DataTypeTok{label =} \KeywordTok{round}\NormalTok{(cdf,}\DecValTok{2}\NormalTok{), }\DataTypeTok{y =}\NormalTok{ cdf }\OperatorTok{+}\StringTok{ }\FloatTok{0.01}\NormalTok{),}
    \DataTypeTok{position =} \KeywordTok{position_dodge}\NormalTok{(}\FloatTok{0.9}\NormalTok{),}
    \DataTypeTok{size =} \DecValTok{3}\NormalTok{,}
    \DataTypeTok{vjust =} \DecValTok{0}
\NormalTok{  ) }\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \StringTok{"Cumulative Probability of X = x failed trials to achieve 3rd success"}\NormalTok{,}
       \DataTypeTok{subtitle =} \StringTok{"NB(3,.2)"}\NormalTok{,}
       \DataTypeTok{x =} \StringTok{"Failed Trials (x)"}\NormalTok{,}
       \DataTypeTok{y =} \StringTok{"probability"}\NormalTok{) }
\end{Highlighting}
\end{Shaded}

\includegraphics{bioinfBookXIE186_files/figure-latex/unnamed-chunk-145-1.pdf}

\hypertarget{why-is-it-called-negative-binomial}{%
\subsubsection{Why is it called Negative Binomial?}\label{why-is-it-called-negative-binomial}}

The ``negative'' part of negative binomial stems from the fact that one facet of the binomial distribution is reversed: in a binomial experiment, we count the number of successes in a fixed number of trials. In a negative binomial experiment, we're counting the failures, or how many cards it takes you to pick two aces.

\hypertarget{part-python}{%
\part{Python}\label{part-python}}

\hypertarget{first-perl-program-1}{%
\chapter{First Perl Program}\label{first-perl-program-1}}

As all other programming books, we begin with a ``Hello world'' program.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{#!/usr/bin/python}
\CommentTok{#Printing a line of text “Hello, Bioinformatics”}
\FunctionTok{print}\NormalTok{(}\KeywordTok{"}\StringTok{Hello, Bioinformatics!}\CharTok{\textbackslash{}n}\KeywordTok{"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

This program show how to display a line a text in Perl. It have several features. We go through each line in detail.

Line 1 is what we call shebang line. This line starts with shebang construct (\texttt{\#!}). \texttt{/usr/bin/perl} indicates the path of the Perl interpreter.

Line 3 shows how to print a line of text in Perl. Nearly all programming language use print to display texts on the screen. Here, print is a built-in function in Perl. It print the string of characters (its arguments) between quotation marks (``'' or `').

\bibliography{book.bib,packages.bib}


\end{document}
