<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 13 Input and output in Perl | All things considered for Bioinformatics</title>
  <meta name="description" content="This is a book for people who wants to learn Bioinformatics from scratch">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 13 Input and output in Perl | All things considered for Bioinformatics" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="images/cover.jpg" />
  <meta property="og:description" content="This is a book for people who wants to learn Bioinformatics from scratch" />
  <meta name="github-repo" content="xie186/Novice2Expert4Bioinformatics" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Input and output in Perl | All things considered for Bioinformatics" />
  
  <meta name="twitter:description" content="This is a book for people who wants to learn Bioinformatics from scratch" />
  <meta name="twitter:image" content="images/cover.jpg" />

<meta name="author" content="Shaojun Xie (解少俊)">


<meta name="date" content="2019-04-15">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="string-manipulation.html">
<link rel="next" href="practical-perl-program.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">From Novice to Expert in Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Linux</b></span></li>
<li class="chapter" data-level="1" data-path="why-linux.html"><a href="why-linux.html"><i class="fa fa-check"></i><b>1</b> Why Linux?</a><ul>
<li class="chapter" data-level="1.1" data-path="why-linux.html"><a href="why-linux.html#what-is-linux"><i class="fa fa-check"></i><b>1.1</b> What is Linux</a></li>
<li class="chapter" data-level="1.2" data-path="why-linux.html"><a href="why-linux.html#linux-for-bioinformatics"><i class="fa fa-check"></i><b>1.2</b> Linux for bioinformatics</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html"><i class="fa fa-check"></i><b>2</b> Connecting to Linux</a><ul>
<li class="chapter" data-level="2.1" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#user-interfaces"><i class="fa fa-check"></i><b>2.1</b> User interfaces</a></li>
<li class="chapter" data-level="2.2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#how-to-connect"><i class="fa fa-check"></i><b>2.2</b> How to connect</a><ul>
<li class="chapter" data-level="2.2.1" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#digitalocean-droplet"><i class="fa fa-check"></i><b>2.2.1</b> DigitalOcean droplet</a></li>
<li class="chapter" data-level="2.2.2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#aws-free-tier-offering"><i class="fa fa-check"></i><b>2.2.2</b> AWS Free Tier offering</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html"><i class="fa fa-check"></i><b>3</b> File system and commands of Linux</a><ul>
<li class="chapter" data-level="3.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#path"><i class="fa fa-check"></i><b>3.1</b> Path</a><ul>
<li class="chapter" data-level="3.1.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#relative-and-absolute-path"><i class="fa fa-check"></i><b>3.1.1</b> Relative and absolute path</a></li>
<li class="chapter" data-level="3.1.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#comand-ls"><i class="fa fa-check"></i><b>3.1.2</b> Comand <code>ls</code></a></li>
<li class="chapter" data-level="3.1.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cd"><i class="fa fa-check"></i><b>3.1.3</b> Command <code>cd</code></a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#manipulations-of-files-and-directories"><i class="fa fa-check"></i><b>3.2</b> Manipulations of files and directories</a><ul>
<li class="chapter" data-level="3.2.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cp"><i class="fa fa-check"></i><b>3.2.1</b> Command <code>cp</code></a></li>
<li class="chapter" data-level="3.2.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-mv"><i class="fa fa-check"></i><b>3.2.2</b> Command <code>mv</code></a></li>
<li class="chapter" data-level="3.2.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-mkdir"><i class="fa fa-check"></i><b>3.2.3</b> Command <code>mkdir</code></a></li>
<li class="chapter" data-level="3.2.4" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-rm"><i class="fa fa-check"></i><b>3.2.4</b> Command ‘rm’</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#viewing-text-files-in-linux"><i class="fa fa-check"></i><b>3.3</b> Viewing text files in Linux</a><ul>
<li class="chapter" data-level="3.3.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cat"><i class="fa fa-check"></i><b>3.3.1</b> Command <code>cat</code></a></li>
<li class="chapter" data-level="3.3.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-more-and-less"><i class="fa fa-check"></i><b>3.3.2</b> Command <code>more</code> and <code>less</code></a></li>
<li class="chapter" data-level="3.3.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-head-and-tail"><i class="fa fa-check"></i><b>3.3.3</b> Command <code>head</code> and <code>tail</code></a></li>
<li class="chapter" data-level="3.3.4" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#auto-completion"><i class="fa fa-check"></i><b>3.3.4</b> Auto-completion</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#understand-standard-input-and-stardard-output"><i class="fa fa-check"></i><b>3.4</b> Understand standard input and stardard output</a><ul>
<li class="chapter" data-level="3.4.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stdin"><i class="fa fa-check"></i><b>3.4.1</b> STDIN</a></li>
<li class="chapter" data-level="3.4.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stdout"><i class="fa fa-check"></i><b>3.4.2</b> STDOUT</a></li>
<li class="chapter" data-level="3.4.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stderr"><i class="fa fa-check"></i><b>3.4.3</b> STDERR</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#find-disk-usage-of-files-and-directories"><i class="fa fa-check"></i><b>3.5</b> Find Disk Usage of Files and Directories</a></li>
<li class="chapter" data-level="3.6" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#advanced-topic"><i class="fa fa-check"></i><b>3.6</b> Advanced topic</a><ul>
<li class="chapter" data-level="3.6.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#linux-md5sum-command"><i class="fa fa-check"></i><b>3.6.1</b> Linux <code>md5sum</code> Command</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#filtering-output"><i class="fa fa-check"></i><b>3.7</b> Filtering Output</a><ul>
<li class="chapter" data-level="3.7.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#cut"><i class="fa fa-check"></i><b>3.7.1</b> <code>cut</code></a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#check-you-job-status"><i class="fa fa-check"></i><b>3.8</b> Check you job status</a></li>
<li class="chapter" data-level="3.9" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#finding-things"><i class="fa fa-check"></i><b>3.9</b> Finding Things</a><ul>
<li class="chapter" data-level="3.9.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#list-files-bigger-than-filesize-specified"><i class="fa fa-check"></i><b>3.9.1</b> List files bigger than filesize specified</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html"><i class="fa fa-check"></i><b>4</b> Achiving and compressing files</a><ul>
<li class="chapter" data-level="4.1" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#common-compressed-file-format"><i class="fa fa-check"></i><b>4.1</b> Common compressed file format</a></li>
<li class="chapter" data-level="4.2" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#how-to-work-with-different-format"><i class="fa fa-check"></i><b>4.2</b> How to work with different format</a><ul>
<li class="chapter" data-level="4.2.1" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#gz"><i class="fa fa-check"></i><b>4.2.1</b> <code>*.gz</code></a></li>
<li class="chapter" data-level="4.2.2" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#tar.gz"><i class="fa fa-check"></i><b>4.2.2</b> <code>*.tar.gz</code></a></li>
<li class="chapter" data-level="4.2.3" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#zip"><i class="fa fa-check"></i><b>4.2.3</b> <code>*.zip</code></a></li>
<li class="chapter" data-level="4.2.4" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#tar.bz2"><i class="fa fa-check"></i><b>4.2.4</b> <code>*.tar.bz2</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html"><i class="fa fa-check"></i><b>5</b> Process management in Linux</a><ul>
<li class="chapter" data-level="5.1" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#top"><i class="fa fa-check"></i><b>5.1</b> <code>top</code></a></li>
<li class="chapter" data-level="5.2" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#ps"><i class="fa fa-check"></i><b>5.2</b> <code>ps</code></a></li>
<li class="chapter" data-level="5.3" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#kill"><i class="fa fa-check"></i><b>5.3</b> <code>kill</code></a></li>
<li class="chapter" data-level="5.4" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#df"><i class="fa fa-check"></i><b>5.4</b> <code>df</code></a></li>
<li class="chapter" data-level="5.5" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#advanced-topic-free"><i class="fa fa-check"></i><b>5.5</b> Advanced topic <code>free</code></a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="file-transfer.html"><a href="file-transfer.html"><i class="fa fa-check"></i><b>6</b> File transfer</a><ul>
<li class="chapter" data-level="6.1" data-path="file-transfer.html"><a href="file-transfer.html#transferring-files-between-local-computer-and-linux-server"><i class="fa fa-check"></i><b>6.1</b> Transferring files between local computer and Linux server</a><ul>
<li class="chapter" data-level="6.1.1" data-path="file-transfer.html"><a href="file-transfer.html#use-command-line-tools"><i class="fa fa-check"></i><b>6.1.1</b> Use command line tools</a></li>
<li class="chapter" data-level="6.1.2" data-path="file-transfer.html"><a href="file-transfer.html#download-files"><i class="fa fa-check"></i><b>6.1.2</b> Download files</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html"><i class="fa fa-check"></i><b>7</b> Install Bioinformatics software in Linux</a><ul>
<li class="chapter" data-level="7.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#installation-from-source-code"><i class="fa fa-check"></i><b>7.1</b> Installation from source code</a><ul>
<li class="chapter" data-level="7.1.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-bwa"><i class="fa fa-check"></i><b>7.1.1</b> Install <code>bwa</code></a></li>
<li class="chapter" data-level="7.1.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-samtools"><i class="fa fa-check"></i><b>7.1.2</b> Install <code>samtools</code></a></li>
<li class="chapter" data-level="7.1.3" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#align-reads-to-genome-using-bwa-and-store-the-alignment-results-in-sambam-files"><i class="fa fa-check"></i><b>7.1.3</b> Align reads to genome using <code>bwa</code> and store the alignment results in SAM/BAM files</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#installing-a-precompiled-binary-executable"><i class="fa fa-check"></i><b>7.2</b> Installing a precompiled binary (executable)</a><ul>
<li class="chapter" data-level="7.2.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-bwa-1"><i class="fa fa-check"></i><b>7.2.1</b> Install <code>bwa</code></a></li>
<li class="chapter" data-level="7.2.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-with-conda-recommended-way"><i class="fa fa-check"></i><b>7.2.2</b> Install with <code>conda</code> (recommended way)</a></li>
<li class="chapter" data-level="7.2.3" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-ussing-docker-advance-topic"><i class="fa fa-check"></i><b>7.2.3</b> Install ussing <code>Docker</code> (advance topic)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html"><i class="fa fa-check"></i><b>8</b> Text editor in Linux</a><ul>
<li class="chapter" data-level="8.1" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#basic-vi-skills"><i class="fa fa-check"></i><b>8.1</b> Basic vi skills</a></li>
<li class="chapter" data-level="8.2" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#create-new-text-file-with-vi"><i class="fa fa-check"></i><b>8.2</b> Create new text file with <code>vi</code></a></li>
<li class="chapter" data-level="8.3" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#an-example-for-using-editor-r"><i class="fa fa-check"></i><b>8.3</b> An example for using editor R</a></li>
</ul></li>
<li class="part"><span><b>II Perl</b></span></li>
<li class="chapter" data-level="9" data-path="first-perl-program.html"><a href="first-perl-program.html"><i class="fa fa-check"></i><b>9</b> First Perl Program</a><ul>
<li class="chapter" data-level="9.1" data-path="first-perl-program.html"><a href="first-perl-program.html#first-program"><i class="fa fa-check"></i><b>9.1</b> First Program</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="varible-in-perl.html"><a href="varible-in-perl.html"><i class="fa fa-check"></i><b>10</b> Varible in Perl</a><ul>
<li class="chapter" data-level="10.1" data-path="varible-in-perl.html"><a href="varible-in-perl.html#scalar-variable"><i class="fa fa-check"></i><b>10.1</b> Scalar variable</a></li>
<li class="chapter" data-level="10.2" data-path="varible-in-perl.html"><a href="varible-in-perl.html#arithmetic-operations-in-perl"><i class="fa fa-check"></i><b>10.2</b> Arithmetic operations in Perl</a><ul>
<li class="chapter" data-level="10.2.1" data-path="varible-in-perl.html"><a href="varible-in-perl.html#shorthand-operations"><i class="fa fa-check"></i><b>10.2.1</b> Shorthand operations</a></li>
<li class="chapter" data-level="10.2.2" data-path="varible-in-perl.html"><a href="varible-in-perl.html#auto-increment-and-auto-decrement"><i class="fa fa-check"></i><b>10.2.2</b> Auto increment and auto decrement</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="varible-in-perl.html"><a href="varible-in-perl.html#use-strict-user-warnings-and-my"><i class="fa fa-check"></i><b>10.3</b> <code>use strict</code>; <code>user warnings</code> and <code>my</code></a></li>
<li class="chapter" data-level="10.4" data-path="varible-in-perl.html"><a href="varible-in-perl.html#array"><i class="fa fa-check"></i><b>10.4</b> Array</a><ul>
<li class="chapter" data-level="10.4.1" data-path="varible-in-perl.html"><a href="varible-in-perl.html#init-an-array"><i class="fa fa-check"></i><b>10.4.1</b> Init an array</a></li>
<li class="chapter" data-level="10.4.2" data-path="varible-in-perl.html"><a href="varible-in-perl.html#array-index"><i class="fa fa-check"></i><b>10.4.2</b> Array index</a></li>
<li class="chapter" data-level="10.4.3" data-path="varible-in-perl.html"><a href="varible-in-perl.html#length-of-the-array"><i class="fa fa-check"></i><b>10.4.3</b> Length of the array</a></li>
<li class="chapter" data-level="10.4.4" data-path="varible-in-perl.html"><a href="varible-in-perl.html#sort-arrays-in-perl"><i class="fa fa-check"></i><b>10.4.4</b> Sort Arrays in Perl</a></li>
<li class="chapter" data-level="10.4.5" data-path="varible-in-perl.html"><a href="varible-in-perl.html#use-push-pop-shift-and-unshift-in-perl"><i class="fa fa-check"></i><b>10.4.5</b> Use <code>push</code>, <code>pop</code>, <code>shift</code> and <code>unshift</code> in Perl</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="varible-in-perl.html"><a href="varible-in-perl.html#hash-in-perl"><i class="fa fa-check"></i><b>10.5</b> Hash in Perl</a><ul>
<li class="chapter" data-level="10.5.1" data-path="varible-in-perl.html"><a href="varible-in-perl.html#sort-keys-numerically-using-sort-ab"><i class="fa fa-check"></i><b>10.5.1</b> Sort keys numerically using <code>sort {$a&lt;=&gt;$b}</code></a></li>
<li class="chapter" data-level="10.5.2" data-path="varible-in-perl.html"><a href="varible-in-perl.html#use-exists-function-on-a-hash"><i class="fa fa-check"></i><b>10.5.2</b> Use <code>exists</code> function on a hash</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="control-structure.html"><a href="control-structure.html"><i class="fa fa-check"></i><b>11</b> Control structure</a><ul>
<li class="chapter" data-level="11.1" data-path="control-structure.html"><a href="control-structure.html#for-loop"><i class="fa fa-check"></i><b>11.1</b> <code>for</code> loop</a></li>
<li class="chapter" data-level="11.2" data-path="control-structure.html"><a href="control-structure.html#foreach-loop"><i class="fa fa-check"></i><b>11.2</b> <code>foreach</code> loop</a></li>
<li class="chapter" data-level="11.3" data-path="control-structure.html"><a href="control-structure.html#while-loop"><i class="fa fa-check"></i><b>11.3</b> <code>while</code> loop</a></li>
<li class="chapter" data-level="11.4" data-path="control-structure.html"><a href="control-structure.html#statement-if-else"><i class="fa fa-check"></i><b>11.4</b> Statement if-else</a></li>
<li class="chapter" data-level="11.5" data-path="control-structure.html"><a href="control-structure.html#operator-last-and-next"><i class="fa fa-check"></i><b>11.5</b> Operator <code>last</code> and <code>next</code></a></li>
<li class="chapter" data-level="11.6" data-path="control-structure.html"><a href="control-structure.html#operator-redo"><i class="fa fa-check"></i><b>11.6</b> Operator <code>redo</code></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="string-manipulation.html"><a href="string-manipulation.html"><i class="fa fa-check"></i><b>12</b> String manipulation</a><ul>
<li class="chapter" data-level="12.1" data-path="string-manipulation.html"><a href="string-manipulation.html#string-concatenation"><i class="fa fa-check"></i><b>12.1</b> String concatenation</a></li>
<li class="chapter" data-level="12.2" data-path="string-manipulation.html"><a href="string-manipulation.html#substring-extraction"><i class="fa fa-check"></i><b>12.2</b> Substring extraction</a></li>
<li class="chapter" data-level="12.3" data-path="string-manipulation.html"><a href="string-manipulation.html#substring-search"><i class="fa fa-check"></i><b>12.3</b> Substring search</a></li>
<li class="chapter" data-level="12.4" data-path="string-manipulation.html"><a href="string-manipulation.html#split-string"><i class="fa fa-check"></i><b>12.4</b> Split String</a></li>
<li class="chapter" data-level="12.5" data-path="string-manipulation.html"><a href="string-manipulation.html#regular-expression"><i class="fa fa-check"></i><b>12.5</b> Regular expression</a><ul>
<li class="chapter" data-level="12.5.1" data-path="string-manipulation.html"><a href="string-manipulation.html#pattern-matching"><i class="fa fa-check"></i><b>12.5.1</b> Pattern matching</a></li>
<li class="chapter" data-level="12.5.2" data-path="string-manipulation.html"><a href="string-manipulation.html#pattern-substitution"><i class="fa fa-check"></i><b>12.5.2</b> Pattern substitution</a></li>
<li class="chapter" data-level="12.5.3" data-path="string-manipulation.html"><a href="string-manipulation.html#modifiers-to-pattern-matching-and-substitution"><i class="fa fa-check"></i><b>12.5.3</b> Modifiers to pattern matching and substitution</a></li>
<li class="chapter" data-level="12.5.4" data-path="string-manipulation.html"><a href="string-manipulation.html#greedy-or-non-greedy"><i class="fa fa-check"></i><b>12.5.4</b> Greedy or non-greedy</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html"><i class="fa fa-check"></i><b>13</b> Input and output in Perl</a><ul>
<li class="chapter" data-level="13.1" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#input"><i class="fa fa-check"></i><b>13.1</b> Input</a><ul>
<li class="chapter" data-level="13.1.1" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#standard-input"><i class="fa fa-check"></i><b>13.1.1</b> Standard input</a></li>
<li class="chapter" data-level="13.1.2" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#input-from-a-file-on-the-disk"><i class="fa fa-check"></i><b>13.1.2</b> Input from a file on the disk</a></li>
<li class="chapter" data-level="13.1.3" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#special-variable-_"><i class="fa fa-check"></i><b>13.1.3</b> Special variable <code>$_</code></a></li>
<li class="chapter" data-level="13.1.4" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#the-input-record-separator"><i class="fa fa-check"></i><b>13.1.4</b> The input record separator: <code>$/</code></a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#output-to-a-file-on-the-disk"><i class="fa fa-check"></i><b>13.2</b> Output to a file on the disk</a></li>
<li class="chapter" data-level="13.3" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#arguments-from-command-line"><i class="fa fa-check"></i><b>13.3</b> Arguments from command line</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="practical-perl-program.html"><a href="practical-perl-program.html"><i class="fa fa-check"></i><b>14</b> Practical Perl program</a><ul>
<li class="chapter" data-level="14.1" data-path="practical-perl-program.html"><a href="practical-perl-program.html#add-annotation-information-to-deseq2-results"><i class="fa fa-check"></i><b>14.1</b> Add annotation information to DESeq2 results</a></li>
<li class="chapter" data-level="14.2" data-path="practical-perl-program.html"><a href="practical-perl-program.html#merge-overlap-genomic-regions"><i class="fa fa-check"></i><b>14.2</b> Merge overlap genomic regions</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="perl-modules.html"><a href="perl-modules.html"><i class="fa fa-check"></i><b>15</b> Perl modules</a><ul>
<li class="chapter" data-level="15.1" data-path="perl-modules.html"><a href="perl-modules.html#what-is-a-perl-module"><i class="fa fa-check"></i><b>15.1</b> What is a Perl module</a></li>
<li class="chapter" data-level="15.2" data-path="perl-modules.html"><a href="perl-modules.html#how-to-install-a-perl-module"><i class="fa fa-check"></i><b>15.2</b> How to install a Perl module</a></li>
<li class="chapter" data-level="15.3" data-path="perl-modules.html"><a href="perl-modules.html#how-to-use-a-perl-module"><i class="fa fa-check"></i><b>15.3</b> How to use a Perl module</a></li>
<li class="chapter" data-level="15.4" data-path="perl-modules.html"><a href="perl-modules.html#how-to-use-bioperl-module"><i class="fa fa-check"></i><b>15.4</b> How to use BioPerl module</a></li>
<li class="chapter" data-level="15.5" data-path="perl-modules.html"><a href="perl-modules.html#how-to-write-a-module"><i class="fa fa-check"></i><b>15.5</b> How to write a module</a></li>
</ul></li>
<li class="part"><span><b>III Statistics and R</b></span></li>
<li class="chapter" data-level="16" data-path="r-introduction.html"><a href="r-introduction.html"><i class="fa fa-check"></i><b>16</b> R introduction</a><ul>
<li class="chapter" data-level="16.1" data-path="r-introduction.html"><a href="r-introduction.html#basic-r-function"><i class="fa fa-check"></i><b>16.1</b> Basic R function</a></li>
<li class="chapter" data-level="16.2" data-path="r-introduction.html"><a href="r-introduction.html#producing-simple-graphs-with-r"><i class="fa fa-check"></i><b>16.2</b> Producing Simple Graphs with R</a><ul>
<li class="chapter" data-level="16.2.1" data-path="r-introduction.html"><a href="r-introduction.html#line-charts"><i class="fa fa-check"></i><b>16.2.1</b> Line Charts</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html"><i class="fa fa-check"></i><b>17</b> Producing simple graphs</a><ul>
<li class="chapter" data-level="17.1" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#producing-simple-graphs-using-r"><i class="fa fa-check"></i><b>17.1</b> Producing simple graphs using R</a><ul>
<li class="chapter" data-level="17.1.1" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#line-charts-1"><i class="fa fa-check"></i><b>17.1.1</b> Line Charts</a></li>
<li class="chapter" data-level="17.1.2" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#bar-charts"><i class="fa fa-check"></i><b>17.1.2</b> Bar Charts</a></li>
<li class="chapter" data-level="17.1.3" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#histograms"><i class="fa fa-check"></i><b>17.1.3</b> Histograms</a></li>
<li class="chapter" data-level="17.1.4" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#pie-charts"><i class="fa fa-check"></i><b>17.1.4</b> Pie Charts</a></li>
<li class="chapter" data-level="17.1.5" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#dotcharts"><i class="fa fa-check"></i><b>17.1.5</b> Dotcharts</a></li>
<li class="chapter" data-level="17.1.6" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#misc"><i class="fa fa-check"></i><b>17.1.6</b> Misc</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="ggplot2.html"><a href="ggplot2.html"><i class="fa fa-check"></i><b>18</b> <code>ggplot2</code></a><ul>
<li class="chapter" data-level="18.1" data-path="ggplot2.html"><a href="ggplot2.html#ggplot2-1"><i class="fa fa-check"></i><b>18.1</b> ggplot2</a></li>
<li class="chapter" data-level="18.2" data-path="ggplot2.html"><a href="ggplot2.html#ggplot2-practical"><i class="fa fa-check"></i><b>18.2</b> ggplot2 practical</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html"><i class="fa fa-check"></i><b>19</b> Heatmap Tutorial</a><ul>
<li class="chapter" data-level="19.1" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#install-pheatmap-package"><i class="fa fa-check"></i><b>19.1</b> Install <code>pheatmap</code> package</a></li>
<li class="chapter" data-level="19.2" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#draw-a-heatmap-for-gene-expression-of-rna-seq-data"><i class="fa fa-check"></i><b>19.2</b> Draw a heatmap for gene expression of RNA-seq data</a></li>
<li class="chapter" data-level="19.3" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#add-the-annotation"><i class="fa fa-check"></i><b>19.3</b> Add the annotation</a></li>
<li class="chapter" data-level="19.4" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#section"><i class="fa fa-check"></i><b>19.4</b> </a></li>
<li class="chapter" data-level="19.5" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#transfrom-the-data"><i class="fa fa-check"></i><b>19.5</b> Transfrom the data</a></li>
<li class="chapter" data-level="19.6" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-add-annotations"><i class="fa fa-check"></i><b>19.6</b> How to add annotations</a></li>
<li class="chapter" data-level="19.7" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-cut-the-trees"><i class="fa fa-check"></i><b>19.7</b> How to cut the trees</a></li>
<li class="chapter" data-level="19.8" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-get-the-cluster-information-from-the-heatmap"><i class="fa fa-check"></i><b>19.8</b> How to get the cluster information from the heatmap</a></li>
<li class="chapter" data-level="19.9" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#section-1"><i class="fa fa-check"></i><b>19.9</b> </a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="preparation-of-figures-for-manuscript.html"><a href="preparation-of-figures-for-manuscript.html"><i class="fa fa-check"></i><b>20</b> Preparation of figures for manuscript</a><ul>
<li class="chapter" data-level="20.1" data-path="preparation-of-figures-for-manuscript.html"><a href="preparation-of-figures-for-manuscript.html#test1"><i class="fa fa-check"></i><b>20.1</b> test1</a></li>
</ul></li>
<li class="part"><span><b>IV Omics data</b></span></li>
<li class="chapter" data-level="21" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html"><i class="fa fa-check"></i><b>21</b> Introduction to NGS</a><ul>
<li class="chapter" data-level="21.1" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#introduction-of-biology-for-bioinformatics"><i class="fa fa-check"></i><b>21.1</b> Introduction of Biology for Bioinformatics</a><ul>
<li class="chapter" data-level="21.1.1" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#what-is-dna"><i class="fa fa-check"></i><b>21.1.1</b> What is DNA</a></li>
<li class="chapter" data-level="21.1.2" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#what-is-genome"><i class="fa fa-check"></i><b>21.1.2</b> What is genome</a></li>
<li class="chapter" data-level="21.1.3" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#how-to-assemble-a-genome"><i class="fa fa-check"></i><b>21.1.3</b> How to assemble a genome</a></li>
<li class="chapter" data-level="21.1.4" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#what-is-gene"><i class="fa fa-check"></i><b>21.1.4</b> What is gene</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#what-is-ngs"><i class="fa fa-check"></i><b>21.2</b> What is NGS</a></li>
<li class="chapter" data-level="21.3" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#application-of-ngs"><i class="fa fa-check"></i><b>21.3</b> Application of NGS</a></li>
<li class="chapter" data-level="21.4" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#section-2"><i class="fa fa-check"></i><b>21.4</b> </a></li>
<li class="chapter" data-level="21.5" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#data-file-formats-to-store-the-genomic-information"><i class="fa fa-check"></i><b>21.5</b> Data file formats to store the genomic information</a><ul>
<li class="chapter" data-level="21.5.1" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#fasta-file"><i class="fa fa-check"></i><b>21.5.1</b> <code>fasta</code> file</a></li>
<li class="chapter" data-level="21.5.2" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#gff-file"><i class="fa fa-check"></i><b>21.5.2</b> GFF file</a></li>
<li class="chapter" data-level="21.5.3" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#gtf-file"><i class="fa fa-check"></i><b>21.5.3</b> GTF file</a></li>
<li class="chapter" data-level="21.5.4" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#fastq-file"><i class="fa fa-check"></i><b>21.5.4</b> FASTQ file</a></li>
<li class="chapter" data-level="21.5.5" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#sambam-file"><i class="fa fa-check"></i><b>21.5.5</b> SAM/BAM file</a></li>
<li class="chapter" data-level="21.5.6" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#vcf-file"><i class="fa fa-check"></i><b>21.5.6</b> VCF file</a></li>
<li class="chapter" data-level="21.5.7" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#bed-format"><i class="fa fa-check"></i><b>21.5.7</b> BED format</a></li>
<li class="chapter" data-level="21.5.8" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#bedgraph-format"><i class="fa fa-check"></i><b>21.5.8</b> bedGraph format</a></li>
</ul></li>
<li class="chapter" data-level="21.6" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#usefull-links"><i class="fa fa-check"></i><b>21.6</b> Usefull links:</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="bs-seq.html"><a href="bs-seq.html"><i class="fa fa-check"></i><b>22</b> BS-seq</a><ul>
<li class="chapter" data-level="22.1" data-path="bs-seq.html"><a href="bs-seq.html#section-3"><i class="fa fa-check"></i><b>22.1</b> </a></li>
<li class="chapter" data-level="22.2" data-path="bs-seq.html"><a href="bs-seq.html#what-is-bs-seq"><i class="fa fa-check"></i><b>22.2</b> What is BS-seq</a></li>
<li class="chapter" data-level="22.3" data-path="bs-seq.html"><a href="bs-seq.html#section-4"><i class="fa fa-check"></i><b>22.3</b> </a></li>
<li class="chapter" data-level="22.4" data-path="bs-seq.html"><a href="bs-seq.html#how-to-analyze-bs-seq-data"><i class="fa fa-check"></i><b>22.4</b> How to analyze BS-seq data</a><ul>
<li class="chapter" data-level="22.4.1" data-path="bs-seq.html"><a href="bs-seq.html#section-5"><i class="fa fa-check"></i><b>22.4.1</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="start-a-project.html"><a href="start-a-project.html"><i class="fa fa-check"></i><b>23</b> Start a project</a><ul>
<li class="chapter" data-level="23.1" data-path="start-a-project.html"><a href="start-a-project.html#experimetal-design"><i class="fa fa-check"></i><b>23.1</b> Experimetal design</a><ul>
<li class="chapter" data-level="23.1.1" data-path="start-a-project.html"><a href="start-a-project.html#how-many-biological-replicates"><i class="fa fa-check"></i><b>23.1.1</b> How many biological replicates</a></li>
<li class="chapter" data-level="23.1.2" data-path="start-a-project.html"><a href="start-a-project.html#how-big"><i class="fa fa-check"></i><b>23.1.2</b> How big</a></li>
<li class="chapter" data-level="23.1.3" data-path="start-a-project.html"><a href="start-a-project.html#how-much-data-to-generate"><i class="fa fa-check"></i><b>23.1.3</b> How much data to generate</a></li>
<li class="chapter" data-level="23.1.4" data-path="start-a-project.html"><a href="start-a-project.html#section-6"><i class="fa fa-check"></i><b>23.1.4</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="24" data-path="capstone-project.html"><a href="capstone-project.html"><i class="fa fa-check"></i><b>24</b> Capstone project:</a><ul>
<li class="chapter" data-level="24.1" data-path="capstone-project.html"><a href="capstone-project.html#introduction"><i class="fa fa-check"></i><b>24.1</b> Introduction</a></li>
<li class="chapter" data-level="24.2" data-path="capstone-project.html"><a href="capstone-project.html#method"><i class="fa fa-check"></i><b>24.2</b> Method</a></li>
<li class="chapter" data-level="24.3" data-path="capstone-project.html"><a href="capstone-project.html#pipelines"><i class="fa fa-check"></i><b>24.3</b> Pipelines</a></li>
<li class="chapter" data-level="24.4" data-path="capstone-project.html"><a href="capstone-project.html#section-7"><i class="fa fa-check"></i><b>24.4</b> </a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html"><i class="fa fa-check"></i><b>25</b> Practical data analysis on WGBS</a><ul>
<li class="chapter" data-level="25.1" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html#donwload-the-data"><i class="fa fa-check"></i><b>25.1</b> Donwload the data</a><ul>
<li class="chapter" data-level="25.1.1" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html#reference"><i class="fa fa-check"></i><b>25.1.1</b> Reference</a></li>
<li class="chapter" data-level="25.1.2" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html#now-lets-download-the-wgbs-data"><i class="fa fa-check"></i><b>25.1.2</b> Now let’s download the WGBS data</a></li>
</ul></li>
<li class="chapter" data-level="25.2" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html#convert-sra-files-to-fastq-files"><i class="fa fa-check"></i><b>25.2</b> Convert <code>sra</code> files to <code>fastq</code> files</a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="section-8.html"><a href="section-8.html"><i class="fa fa-check"></i><b>26</b> </a><ul>
<li class="chapter" data-level="26.1" data-path="section-8.html"><a href="section-8.html#including-plots"><i class="fa fa-check"></i><b>26.1</b> Including Plots</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="good-resouces-to-learn-bioinformatics.html"><a href="good-resouces-to-learn-bioinformatics.html"><i class="fa fa-check"></i><b>27</b> Good resouces to learn Bioinformatics</a><ul>
<li class="chapter" data-level="27.1" data-path="good-resouces-to-learn-bioinformatics.html"><a href="good-resouces-to-learn-bioinformatics.html#references"><i class="fa fa-check"></i><b>27.1</b> References:</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="chapter" data-level="28" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html"><i class="fa fa-check"></i><b>28</b> CRISPR-CAS9 screen data processing</a><ul>
<li class="chapter" data-level="28.1" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html#installation-of-softwares-using-conda"><i class="fa fa-check"></i><b>28.1</b> Installation of softwares using <code>conda</code></a></li>
<li class="chapter" data-level="28.2" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html#an-exmaple-of-how-to-process"><i class="fa fa-check"></i><b>28.2</b> An exmaple of how to process</a><ul>
<li class="chapter" data-level="28.2.1" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html#donwload-the-data-1"><i class="fa fa-check"></i><b>28.2.1</b> Donwload the data</a></li>
<li class="chapter" data-level="28.2.2" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html#reference-1"><i class="fa fa-check"></i><b>28.2.2</b> Reference</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="29" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html"><i class="fa fa-check"></i><b>29</b> Commands for Linux administration (Advanced topic)</a><ul>
<li class="chapter" data-level="29.1" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#w"><i class="fa fa-check"></i><b>29.1</b> <code>w</code></a></li>
<li class="chapter" data-level="29.2" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#who"><i class="fa fa-check"></i><b>29.2</b> <code>who</code></a></li>
<li class="chapter" data-level="29.3" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#uptime"><i class="fa fa-check"></i><b>29.3</b> <code>uptime</code></a></li>
<li class="chapter" data-level="29.4" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#whoami"><i class="fa fa-check"></i><b>29.4</b> <code>whoami</code></a></li>
<li class="chapter" data-level="29.5" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#ifconfig"><i class="fa fa-check"></i><b>29.5</b> <code>ifconfig</code></a></li>
<li class="chapter" data-level="29.6" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#useradd-and-passwd"><i class="fa fa-check"></i><b>29.6</b> <code>useradd</code> and <code>passwd</code></a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">All things considered for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="input-and-output-in-perl" class="section level1">
<h1><span class="header-section-number">Chapter 13</span> Input and output in Perl</h1>
<div id="input" class="section level2">
<h2><span class="header-section-number">13.1</span> Input</h2>
<div id="standard-input" class="section level3">
<h3><span class="header-section-number">13.1.1</span> Standard input</h3>
</div>
<div id="input-from-a-file-on-the-disk" class="section level3">
<h3><span class="header-section-number">13.1.2</span> Input from a file on the disk</h3>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="kw">my</span> <span class="dt">$fasta</span> = <span class="kw">&quot;</span><span class="st">./data/test_ref.fa</span><span class="kw">&quot;</span>;

<span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="dt">$!</span><span class="st">: </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;
<span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$line</span> = <span class="kw">&lt;FASTA&gt;</span>){
    <span class="fu">chomp</span> <span class="dt">$line</span>;
    <span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$line</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}
<span class="fu">close</span> FASTA;</code></pre>
<pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">perl</span> code_perl/open_file.pl</code></pre>
<pre><code>## &gt;chr1
## ACGCTAGCTAGTCAGTCGATCGT
## CGTAGCTAGCTAG
## &gt;chr2
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTA
## &gt;chr3
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTA</code></pre>
</div>
<div id="special-variable-_" class="section level3">
<h3><span class="header-section-number">13.1.3</span> Special variable <code>$_</code></h3>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="kw">my</span> <span class="dt">$fasta</span> = <span class="kw">&quot;</span><span class="st">./data/test_ref.fa</span><span class="kw">&quot;</span>;

<span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="dt">$!</span><span class="st">: </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;
<span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$line</span> = <span class="kw">&lt;FASTA&gt;</span>){
    <span class="fu">chomp</span> <span class="dt">$line</span>;
    <span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$line</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}
<span class="fu">close</span> FASTA;</code></pre>
<p>When you are using <code>while</code> and <code>foreach</code>, the content will be assigned to <code>$_</code> if you don’t assign it explicitly to any variable.</p>
<p>For some build-in functions (<code>chomp</code>, <code>length</code>, <code>split</code> and <code>print</code>),</p>
<p>I do NOT suggest you to use <code>$_</code> because <code>$_</code> will reduce the readerbility of the code.</p>
<p>In Perl, $_ is a powerful
In Perl, several functions and operators use this variable as a default, in case no parameter is explicitly used. In general, I’d say you should NOT see $_ in real code. I think the whole point of $_ is that you don’t have to write it explicitly.</p>
</div>
<div id="the-input-record-separator" class="section level3">
<h3><span class="header-section-number">13.1.4</span> The input record separator: <code>$/</code></h3>
<p><code>$/</code> is the input record separator. By default, <code>$/</code> equals newline (<code>\n</code>).</p>
<p>You could actually change the value of <code>$/</code>. For example, by assigning the greate-than ‘&gt;’ like this: <code>$/ = &quot;&gt;&quot;</code>;. Then every call to the read-line operator <code>$chunk = &lt;FILEHANDLE&gt;</code> will read in all the characters up-to and including the first ‘&gt;’.</p>
<p>We could also assign longer strings to <code>$/</code> and then that would be the input record separator.</p>
<div id="how-to-process-fasta-file-by-changing" class="section level4">
<h4><span class="header-section-number">13.1.4.1</span> How to process <code>fasta</code> file by changing <code>$/</code></h4>
<p>FASTA format is a text-based format for representing either nucleotide sequences or peptide sequences, in which base pairs or amino acids are represented using single-letter codes (A, T, G, C, etc.).</p>
<p>In <code>fasta</code> format, each sequence begins with a single-line description, followed by lines of sequence data. The description line is distinguished from the sequence data by a greater-than (“&gt;”) symbol in the first column.</p>
<p>Here is an example of fasta file (Figure <a href="input-and-output-in-perl.html#fig:fastaPYL">13.1</a>).</p>

<div class="figure"><span id="fig:fastaPYL"></span>
<img src="figures/fasta_pyl10_ara.png" alt="Protein sequence of PYL10 in Arabidopsis."  />
<p class="caption">
Figure 13.1: Protein sequence of PYL10 in Arabidopsis.
</p>
</div>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="kw">my</span> <span class="dt">$fasta</span> = <span class="kw">&quot;</span><span class="st">./data/test_ref.fa</span><span class="kw">&quot;</span>;

<span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="dt">$!</span><span class="st">: </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;
<span class="co"># Set the input record separator</span>
<span class="dt">$/</span> = <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">&gt;</span><span class="kw">&quot;</span>;
<span class="co"># Print header</span>
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Chrom</span><span class="ch">\t</span><span class="st">Length</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$chunk</span> = <span class="kw">&lt;FASTA&gt;</span>){
    <span class="co"># Remove &gt; in the first chunk</span>
    <span class="dt">$chunk</span> =~ <span class="kw">s/</span><span class="ch">^</span><span class="ot">&gt;</span><span class="kw">//</span>;
    <span class="co"># Remove \n&gt; from the end of the chunk</span>
    <span class="fu">chomp</span> <span class="dt">$chunk</span>;
    <span class="co"># Assign ID and seq to two variables</span>
    <span class="kw">my</span> (<span class="dt">$chrom</span>, <span class="dt">$seq</span>) = <span class="fu">split</span>(<span class="kw">/</span><span class="bn">\n</span><span class="kw">/</span>, <span class="dt">$chunk</span>, <span class="dv">2</span>);
    <span class="co"># Remove \n in the string</span>
    <span class="dt">$seq</span> =~ <span class="kw">s/</span><span class="bn">\n</span><span class="kw">//g</span>;
    <span class="co"># Calculate the length of sequence.</span>
    <span class="kw">my</span> <span class="dt">$seq_length</span> = <span class="fu">length</span> <span class="dt">$seq</span>;
    <span class="co"># Print ID and length of the sequence.</span>
    <span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$chrom</span><span class="ch">\t</span><span class="dt">$seq_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}
<span class="fu">close</span> FASTA;</code></pre>
<p>You can consider the content in the fasta file as a string (Figure <a href="input-and-output-in-perl.html#fig:fastaString">13.2</a>) sperarated by <code>\n&gt;</code>.</p>
<p>We set the input record separator <code>$/</code> as <code>\n&gt;</code>.</p>
<p>In the first cycle of the <code>while</code> loop, the first chunk is <code>&gt;chr1\nACGCTAGCTAGTCAGTCGATCGT\nCGTAGCTAGCTAG\n&gt;</code>.</p>
<p>You should notice that there is a leading <code>&gt;</code> in the first chunk. So you need to use <code>s/^&gt;//</code> to remove the leading <code>&gt;</code>.</p>
<p>To remove the trailing <code>\n&gt;</code>, you can use <code>chomp $chunk;</code>. Then the string is:</p>
<pre><code>chr1\nACGCTAGCTAGTCAGTCGATCGT\nCGTAGCTAGCTAG</code></pre>
<p>The string before first <code>\n</code> is the sequence ID. The string after the first <code>\n</code> is the sequence. To get the sequnce ID and sequnce and assign to two variables, you can use <code>split(/\n/, $chunk, 2)</code>. Here LIMIT is set to 2, it represents the maximum number (here 2) of fields into which the EXPR may be split.</p>
<p>The goal of this code is to output the sequence IDs and their lengths. Currently the string of the variable (<code>$seq</code>) is <code>ACGCTAGCTAGTCAGTCGATCGT\nCGTAGCTAGCTAG</code>. To do this, you first need to remove <code>\n</code> in the sequence by using <code>$seq =~ s/\n//g;</code>. Now the string of the variable (<code>$seq</code>) is <code>ACGCTAGCTAGTCAGTCGATCGTCGTAGCTAGCTAG</code>. Then built-in function <code>length</code> can be used to calculate the length of the sequence. The function <code>print</code> is used to output the sequnce ID and the length of the sequence.</p>
<p>In the second cycles of the <code>while</code> loop, the chunk is
<code>chr2\nCTGCGGGCTAAATCGATCGATCG\nGTACGTACGAGCTAGCTA\n&gt;</code>.</p>
<p>Unlike the first chunk, there is no leading <code>&gt;</code> in the second chunk. So the code <code>s/^&gt;//</code> will do nothing for the second chunk.</p>
<p>The following steps is the same as you can do for the first cycle.</p>
<p>In the third cycle of the <code>while</code> loop, you can do the same thing on the third sequence.</p>

<div class="figure"><span id="fig:fastaString"></span>
<img src="figures/fasta_string.png" alt="Fasta string."  />
<p class="caption">
Figure 13.2: Fasta string.
</p>
</div>
<pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">perl</span> code_perl/fa_seq_len.pl</code></pre>
<pre><code>## Chrom    Length
## chr1 36
## chr2 41
## chr3 41</code></pre>
</div>
</div>
</div>
<div id="output-to-a-file-on-the-disk" class="section level2">
<h2><span class="header-section-number">13.2</span> Output to a file on the disk</h2>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="kw">my</span> <span class="dt">$fasta</span>   = <span class="kw">&quot;</span><span class="st">./data/test_ref.fa</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$out_len</span> = <span class="kw">&quot;</span><span class="st">./data/test_ref_len.txt</span><span class="kw">&quot;</span>;

<span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">Can&#39;t open file for reading: </span><span class="dt">$!</span><span class="st"> </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;
<span class="fu">open</span> OUT, <span class="kw">&quot;</span><span class="st">+&gt;</span><span class="dt">$out_len</span><span class="kw">&quot;</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">Can&#39;t open file for writing: </span><span class="dt">$!</span><span class="st"> </span><span class="dt">$out_len</span><span class="kw">&quot;</span>;
<span class="co"># Set the input record separator</span>
<span class="dt">$/</span> = <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">&gt;</span><span class="kw">&quot;</span>;
<span class="co"># Print header</span>
<span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="st">Chrom</span><span class="ch">\t</span><span class="st">Length</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$chunk</span> = <span class="kw">&lt;FASTA&gt;</span>){
    <span class="co"># Remove &gt; in the first chunk</span>
    <span class="dt">$chunk</span> =~ <span class="kw">s/</span><span class="ch">^</span><span class="ot">&gt;</span><span class="kw">//</span>;
    <span class="co"># Remove \n&gt; from the end of the chunk</span>
    <span class="fu">chomp</span> <span class="dt">$chunk</span>;
    <span class="co"># Assign ID and seq to two variables</span>
    <span class="kw">my</span> (<span class="dt">$chrom</span>, <span class="dt">$seq</span>) = <span class="fu">split</span>(<span class="kw">/</span><span class="bn">\n</span><span class="kw">/</span>, <span class="dt">$chunk</span>, <span class="dv">2</span>);
    <span class="co"># Remove \n in the string</span>
    <span class="dt">$seq</span> =~ <span class="kw">s/</span><span class="bn">\n</span><span class="kw">//g</span>;
    <span class="co"># Calculate the length of sequence.</span>
    <span class="kw">my</span> <span class="dt">$seq_length</span> = <span class="fu">length</span> <span class="dt">$seq</span>;
    <span class="co"># Print ID and length of the sequence.</span>
    <span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="dt">$chrom</span><span class="ch">\t</span><span class="dt">$seq_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}
<span class="fu">close</span> FASTA;
<span class="fu">close</span> OUT;</code></pre>
<pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">perl</span> code_perl/fa_seq_len_out.pl
<span class="fu">cat</span> ./data/test_ref_len.txt</code></pre>
<pre><code>## Chrom    Length
## chr1 36
## chr2 41
## chr3 41</code></pre>
</div>
<div id="arguments-from-command-line" class="section level2">
<h2><span class="header-section-number">13.3</span> Arguments from command line</h2>
<p>Imagine you have 5 fasta files and you want to calculate sequence lengths for all the 5 files, if you modify the file each time you run the code, it will be very tedious.</p>
<p>Perls provides an array called <code>@ARGV</code>. <code>@ARGV</code> holds all the arguments from the command line. The first one will be <code>$ARGV[0]</code>.</p>
<p><code>@ARGV</code> will automatically hold all the arguments. If no arguments are provided, <code>@ARGV</code> will be empty.</p>
<p>The following example shows you how it looks like in real code.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="kw">my</span> (<span class="dt">$fasta</span>, <span class="dt">$out_len</span>) = <span class="dt">@ARGV</span>; 
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First arguments \$ARGV[0] : </span><span class="dt">$ARGV</span><span class="st">[0]</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second arguments \$ARGV[0]: </span><span class="dt">$ARGV</span><span class="st">[1]</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The variable \$fasta:   </span><span class="dt">$fasta</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The variable \$out_len: </span><span class="dt">$out_len</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">perl</span> code_perl/test_argv.pl test_ref.fa test_ref_len.txt</code></pre>
<pre><code>## First arguments $ARGV[0] : test_ref.fa
## Second arguments $ARGV[0]: test_ref_len.txt
## The variable $fasta:   test_ref.fa
## The variable $out_len: test_ref_len.txt</code></pre>
<p>In the above example, the first argument is <code>$ARGV[0]</code> which stores the file name: test_ref.fa. The second one stores the file name: test_ref_len.txt. It’s highly recommended to assign the values in <code>@ARGV</code> to some variables that the readers (including yourself) can understand the variables based on the names.</p>
<p>Here I’ll show you how to use ARGV using an example. The code below will show you what to do if you want to filter a fasta file based on the sequence length.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="kw">my</span> (<span class="dt">$fasta</span>, <span class="dt">$out_len</span>) = <span class="dt">@ARGV</span>; 

<span class="fu">open</span> FASTA, <span class="dt">$fasta</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">Can&#39;t open file for reading: </span><span class="dt">$!</span><span class="st"> </span><span class="dt">$fasta</span><span class="kw">&quot;</span>;
<span class="fu">open</span> OUT, <span class="kw">&quot;</span><span class="st">+&gt;</span><span class="dt">$out_len</span><span class="kw">&quot;</span> <span class="kw">or</span> <span class="fu">die</span> <span class="kw">&quot;</span><span class="st">Can&#39;t open file for writing: </span><span class="dt">$!</span><span class="st"> </span><span class="dt">$out_len</span><span class="kw">&quot;</span>;
<span class="co"># Set the input record separator</span>
<span class="dt">$/</span> = <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">&gt;</span><span class="kw">&quot;</span>;
<span class="co"># Print header</span>
<span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="st">The input file name is: </span><span class="dt">$fasta</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="st">Chrom</span><span class="ch">\t</span><span class="st">Length</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="kw">while</span>(<span class="kw">my</span> <span class="dt">$chunk</span> = <span class="kw">&lt;FASTA&gt;</span>){
    <span class="co"># Remove &gt; in the first chunk</span>
    <span class="dt">$chunk</span> =~ <span class="kw">s/</span><span class="ch">^</span><span class="ot">&gt;</span><span class="kw">//</span>;
    <span class="co"># Remove \n&gt; from the end of the chunk</span>
    <span class="fu">chomp</span> <span class="dt">$chunk</span>;
    <span class="co"># Assign ID and seq to two variables</span>
    <span class="kw">my</span> (<span class="dt">$chrom</span>, <span class="dt">$seq</span>) = <span class="fu">split</span>(<span class="kw">/</span><span class="bn">\n</span><span class="kw">/</span>, <span class="dt">$chunk</span>, <span class="dv">2</span>);
    <span class="co"># Remove \n in the string</span>
    <span class="dt">$seq</span> =~ <span class="kw">s/</span><span class="bn">\n</span><span class="kw">//g</span>;
    <span class="co"># Calculate the length of sequence.</span>
    <span class="kw">my</span> <span class="dt">$seq_length</span> = <span class="fu">length</span> <span class="dt">$seq</span>;
    <span class="co"># Print ID and length of the sequence.</span>
    <span class="fu">print</span> OUT <span class="kw">&quot;</span><span class="dt">$chrom</span><span class="ch">\t</span><span class="dt">$seq_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}
<span class="fu">close</span> FASTA;
<span class="fu">close</span> OUT;</code></pre>
<p>For example if you want to remove the sequences that is shorter than 30 bps, you can use the following example.</p>
<pre class="sourceCode sh"><code class="sourceCode bash"><span class="bu">printf</span> <span class="st">&quot;Before filtering:\n&quot;</span>
<span class="fu">cat</span> ./data/test_ref2.fa
<span class="fu">perl</span> code_perl/fa_seq_len_out_argv_fil.pl ./data/test_ref2.fa 30 ./data/test_ref2_30.fa

<span class="bu">printf</span> <span class="st">&quot;\nAfter filtering: \n&quot;</span>
<span class="fu">cat</span> ./data/test_ref2_30.fa </code></pre>
<pre><code>## Before filtering:
## &gt;chr1
## ACGCTAGCTAGTCAGTCGATCGT
## CGTAGCTAGCTAG
## &gt;chr2
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAGCTAGCTAA
## &gt;chr3
## CTGCGGGCTAAATCGATCGATCG
## GTACGTACGAG
## &gt;chr4
## CTGCGGGCTAAATCAGCTAA
## &gt;chr5
## CTGCTCGATCGATCGACGAGCTA
## GCTA
## After filtering: 
## The input file name is: ./data/test_ref2.fa
## Chrom    Length
## &gt;chr1 36
## ACGCTAGCTAGTCAGTCGATCGTCGTAGCTAGCTAG
## &gt;chr2 42
## CTGCGGGCTAAATCGATCGATCGGTACGTACGAGCTAGCTAA
## &gt;chr3 34
## CTGCGGGCTAAATCGATCGATCGGTACGTACGAG</code></pre>
<p>If you have multiple files, you can just change the file names in the command line. It’s very handy. You can also change the length cutoff in the command line.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="string-manipulation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="practical-perl-program.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/OmicsAcademy/Novice2Expert4Bioinformatics/edit/master/15_PerlInputOutput.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["bioinfBookXIE186.pdf", "bioinfBookXIE186.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
